{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# T027 · Kinase similarity: ligand-profile\n",
    "\n",
    "Authors:\n",
    "\n",
    "- Talia B. Kimber, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)\n",
    "- Dominique Sydow, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)\n",
    "- Andrea Volkamer, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aim of this talktorial\n",
    "\n",
    "The aim of this talktorial is to investigate kinase similarity through ligand profiling data. In the context of drug design, the following assumption can be made: if a compound was tested as active on two different kinases, it is suspected that these two kinases may have some degree of similarity. The concept of kinase promiscuity is also covered."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contents in *Theory*\n",
    "\n",
    "* Kinase dataset\n",
    "* Bioactivity data\n",
    "* Kinase similarity descriptor: ligand-profile\n",
    "    * Kinase similarity\n",
    "    * Kinase promiscuity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contents in *Practical*\n",
    "\n",
    "* Retrieve the data\n",
    "* Preprocess the data\n",
    "    * Kinases of interest\n",
    "    * Hit or non-hit\n",
    "* Kinase promiscuity\n",
    "* Kinase similarity\n",
    "    * Visualize similarity as kinase matrix\n",
    "    * Save kinase similarity matrix\n",
    "* Kinase distance matrix\n",
    "    * Save kinase distance matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "\n",
    "* Kinase dataset: [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629) \n",
    "* [ChEMBL](https://www.ebi.ac.uk/chembl/) database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Theory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use nine kinases as investigated in [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629). In the study diverse kinase similarity measures were analyzed for different combinations of kinase on- and off-targets to explore the limits of multi-kinase screenings.\n",
    "\n",
    " \n",
    "\n",
    "> We aggregated the investigated kinases in “profiles”. Profile 1 combined **EGFR** and **ErbB2** as targets and **BRAF** as a (general) anti-target. Out of similar considerations, Profile 2 consisted of EGFR and **PI3K** as targets and BRAF as anti-target. This profile is expected to be more challenging as PI3K is an atypical kinase and thus less similar to EGFR than for example ErbB2 used in Profile 1. Profile 3, comprised of EGFR and **VEGFR2** as targets and BRAF as anti-target, was contrasted with the hit rate that we found with a standard docking against the single target VEGFR2 (Profile 4).\n",
    "> To broaden the comparison and obtain an estimate for the promiscuity of each compound, the kinases **CDK2**, **LCK**, **MET** and **p38α** were included in the experimental assay panel and the structure-based bioinformatics comparison as commonly used anti-targets.\n",
    "\n",
    " \n",
    "\n",
    "*Table 1:* \n",
    "Kinases used in this notebook, taken from [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629), with their synonyms, UniProt IDs, kinase groups, and full unabbreviated names.\n",
    "\n",
    " \n",
    "\n",
    "| Kinase                     | Synonyms               | UniProt ID | Group    | Full kinase name                                 |\n",
    "|----------------------------|------------------------|------------|----------|--------------------------------------------------|\n",
    "| EGFR                       | ErbB1                  | P00533     | TK       | Epidermal growth factor receptor                 |\n",
    "| ErbB2                      | Her2                   | P04626     | TK       | Erythroblastic leukemia viral oncogene homolog 2 |\n",
    "| PI3K                       | PI3KCA, p110a          | P42336     | Atypical | Phosphatidylinositol-3-kinase                    |\n",
    "| VEGFR2                     | KDR                    | P35968     | TK       | Vascular endothelial growth factor receptor 2    |\n",
    "| BRAF                       | -                      | P15056     | TKL      | Rapidly accelerated fibrosarcoma isoform B       |\n",
    "| CDK2                       | -                      | P24941     | CMGC     | Cyclic-dependent kinase 2                        |\n",
    "| LCK                        | -                      | P06239     | TK       | Lymphocyte-specific protein tyrosine kinase      |\n",
    "| MET                        | -                      | P08581     | TK       | Mesenchymal-epithelial transition factor         |\n",
    "| p38a                       | MAPK14                 | Q16539     | CMGC     | p38 mitogen activated protein kinase α           |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bioactivity data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to measure kinase similarity through ligand profiling data, bioactivity data is retrieved from the well-known [ChEMBL](https://www.ebi.ac.uk/chembl/) database and the query focuses on human-kinases. Luckily, a curated version of ChEMBL28 is already freely available in the openkinome organization, see https://github.com/openkinome/kinodata.\n",
    "For more details on querying the ChEMBL database, please refer to [Talktorial T001](https://github.com/volkamerlab/teachopencadd/blob/master/teachopencadd/talktorials/T001_query_chembl/talktorial.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In drug design, it is common to binarize the activity of a compound against a target of interest as a \"hit\" or \"non-hit\". Practically speaking, this is done using a cutoff value for measured activity. If the activity is greater than the cutoff, the compound is labeled as active (hit), and inactive (non-hit) otherwise."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase similarity descriptor: ligand-profile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a measure of similarity, we use ligand profiling data in this talktorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Kinase similarity\n",
    "\n",
    "We use the following metric as similarity between kinases $K_i$ and $K_j$:\n",
    "\n",
    "$$\n",
    "\\text{similarity}(K_i, K_j) = \\frac{\\#\\text{ of compounds that were tested as actives on } K_i \\text{ and } K_j}\n",
    "{\\#\\text{ of compounds that were tested on both } K_i \\text{ and } K_j}.\n",
    "$$\n",
    "\n",
    "Assuming that only one compound was tested on two kinases, and that the compound was tested as active for one and inactive for the other, then the similarity between these two kinases would be zero."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kinase promiscuity\n",
    "Computing the similarity between a kinase and itself may be interpreted as kinase promiscuity, where the similarity described above would therefore represent the fraction of active compounds over all tested compounds."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Practical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "HERE = Path(_dh[-1])\n",
    "DATA = HERE / \"data\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We retrieve a pre-curated version of a kinase subset of ChEMBL v.28 freely available at openkinome, see https://github.com/openkinome/kinodata/releases/tag/v0.2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current shape of data: (186972, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>activities.activity_id</th>\n",
       "      <th>assays.chembl_id</th>\n",
       "      <th>target_dictionary.chembl_id</th>\n",
       "      <th>molecule_dictionary.chembl_id</th>\n",
       "      <th>molecule_dictionary.max_phase</th>\n",
       "      <th>activities.standard_type</th>\n",
       "      <th>activities.standard_value</th>\n",
       "      <th>activities.standard_units</th>\n",
       "      <th>compound_structures.canonical_smiles</th>\n",
       "      <th>compound_structures.standard_inchi</th>\n",
       "      <th>component_sequences.sequence</th>\n",
       "      <th>assays.confidence_score</th>\n",
       "      <th>docs.chembl_id</th>\n",
       "      <th>docs.year</th>\n",
       "      <th>docs.authors</th>\n",
       "      <th>UniprotID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>96251</td>\n",
       "      <td>16291323</td>\n",
       "      <td>CHEMBL3705523</td>\n",
       "      <td>CHEMBL2973</td>\n",
       "      <td>CHEMBL3666724</td>\n",
       "      <td>0</td>\n",
       "      <td>pIC50</td>\n",
       "      <td>14.096910</td>\n",
       "      <td>nM</td>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>InChI=1S/C31H33N7O3/c1-2-4-29(40)33-22-6-3-5-2...</td>\n",
       "      <td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>\n",
       "      <td>9</td>\n",
       "      <td>CHEMBL3639077</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>97672</td>\n",
       "      <td>16306943</td>\n",
       "      <td>CHEMBL3705523</td>\n",
       "      <td>CHEMBL2973</td>\n",
       "      <td>CHEMBL1968705</td>\n",
       "      <td>0</td>\n",
       "      <td>pIC50</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>nM</td>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>InChI=1S/C31H33N7O2/c1-2-6-29(39)33-23-8-5-7-2...</td>\n",
       "      <td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>\n",
       "      <td>9</td>\n",
       "      <td>CHEMBL3639077</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>93879</td>\n",
       "      <td>16264754</td>\n",
       "      <td>CHEMBL3705523</td>\n",
       "      <td>CHEMBL2973</td>\n",
       "      <td>CHEMBL3666728</td>\n",
       "      <td>0</td>\n",
       "      <td>pIC50</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>nM</td>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>InChI=1S/C34H40N8O3/c1-5-7-32(43)36-24-9-6-8-2...</td>\n",
       "      <td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>\n",
       "      <td>9</td>\n",
       "      <td>CHEMBL3639077</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100714</td>\n",
       "      <td>16340050</td>\n",
       "      <td>CHEMBL3705523</td>\n",
       "      <td>CHEMBL2973</td>\n",
       "      <td>CHEMBL1997433</td>\n",
       "      <td>0</td>\n",
       "      <td>pIC50</td>\n",
       "      <td>13.958607</td>\n",
       "      <td>nM</td>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>InChI=1S/C28H28N6O3/c1-3-5-26(35)30-20-7-4-6-1...</td>\n",
       "      <td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>\n",
       "      <td>9</td>\n",
       "      <td>CHEMBL3639077</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>95877</td>\n",
       "      <td>16287186</td>\n",
       "      <td>CHEMBL3705523</td>\n",
       "      <td>CHEMBL2973</td>\n",
       "      <td>CHEMBL3666721</td>\n",
       "      <td>0</td>\n",
       "      <td>pIC50</td>\n",
       "      <td>13.920819</td>\n",
       "      <td>nM</td>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>InChI=1S/C32H35N7O2/c1-2-7-30(40)34-24-9-6-8-2...</td>\n",
       "      <td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>\n",
       "      <td>9</td>\n",
       "      <td>CHEMBL3639077</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  activities.activity_id assays.chembl_id  \\\n",
       "0       96251                16291323    CHEMBL3705523   \n",
       "1       97672                16306943    CHEMBL3705523   \n",
       "2       93879                16264754    CHEMBL3705523   \n",
       "3      100714                16340050    CHEMBL3705523   \n",
       "4       95877                16287186    CHEMBL3705523   \n",
       "\n",
       "  target_dictionary.chembl_id molecule_dictionary.chembl_id  \\\n",
       "0                  CHEMBL2973                 CHEMBL3666724   \n",
       "1                  CHEMBL2973                 CHEMBL1968705   \n",
       "2                  CHEMBL2973                 CHEMBL3666728   \n",
       "3                  CHEMBL2973                 CHEMBL1997433   \n",
       "4                  CHEMBL2973                 CHEMBL3666721   \n",
       "\n",
       "   molecule_dictionary.max_phase activities.standard_type  \\\n",
       "0                              0                    pIC50   \n",
       "1                              0                    pIC50   \n",
       "2                              0                    pIC50   \n",
       "3                              0                    pIC50   \n",
       "4                              0                    pIC50   \n",
       "\n",
       "   activities.standard_value activities.standard_units  \\\n",
       "0                  14.096910                        nM   \n",
       "1                  14.000000                        nM   \n",
       "2                  14.000000                        nM   \n",
       "3                  13.958607                        nM   \n",
       "4                  13.920819                        nM   \n",
       "\n",
       "                compound_structures.canonical_smiles  \\\n",
       "0  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...   \n",
       "1  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...   \n",
       "2  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...   \n",
       "3  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...   \n",
       "4  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...   \n",
       "\n",
       "                  compound_structures.standard_inchi  \\\n",
       "0  InChI=1S/C31H33N7O3/c1-2-4-29(40)33-22-6-3-5-2...   \n",
       "1  InChI=1S/C31H33N7O2/c1-2-6-29(39)33-23-8-5-7-2...   \n",
       "2  InChI=1S/C34H40N8O3/c1-5-7-32(43)36-24-9-6-8-2...   \n",
       "3  InChI=1S/C28H28N6O3/c1-3-5-26(35)30-20-7-4-6-1...   \n",
       "4  InChI=1S/C32H35N7O2/c1-2-7-30(40)34-24-9-6-8-2...   \n",
       "\n",
       "                        component_sequences.sequence  assays.confidence_score  \\\n",
       "0  MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...                        9   \n",
       "1  MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...                        9   \n",
       "2  MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...                        9   \n",
       "3  MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...                        9   \n",
       "4  MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...                        9   \n",
       "\n",
       "  docs.chembl_id  docs.year docs.authors UniprotID  \n",
       "0  CHEMBL3639077     2014.0          NaN    O75116  \n",
       "1  CHEMBL3639077     2014.0          NaN    O75116  \n",
       "2  CHEMBL3639077     2014.0          NaN    O75116  \n",
       "3  CHEMBL3639077     2014.0          NaN    O75116  \n",
       "4  CHEMBL3639077     2014.0          NaN    O75116  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"https://github.com/openkinome/kinodata/releases/download/\\\n",
    "v0.2/activities-chembl28_v0.2.zip\"\n",
    "data = pd.read_csv(path, index_col=None)\n",
    "print(f\"Current shape of data: {data.shape}\")\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocess the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We look at the type of activity and the associated units."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activities: {'pIC50', 'pKd', 'pKi'}\n",
      "Units: {'nM'}\n"
     ]
    }
   ],
   "source": [
    "print(f\"Activities: {set(data['activities.standard_type'])}\\n\"\n",
    "      f\"Units: {set(data['activities.standard_units'])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's keep the entries which have pIC50 values only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data[data[\"activities.standard_type\"] == \"pIC50\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'activities.activity_id', 'assays.chembl_id',\n",
       "       'target_dictionary.chembl_id', 'molecule_dictionary.chembl_id',\n",
       "       'molecule_dictionary.max_phase', 'activities.standard_type',\n",
       "       'activities.standard_value', 'activities.standard_units',\n",
       "       'compound_structures.canonical_smiles',\n",
       "       'compound_structures.standard_inchi', 'component_sequences.sequence',\n",
       "       'assays.confidence_score', 'docs.chembl_id', 'docs.year',\n",
       "       'docs.authors', 'UniprotID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The dataframe contains many columns that won't be necessary for the rest of the notebook which are therefore removed. \n",
    "Only relevant information is kept, namely the canonical SMILES of the compound, the measured activity and the UniProt ID of the kinase. These columns are renamed for readability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data[[\"compound_structures.canonical_smiles\",\n",
    "             \"activities.standard_value\",\n",
    "             \"UniprotID\"]]\n",
    "data = data.rename(columns={\"compound_structures.canonical_smiles\": \"smiles\",\n",
    "                            \"activities.standard_value\": \"activity_value\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current shape of data: (159978, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>smiles</th>\n",
       "      <th>activity_value</th>\n",
       "      <th>UniprotID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>14.096910</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>13.958607</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>\n",
       "      <td>13.920819</td>\n",
       "      <td>O75116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              smiles  activity_value UniprotID\n",
       "0  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...       14.096910    O75116\n",
       "1  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...       14.000000    O75116\n",
       "2  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...       14.000000    O75116\n",
       "3  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...       13.958607    O75116\n",
       "4  CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...       13.920819    O75116"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Current shape of data: {data.shape}\")\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also drop NA values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current shape of data: (159823, 3)\n"
     ]
    }
   ],
   "source": [
    "data = data.dropna()\n",
    "print(f\"Current shape of data: {data.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kinases of interest\n",
    "\n",
    "We focus on the kinases of interest and map them to their [UniProt](https://www.uniprot.org/) IDs, using Table 1 from the theory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_to_uniprot = {'EGFR': 'P00533',\n",
    "                   'ErbB2': 'P04626',\n",
    "                   'BRAF': 'P15056',\n",
    "                   'CDK2': 'P24941',\n",
    "                   'LCK': 'P06239',\n",
    "                   'MET': 'P08581',\n",
    "                   'p38a': 'Q16539',\n",
    "                   'KDR': 'P35968',\n",
    "                   'p110a': 'P42336'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We keep data for these kinases only:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current shape of data: (33169, 3)\n"
     ]
    }
   ],
   "source": [
    "data = data[data[\"UniprotID\"].isin(name_to_uniprot.values())]\n",
    "print(f\"Current shape of data: {data.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>smiles</th>\n",
       "      <th>activity_value</th>\n",
       "      <th>UniprotID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Brc1cccc(Nc2ncnc3cc4ccccc4cc23)c1</td>\n",
       "      <td>11.522879</td>\n",
       "      <td>P00533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>CCOc1cc2ncnc(Nc3cccc(Br)c3)c2cc1OCC</td>\n",
       "      <td>11.221849</td>\n",
       "      <td>P00533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>CN(C)c1cc2c(Nc3cccc(Br)c3)ncnc2cn1</td>\n",
       "      <td>11.221849</td>\n",
       "      <td>P00533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Brc1cccc(Nc2ncnc3cc4[nH]cnc4cc23)c1</td>\n",
       "      <td>11.096910</td>\n",
       "      <td>P00533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>CNc1cc2c(Nc3cccc(Br)c3)ncnc2cn1</td>\n",
       "      <td>11.096910</td>\n",
       "      <td>P00533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  smiles  activity_value UniprotID\n",
       "57     Brc1cccc(Nc2ncnc3cc4ccccc4cc23)c1       11.522879    P00533\n",
       "98   CCOc1cc2ncnc(Nc3cccc(Br)c3)c2cc1OCC       11.221849    P00533\n",
       "101   CN(C)c1cc2c(Nc3cccc(Br)c3)ncnc2cn1       11.221849    P00533\n",
       "139  Brc1cccc(Nc2ncnc3cc4[nH]cnc4cc23)c1       11.096910    P00533\n",
       "140      CNc1cc2c(Nc3cccc(Br)c3)ncnc2cn1       11.096910    P00533"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at EGFR data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "EGFR_data = data[data[\"UniprotID\"] == \"P00533\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some compounds have been tested several times against EGFR, as shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('COc1cc2ncnc(Nc3ccc(F)c(Cl)c3)c2cc1OCCCN1CCOCC1', 40),\n",
       " ('C#Cc1cccc(Nc2ncnc3cc(OCCOC)c(OCCOC)cc23)c1', 24),\n",
       " ('C=CC(=O)Nc1cc(Nc2nccc(-c3cn(C)c4ccccc34)n2)c(OC)cc1N(C)CCN(C)C', 12),\n",
       " ('C=CC(=O)Nc1cccc(Oc2nc(Nc3ccc(N4CCN(C)CC4)cc3OC)ncc2Cl)c1', 10),\n",
       " ('CN[C@@H]1C[C@H]2O[C@@](C)([C@@H]1OC)n1c3ccccc3c3c4c(c5c6ccccc6n2c5c31)C(=O)NC4',\n",
       "  7)]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measured_compounds = Counter(EGFR_data[\"smiles\"])\n",
    "measured_compounds.most_common()[0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a simple workaround, we keep the value for which the compound has the best activity value, i.e., highest pIC50 value here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>smiles</th>\n",
       "      <th>activity_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1ccc(C(...</td>\n",
       "      <td>5.336488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1cccc2c...</td>\n",
       "      <td>5.996539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.CC[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1...</td>\n",
       "      <td>8.397940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>\n",
       "      <td>7.207608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>\n",
       "      <td>8.420216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniprotID                                             smiles  activity_value\n",
       "0    P00533  Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1ccc(C(...        5.336488\n",
       "1    P00533  Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1cccc2c...        5.996539\n",
       "2    P00533  Br.CC[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1...        8.397940\n",
       "3    P00533  Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...        7.207608\n",
       "4    P00533  Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...        8.420216"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = data.groupby([\"UniprotID\",\n",
    "                     \"smiles\"])['activity_value'].max().reset_index()\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Hit or non-hit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we binarize the pIC50 values to obtain hit or non-hit using a cut-off. We use a pIC50 cut-off of $6.3$, similarly to the [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff = 6.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binarize_pic50(pic50_value, threshold):\n",
    "    \"\"\"\n",
    "    Binarizes a scalar value given a threshold.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    pic50_value : float\n",
    "        The measurement pIC50 value of a kinase-ligand pair.\n",
    "    threshold : float\n",
    "        The cutoff to determine activity.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    1 if the pIC50 value is above the threshold, which indicates activity.\n",
    "    0 otherwise.\n",
    "    \"\"\"\n",
    "    if pic50_value >= threshold:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"activity_binary\"] = data[\"activity_value\"].apply(binarize_pic50,\n",
    "                                                       args=(cutoff, ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current shape of data: (32688, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>smiles</th>\n",
       "      <th>activity_value</th>\n",
       "      <th>activity_binary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1ccc(C(...</td>\n",
       "      <td>5.336488</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1cccc2c...</td>\n",
       "      <td>5.996539</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.CC[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1...</td>\n",
       "      <td>8.397940</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>\n",
       "      <td>7.207608</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>\n",
       "      <td>8.420216</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  UniprotID                                             smiles  \\\n",
       "0    P00533  Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1ccc(C(...   \n",
       "1    P00533  Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1cccc2c...   \n",
       "2    P00533  Br.CC[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1...   \n",
       "3    P00533  Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...   \n",
       "4    P00533  Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...   \n",
       "\n",
       "   activity_value  activity_binary  \n",
       "0        5.336488                0  \n",
       "1        5.996539                0  \n",
       "2        8.397940                1  \n",
       "3        7.207608                1  \n",
       "4        8.420216                1  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Current shape of data: {data.shape}\")\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase promiscuity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now look at the kinase promiscuity.\n",
    "\n",
    "For a given kinase, three values are computed:\n",
    "\n",
    "1. the total number of measured compounds against the given kinase,\n",
    "2. the number of active compounds against the kinase, and\n",
    "3. the fraction of active compounds, i.e. the ratio of active compounds over the total number of measured compounds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kinase_to_activity_numbers(uniprot_id, activity_dataframe):\n",
    "    \"\"\"\n",
    "    Retrieve the three values for a given kinase.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    uniprot_id : str\n",
    "        The uniprot id of the kinase of interest, e.g. \"P00533\" for \"EGFR\".\n",
    "\n",
    "    activity_dataframe :  pd.DataFrame\n",
    "        The dataframe with activity values for kinases.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    tuple : (int, int, float)\n",
    "        The three metrics:\n",
    "        1. The total number of measured compounds against the kinase\n",
    "        2. The number of active compounds against the kinase.\n",
    "        3. The fraction of active compounds against the kinase.\n",
    "    \"\"\"\n",
    "    kinase_data = activity_dataframe[activity_dataframe[\"UniprotID\"]\n",
    "                                     == uniprot_id]\n",
    "    total_measured_compounds = len(kinase_data)\n",
    "    active_compounds = len(kinase_data[kinase_data[\"activity_binary\"] == 1])\n",
    "    if total_measured_compounds > 0:\n",
    "        fraction = active_compounds/total_measured_compounds\n",
    "    else:\n",
    "        print(\"No compounds were measured for this kinase.\")\n",
    "        fraction = np.nan\n",
    "    return (total_measured_compounds, active_compounds, fraction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see what information we get for EGFR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of measured compounds: \t5869 \n",
      "Number of active compounds: \t\t3574 \n",
      "Fraction of active compounds: \t\t0.61 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "uniprot_id = name_to_uniprot[\"EGFR\"]\n",
    "EGFR_metrics = kinase_to_activity_numbers(uniprot_id, data)\n",
    "print(f\"Total number of measured compounds: \\t\"\n",
    "      f\"{EGFR_metrics[0]} \\n\"\n",
    "      f\"Number of active compounds: \\t\\t\"\n",
    "      f\"{EGFR_metrics[1]} \\n\"\n",
    "      f\"Fraction of active compounds: \\t\\t\"\n",
    "      f\"{EGFR_metrics[2]:.2f} \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's create a table from these values for all kinases:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_numbers_to_table(activity_dataframe):\n",
    "    \"\"\"\n",
    "    Create a table with all three values for all kinases.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    activity_dataframe :  pd.DataFrame\n",
    "        The dataframe with activity values for kinases.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    promiscuity_table : pd.DataFrame\n",
    "        A dataframe with the kinases as rows and values as columns.\n",
    "    \"\"\"\n",
    "    promiscuity_table = pd.DataFrame(index=name_to_uniprot.keys(),\n",
    "                                     columns=[\"total\", \"actives\", \"fraction\"])\n",
    "    for name, uniprot_id in name_to_uniprot.items():\n",
    "        values = kinase_to_activity_numbers(uniprot_id, activity_dataframe)\n",
    "        promiscuity_table.loc[name] = values\n",
    "    return promiscuity_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total</th>\n",
       "      <th>actives</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>5869</td>\n",
       "      <td>3574</td>\n",
       "      <td>0.608962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ErbB2</th>\n",
       "      <td>1694</td>\n",
       "      <td>1025</td>\n",
       "      <td>0.605077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF</th>\n",
       "      <td>3682</td>\n",
       "      <td>2988</td>\n",
       "      <td>0.811515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDK2</th>\n",
       "      <td>1505</td>\n",
       "      <td>819</td>\n",
       "      <td>0.544186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LCK</th>\n",
       "      <td>1537</td>\n",
       "      <td>930</td>\n",
       "      <td>0.605075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET</th>\n",
       "      <td>2821</td>\n",
       "      <td>2235</td>\n",
       "      <td>0.792272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p38a</th>\n",
       "      <td>3609</td>\n",
       "      <td>2760</td>\n",
       "      <td>0.764755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDR</th>\n",
       "      <td>7624</td>\n",
       "      <td>5312</td>\n",
       "      <td>0.696747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p110a</th>\n",
       "      <td>4347</td>\n",
       "      <td>2780</td>\n",
       "      <td>0.639522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total actives  fraction\n",
       "EGFR   5869    3574  0.608962\n",
       "ErbB2  1694    1025  0.605077\n",
       "BRAF   3682    2988  0.811515\n",
       "CDK2   1505     819  0.544186\n",
       "LCK    1537     930  0.605075\n",
       "MET    2821    2235  0.792272\n",
       "p38a   3609    2760  0.764755\n",
       "KDR    7624    5312  0.696747\n",
       "p110a  4347    2780  0.639522"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kinase_promiscuity_table = from_numbers_to_table(data)\n",
    "kinase_promiscuity_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's beautify the table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "zero-size array to reduction operation fmin which has no identity",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/IPython/core/formatters.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    343\u001b[0m             \u001b[0mmethod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_real_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprint_method\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    344\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmethod\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 345\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    346\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    347\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/pandas/io/formats/style.py\u001b[0m in \u001b[0;36m_repr_html_\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    203\u001b[0m         \u001b[0mHooks\u001b[0m \u001b[0minto\u001b[0m \u001b[0mJupyter\u001b[0m \u001b[0mnotebook\u001b[0m \u001b[0mrich\u001b[0m \u001b[0mdisplay\u001b[0m \u001b[0msystem\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    204\u001b[0m         \"\"\"\n\u001b[0;32m--> 205\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrender\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    206\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m     @doc(\n",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/pandas/io/formats/style.py\u001b[0m in \u001b[0;36mrender\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    617\u001b[0m         \u001b[0;34m*\u001b[0m \u001b[0mtable_attributes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    618\u001b[0m         \"\"\"\n\u001b[0;32m--> 619\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_compute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    620\u001b[0m         \u001b[0;31m# TODO: namespace all the pandas keys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    621\u001b[0m         \u001b[0md\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_translate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/pandas/io/formats/style.py\u001b[0m in \u001b[0;36m_compute\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    703\u001b[0m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    704\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_todo\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 705\u001b[0;31m             \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    706\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    707\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/pandas/io/formats/style.py\u001b[0m in \u001b[0;36m_apply\u001b[0;34m(self, func, axis, subset, **kwargs)\u001b[0m\n\u001b[1;32m    720\u001b[0m             \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    721\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 722\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    723\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    724\u001b[0m                 raise TypeError(\n",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/pandas/io/formats/style.py\u001b[0m in \u001b[0;36m_highlight_extrema\u001b[0;34m(data, color, max_)\u001b[0m\n\u001b[1;32m   1547\u001b[0m             \u001b[0mextrema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnanmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1548\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1549\u001b[0;31m             \u001b[0mextrema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnanmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1550\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1551\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Series from .apply\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<__array_function__ internals>\u001b[0m in \u001b[0;36mnanmin\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;32m~/.local/miniconda/envs/teachopencadd_review/lib/python3.8/site-packages/numpy/lib/nanfunctions.py\u001b[0m in \u001b[0;36mnanmin\u001b[0;34m(a, axis, out, keepdims)\u001b[0m\n\u001b[1;32m    317\u001b[0m         \u001b[0;31m# Fast, but not safe for subclasses of ndarray, or object arrays,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    318\u001b[0m         \u001b[0;31m# which do not implement isnan (gh-9009), or fmin correctly (gh-8975)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 319\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfmin\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    320\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misnan\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    321\u001b[0m             warnings.warn(\"All-NaN slice encountered\", RuntimeWarning,\n",
      "\u001b[0;31mValueError\u001b[0m: zero-size array to reduction operation fmin which has no identity"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fd9e5ed4be0>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kinase_promiscuity_table.style.\\\n",
    "    format(\"{:.3f}\", subset=[\"fraction\"]).\\\n",
    "    background_gradient(cmap='Purples', subset=[\"fraction\"]).\\\n",
    "    highlight_min(color=\"yellow\", axis=None).\\\n",
    "    highlight_max(color=\"red\", subset=[\"fraction\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the table, we notice that CDK2 and BRAF are the least and most promiscuous kinases, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Kinase similarity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now investigate how we can use the similarity measure discussed in the theory to compare kinases. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def similarity_ligand_profile(uniprot_id1, uniprot_id2, activity_dataframe):\n",
    "    \"\"\"\n",
    "    Compute the similarity between two kinases using ligand profiling.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    uniprot_id1 : str\n",
    "        Uniprot id of first kinase of interest.\n",
    "    uniprot_id2 : str\n",
    "        Uniprot id of second kinase of interest.\n",
    "\n",
    "    activity_dataframe :  pd.DataFrame\n",
    "        The dataframe with activity values for kinases.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    fraction : float\n",
    "        The metric for kinase similariy,\n",
    "        i.e. number of active compounds on both kinases\n",
    "        over number of measured compounds on both kinases.\n",
    "    \"\"\"\n",
    "    if uniprot_id1 == uniprot_id2:\n",
    "        (total_compounds,\n",
    "         active_compounds,\n",
    "         fraction) = kinase_to_activity_numbers(uniprot_id1,\n",
    "                                                activity_dataframe)\n",
    "        return fraction\n",
    "    else:\n",
    "        # Data for the two kinases only\n",
    "        reduced_data = activity_dataframe[activity_dataframe\n",
    "                                          [\"UniprotID\"].isin([uniprot_id1,\n",
    "                                                              uniprot_id2])]\n",
    "\n",
    "        # Look at active compounds only\n",
    "        active_entries = reduced_data[reduced_data[\"activity_binary\"] == 1]\n",
    "        # Group by compounds\n",
    "        compounds = active_entries.groupby(\"smiles\").size()\n",
    "        # Look at the number of active compounds measured on both kinases\n",
    "        active_compounds_on_both = compounds[compounds == 2].shape\n",
    "\n",
    "        # Look at all tested compounds\n",
    "        compounds = reduced_data.groupby(\"smiles\").size()\n",
    "        # Look at the number of compounds measured on both kinases\n",
    "        measured_compounds_on_both = compounds[compounds == 2].shape\n",
    "\n",
    "        if measured_compounds_on_both[0] > 0:\n",
    "            fraction = (active_compounds_on_both[0] /\n",
    "                        measured_compounds_on_both[0])\n",
    "        else:\n",
    "            print(f\"No compounds were measured on both kinases, \"\n",
    "                  f\"namely {uniprot_id1} and {uniprot_id2}.\")\n",
    "            fraction = np.nan\n",
    "        return fraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at the similarity between EGFR and MET:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ligand profile similarity between EGFR and MET: 0.23.\n"
     ]
    }
   ],
   "source": [
    "uniprot_id1=name_to_uniprot[\"EGFR\"]\n",
    "uniprot_id2=name_to_uniprot[\"MET\"]\n",
    "similarity_EGFR_MET = similarity_ligand_profile(uniprot_id1, uniprot_id2, data)\n",
    "print(f\"Ligand profile similarity between EGFR and MET: \"\n",
    "      f\"{similarity_EGFR_MET:.2f}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualize similarity as kinase matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No compounds were measured on both kinases, namely P04626 and P42336.\n",
      "No compounds were measured on both kinases, namely P42336 and P04626.\n"
     ]
    }
   ],
   "source": [
    "kinase_similarity_matrix = np.zeros((len(name_to_uniprot),\n",
    "                                     len(name_to_uniprot)))\n",
    "for i, kinase_name1 in enumerate(name_to_uniprot):\n",
    "    for j, kinase_name2 in enumerate(name_to_uniprot):\n",
    "        uniprot_id1=name_to_uniprot[kinase_name1]\n",
    "        uniprot_id2=name_to_uniprot[kinase_name2]\n",
    "        kinase_similarity_matrix[i, j] = similarity_ligand_profile(\n",
    "            uniprot_id1,\n",
    "            uniprot_id2,\n",
    "            data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EGFR</th>\n",
       "      <th>ErbB2</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>CDK2</th>\n",
       "      <th>LCK</th>\n",
       "      <th>MET</th>\n",
       "      <th>p38a</th>\n",
       "      <th>KDR</th>\n",
       "      <th>p110a</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.608962</td>\n",
       "      <td>0.556918</td>\n",
       "      <td>0.419355</td>\n",
       "      <td>0.116279</td>\n",
       "      <td>0.238462</td>\n",
       "      <td>0.225806</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.348214</td>\n",
       "      <td>0.071429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ErbB2</th>\n",
       "      <td>0.556918</td>\n",
       "      <td>0.605077</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.397790</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF</th>\n",
       "      <td>0.419355</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.811515</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.511628</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.644444</td>\n",
       "      <td>0.748120</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDK2</th>\n",
       "      <td>0.116279</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.544186</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.601695</td>\n",
       "      <td>0.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LCK</th>\n",
       "      <td>0.238462</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.511628</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.605075</td>\n",
       "      <td>0.287879</td>\n",
       "      <td>0.489362</td>\n",
       "      <td>0.432692</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET</th>\n",
       "      <td>0.225806</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.287879</td>\n",
       "      <td>0.792272</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.539359</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p38a</th>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.644444</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.489362</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.764755</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDR</th>\n",
       "      <td>0.348214</td>\n",
       "      <td>0.397790</td>\n",
       "      <td>0.748120</td>\n",
       "      <td>0.601695</td>\n",
       "      <td>0.432692</td>\n",
       "      <td>0.539359</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.696747</td>\n",
       "      <td>0.180791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p110a</th>\n",
       "      <td>0.071429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.180791</td>\n",
       "      <td>0.639522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           EGFR     ErbB2      BRAF      CDK2       LCK       MET      p38a  \\\n",
       "EGFR   0.608962  0.556918  0.419355  0.116279  0.238462  0.225806  0.315789   \n",
       "ErbB2  0.556918  0.605077  0.250000  0.153846  0.176471  0.035714  0.125000   \n",
       "BRAF   0.419355  0.250000  0.811515  0.066667  0.511628  0.111111  0.644444   \n",
       "CDK2   0.116279  0.153846  0.066667  0.544186  0.166667  0.086957  0.083333   \n",
       "LCK    0.238462  0.176471  0.511628  0.166667  0.605075  0.287879  0.489362   \n",
       "MET    0.225806  0.035714  0.111111  0.086957  0.287879  0.792272  0.047619   \n",
       "p38a   0.315789  0.125000  0.644444  0.083333  0.489362  0.047619  0.764755   \n",
       "KDR    0.348214  0.397790  0.748120  0.601695  0.432692  0.539359  0.500000   \n",
       "p110a  0.071429       NaN  0.166667  0.266667  0.000000  0.000000  0.000000   \n",
       "\n",
       "            KDR     p110a  \n",
       "EGFR   0.348214  0.071429  \n",
       "ErbB2  0.397790       NaN  \n",
       "BRAF   0.748120  0.166667  \n",
       "CDK2   0.601695  0.266667  \n",
       "LCK    0.432692  0.000000  \n",
       "MET    0.539359  0.000000  \n",
       "p38a   0.500000  0.000000  \n",
       "KDR    0.696747  0.180791  \n",
       "p110a  0.180791  0.639522  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kinase_similarity_matrix = pd.DataFrame(data=kinase_similarity_matrix,\n",
    "                                        index=name_to_uniprot.keys(),\n",
    "                                        columns=name_to_uniprot.keys())\n",
    "kinase_similarity_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "#T_9e788_row0_col0,#T_9e788_row1_col1,#T_9e788_row2_col2,#T_9e788_row2_col7,#T_9e788_row4_col4,#T_9e788_row5_col5,#T_9e788_row6_col6,#T_9e788_row7_col3,#T_9e788_row8_col8{\n",
       "            background-color:  #008000;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_9e788_row0_col1,#T_9e788_row7_col2{\n",
       "            background-color:  #138913;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_9e788_row0_col2,#T_9e788_row5_col4{\n",
       "            background-color:  #7bbc7b;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row0_col3{\n",
       "            background-color:  #d5e9d5;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row0_col4{\n",
       "            background-color:  #8fc68f;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row0_col5,#T_9e788_row7_col8{\n",
       "            background-color:  #a8d2a8;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row0_col6{\n",
       "            background-color:  #8ac48a;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row0_col7{\n",
       "            background-color:  #a6d1a6;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row0_col8,#T_9e788_row3_col5{\n",
       "            background-color:  #d1e6d1;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col0{\n",
       "            background-color:  #168b16;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col2,#T_9e788_row4_col1{\n",
       "            background-color:  #b1d7b1;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col3,#T_9e788_row1_col6{\n",
       "            background-color:  #c5e0c5;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col4{\n",
       "            background-color:  #a6d2a6;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col5{\n",
       "            background-color:  #e0eee0;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col7{\n",
       "            background-color:  #91c791;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row1_col8,#T_9e788_row8_col1{\n",
       "            background-color:  #000000;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_9e788_row2_col0{\n",
       "            background-color:  #53a953;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row2_col1{\n",
       "            background-color:  #92c892;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row2_col3,#T_9e788_row3_col2,#T_9e788_row4_col8,#T_9e788_row5_col1,#T_9e788_row5_col8,#T_9e788_row6_col8,#T_9e788_row8_col0,#T_9e788_row8_col4,#T_9e788_row8_col5,#T_9e788_row8_col6,#T_9e788_row8_col7{\n",
       "            background-color:  #ebf3eb;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row2_col4{\n",
       "            background-color:  #249224;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row2_col5{\n",
       "            background-color:  #cae3ca;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row2_col6{\n",
       "            background-color:  #259225;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row2_col8{\n",
       "            background-color:  #aed5ae;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col0{\n",
       "            background-color:  #d7e9d7;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col1{\n",
       "            background-color:  #badbba;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col3{\n",
       "            background-color:  #198c19;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col4{\n",
       "            background-color:  #aad3aa;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col6{\n",
       "            background-color:  #d2e7d2;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col7{\n",
       "            background-color:  #3d9e3d;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row3_col8{\n",
       "            background-color:  #89c389;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row4_col0{\n",
       "            background-color:  #a2cfa2;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row4_col2{\n",
       "            background-color:  #5fae5f;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row4_col3{\n",
       "            background-color:  #bfdebf;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row4_col5{\n",
       "            background-color:  #95c995;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row4_col6{\n",
       "            background-color:  #55a955;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row4_col7{\n",
       "            background-color:  #83c083;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row5_col0{\n",
       "            background-color:  #a7d2a7;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row5_col2,#T_9e788_row5_col6,#T_9e788_row6_col5{\n",
       "            background-color:  #ddecdd;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row5_col3{\n",
       "            background-color:  #e2efe2;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row5_col7,#T_9e788_row7_col1{\n",
       "            background-color:  #56aa56;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row6_col0{\n",
       "            background-color:  #80bf80;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row6_col1{\n",
       "            background-color:  #c6e1c6;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row6_col2{\n",
       "            background-color:  #349a34;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row6_col3{\n",
       "            background-color:  #e4f0e4;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row6_col4{\n",
       "            background-color:  #2c962c;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row6_col7{\n",
       "            background-color:  #66b266;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row7_col0{\n",
       "            background-color:  #72b872;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row7_col4{\n",
       "            background-color:  #42a042;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row7_col5{\n",
       "            background-color:  #4ba54a;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row7_col6{\n",
       "            background-color:  #51a851;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row7_col7{\n",
       "            background-color:  #158a15;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row8_col2{\n",
       "            background-color:  #cbe4cb;\n",
       "            color:  #000000;\n",
       "        }#T_9e788_row8_col3{\n",
       "            background-color:  #93c893;\n",
       "            color:  #000000;\n",
       "        }</style><table id=\"T_9e788_\" ><thead>    <tr>        <th class=\"blank level0\" ></th>        <th class=\"col_heading level0 col0\" >EGFR</th>        <th class=\"col_heading level0 col1\" >ErbB2</th>        <th class=\"col_heading level0 col2\" >BRAF</th>        <th class=\"col_heading level0 col3\" >CDK2</th>        <th class=\"col_heading level0 col4\" >LCK</th>        <th class=\"col_heading level0 col5\" >MET</th>        <th class=\"col_heading level0 col6\" >p38a</th>        <th class=\"col_heading level0 col7\" >KDR</th>        <th class=\"col_heading level0 col8\" >p110a</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_9e788_level0_row0\" class=\"row_heading level0 row0\" >EGFR</th>\n",
       "                        <td id=\"T_9e788_row0_col0\" class=\"data row0 col0\" >0.609</td>\n",
       "                        <td id=\"T_9e788_row0_col1\" class=\"data row0 col1\" >0.557</td>\n",
       "                        <td id=\"T_9e788_row0_col2\" class=\"data row0 col2\" >0.419</td>\n",
       "                        <td id=\"T_9e788_row0_col3\" class=\"data row0 col3\" >0.116</td>\n",
       "                        <td id=\"T_9e788_row0_col4\" class=\"data row0 col4\" >0.238</td>\n",
       "                        <td id=\"T_9e788_row0_col5\" class=\"data row0 col5\" >0.226</td>\n",
       "                        <td id=\"T_9e788_row0_col6\" class=\"data row0 col6\" >0.316</td>\n",
       "                        <td id=\"T_9e788_row0_col7\" class=\"data row0 col7\" >0.348</td>\n",
       "                        <td id=\"T_9e788_row0_col8\" class=\"data row0 col8\" >0.071</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row1\" class=\"row_heading level0 row1\" >ErbB2</th>\n",
       "                        <td id=\"T_9e788_row1_col0\" class=\"data row1 col0\" >0.557</td>\n",
       "                        <td id=\"T_9e788_row1_col1\" class=\"data row1 col1\" >0.605</td>\n",
       "                        <td id=\"T_9e788_row1_col2\" class=\"data row1 col2\" >0.250</td>\n",
       "                        <td id=\"T_9e788_row1_col3\" class=\"data row1 col3\" >0.154</td>\n",
       "                        <td id=\"T_9e788_row1_col4\" class=\"data row1 col4\" >0.176</td>\n",
       "                        <td id=\"T_9e788_row1_col5\" class=\"data row1 col5\" >0.036</td>\n",
       "                        <td id=\"T_9e788_row1_col6\" class=\"data row1 col6\" >0.125</td>\n",
       "                        <td id=\"T_9e788_row1_col7\" class=\"data row1 col7\" >0.398</td>\n",
       "                        <td id=\"T_9e788_row1_col8\" class=\"data row1 col8\" >nan</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row2\" class=\"row_heading level0 row2\" >BRAF</th>\n",
       "                        <td id=\"T_9e788_row2_col0\" class=\"data row2 col0\" >0.419</td>\n",
       "                        <td id=\"T_9e788_row2_col1\" class=\"data row2 col1\" >0.250</td>\n",
       "                        <td id=\"T_9e788_row2_col2\" class=\"data row2 col2\" >0.812</td>\n",
       "                        <td id=\"T_9e788_row2_col3\" class=\"data row2 col3\" >0.067</td>\n",
       "                        <td id=\"T_9e788_row2_col4\" class=\"data row2 col4\" >0.512</td>\n",
       "                        <td id=\"T_9e788_row2_col5\" class=\"data row2 col5\" >0.111</td>\n",
       "                        <td id=\"T_9e788_row2_col6\" class=\"data row2 col6\" >0.644</td>\n",
       "                        <td id=\"T_9e788_row2_col7\" class=\"data row2 col7\" >0.748</td>\n",
       "                        <td id=\"T_9e788_row2_col8\" class=\"data row2 col8\" >0.167</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row3\" class=\"row_heading level0 row3\" >CDK2</th>\n",
       "                        <td id=\"T_9e788_row3_col0\" class=\"data row3 col0\" >0.116</td>\n",
       "                        <td id=\"T_9e788_row3_col1\" class=\"data row3 col1\" >0.154</td>\n",
       "                        <td id=\"T_9e788_row3_col2\" class=\"data row3 col2\" >0.067</td>\n",
       "                        <td id=\"T_9e788_row3_col3\" class=\"data row3 col3\" >0.544</td>\n",
       "                        <td id=\"T_9e788_row3_col4\" class=\"data row3 col4\" >0.167</td>\n",
       "                        <td id=\"T_9e788_row3_col5\" class=\"data row3 col5\" >0.087</td>\n",
       "                        <td id=\"T_9e788_row3_col6\" class=\"data row3 col6\" >0.083</td>\n",
       "                        <td id=\"T_9e788_row3_col7\" class=\"data row3 col7\" >0.602</td>\n",
       "                        <td id=\"T_9e788_row3_col8\" class=\"data row3 col8\" >0.267</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row4\" class=\"row_heading level0 row4\" >LCK</th>\n",
       "                        <td id=\"T_9e788_row4_col0\" class=\"data row4 col0\" >0.238</td>\n",
       "                        <td id=\"T_9e788_row4_col1\" class=\"data row4 col1\" >0.176</td>\n",
       "                        <td id=\"T_9e788_row4_col2\" class=\"data row4 col2\" >0.512</td>\n",
       "                        <td id=\"T_9e788_row4_col3\" class=\"data row4 col3\" >0.167</td>\n",
       "                        <td id=\"T_9e788_row4_col4\" class=\"data row4 col4\" >0.605</td>\n",
       "                        <td id=\"T_9e788_row4_col5\" class=\"data row4 col5\" >0.288</td>\n",
       "                        <td id=\"T_9e788_row4_col6\" class=\"data row4 col6\" >0.489</td>\n",
       "                        <td id=\"T_9e788_row4_col7\" class=\"data row4 col7\" >0.433</td>\n",
       "                        <td id=\"T_9e788_row4_col8\" class=\"data row4 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row5\" class=\"row_heading level0 row5\" >MET</th>\n",
       "                        <td id=\"T_9e788_row5_col0\" class=\"data row5 col0\" >0.226</td>\n",
       "                        <td id=\"T_9e788_row5_col1\" class=\"data row5 col1\" >0.036</td>\n",
       "                        <td id=\"T_9e788_row5_col2\" class=\"data row5 col2\" >0.111</td>\n",
       "                        <td id=\"T_9e788_row5_col3\" class=\"data row5 col3\" >0.087</td>\n",
       "                        <td id=\"T_9e788_row5_col4\" class=\"data row5 col4\" >0.288</td>\n",
       "                        <td id=\"T_9e788_row5_col5\" class=\"data row5 col5\" >0.792</td>\n",
       "                        <td id=\"T_9e788_row5_col6\" class=\"data row5 col6\" >0.048</td>\n",
       "                        <td id=\"T_9e788_row5_col7\" class=\"data row5 col7\" >0.539</td>\n",
       "                        <td id=\"T_9e788_row5_col8\" class=\"data row5 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row6\" class=\"row_heading level0 row6\" >p38a</th>\n",
       "                        <td id=\"T_9e788_row6_col0\" class=\"data row6 col0\" >0.316</td>\n",
       "                        <td id=\"T_9e788_row6_col1\" class=\"data row6 col1\" >0.125</td>\n",
       "                        <td id=\"T_9e788_row6_col2\" class=\"data row6 col2\" >0.644</td>\n",
       "                        <td id=\"T_9e788_row6_col3\" class=\"data row6 col3\" >0.083</td>\n",
       "                        <td id=\"T_9e788_row6_col4\" class=\"data row6 col4\" >0.489</td>\n",
       "                        <td id=\"T_9e788_row6_col5\" class=\"data row6 col5\" >0.048</td>\n",
       "                        <td id=\"T_9e788_row6_col6\" class=\"data row6 col6\" >0.765</td>\n",
       "                        <td id=\"T_9e788_row6_col7\" class=\"data row6 col7\" >0.500</td>\n",
       "                        <td id=\"T_9e788_row6_col8\" class=\"data row6 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row7\" class=\"row_heading level0 row7\" >KDR</th>\n",
       "                        <td id=\"T_9e788_row7_col0\" class=\"data row7 col0\" >0.348</td>\n",
       "                        <td id=\"T_9e788_row7_col1\" class=\"data row7 col1\" >0.398</td>\n",
       "                        <td id=\"T_9e788_row7_col2\" class=\"data row7 col2\" >0.748</td>\n",
       "                        <td id=\"T_9e788_row7_col3\" class=\"data row7 col3\" >0.602</td>\n",
       "                        <td id=\"T_9e788_row7_col4\" class=\"data row7 col4\" >0.433</td>\n",
       "                        <td id=\"T_9e788_row7_col5\" class=\"data row7 col5\" >0.539</td>\n",
       "                        <td id=\"T_9e788_row7_col6\" class=\"data row7 col6\" >0.500</td>\n",
       "                        <td id=\"T_9e788_row7_col7\" class=\"data row7 col7\" >0.697</td>\n",
       "                        <td id=\"T_9e788_row7_col8\" class=\"data row7 col8\" >0.181</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_9e788_level0_row8\" class=\"row_heading level0 row8\" >p110a</th>\n",
       "                        <td id=\"T_9e788_row8_col0\" class=\"data row8 col0\" >0.071</td>\n",
       "                        <td id=\"T_9e788_row8_col1\" class=\"data row8 col1\" >nan</td>\n",
       "                        <td id=\"T_9e788_row8_col2\" class=\"data row8 col2\" >0.167</td>\n",
       "                        <td id=\"T_9e788_row8_col3\" class=\"data row8 col3\" >0.267</td>\n",
       "                        <td id=\"T_9e788_row8_col4\" class=\"data row8 col4\" >0.000</td>\n",
       "                        <td id=\"T_9e788_row8_col5\" class=\"data row8 col5\" >0.000</td>\n",
       "                        <td id=\"T_9e788_row8_col6\" class=\"data row8 col6\" >0.000</td>\n",
       "                        <td id=\"T_9e788_row8_col7\" class=\"data row8 col7\" >0.181</td>\n",
       "                        <td id=\"T_9e788_row8_col8\" class=\"data row8 col8\" >0.640</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fd9e5b8bbb0>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show matrix with background gradient\n",
    "cm = sns.light_palette(\"green\", as_cmap=True)\n",
    "kinase_similarity_matrix.style.\\\n",
    "    background_gradient(cmap=cm).\\\n",
    "    format(\"{:.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the diagonal contains the previously discussed promiscuity values.\n",
    "\n",
    "As mentioned above, no compounds were measured on both ErbB2 and p110a and therefore create a `np.nan` entry which can be problematic for algorithmic reason.\n",
    "\n",
    "As a simple workaround, we will fill the NA values with zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "#T_07da8_row0_col0,#T_07da8_row1_col1,#T_07da8_row2_col2,#T_07da8_row2_col7,#T_07da8_row4_col4,#T_07da8_row5_col5,#T_07da8_row6_col6,#T_07da8_row7_col3,#T_07da8_row8_col8{\n",
       "            background-color:  #008000;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_07da8_row0_col1{\n",
       "            background-color:  #128912;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_07da8_row0_col2,#T_07da8_row5_col4{\n",
       "            background-color:  #7bbc7b;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row0_col3{\n",
       "            background-color:  #d5e9d5;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row0_col4{\n",
       "            background-color:  #8fc68f;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row0_col5,#T_07da8_row7_col8{\n",
       "            background-color:  #a8d2a8;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row0_col6,#T_07da8_row2_col1{\n",
       "            background-color:  #8ac48a;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row0_col7{\n",
       "            background-color:  #a6d1a6;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row0_col8,#T_07da8_row3_col5{\n",
       "            background-color:  #d1e6d1;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col0{\n",
       "            background-color:  #168b16;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col2{\n",
       "            background-color:  #b1d7b1;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col3,#T_07da8_row1_col6{\n",
       "            background-color:  #c5e0c5;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col4,#T_07da8_row4_col1{\n",
       "            background-color:  #a6d2a6;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col5{\n",
       "            background-color:  #e0eee0;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col7{\n",
       "            background-color:  #91c791;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row1_col8,#T_07da8_row2_col3,#T_07da8_row3_col2,#T_07da8_row4_col8,#T_07da8_row5_col8,#T_07da8_row6_col8,#T_07da8_row8_col0,#T_07da8_row8_col1,#T_07da8_row8_col4,#T_07da8_row8_col5,#T_07da8_row8_col6,#T_07da8_row8_col7{\n",
       "            background-color:  #ebf3eb;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row2_col0{\n",
       "            background-color:  #53a953;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row2_col4{\n",
       "            background-color:  #249224;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row2_col5{\n",
       "            background-color:  #cae3ca;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row2_col6{\n",
       "            background-color:  #259225;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row2_col8{\n",
       "            background-color:  #aed5ae;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col0{\n",
       "            background-color:  #d7e9d7;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col1{\n",
       "            background-color:  #afd6af;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col3{\n",
       "            background-color:  #198c19;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col4{\n",
       "            background-color:  #aad3aa;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col6{\n",
       "            background-color:  #d2e7d2;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col7{\n",
       "            background-color:  #3d9e3d;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row3_col8{\n",
       "            background-color:  #89c389;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row4_col0{\n",
       "            background-color:  #a2cfa2;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row4_col2{\n",
       "            background-color:  #5fae5f;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row4_col3{\n",
       "            background-color:  #bfdebf;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row4_col5{\n",
       "            background-color:  #95c995;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row4_col6{\n",
       "            background-color:  #55a955;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row4_col7{\n",
       "            background-color:  #83c083;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row5_col0{\n",
       "            background-color:  #a7d2a7;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row5_col1,#T_07da8_row5_col2,#T_07da8_row5_col6,#T_07da8_row6_col5{\n",
       "            background-color:  #ddecdd;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row5_col3{\n",
       "            background-color:  #e2efe2;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row5_col7{\n",
       "            background-color:  #56aa56;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row6_col0{\n",
       "            background-color:  #80bf80;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row6_col1{\n",
       "            background-color:  #bbdcbb;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row6_col2{\n",
       "            background-color:  #349a34;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row6_col3{\n",
       "            background-color:  #e4f0e4;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row6_col4{\n",
       "            background-color:  #2c962c;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row6_col7{\n",
       "            background-color:  #66b266;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row7_col0{\n",
       "            background-color:  #72b872;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row7_col1{\n",
       "            background-color:  #50a750;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row7_col2{\n",
       "            background-color:  #138913;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_07da8_row7_col4{\n",
       "            background-color:  #42a042;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row7_col5{\n",
       "            background-color:  #4ba54a;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row7_col6{\n",
       "            background-color:  #51a851;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row7_col7{\n",
       "            background-color:  #158a15;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row8_col2{\n",
       "            background-color:  #cbe4cb;\n",
       "            color:  #000000;\n",
       "        }#T_07da8_row8_col3{\n",
       "            background-color:  #93c893;\n",
       "            color:  #000000;\n",
       "        }</style><table id=\"T_07da8_\" ><thead>    <tr>        <th class=\"blank level0\" ></th>        <th class=\"col_heading level0 col0\" >EGFR</th>        <th class=\"col_heading level0 col1\" >ErbB2</th>        <th class=\"col_heading level0 col2\" >BRAF</th>        <th class=\"col_heading level0 col3\" >CDK2</th>        <th class=\"col_heading level0 col4\" >LCK</th>        <th class=\"col_heading level0 col5\" >MET</th>        <th class=\"col_heading level0 col6\" >p38a</th>        <th class=\"col_heading level0 col7\" >KDR</th>        <th class=\"col_heading level0 col8\" >p110a</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_07da8_level0_row0\" class=\"row_heading level0 row0\" >EGFR</th>\n",
       "                        <td id=\"T_07da8_row0_col0\" class=\"data row0 col0\" >0.609</td>\n",
       "                        <td id=\"T_07da8_row0_col1\" class=\"data row0 col1\" >0.557</td>\n",
       "                        <td id=\"T_07da8_row0_col2\" class=\"data row0 col2\" >0.419</td>\n",
       "                        <td id=\"T_07da8_row0_col3\" class=\"data row0 col3\" >0.116</td>\n",
       "                        <td id=\"T_07da8_row0_col4\" class=\"data row0 col4\" >0.238</td>\n",
       "                        <td id=\"T_07da8_row0_col5\" class=\"data row0 col5\" >0.226</td>\n",
       "                        <td id=\"T_07da8_row0_col6\" class=\"data row0 col6\" >0.316</td>\n",
       "                        <td id=\"T_07da8_row0_col7\" class=\"data row0 col7\" >0.348</td>\n",
       "                        <td id=\"T_07da8_row0_col8\" class=\"data row0 col8\" >0.071</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row1\" class=\"row_heading level0 row1\" >ErbB2</th>\n",
       "                        <td id=\"T_07da8_row1_col0\" class=\"data row1 col0\" >0.557</td>\n",
       "                        <td id=\"T_07da8_row1_col1\" class=\"data row1 col1\" >0.605</td>\n",
       "                        <td id=\"T_07da8_row1_col2\" class=\"data row1 col2\" >0.250</td>\n",
       "                        <td id=\"T_07da8_row1_col3\" class=\"data row1 col3\" >0.154</td>\n",
       "                        <td id=\"T_07da8_row1_col4\" class=\"data row1 col4\" >0.176</td>\n",
       "                        <td id=\"T_07da8_row1_col5\" class=\"data row1 col5\" >0.036</td>\n",
       "                        <td id=\"T_07da8_row1_col6\" class=\"data row1 col6\" >0.125</td>\n",
       "                        <td id=\"T_07da8_row1_col7\" class=\"data row1 col7\" >0.398</td>\n",
       "                        <td id=\"T_07da8_row1_col8\" class=\"data row1 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row2\" class=\"row_heading level0 row2\" >BRAF</th>\n",
       "                        <td id=\"T_07da8_row2_col0\" class=\"data row2 col0\" >0.419</td>\n",
       "                        <td id=\"T_07da8_row2_col1\" class=\"data row2 col1\" >0.250</td>\n",
       "                        <td id=\"T_07da8_row2_col2\" class=\"data row2 col2\" >0.812</td>\n",
       "                        <td id=\"T_07da8_row2_col3\" class=\"data row2 col3\" >0.067</td>\n",
       "                        <td id=\"T_07da8_row2_col4\" class=\"data row2 col4\" >0.512</td>\n",
       "                        <td id=\"T_07da8_row2_col5\" class=\"data row2 col5\" >0.111</td>\n",
       "                        <td id=\"T_07da8_row2_col6\" class=\"data row2 col6\" >0.644</td>\n",
       "                        <td id=\"T_07da8_row2_col7\" class=\"data row2 col7\" >0.748</td>\n",
       "                        <td id=\"T_07da8_row2_col8\" class=\"data row2 col8\" >0.167</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row3\" class=\"row_heading level0 row3\" >CDK2</th>\n",
       "                        <td id=\"T_07da8_row3_col0\" class=\"data row3 col0\" >0.116</td>\n",
       "                        <td id=\"T_07da8_row3_col1\" class=\"data row3 col1\" >0.154</td>\n",
       "                        <td id=\"T_07da8_row3_col2\" class=\"data row3 col2\" >0.067</td>\n",
       "                        <td id=\"T_07da8_row3_col3\" class=\"data row3 col3\" >0.544</td>\n",
       "                        <td id=\"T_07da8_row3_col4\" class=\"data row3 col4\" >0.167</td>\n",
       "                        <td id=\"T_07da8_row3_col5\" class=\"data row3 col5\" >0.087</td>\n",
       "                        <td id=\"T_07da8_row3_col6\" class=\"data row3 col6\" >0.083</td>\n",
       "                        <td id=\"T_07da8_row3_col7\" class=\"data row3 col7\" >0.602</td>\n",
       "                        <td id=\"T_07da8_row3_col8\" class=\"data row3 col8\" >0.267</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row4\" class=\"row_heading level0 row4\" >LCK</th>\n",
       "                        <td id=\"T_07da8_row4_col0\" class=\"data row4 col0\" >0.238</td>\n",
       "                        <td id=\"T_07da8_row4_col1\" class=\"data row4 col1\" >0.176</td>\n",
       "                        <td id=\"T_07da8_row4_col2\" class=\"data row4 col2\" >0.512</td>\n",
       "                        <td id=\"T_07da8_row4_col3\" class=\"data row4 col3\" >0.167</td>\n",
       "                        <td id=\"T_07da8_row4_col4\" class=\"data row4 col4\" >0.605</td>\n",
       "                        <td id=\"T_07da8_row4_col5\" class=\"data row4 col5\" >0.288</td>\n",
       "                        <td id=\"T_07da8_row4_col6\" class=\"data row4 col6\" >0.489</td>\n",
       "                        <td id=\"T_07da8_row4_col7\" class=\"data row4 col7\" >0.433</td>\n",
       "                        <td id=\"T_07da8_row4_col8\" class=\"data row4 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row5\" class=\"row_heading level0 row5\" >MET</th>\n",
       "                        <td id=\"T_07da8_row5_col0\" class=\"data row5 col0\" >0.226</td>\n",
       "                        <td id=\"T_07da8_row5_col1\" class=\"data row5 col1\" >0.036</td>\n",
       "                        <td id=\"T_07da8_row5_col2\" class=\"data row5 col2\" >0.111</td>\n",
       "                        <td id=\"T_07da8_row5_col3\" class=\"data row5 col3\" >0.087</td>\n",
       "                        <td id=\"T_07da8_row5_col4\" class=\"data row5 col4\" >0.288</td>\n",
       "                        <td id=\"T_07da8_row5_col5\" class=\"data row5 col5\" >0.792</td>\n",
       "                        <td id=\"T_07da8_row5_col6\" class=\"data row5 col6\" >0.048</td>\n",
       "                        <td id=\"T_07da8_row5_col7\" class=\"data row5 col7\" >0.539</td>\n",
       "                        <td id=\"T_07da8_row5_col8\" class=\"data row5 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row6\" class=\"row_heading level0 row6\" >p38a</th>\n",
       "                        <td id=\"T_07da8_row6_col0\" class=\"data row6 col0\" >0.316</td>\n",
       "                        <td id=\"T_07da8_row6_col1\" class=\"data row6 col1\" >0.125</td>\n",
       "                        <td id=\"T_07da8_row6_col2\" class=\"data row6 col2\" >0.644</td>\n",
       "                        <td id=\"T_07da8_row6_col3\" class=\"data row6 col3\" >0.083</td>\n",
       "                        <td id=\"T_07da8_row6_col4\" class=\"data row6 col4\" >0.489</td>\n",
       "                        <td id=\"T_07da8_row6_col5\" class=\"data row6 col5\" >0.048</td>\n",
       "                        <td id=\"T_07da8_row6_col6\" class=\"data row6 col6\" >0.765</td>\n",
       "                        <td id=\"T_07da8_row6_col7\" class=\"data row6 col7\" >0.500</td>\n",
       "                        <td id=\"T_07da8_row6_col8\" class=\"data row6 col8\" >0.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row7\" class=\"row_heading level0 row7\" >KDR</th>\n",
       "                        <td id=\"T_07da8_row7_col0\" class=\"data row7 col0\" >0.348</td>\n",
       "                        <td id=\"T_07da8_row7_col1\" class=\"data row7 col1\" >0.398</td>\n",
       "                        <td id=\"T_07da8_row7_col2\" class=\"data row7 col2\" >0.748</td>\n",
       "                        <td id=\"T_07da8_row7_col3\" class=\"data row7 col3\" >0.602</td>\n",
       "                        <td id=\"T_07da8_row7_col4\" class=\"data row7 col4\" >0.433</td>\n",
       "                        <td id=\"T_07da8_row7_col5\" class=\"data row7 col5\" >0.539</td>\n",
       "                        <td id=\"T_07da8_row7_col6\" class=\"data row7 col6\" >0.500</td>\n",
       "                        <td id=\"T_07da8_row7_col7\" class=\"data row7 col7\" >0.697</td>\n",
       "                        <td id=\"T_07da8_row7_col8\" class=\"data row7 col8\" >0.181</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_07da8_level0_row8\" class=\"row_heading level0 row8\" >p110a</th>\n",
       "                        <td id=\"T_07da8_row8_col0\" class=\"data row8 col0\" >0.071</td>\n",
       "                        <td id=\"T_07da8_row8_col1\" class=\"data row8 col1\" >0.000</td>\n",
       "                        <td id=\"T_07da8_row8_col2\" class=\"data row8 col2\" >0.167</td>\n",
       "                        <td id=\"T_07da8_row8_col3\" class=\"data row8 col3\" >0.267</td>\n",
       "                        <td id=\"T_07da8_row8_col4\" class=\"data row8 col4\" >0.000</td>\n",
       "                        <td id=\"T_07da8_row8_col5\" class=\"data row8 col5\" >0.000</td>\n",
       "                        <td id=\"T_07da8_row8_col6\" class=\"data row8 col6\" >0.000</td>\n",
       "                        <td id=\"T_07da8_row8_col7\" class=\"data row8 col7\" >0.181</td>\n",
       "                        <td id=\"T_07da8_row8_col8\" class=\"data row8 col8\" >0.640</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fd9e5b8bd90>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kinase_similarity_matrix = kinase_similarity_matrix.fillna(0)\n",
    "\n",
    "kinase_similarity_matrix.style.\\\n",
    "    background_gradient(cmap=cm).\\\n",
    "    format(\"{:.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save kinase similarity matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "kinase_similarity_matrix.to_csv(DATA /\n",
    "                                \"kinase_similarity_matrix_ligand_profile.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Kinase distance matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to apply some clustering algorithm to assess the similarity between kinases, it is necessary to start with a distance matrix. In the case of the similarity matrix above, this matrix is not a distance matrix. For example, the diagonal elements are not zero.\n",
    "\n",
    "Since all entries are between $0$ and $1$, the similarity matrix $SM$ can be converted to a distance matrix $DM$ using $$ DM = 1-SM.$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The values of the similarity matrix lie between: 0.00 and 0.81\n"
     ]
    }
   ],
   "source": [
    "print(f\"The values of the similarity matrix lie between: \"\n",
    "      f\"{kinase_similarity_matrix.min().min():.2f}\"\n",
    "      f\" and {kinase_similarity_matrix.max().max():.2f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "kinase_distance_matrix = 1 - kinase_similarity_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we set the diagonal values to $0$ and we obtain the kinase distance matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.fill_diagonal(kinase_distance_matrix.values, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "#T_1426b_row0_col0,#T_1426b_row1_col1,#T_1426b_row2_col2,#T_1426b_row3_col3,#T_1426b_row4_col4,#T_1426b_row5_col5,#T_1426b_row6_col6,#T_1426b_row7_col7,#T_1426b_row8_col8{\n",
       "            background-color:  #ebf3eb;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col1{\n",
       "            background-color:  #83c083;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col2{\n",
       "            background-color:  #58ab58;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col3{\n",
       "            background-color:  #0c860c;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row0_col4{\n",
       "            background-color:  #389b38;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col5{\n",
       "            background-color:  #349a34;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col6{\n",
       "            background-color:  #4aa44a;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col7{\n",
       "            background-color:  #309730;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row0_col8{\n",
       "            background-color:  #118811;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row1_col0{\n",
       "            background-color:  #7abc7a;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row1_col2{\n",
       "            background-color:  #2e972e;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row1_col3{\n",
       "            background-color:  #158a15;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row1_col4,#T_1426b_row4_col1{\n",
       "            background-color:  #299429;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row1_col5,#T_1426b_row5_col1{\n",
       "            background-color:  #088408;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row1_col6,#T_1426b_row6_col1{\n",
       "            background-color:  #1d8e1d;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row1_col7,#T_1426b_row6_col0{\n",
       "            background-color:  #3e9e3e;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row1_col8,#T_1426b_row2_col3,#T_1426b_row3_col2,#T_1426b_row4_col8,#T_1426b_row5_col8,#T_1426b_row6_col8,#T_1426b_row8_col0,#T_1426b_row8_col1,#T_1426b_row8_col4,#T_1426b_row8_col5,#T_1426b_row8_col6,#T_1426b_row8_col7{\n",
       "            background-color:  #008000;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row2_col0{\n",
       "            background-color:  #57ab57;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row2_col1{\n",
       "            background-color:  #3a9c3a;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row2_col4,#T_1426b_row3_col7{\n",
       "            background-color:  #78bb78;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row2_col5{\n",
       "            background-color:  #1a8d1a;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row2_col6{\n",
       "            background-color:  #97ca97;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row2_col7{\n",
       "            background-color:  #a3d0a3;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row2_col8,#T_1426b_row3_col4,#T_1426b_row5_col0{\n",
       "            background-color:  #279327;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row3_col0,#T_1426b_row5_col2,#T_1426b_row5_col6,#T_1426b_row6_col5{\n",
       "            background-color:  #0b850b;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row3_col1{\n",
       "            background-color:  #249224;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row3_col5{\n",
       "            background-color:  #148a14;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row3_col6{\n",
       "            background-color:  #138913;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row3_col8{\n",
       "            background-color:  #3f9f3f;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row4_col0,#T_1426b_row7_col8{\n",
       "            background-color:  #2a952a;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row4_col2{\n",
       "            background-color:  #70b770;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row4_col3,#T_1426b_row8_col2{\n",
       "            background-color:  #198c19;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row4_col5,#T_1426b_row5_col4{\n",
       "            background-color:  #43a143;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row4_col6,#T_1426b_row6_col4{\n",
       "            background-color:  #73b873;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row4_col7{\n",
       "            background-color:  #48a348;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row5_col3{\n",
       "            background-color:  #058205;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row5_col7{\n",
       "            background-color:  #67b267;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row6_col2{\n",
       "            background-color:  #91c791;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row6_col3{\n",
       "            background-color:  #048204;\n",
       "            color:  #f1f1f1;\n",
       "        }#T_1426b_row6_col7{\n",
       "            background-color:  #5bad5b;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col0{\n",
       "            background-color:  #46a246;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col1{\n",
       "            background-color:  #5dae5d;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col2{\n",
       "            background-color:  #abd4ab;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col3{\n",
       "            background-color:  #86c286;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col4{\n",
       "            background-color:  #65b265;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col5{\n",
       "            background-color:  #7fbe7f;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row7_col6{\n",
       "            background-color:  #75b975;\n",
       "            color:  #000000;\n",
       "        }#T_1426b_row8_col3{\n",
       "            background-color:  #329832;\n",
       "            color:  #000000;\n",
       "        }</style><table id=\"T_1426b_\" ><thead>    <tr>        <th class=\"blank level0\" ></th>        <th class=\"col_heading level0 col0\" >EGFR</th>        <th class=\"col_heading level0 col1\" >ErbB2</th>        <th class=\"col_heading level0 col2\" >BRAF</th>        <th class=\"col_heading level0 col3\" >CDK2</th>        <th class=\"col_heading level0 col4\" >LCK</th>        <th class=\"col_heading level0 col5\" >MET</th>        <th class=\"col_heading level0 col6\" >p38a</th>        <th class=\"col_heading level0 col7\" >KDR</th>        <th class=\"col_heading level0 col8\" >p110a</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_1426b_level0_row0\" class=\"row_heading level0 row0\" >EGFR</th>\n",
       "                        <td id=\"T_1426b_row0_col0\" class=\"data row0 col0\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row0_col1\" class=\"data row0 col1\" >0.443</td>\n",
       "                        <td id=\"T_1426b_row0_col2\" class=\"data row0 col2\" >0.581</td>\n",
       "                        <td id=\"T_1426b_row0_col3\" class=\"data row0 col3\" >0.884</td>\n",
       "                        <td id=\"T_1426b_row0_col4\" class=\"data row0 col4\" >0.762</td>\n",
       "                        <td id=\"T_1426b_row0_col5\" class=\"data row0 col5\" >0.774</td>\n",
       "                        <td id=\"T_1426b_row0_col6\" class=\"data row0 col6\" >0.684</td>\n",
       "                        <td id=\"T_1426b_row0_col7\" class=\"data row0 col7\" >0.652</td>\n",
       "                        <td id=\"T_1426b_row0_col8\" class=\"data row0 col8\" >0.929</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row1\" class=\"row_heading level0 row1\" >ErbB2</th>\n",
       "                        <td id=\"T_1426b_row1_col0\" class=\"data row1 col0\" >0.443</td>\n",
       "                        <td id=\"T_1426b_row1_col1\" class=\"data row1 col1\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row1_col2\" class=\"data row1 col2\" >0.750</td>\n",
       "                        <td id=\"T_1426b_row1_col3\" class=\"data row1 col3\" >0.846</td>\n",
       "                        <td id=\"T_1426b_row1_col4\" class=\"data row1 col4\" >0.824</td>\n",
       "                        <td id=\"T_1426b_row1_col5\" class=\"data row1 col5\" >0.964</td>\n",
       "                        <td id=\"T_1426b_row1_col6\" class=\"data row1 col6\" >0.875</td>\n",
       "                        <td id=\"T_1426b_row1_col7\" class=\"data row1 col7\" >0.602</td>\n",
       "                        <td id=\"T_1426b_row1_col8\" class=\"data row1 col8\" >1.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row2\" class=\"row_heading level0 row2\" >BRAF</th>\n",
       "                        <td id=\"T_1426b_row2_col0\" class=\"data row2 col0\" >0.581</td>\n",
       "                        <td id=\"T_1426b_row2_col1\" class=\"data row2 col1\" >0.750</td>\n",
       "                        <td id=\"T_1426b_row2_col2\" class=\"data row2 col2\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row2_col3\" class=\"data row2 col3\" >0.933</td>\n",
       "                        <td id=\"T_1426b_row2_col4\" class=\"data row2 col4\" >0.488</td>\n",
       "                        <td id=\"T_1426b_row2_col5\" class=\"data row2 col5\" >0.889</td>\n",
       "                        <td id=\"T_1426b_row2_col6\" class=\"data row2 col6\" >0.356</td>\n",
       "                        <td id=\"T_1426b_row2_col7\" class=\"data row2 col7\" >0.252</td>\n",
       "                        <td id=\"T_1426b_row2_col8\" class=\"data row2 col8\" >0.833</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row3\" class=\"row_heading level0 row3\" >CDK2</th>\n",
       "                        <td id=\"T_1426b_row3_col0\" class=\"data row3 col0\" >0.884</td>\n",
       "                        <td id=\"T_1426b_row3_col1\" class=\"data row3 col1\" >0.846</td>\n",
       "                        <td id=\"T_1426b_row3_col2\" class=\"data row3 col2\" >0.933</td>\n",
       "                        <td id=\"T_1426b_row3_col3\" class=\"data row3 col3\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row3_col4\" class=\"data row3 col4\" >0.833</td>\n",
       "                        <td id=\"T_1426b_row3_col5\" class=\"data row3 col5\" >0.913</td>\n",
       "                        <td id=\"T_1426b_row3_col6\" class=\"data row3 col6\" >0.917</td>\n",
       "                        <td id=\"T_1426b_row3_col7\" class=\"data row3 col7\" >0.398</td>\n",
       "                        <td id=\"T_1426b_row3_col8\" class=\"data row3 col8\" >0.733</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row4\" class=\"row_heading level0 row4\" >LCK</th>\n",
       "                        <td id=\"T_1426b_row4_col0\" class=\"data row4 col0\" >0.762</td>\n",
       "                        <td id=\"T_1426b_row4_col1\" class=\"data row4 col1\" >0.824</td>\n",
       "                        <td id=\"T_1426b_row4_col2\" class=\"data row4 col2\" >0.488</td>\n",
       "                        <td id=\"T_1426b_row4_col3\" class=\"data row4 col3\" >0.833</td>\n",
       "                        <td id=\"T_1426b_row4_col4\" class=\"data row4 col4\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row4_col5\" class=\"data row4 col5\" >0.712</td>\n",
       "                        <td id=\"T_1426b_row4_col6\" class=\"data row4 col6\" >0.511</td>\n",
       "                        <td id=\"T_1426b_row4_col7\" class=\"data row4 col7\" >0.567</td>\n",
       "                        <td id=\"T_1426b_row4_col8\" class=\"data row4 col8\" >1.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row5\" class=\"row_heading level0 row5\" >MET</th>\n",
       "                        <td id=\"T_1426b_row5_col0\" class=\"data row5 col0\" >0.774</td>\n",
       "                        <td id=\"T_1426b_row5_col1\" class=\"data row5 col1\" >0.964</td>\n",
       "                        <td id=\"T_1426b_row5_col2\" class=\"data row5 col2\" >0.889</td>\n",
       "                        <td id=\"T_1426b_row5_col3\" class=\"data row5 col3\" >0.913</td>\n",
       "                        <td id=\"T_1426b_row5_col4\" class=\"data row5 col4\" >0.712</td>\n",
       "                        <td id=\"T_1426b_row5_col5\" class=\"data row5 col5\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row5_col6\" class=\"data row5 col6\" >0.952</td>\n",
       "                        <td id=\"T_1426b_row5_col7\" class=\"data row5 col7\" >0.461</td>\n",
       "                        <td id=\"T_1426b_row5_col8\" class=\"data row5 col8\" >1.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row6\" class=\"row_heading level0 row6\" >p38a</th>\n",
       "                        <td id=\"T_1426b_row6_col0\" class=\"data row6 col0\" >0.684</td>\n",
       "                        <td id=\"T_1426b_row6_col1\" class=\"data row6 col1\" >0.875</td>\n",
       "                        <td id=\"T_1426b_row6_col2\" class=\"data row6 col2\" >0.356</td>\n",
       "                        <td id=\"T_1426b_row6_col3\" class=\"data row6 col3\" >0.917</td>\n",
       "                        <td id=\"T_1426b_row6_col4\" class=\"data row6 col4\" >0.511</td>\n",
       "                        <td id=\"T_1426b_row6_col5\" class=\"data row6 col5\" >0.952</td>\n",
       "                        <td id=\"T_1426b_row6_col6\" class=\"data row6 col6\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row6_col7\" class=\"data row6 col7\" >0.500</td>\n",
       "                        <td id=\"T_1426b_row6_col8\" class=\"data row6 col8\" >1.000</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row7\" class=\"row_heading level0 row7\" >KDR</th>\n",
       "                        <td id=\"T_1426b_row7_col0\" class=\"data row7 col0\" >0.652</td>\n",
       "                        <td id=\"T_1426b_row7_col1\" class=\"data row7 col1\" >0.602</td>\n",
       "                        <td id=\"T_1426b_row7_col2\" class=\"data row7 col2\" >0.252</td>\n",
       "                        <td id=\"T_1426b_row7_col3\" class=\"data row7 col3\" >0.398</td>\n",
       "                        <td id=\"T_1426b_row7_col4\" class=\"data row7 col4\" >0.567</td>\n",
       "                        <td id=\"T_1426b_row7_col5\" class=\"data row7 col5\" >0.461</td>\n",
       "                        <td id=\"T_1426b_row7_col6\" class=\"data row7 col6\" >0.500</td>\n",
       "                        <td id=\"T_1426b_row7_col7\" class=\"data row7 col7\" >0.000</td>\n",
       "                        <td id=\"T_1426b_row7_col8\" class=\"data row7 col8\" >0.819</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_1426b_level0_row8\" class=\"row_heading level0 row8\" >p110a</th>\n",
       "                        <td id=\"T_1426b_row8_col0\" class=\"data row8 col0\" >0.929</td>\n",
       "                        <td id=\"T_1426b_row8_col1\" class=\"data row8 col1\" >1.000</td>\n",
       "                        <td id=\"T_1426b_row8_col2\" class=\"data row8 col2\" >0.833</td>\n",
       "                        <td id=\"T_1426b_row8_col3\" class=\"data row8 col3\" >0.733</td>\n",
       "                        <td id=\"T_1426b_row8_col4\" class=\"data row8 col4\" >1.000</td>\n",
       "                        <td id=\"T_1426b_row8_col5\" class=\"data row8 col5\" >1.000</td>\n",
       "                        <td id=\"T_1426b_row8_col6\" class=\"data row8 col6\" >1.000</td>\n",
       "                        <td id=\"T_1426b_row8_col7\" class=\"data row8 col7\" >0.819</td>\n",
       "                        <td id=\"T_1426b_row8_col8\" class=\"data row8 col8\" >0.000</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fd9e9fd2700>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kinase_distance_matrix.style.\\\n",
    "    background_gradient(cmap=cm).\\\n",
    "    format(\"{:.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save kinase distance matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "kinase_distance_matrix.to_csv(DATA /\n",
    "                              \"kinase_distance_matrix_ligand_profile.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Discussion\n",
    "\n",
    "In this talktorial, we investigate how activity data can be used as a measure of similarity between kinases. The fraction of compounds tested as actives over the total number of measured compounds is a way of accessing the similarity. Moreover, using the same rationale, the promiscuity of a kinase can quantified using the ratio of active compounds over measured compounds.\n",
    "\n",
    "The kinase distance matrix above will be reloaded in Talktorial T028, where we compare kinase similarities from different perspectives, including the ligand profile perspective we have talked about in this talktorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quiz\n",
    "\n",
    "1. Is there be an optimal way to deal with multiple kinase-ligand measurements?\n",
    "2. Can promiscuity be fairly compared between two kinases if one has been tested against many compounds whereas the other only against very few? \n",
    "3. Using the similarity described in this talktorial, what does it mean that two kinases have a similarity of $0$, as is the case for p110a and LCK?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc-autonumbering": true,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
