{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0AT1L4_fX-My"
   },
   "source": [
    "# T04 - Fully Automated Drug Design Pipeline\n",
    "\n",
    "**Authors:**\n",
    "- Armin Ariamajd, CADD seminar, 2021, Charité/Freie Universität Berlin\n",
    "- Melanie Vogel,  CADD seminar, 2021, Charité/Freie Universität Berlin\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ywtrPI2vYQ2F"
   },
   "source": [
    "## Aim of This Talktorial\n",
    "<br>\n",
    "<div style=\"text-align: justify\">     \n",
    "In this talktorial we will learn to develop an <b>automated structure-based virtual screening pipeline, particularly suited for hit expansion and lead optimization</b> phases of a drug discovery project, where a known promising ligand (i.e. an initial hit or lead compound) needs to be structurally modified in order to improve its binding affinity and selectivity for the target protein. The general architecture of the pipeline can thus be summarized as follows (Figure 1):\n",
    "</div>\n",
    "<br>\n",
    "\n",
    "* **Input**\n",
    "    * Target protein structure and a promising ligand (e.g. lead or hit compound), plus specifications of the processes that need to be performed.\n",
    "<br><br>\n",
    "* **Processes**\n",
    "    1. Detection of the most druggable binding site on the given protein structure (if not defined in the input data).\n",
    "    2. Finding derivatives and structural analogs of the given ligand by performing similarity search on available compound databases, and filtering the results to obtain the most drug-like analogs.\n",
    "    3. Performing docking experiments on the selected binding site, using the found analogs.\n",
    "    4. Analyzing the protein–ligand interactions in the calculated binding modes of each analog.\n",
    "    5. Visualizing the binding modes and the corresponding interactions.\n",
    "<br><br>\n",
    "* **Output**\n",
    "    * Optimized ligand structure(s) in terms of binding affinity and selectivity.\n",
    "\n",
    "<br>\n",
    "\n",
    "<p style=\"text-align:center;\"><img src=\"images/fig1.png\"/></p>\n",
    "\n",
    "**Figure 1.** General architecture of the automated structure-based virtual screening pipeline.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MgOe9WJjYaw6"
   },
   "source": [
    "### Table of Contents\n",
    "\n",
    "1. [**Theory**](#theory)\n",
    "    * [Drug Design Pipeline](#drug_design_theory)\n",
    "    * [Binding Site Detection](#binding_site_theory) \n",
    "    * [Chemical Similarity](#similarity_theory)\n",
    "    * [Molecular Docking](#docking_theory)\n",
    "    * [Protein-Ligand Interactions](#interactions_theory)\n",
    "    * [Visual Inspection of The Docking Results](#visual_theory)\n",
    "<br><br>\n",
    "2. [**Practical**](#practical)\n",
    "    * [Outline of the Virtual Screening Pipeline](#1_practical)\n",
    "    * [Reading the Input Data and Initializing Output Paths](#2_practical)\n",
    "    * [Processing the Input Protein Data](#3_practical)\n",
    "    * [Processing the Input Ligand Data](#4_practical)\n",
    "    * [Binding Site Detection](#5_practical)\n",
    "    * [Ligand Similarity Search](#6_practical)\n",
    "    * [Molecular Docking](#7_practical)\n",
    "    * [Analysis of Protein–Ligand Interactions](#8_practical)\n",
    "    * [Selection of the Best Optimized Ligand](#9_practical)\n",
    "    * [Putting The Pieces Together: A Fully Automated Pipeline](#10_practical)\n",
    "<br><br>\n",
    "3. [**Discussion**](#discussion)\n",
    "4. [**Quiz**](#quiz)\n",
    "5. [**Supplementary Information**](#supp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "* ***TeachOpenCADD*** Teaching Platform\n",
    "    1. Journal article on *TeachOpenCADD* teaching platform for computer-aided drug design: [D. Sydow <i>et al.</i>, <i>J. Cheminform.</i> <b>2019</b>, 11, 29.](https://doi.org/10.1186/s13321-019-0351-x)\n",
    "    2. [*TeachOpenCADD* website](https://projects.volkamerlab.org/teachopencadd/index.html) at [Volkamer lab](https://volkamerlab.org/)\n",
    "    2. This talktorial is inspired by the *TeachOpenCADD* Talktorials [T013 - T017](https://github.com/volkamerlab/teachopencadd/tree/t011-base/teachopencadd/talktorials)\n",
    "<br><br>\n",
    "* **Drug Design Pipeline**\n",
    "    3. Book on drug design: [G. Klebe, <i>Drug Design</i>, Springer, <b>2013</b>.](https://doi.org/10.1007/978-3-642-17907-5)\n",
    "    4. Review article on early stages of drug discovery: [Hughes <i>et al.</i>, Br. J. Pharmacol.</i> <b>2011</b>, 162, 1239-1249.](https://doi.org/10.1111/j.1476-5381.2010.01127.x)\n",
    "    5. Review article on computational drug design: [Sliwoski <i>et al.</i>, <i>Pharmacol. Rev.</i> <b>2014</b>, 66, 334-395.](https://doi.org/10.1124/pr.112.007336)\n",
    "    6. Review article on computational drug discovery: [Leelananda <i>et al.</i>, <i>Beilstein J. Org. Chem.</i> <b>2016</b>, 12, 2694-2718.](https://doi.org/10.3762/bjoc.12.267)\n",
    "    7. Review article on free software for building a virtual screening pipeline: [Glaab, <i>Brief. Bioinform.</i> <b>2016</b>, 17, 352-366.](https://doi.org/10.1093/bib/bbv037)\n",
    "    8. Review article on automating drug discovery: [Schneider, <i>Nat. Rev. Drug Discov.</i> <b>2018</b>, 17, 97-113.](https://doi.org/10.1038/nrd.2017.232)\n",
    "    9. Review article on structure-based drug discovery: [Batool <i>et al.</i>, <i>Int. J. Mol. Sci.</i> <b>2019</b>, 20, 2783.](https://doi.org/10.3390/ijms20112783)\n",
    "<br><br>\n",
    "* **Binding Site Detection and The *DoGSiteScorer* Program** \n",
    "    10. Book chapter on prediction and analysis of binding sites: [Volkamer <i>et al.</i>, <i>Applied Chemoinformatics</i>, Wiley, <b>2018</b>, pp. 283-311.](https://doi.org/10.1002/9783527806539.ch6g)\n",
    "    11. Journal article on binding-site and druggability predictions using *DoGSiteScorer*: [Volkamer <i>et al.</i>, <i>J. Chem. Inf. Model.</i> <b>2012</b>, <i>52</i>, 360-372.](https://doi.org/10.1021/ci200454v)\n",
    "    12. Journal article describing the *ProteinsPlus* web-portal: [R. Fahrrolfes <i>et al.</i>, <i>Nucleic Acids Res.</i> <b>2017</b>, 45, W337-W343.](https://doi.org/10.1093/nar/gkx333)\n",
    "    13. [*ProteinsPlus* website](https://proteins.plus/), and information regarding the usage of its *DoGSiteScorer* [REST-API](https://proteins.plus/help/dogsite_rest)\n",
    "    14. *TeachOpenCADD* Talktorial on binding-site detection: [Talktorial T014](https://github.com/volkamerlab/teachopencadd/tree/t011-base/teachopencadd/talktorials/T014_binding_site_detection)\n",
    "    15. *TeachOpenCADD* talktorial on querying online API web-services: [Talktorial T011](https://github.com/volkamerlab/teachopencadd/tree/t011-base/teachopencadd/talktorials/T011_query_online_api_webservices)\n",
    "<br><br>\n",
    "* **Chemical Similarity and The *PubChem* Online Database**\n",
    "    16. Review article on molecular similarity in medicinal chemistry: [G. Maggiora <i>et al.</i>, <i>J. Med. Chem.</i> <b>2014</b>, 57, 3186-3204.](https://doi.org/10.1021/jm401411z)\n",
    "    17. Journal article on extended-connectivity fingerprints: [D. Rogers <i>et al.</i>, <i>J. Chem. Inf. Model.</i> <b>2010</b>, 50, 742-754.](https://doi.org/10.1021/ci100050t)\n",
    "    18. Journal article describing the latest developments of the *PubChem* web-services: [S. Kim <i>et al.</i>, <i>Nucleic Acids Res.</i> <b>2019</b>, 47, D1102-D1109.](https://doi.org/10.1093/nar/gky1033)\n",
    "    19. [*PubChem* website](https://pubchem.ncbi.nlm.nih.gov/), and information regarding the usage of its [APIs](https://pubchemdocs.ncbi.nlm.nih.gov/programmatic-access)\n",
    "    20. Description of *PubChem*'s [custom substructure fingerprint](https://ftp.ncbi.nlm.nih.gov/pubchem/specifications/pubchem_fingerprints.pdf) and [*Tanimoto* similarity measure](https://jcheminf.biomedcentral.com/articles/10.1186/s13321-016-0163-1) used in its similarity search engine.  \n",
    "    20. *TeachOpenCADD* talktorial on compound similarity: [Talktorial T004](https://github.com/volkamerlab/teachopencadd/blob/master/teachopencadd/talktorials/T004_compound_similarity/talktorial.ipynb)\n",
    "    21. *TeachOpenCADD* talktorial on data acquisition from PubChem: [Talktorial T013](https://github.com/volkamerlab/teachopencadd/blob/t011-base/teachopencadd/talktorials/T013_query_pubchem/talktorial.ipynb) \n",
    "<br><br>\n",
    "* **Molecular Docking and The *Smina* Program**\n",
    "    22. Review article on molecular docking algorithms: [X. Y. Meng <i>et al.</i>, <i>Curr. Comput. Aided Drug Des.</i> <b>2011</b>, 7, 146-157.](https://doi.org/10.2174/157340911795677602)\n",
    "    23. Review article on different software used for molecular docking: [N. S. Pagadala <i>et al.</i>, <i>Biophys. Rev.</i> <b>2017</b>, 9, 91-102.](https://doi.org/10.1007/s12551-016-0247-1)\n",
    "    24. Review article on evaluation and comparison of different docking algorithms: [G. L. Warren <i> et al.</i>, <i>J. Med. Chem.</i> <b>2006</b>, 49, 5912-5931.](https://doi.org/10.1021/jm050362n)\n",
    "    25. Review article on evaluation of ten docking programs on a diverse set of protein-ligand complexes: [Z. Wang <i> et al.</i>, Phys. Chem. Chem. Phys.</i> <b>2016</b>, 18, 12964-12975.](https://doi.org/10.1039/C6CP01555G)\n",
    "    26. Journal article describing the Smina docking program and its scoring function: [D. R. Koes <i>et al.</i>, <i>J. Chem. Inf. Model.</i> <b>2013</b>, 53, 1893-1904.](https://doi.org/10.1021/ci300604z) \n",
    "    27. [*OpenBabel* documentation](http://openbabel.org/wiki/Main_Page)\n",
    "    28. [*Smina* documentation](https://sourceforge.net/projects/smina/)\n",
    "    28. *TeachOpenCADD* talktorial on protein–ligand docking: [Talktorial T015](https://github.com/volkamerlab/teachopencadd/blob/t011-base/teachopencadd/talktorials/T015_protein_ligand_docking/talktorial.ipynb)\n",
    "<br><br>\n",
    "* **Protein-Ligand Interactions and the *PLIP* Program**\n",
    "    29. Review article on protein-ligand interactions: [X. Du <i>et al.</i>, <i>Int. J. Mol. Sci.</i> <b>2016</b>, 17, 144.](https://doi.org/10.3390/ijms17020144)\n",
    "    30. Journal article analyzing the types and frequencies of different protein-ligand interactions in available protein-ligand complex structures: [R. Ferreira de Freitas <i> et al.</i>, <i>Med. Chem. Commun.</i> <b>2017</b>, 8, 1970-1981.](https://doi.org/10.1039/C7MD00381A)\n",
    "    31. Journal article describing the *PLIP* algorithm: [S. Salentin <i>et al.</i>, <i>Nucleic Acids Res.</i> <b>2015</b>, 43, W443-447.](https://doi.org/10.1093/nar/gkv315)\n",
    "    32. [*PLIP* website](https://plip-tool.biotec.tu-dresden.de/plip-web/plip/index)\n",
    "    34. [*PLIP* documentation](https://github.com/pharmai/plip)\n",
    "    33. *TeachOpenCADD* talktorial on protein-ligand interactions: [Talktorial T016](https://github.com/volkamerlab/teachopencadd/blob/t011-base/teachopencadd/talktorials/T016_protein_ligand_interactions/talktorial.ipynb)\n",
    "<br><br>\n",
    "* **Visual Inspection of Docking Results and the *NGLView* Program**\n",
    "    34. Journal article describing the NGLView program: [H. Nguyen <i>et al.</i>, <i>Bioinformatics</i> <b>2018</b>, 34, 1241-1242.](https://doi.org/10.1093/bioinformatics/btx789)\n",
    "    35. [*NGLView* documentation](http://nglviewer.org/nglview/latest/api.html)\n",
    "    36. *TeachOpenCADD* talktorial on advanced NGLView usage: [Talktorial T017](https://github.com/volkamerlab/teachopencadd/blob/t011-base/teachopencadd/talktorials/T017_advanced_nglview_usage/talktorial.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "NsYZPbWzaJKc"
   },
   "source": [
    "<a id='theory'></a>\n",
    "## Theory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='drug_design_theory'></a>\n",
    "### Drug Design Pipeline\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "Modern drug discovery and development is a complex process, which can be categorized into several phases (Figure 2). The whole process is highly time- and capital-intensive, and on average it takes about 2–4 years and hundreds of millions of dollars only to get to the pre-clinical phase. Moreover, it is not uncommon for the drug candidates to fail in the pre-clinical and clinical phases (e.g. due to ineffectiveness or side-effects in human subjects, which could not have been predicted beforehand), so that the process has to be repeated. Therefore, development of computer-aided drug design pipelines that can greatly reduce the time and cost of drug discovery projects is an attractive alternative and an active research area.\n",
    "</div>  \n",
    "<br>\n",
    "<p style=\"text-align:center;\"><img src=\"images/fig2.png\" width=\"800\" height=auto/></p>\n",
    "\n",
    "**Figure 2.** Schematic representation of the main phases in a modern drug discovery pipeline.\n",
    "<br><br>\n",
    "<div style=\"text-align: justify\">\n",
    "    In this talktorial, we are only going to focus on the <b>hit-to-lead</b> (also known as lead generation) and <b>lead optimization</b> phases of the pipeline. In both phases, a similar procedure takes place, where the initial hit/lead compound is derivatized to synthesize a set of structural analogs, which are then tested against the target protein in screening experiments, with the goal of finding analogs with improved binding affinities, selectivities, physiochemical and pharmacokinetic properties. Hereby, we are going to implement this procedure <i>in silico</i>; Given a target protein structure and a hit/lead compound, instead of chemically synthesizing a variety of analogs and testing their potency in high-throughput screening experiments, we will obtain these analogs via a similarity-search algorithm. We can then calculate their physiochemical properties and select the best analogs to perform virtual docking experiment on, in order obtain their binding affinities to the target protein. Subsequently, the analogs with the highest calculated binding affinities are selected, and we investigate their binding modes to choose those compounds that exhibit the desired protein-ligand interactions, e.g. in order to obtain higher selectivities for the target protein as well."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "S3UoQg9yaO22"
   },
   "source": [
    "<a id='binding_site_theory'></a>\n",
    "### Binding Site Detection\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "Binding sites (also known as binding pockets) are cavities in the 3-dimensional structure of a protein in its native state (Figure 3). These are mostly found on the surface of the protein structure, and are the main regions through which the protein interacts with other entities, such as small molecules (also called ligands). For a constructive interaction, the two binding partners need to have complementary steric and electronic properties (cf. lock-and-key principle, induced-fit model). Therefore, attractive intramolecular interactions between the residues of the binding pocket and the ligand is one of the key factors in determining the ligand's potency as a drug. Furthermore, a protein may possess several binding sites, and interaction of a ligand with each of these binding sites may result in a different modulation of the protein structure, and thus its function. Consequently, detailed knowledge of the target protein's binding sites is of utmost importance in a drug discovery pipeline in order to be able to design ligands that interact most favorably with a specific binding site of the protein, leading to the desired modulation (e.g. inhibition or stimulation of the protein's function). On the other hand, it is also necessary to limit the docking experiments to the binding site of interest, so that the calculated affinities correspond to binding of the ligand to the desired pocket, and not just anywhere on the protein structure. This is not a problem when a crystal structure of the target protein is available, in which a ligand has been co-crystallized in the desired binding site. In that case, the residues surrounding the co-crystallized ligand can be simply used to define the binding pocket. However, <b>when the available protein structure does not contain a co-crystallized ligand in the desired binding pocket, it is essential for the docking experiment to first detect and define a druggable pocket in the protein</b>. \n",
    "</div>\n",
    "<p style=\"text-align:center;\">\n",
    "<img src=\"images/fig3.png\" width=\"600\" height=auto class=\"center\"/>\n",
    "</p>\n",
    "\n",
    "**Figure 3.** Crystal structure of a protein (EGFR; PDB-code: 3W32) with a co-crystallized ligand in its main (orthosteric) binding site. The protein's surface is shown in gray. The binding site is colored blue. The ligand's carbon atoms are colored green. The image was created using *PyMol*."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Binding Site Detection With *DoGSiteScorer*\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "For binding site detection in this talktorial, we will use the <b><i>DoGSiteScorer</i></b> functionality of the <b><i><a href=\"https://proteins.plus/\">ProteinsPlus</a></i></b> webserver, which uses a <b>geometry- and grid-based detection algorithm</b>. More specifically, the protein is embedded into a Cartesian 3D-grid, where each grid point is labeled as either free or occupied, depending on whether it lies within the van-der-Waals radius of any protein atoms. Subsequently, an edge-detection algorithm used in image processing, called <b>Difference of Gaussians</b> (DoG; hence the name <i>DoGSiteScorer</i>), is utilized to identify the protrusions on the protein surface. By doing so, the cavities on the protein surface that can accommodate a spheric object are identified per grid point. Lastly, cavities on neighboring grid-points are clustered together based on specific cut-off criteria, resulting in defined sub-pockets, which are then merged into pockets (Figure 4). For each (sub-)pocket, the algorithm then calculates several <b>descriptors</b>, e.g. volume, surface area, depth, hydrophobicity, number of hydrogen-bond donors/acceptors and amino acid count. Moreover, <i>DoGSiteScorer</i> also calculates two different <b>druggability</b> estimates for each (sub-)pocket; A <i>simple druggability score</i> is calculated based on a linear combination of the three descriptors describing volume, hydrophobicity and enclosure. In addition, another score is calculated by incorporating a subset of meaningful descriptors into a support vector machine (SVM) model, trained and tested on the freely available (non-redundant) druggability dataset consisting of 1069 targets. Both calculated druggability scores are between 0 and 1, where a higher score corresponds to a more druggable binding site. Using these calculated descriptors and druggability estimates, we can then choose the most suitable pocket depending on the specifics of the project in hand.\n",
    "</div>\n",
    "<br>\n",
    "<p style=\"text-align:center;\"><img src=\"images/fig4.png\" width=\"600\" height=auto/></p>\n",
    "<div style=\"text-align: justify\"> \n",
    "<b>Figure 4.</b> Visualization of some of the sub-pockets (colored meshed volumes) for the EGFR protein (PDB-code: 3W32) as detected by the <i>DoGSiteScorer</i> web-service. The co-crystallized ligand is mostly contained in the purple sub-pocket.  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-sSc9YhNbydI"
   },
   "source": [
    "<a id='similarity_theory'></a>\n",
    "### Chemical Similarity\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "As described above, in the hit expansion and lead optimization steps of an experimental drug design pipeline, several derivatives of the initial hit/lead compound are chemically synthesized, in order to find the most suitable derivative in terms of potency, selectivity and physiochemical properties. This process can be considerably accelerated in a virtual screening pipeline, where the derivatives can be obtained by performing a <b>similarity search</b> on databases of existing chemical compounds. To do so, first a <b>similarity measure</b> is required. This should be a numerical description, which can be calculated for each compound, so that they can be compared to each other based on the value of their descriptor. In addition, a scoring function is needed to calculate the <b>similarity score</b> between two compounds, using their descriptor values. The simplest descriptors for a molecule are the so-called <b>1D-descriptors</b>, which are scalar values corresponding to a certain property of the molecule, e.g. molecular weight, octanol-water partition coefficient (<i>logP</i>; a measure for lipophilicity/hydrophilicity), total polar surface area etc. However, these descriptors usually do not contain enough information to assess the structural and chemical similarity of two compounds. For this purpose, usually <b>2D-descriptors</b> (also called <b>molecular fingerprints</b>) are used, which are vectors that can represent a specific molecular structure in much more detail, using a set of scalars. A variety of different algorithms are available for generation of 2D-descriptors from chemical structures, e.g. <b><i>MACCS</i></b> <b>structural keys</b> and <b><i>ECFP/Morgan</i></b> <b>fingerprints</b>. Generally, these algorithms work by extracting a set of specific features from the structure (Figure 5), generating a numerical representation for each feature, and using these to produce either a bit-vector where each component is a bit defining the presence/absence of a particular feature (e.g. Morgan fingerprints), or a count-vector where each value corresponds to the number of times a specific feature is present in the structure (e.g. MACCS keys). Moreover, there are also different scoring functions (i.e. comparison metrics) to calculate the similarity between two molecules based on their 2D-descriptors. These include <b><i>Euclidean</i></b>-<b>distance</b> and <b><i>Manhattan</i></b>-<b>distance</b> calculations, where both presence and absence of attributes are considered, or <b><i>Tanimoto</i></b> and <b><i>Dice</i></b> coefficients, which only consider the presence of attributes. It should be noted that there is no single correct approach to calculate molecular similarity, and depending on the purpose of the project different descriptors and metrics may be used, which can generally result in vastly different similarity scores.\n",
    "</div>\n",
    "<br>\n",
    "<p style=\"text-align:center;\">\n",
    "<img src=\"images/fig5.png\" width=\"350\" height=auto class=\"center\"/>\n",
    "</p>\n",
    "<div style=\"text-align: justify\"> \n",
    "<b>Figure 5.</b> A simplified depiction of the process of calculating the similarity between two compounds. First, the structure of each compound is encoded into a molecular fingerprint bit-vector, where each bit corresponds -for example- to the presence/absence of a particular fragment in the structure. These fingerprints can then be compared using different similarity metrics in order to calculate a similarity score.\n",
    "    </div>\n",
    "<br>\n",
    "\n",
    "<div style=\"text-align: justify\"> \n",
    "In this talktorial, we will use the <b><i><a href=\"https://pubchem.ncbi.nlm.nih.gov/\">PubChem</a></i></b> web-services for performing the similarity search on the input ligand. <i>PubChem</i>, which is maintained by the <a href=\"https://www.ncbi.nlm.nih.gov/\">U.S. National Center for Biotechnology Information (NCBI)</a> contains an open database with 110 million chemical compounds and their properties (e.g. identifiers, physiochemical properties, biological activities etc.), which can be accessed through both a web-based interface, and several different web-service APIs (here, we will use their <a href=\"https://pubchemdocs.ncbi.nlm.nih.gov/programmatic-access\">PUG-REST API</a>). It also allows for directly performing similarity searches on the database, using a <a href=\"https://ftp.ncbi.nlm.nih.gov/pubchem/specifications/pubchem_fingerprints.pdf\">custom substructure fingerprint</a> as the 2D-descriptor, and the <a href=\"https://jcheminf.biomedcentral.com/articles/10.1186/s13321-016-0163-1\"><i>Tanimoto</i> similarity measure</a> as the metric. Therefore, by submitting a compound's identifier (e.g. SMILES, CID, InChI etc.) to the <i>PubChem</i>'s API, and providing a similarity threshold and the desired number of maximum results, a certain number of compounds within the given similarity threshold can be obtained.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "K5OXBSFIijNr"
   },
   "source": [
    "<a id='docking_theory'></a>\n",
    "### Molecular Docking\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "After defining an appropriate binding site in the target protein and obtaining a set of analogs for the ligand of interest, the next step is to assess the suitability of each analog in terms of its binding mode and -affinity. This can be done using a molecular docking algorithm, which works by <b>sampling the ligand's conformational space</b> inside the protein's binding site, and evaluating the energetics of protein-ligand interactions for each generated conformation, using a <b>scoring function</b>. Doing so, the <b>binding affinity</b> of each conformation is calculated, which can then be used to determine the energetically most-favorable <b>binding modes</b> (also called docking poses) of a ligand. In order to provide accurate results, most docking programs require some preparation of the protein and ligand structures. For example: \n",
    "</div>\n",
    "\n",
    "* Hydrogen atoms that are usually absent in crystal structures should be added to the protein. \n",
    "* Correct protonation state of each atom should be calculated based on a given pH value (usually the physiological pH, i.e. 7.4). \n",
    "* Partial charges should be assigned to all atoms. \n",
    "* For ligands, which are usually inputted only by an identifier (e.g. their SMILES), a reasonable low-energy conformer should also be generated, which is then used as the starting point in the conformational sampling process. \n",
    "\n",
    "<div style=\"text-align: justify\"> \n",
    "However, most of these calculations cannot be accurately performed, since they are either computationally costly (e.g. calculating the lowest-energy conformation of the ligand) or require information that are not available beforehand (e.g. the protonation states of the protein and ligand depend on the interactions between the two, and cannot be accurately calculated without knowing the actual docking pose of the ligand). These inaccuracies, along with other limitations of docking programs (e.g. computational cost of sampling all relevant conformations of the ligand, or inability in accounting for the structural changes in both protein and ligand resulting from their interactions) can result in somewhat unreliable docking results. For examples, in many cases, the docking pose with the highest calculated binding affinity does not correspond to the experimentally determined docking pose of the ligand (Figure 6).\n",
    "    </div>\n",
    "<br>\n",
    "\n",
    "<p style=\"text-align:center;\">\n",
    "<img src=\"images/fig6.png\" width=\"750\" height=auto class=\"center\"/>\n",
    "</p>\n",
    "<div style=\"text-align: justify\"> \n",
    "<b>Figure 6.</b> An example of two generated docking poses (red) in a re-docking experiment performed using the <i>Smina</i> program, superimposed over the corresponding protein structure (EGFR; PDB-code: 3W32) and the co-crystallized ligand in its native binding mode (green). While the generated docking pose shown on the left is calculated to have a higher binding affinity, it also displays a higher distance-RMSD to the native docking pose. The image was created using <i>PyMol</i>.\n",
    "    </div>\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "In this talktorial, we will use the <b><i><a href=\"https://sourceforge.net/projects/smina/\">Smina</a></i></b> docking program, which is an open-source fork of the docking program <b><i><a href=\"http://vina.scripps.edu/\">Autodock Vina</a></i></b>. Other than being open-source, <i>Smina</i> is more focused on improved scoring functions and minimization; It uses a custom empirical scoring function as default, but supports user-parameterized scoring functions as well. In addition, in order to prepare the protein and ligand structures for the docking experiment (as described above), we will also use the <b><i><a href=\"https://github.com/pybel/pybel\">Pybel</a></i></b> package - a <i>Python</i> package for the <b><i><a href=\"http://openbabel.org/wiki/Main_Page\">OpenBabel</a></i></b> program.\n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HemOC7hZinzP"
   },
   "source": [
    "<a id='interactions_theory'></a>\n",
    "### Protein–Ligand Interactions\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "As mentioned earlier, the number and type of non-covalent intramolecular interactions between the protein and the ligand is a main determining factor of the ligand's potency as a drug. These interactions, which are mostly governed by steric and electronic properties of the two interacting partners, can be categorized into several main types (Figure 7):\n",
    "</div>\n",
    "\n",
    "* Hydrophobic interactions\n",
    "* Hydrogen bonding (direct or water-bridged)\n",
    "* Metal complexation\n",
    "* Electrostatic interactions\n",
    "* $\\pi$–$\\pi$- and cation–$\\pi$ interactions\n",
    "* Halogen bonds\n",
    "\n",
    "<p style=\"text-align:center;\">\n",
    "<img src=\"images/fig7.gif\" width=\"500\" height=auto class=\"center\"/>\n",
    "</p>\n",
    "<div style=\"text-align: justify\"> \n",
    "<b>Figure 7.</b> Most common non-covalent intramolecular interactions in available protein-ligand complexes in the <i>Protein Data Bank</i> (PDB), and their frequency distribution. Reprinted from reference [<a href=\"https://doi.org/10.1039/C7MD00381A\">30</a>].\n",
    "</div>\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "While these interactions are already taken into account implicitly by the scoring function of the docking algorithm, it is still useful to explicitly analyze them in the binding modes generated by the docking experiment. This information can be used to validate the calculated binding poses, or to further narrow down the choice of an optimal lead compound in terms of selectivity, e.g. by choosing those derivatives that exhibit interactions with specific mutated or non-conserved residues in the protein. There are several programs available for assessing protein–ligand interactions, e.g. the <b><i><a href=\"https://klifs.net/index.php\">KLIFS</a></i></b> webserver, which is a kinase-centric tool that can identify interactions in the PDB entries containing kinase–ligand complexes. Another more general tool that we are also going to use here is the <b><i><a href=\"https://plip-tool.biotec.tu-dresden.de/plip-web/plip/index\">Protein–Ligand Interaction Profiler (PLIP)</a></i></b>, which is an open-source program, with both a webserver and an available package for <i>Python</i>. <i>PLIP</i> can analyze protein-ligand interactions in any given protein-ligand complex structure, by first determining a distance cut-off value based on the ligand's structure, and subsequently selecting pairs of atoms (one from the protein and one from the ligand) that lie within this cut-off value. It then identifies the potential interactions between selected pairs based on several electronic and geometric considerations. Thus, for each detected interaction, a set of information is outputted, including the interaction type, the involved atoms in the ligand and protein, as well as several properties specific to each interaction type. These information can then be used to visualize the protein-ligand interactions (Figure 8) or to analyze them algorithmically.\n",
    "</div>\n",
    "<br>\n",
    "<p style=\"text-align:center;\">\n",
    "<img src=\"images/fig8.png\" width=\"600\" height=auto class=\"center\"/>\n",
    "</p>\n",
    "<div style=\"text-align: justify\">   \n",
    "<b>Figure 8.</b> Visualization of the protein-ligand interactions in a protein-ligand complex structure (EGFR; PDB-code: 3W32) detected by the <i>PLIP</i> web-service. From the protein, only the interacting residues are shown, with their carbon atoms colored blue. The ligand's carbon atoms are colored brown. Hydrophobic interactions are shown as gray dashed lines. Hydrogen bonds are depicted as blue lines. $\\pi$-stacking interactions are shown as green dashed lines. Halogen bonds are displayed using cyan lines.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "saNArWyiQMY1"
   },
   "source": [
    "<a id='visual_theory'></a>\n",
    "### Visual Inspection of The Docking Results\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Visual inspection of the calculated docking poses and their corresponding protein-ligand interactions is especially important due to the current limitations of docking algorithms, to an extent where most experts find docking scores to be the least important criterion for selecting docking poses. Instead, some of the usual considerations when selecting a binding mode is as follows:\n",
    "</div>\n",
    "\n",
    "* Similarity to experimentally observed binding modes in available crystal structures of the target protein\n",
    "* Steric and electronic complementarity\n",
    "* Minimized number of unsaturated hydrogen bonds\n",
    "* No solvent-exposed hydrophobic moieties in the ligand\n",
    "* displacement of, or interactions with water molecules in the pocket\n",
    "* steric strain induced by the ligand binding\n",
    "\n",
    "<div style=\"text-align: justify\"> \n",
    "However, since manual inspection is a time-consuming process, it can only be performed on a small subset of calculated docking poses, which are usually those with the highest calculated binding affinities. To visualize the results, we use <b><i>NGLView</i></b>, which is a <i>Jupyter</i> widget (using a <i>Python</i> wrapper for the <i>Javascript</i>-based <i>NGL</i> library) allowing for visualization of structures within a <i>Jupyter</i> notebook in an interactive 3D view.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='practical'></a>\n",
    "## Practical"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this section, we will implement and demonstrate the automated structure-based virtual screening pipeline step by step.\n",
    "\n",
    "First of all, we import all the dependencies that we are going to need for the development of the pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard library:\n",
    "from enum import Enum  # for creating enumeration classes\n",
    "import gzip  # for decompressing .gz files downloaded from DoGSiteScorer\n",
    "import io  # for creating file-like objects from strings of data (needed as input for some functions)\n",
    "import logging  # for setting the logging level of some packages (i.e. to disable excessive logging default to some packages e.g. PLIP)\n",
    "from pathlib import Path  # for creating folders and handling local paths\n",
    "import subprocess  # for creating shell processes (needed to communicate with Smina program)\n",
    "import time  # for creating pauses during the runtime (to wait for the response of API requests)\n",
    "from urllib.parse import quote  # for url quoting\n",
    "\n",
    "# 3rd-party packages:\n",
    "from biopandas.pdb import PandasPdb  # for working with PDB files\n",
    "from IPython.display import (\n",
    "    Markdown,\n",
    "    Image,\n",
    ")  # ror more display options in the Jupyter Notebook\n",
    "from ipywidgets import (\n",
    "    AppLayout,\n",
    "    Layout,\n",
    "    Select,\n",
    "    Button,\n",
    ")  # for interactive outputs in the Jupyter Notebook\n",
    "import matplotlib as mpl  # for changing the display settings of plots (see bottom of the cell: Settings)\n",
    "from matplotlib import (\n",
    "    colors,\n",
    ")  # for plotting color-maps (for visualization of protein-ligand interactions)\n",
    "import matplotlib.pyplot as plt  # for plotting of data\n",
    "import nglview as nv  # for visualization of the protein and protein-related data (e.g. binding sites, docking poses)\n",
    "import numpy as np  # for some more functionalities when using Pandas (e.g. for handling NaN values)\n",
    "from openbabel import (\n",
    "    pybel,\n",
    ")  # for preparing protein and ligand for docking, and other manipulations of PDB files\n",
    "import pandas as pd  # for creating dataframes and handling data\n",
    "import plip  # for changing the logging setting of the package (see bottom of the cell: Settings)\n",
    "from plip.structure.preparation import (\n",
    "    PDBComplex,\n",
    ")  # for calculating protein-ligand interactions\n",
    "from plip.exchange.report import (\n",
    "    BindingSiteReport,\n",
    ")  # for calculating protein-ligand interactions\n",
    "import pypdb  # for communicating with the RCSB Protein Data Bank (PDB) to fetch PDB files\n",
    "from rdkit import (\n",
    "    Chem,\n",
    ")  # for handling ligand data and calculating ligand-related properties\n",
    "from rdkit.Chem import Draw, AllChem, Descriptors, PandasTools, rdFMCS\n",
    "import requests  # for communicating with web-service APIs\n",
    "\n",
    "# In-house packages:\n",
    "from opencadd.structure.core import Structure  # for manipulating PDB files\n",
    "\n",
    "# Settings:\n",
    "logging.getLogger(plip.__name__).setLevel(\n",
    "    logging.WARNING\n",
    ")  # disabling excessive INFO logs of the PLIP package\n",
    "mpl.rcParams[\"figure.dpi\"] = 300  # for plots with higher resolution\n",
    "mpl.rcParams[\"agg.path.chunksize\"] = 10000  # for handling plots with large number of data points\n",
    "pd.set_option(\"display.max_columns\", 100)  # showing 100 columns at most\n",
    "pd.set_option(\"display.max_colwidth\", 200)  # increasing the maximum column width\n",
    "pd.set_option(\"display.width\", None)  # showing each row in a single line"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='1_practical'></a>\n",
    "### Outline of the Virtual Screening Pipeline\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Since this is a relatively large project with several different functionalities, it is a good practice to use classes. By doing so, possible name collisions are avoided, and the code will be more well-structured, and thus easier to follow, maintain, reuse and expand. Therefore, we are going to implement the following classes, each handling a specific part of the pipeline:</div>\n",
    "\n",
    "* ***Specs***: Reads and internalizes all the specifications in the input data.\n",
    "* ***Protein***: Creates a protein object, containing all the protein data and required functionalities to process the protein.\n",
    "* ***Ligand***: Creates a ligand object, containing all the ligand data and required functionalities to process the ligand.\n",
    "* ***BindingSiteDetection***: Automatically performs all the necessary processes for binding-site detection.\n",
    "* ***LigandSimilaritySearch***: Automatically performs all the necessary processes for finding analogs.\n",
    "* ***Docking***: Carries out the docking experiments for all provided analogs.\n",
    "* ***InteractionAnalysis***: Analyzes the protein-ligand interactions in the calculated docking modes.\n",
    "* ***OptimizedLigands***: Analyzes the whole data obtained throughout the pipeline to choose the best analogs according to the input specifications.\n",
    "\n",
    "<div style=\"text-align: justify\"> \n",
    "Now if we didn't want to present our code step by step, we would have started by implementing the above classes first, and then at then end, use them to build the pipeline. While this approach would be much more convenient for programming, it is not appropriate for presentation purposes. First of all, the program would not be presentable until the end due to the bottom-to-top approach. Moreover, it would result in large blocks of code, preventing us from demonstrating the workings of the program in a step-by-step fashion. To circumvent this problem, we are going to structure the code as follows:</div>\n",
    "\n",
    "1. For each specific functionality of the pipeline (e.g. reading input data, handling PDB files etc.) a separate helper-class will be defined, containing all the required functions as static methods.\n",
    "2. All functions of the helper-classes can then be separately demonstrated (in [***Supplementary Information***](#supp)). \n",
    "3. Using these helper-classes, the main classes of the pipeline can then be developed and demonstrated in a step-by-step fashion.\n",
    "4. At the end, when all the necessary parts of the pipeline have been developed, a function called <b><i>run_pipeline</i></b> will be defined, which automatically carries out all of the processes we demonstrated through the talktorial, and generates the required output, merely by providing it with an input specification file for the project. \n",
    "<br>\n",
    "\n",
    "<div style=\"text-align: justify\"> \n",
    "By doing so, the pipeline can be implemented and executed step by step, so that the inner workings of the code can be better presented, and simultaneously we will also be able to have a fully-automated pipeline at the end. Therefore, for our demonstration we will first define a container class <b><i>LeadOptimizationPipeline</i></b> with only a single instance attribute called <b><i>name</i></b>, which allows us to instantiate the class and create a new pipeline project by providing the project's name. For the rest of the talktorial we are then going to define the other mentioned classes, instantiate them with the necessary input data, and assign that instance to our created project, in order to have all the generated data always organized in one place."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class LeadOptimizationPipeline:\n",
    "    def __init__(self, project_name):\n",
    "        self.name = project_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Creating a New Project: Instantiating The ***LeadOptimizationPipeline*** Class \n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "    We can now create an instance of the <b><i>LeadOptimizationPipeline</i></b> class, which we will call <b><i>project1</i></b>. For demonstrating the pipeline, we have chosen the same protein and ligand illustrated in <i>figure 1</i>, i.e. the epidermal growth factor receptor (EGFR) as our target protein, and the ligand with the ChEMBL-ID CHEMBL328216. Thus, we will simply name our lead optimization project <b>Project1_EGFR_CHEMBL328216</b>:</div>\n",
    "<br>\n",
    "\n",
    "***Note:*** For more clarity, we will always explicitly write the names of all input parameters for each function and class when using them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1 = LeadOptimizationPipeline(project_name=\"Project1_EGFR_CHEMBL328216\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The instance-attribute ***name*** can then be accessed as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='2_practical'></a>\n",
    "### Reading the Input Data and Initializing Output Paths\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "The first thing the pipeline should be able to do is to read and process the input data for the protein and the ligand, as well as specifications for the processes that need to be performed on them. Since this involves many parameters, it is best to use a file to store all the necessary input data for a specific project. Thus for each project, the user only has to fill in a template input file with all the necessary data, and then specify the filepath of the input data when running the program. Here, we use a CSV file, which is stored in the folder <b><i>data</i></b>, under the name <b><i>InputData_Template</i></b>. For demonstration purposes, we can open the empty template file here to have a closer look:</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pd.read_csv(\"data/InputData_Template.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen, the table contains four columns:\n",
    "* ***Subject***: Specifies the subject of the input parameter. As discussed, we need to input a ***Protein***, a ***Ligand*** and a set of specifications corresponding to each part of the pipeline, namely ***Binding Site***, ***Ligand Similarity Search***, ***Docking***, ***Interaction Analysis*** and ***Optimized Ligand***. \n",
    "\n",
    "* ***Property***: Specifies a specific property of the ***Subject***. Required properties are marked with an asterisk. All other properties are optional (i.e. have default values set in the program), and some are dependent on other properties. For example, If the ***Binding Site Definition Method*** is not ***'coordinates'***, then there is no need to enter the value for the ***Binding Site Coordinates*** row. \n",
    "\n",
    "* ***Value***: The only column that should be filled by the user. Each value corresponds to a specific ***Property*** of a specific ***Subject***.\n",
    "\n",
    "* ***Description***: Provides a short description as to what input data is expected in each specific row, and when it should be provided.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Defining The Required Constants For Handling the Input Data: The ***Consts*** Data-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "In order to import and process the input file and extract all the different parameters, we will use the <i>Pandas</i> package, which can directly read CSV files and transform them into a <i>DataFrame</i> object – the <i>Pandas</i> equivalence of a table in a database. In a <i>quick-and-dirty</i> approach, the input parameters are then called - whenever needed in the code - using their respective index- and column names in the dataframe. However, this will lead to a code that is not easily maintainable or expandable, since a small change in the input data will require the whole code to be revised accordingly. A much more efficient approach is to first internalize all the input data in a data-class, so that the rest of the program needs only to communicate with this class and not the input dataframe. There are several advantages to this approach. First, any error in the input data is recognized in the beginning, before any process is performed. Also, during the initial processing, all missing values can be replaced with the default values at once. Furthermore, in the next versions of the program, any change in the input data needs only to be accounted for in this class, and not in the entire code. Also, the class gives an overview of all the possible inputs for the program. For these purposes, we define a data-class called <i><b>Consts</b></i>, which contains all the possible keywords in the input dataframe, such as the column names, index names, subject names, properties, etc. These keywords are all stored in their respective sub-classes as <b>Enumerations</b>, so that the rest of the code needs only to refer to the enumeration names, and not their values, which are subject to change in the future versions of the program.\n",
    "    </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Consts:\n",
    "    \"\"\"\n",
    "    Data-class containing the required constants for defining the expected input data.\n",
    "    The rest of the code will refer to these constants.\n",
    "    \"\"\"\n",
    "\n",
    "    # Constants for the input dataframe\n",
    "    class DataFrame:\n",
    "\n",
    "        # Name of the columns\n",
    "        class ColumnNames(Enum):\n",
    "            SUBJECT = \"Subject\"\n",
    "            PROPERTY = \"Property\"\n",
    "            VALUE = \"Value\"\n",
    "            DESCRIPTION = \"Description\"\n",
    "\n",
    "        # Name of the subjects\n",
    "        class SubjectNames(Enum):\n",
    "            PROTEIN = \"Protein\"\n",
    "            LIGAND = \"Ligand\"\n",
    "            BINDING_SITE = \"Binding Site\"\n",
    "            LIGAND_SIMILARITY_SEARCH = \"Ligand Similarity Search\"\n",
    "            DOCKING = \"Docking\"\n",
    "            INTERACTION_ANALYSIS = \"Interaction Analysis\"\n",
    "            OPTIMIZED_LIGAND = \"Optimized Ligand\"\n",
    "\n",
    "    # Constants for the input protein data\n",
    "    # i.e. name of protein properties, and their respective allowed values\n",
    "    class Protein:\n",
    "        class Properties(Enum):\n",
    "            INPUT_TYPE = \"Input Type*\"\n",
    "            INPUT = \"Input Value*\"\n",
    "\n",
    "        class InputTypes(Enum):\n",
    "            PDB_CODE = \"pdb_code\"\n",
    "            PDB_FILEPATH = \"pdb_filepath\"\n",
    "\n",
    "    # Constants for the input ligand data\n",
    "    class Ligand:\n",
    "        class Properties(Enum):\n",
    "            INPUT_TYPE = \"Input Type*\"\n",
    "            INPUT = \"Input Value*\"\n",
    "\n",
    "        class InputTypes(Enum):\n",
    "            NAME = \"name\"\n",
    "            IUPAC_NAME = \"iupac_name\"\n",
    "            SMILES = \"smiles\"\n",
    "            CID = \"cid\"\n",
    "            INCHI = \"inchi\"\n",
    "            INCHIKEY = \"inchikey\"\n",
    "\n",
    "    # Constants for the input specification data regarding binding-site\n",
    "    class BindingSite:\n",
    "        class Properties(Enum):\n",
    "            DEFINITION_METHOD = \"Definition Method\"\n",
    "            COORDINATES = \"Coordinates\"\n",
    "            LIGAND = \"LIGAND\"\n",
    "            DETECTION_METHOD = \"Detection Method\"\n",
    "            SELECTION_METHOD = \"Selection Method\"\n",
    "            SELECTION_CRITERIA = \"Selection Criteria\"\n",
    "            PROTEIN_CHAIN_ID = \"Protein Chain-ID\"\n",
    "            PROTEIN_LIGAND_ID = \"Protein Ligand-ID\"\n",
    "\n",
    "        class DefinitionMethods(Enum):\n",
    "            DETECTION = \"detection\"\n",
    "            LIGAND = \"ligand\"\n",
    "            COORDINATES = \"coordinates\"\n",
    "\n",
    "        class DetectionMethods(Enum):\n",
    "            DOGSITESCORER = \"dogsitescorer\"\n",
    "\n",
    "        class SelectionMethods(Enum):\n",
    "            SORTING = \"sorting\"\n",
    "            FUNCTION = \"function\"\n",
    "\n",
    "        class SelectionCriteria(Enum):\n",
    "            LIGAND_COVERAGE = \"lig_cov\"\n",
    "            POCKET_COVERAGE = \"poc_cov\"\n",
    "            VOLUME = \"volume\"\n",
    "            ENCLOSURE = \"enclosure\"\n",
    "            SURFACE = \"surface\"\n",
    "            DEPTH = \"depth\"\n",
    "            SURFACE_VOLUME_RATIO = \"surf/vol\"\n",
    "            ELLIPSOID_MAIN_AXIS_C_A_RATIO = \"ell c/a\"\n",
    "            ELLIPSOID_MAIN_AXIS_B_A_RATIO = \"ell b/a\"\n",
    "            NUM_POCKET_ATOMS = \"siteAtms\"\n",
    "            NUM_CARBONS = \"Cs\"\n",
    "            NUM_NITROGENS = \"Ns\"\n",
    "            NUM_OXYGENS = \"Os\"\n",
    "            NUM_SULFURS = \"Ss\"\n",
    "            NUM_OTHER_ELEM = \"Xs\"\n",
    "            NUM_H_ACCEPTORS = \"accept\"\n",
    "            NUM_H_DONORS = \"donor\"\n",
    "            NUM_HYDROPHOBIC_INTERACTIONS = \"hydrophobic_interactions\"\n",
    "            NUM_HYDROPHOBICITY = \"hydrophobicity\"\n",
    "            NUM_METALS = \"metal\"\n",
    "            NUM_NEGATIVE_AA = \"negAA\"\n",
    "            NUM_POSITIVE_AA = \"posAA\"\n",
    "            NUM_POLAR_AA = \"polarAA\"\n",
    "            NUM_APOLAR_AA = \"apolarAA\"\n",
    "            SIMPLE_SCORE = \"simpleScore\"\n",
    "            DRUG_SCORE = \"drugScore\"\n",
    "\n",
    "    # Constants for the input specification data regarding ligand similarity search\n",
    "    class LigandSimilaritySearch:\n",
    "        class Properties(Enum):\n",
    "            SEARCH_ENGINE = \"Search Engine\"\n",
    "            MIN_SIMILARITY_PERCENT = \"Minumum Similarity [%]\"\n",
    "            MAX_NUM_RESULTS = \"Maximum Number of Results\"\n",
    "            MAX_NUM_DRUGLIKE = \"Maximum Number of Most Drug-Like Analogs to Continue With\"\n",
    "\n",
    "        class SearchEngines(Enum):\n",
    "            PUBCHEM = \"pubchem\"\n",
    "\n",
    "    # Constants for the input specification data regarding docking\n",
    "    class Docking:\n",
    "        class Properties(Enum):\n",
    "            PROGRAM = \"Program\"\n",
    "            NUM_POSES_PER_LIGAND = \"Number of Docking Poses per Ligand\"\n",
    "            EXHAUSTIVENESS = \"Exhaustiveness\"\n",
    "            RANDOM_SEED = \"Random Seed\"\n",
    "\n",
    "        class Programs(Enum):\n",
    "            SMINA = \"smina\"\n",
    "\n",
    "    # Constants for the input specification data regarding interaction analysis\n",
    "    class InteractionAnalysis:\n",
    "        class Properties(Enum):\n",
    "            PROGRAM = \"Program\"\n",
    "\n",
    "        class Programs(Enum):\n",
    "            PLIP = \"plip\"\n",
    "\n",
    "    class OptimizedLigands:\n",
    "        class Properties(Enum):\n",
    "            NUM_RESULTS = \"Number of Results\"\n",
    "            SELECTION_METHOD = \"Selection Method\"\n",
    "            SELECTION_CRITERIA = \"Selection Criteria\"\n",
    "\n",
    "        class SelectionMethods(Enum):\n",
    "            SORTING = \"sorting\"\n",
    "            FUNCTION = \"function\"\n",
    "\n",
    "        class SelectionCriteria(Enum):\n",
    "            BINDING_AFFINITY = \"affinity\"\n",
    "            NUM_H_BONDS = \"h_bond\"\n",
    "            NUM_HYDROPHOBIC_INTERACTIONS = \"hydrophobic\"\n",
    "            NUM_SALT_BRIDGES = \"salt_bridge\"\n",
    "            NUM_WATER_BRIDGES = \"water_bridge\"\n",
    "            NUM_PI_STACKINGS = \"pi_stacking\"\n",
    "            NUM_CATION_PI = \"pi_cation\"\n",
    "            NUM_HALGON_BONDS = \"halogen\"\n",
    "            NUM_METAL_BONDS = \"metal\"\n",
    "            NUM_ALL_INTERACTIONS = \"total_num_interactions\"\n",
    "            DRUG_SCORE_LIPINSKI = \"drug_score_lipinski\"\n",
    "            DRUG_SCORE_QED = \"drug_score_qed\"\n",
    "            DRUG_SCORE_CUSTOM = \"drug_score_custom\"\n",
    "            DRUG_SCORE_TOTAL = \"drug_score_total\"\n",
    "\n",
    "    # Constants for the input data regarding output paths\n",
    "    class Output:\n",
    "        class FolderNames(Enum):\n",
    "            PROTEIN = \"1_Protein\"\n",
    "            LIGAND = \"2_Ligand\"\n",
    "            BINDING_SITE_DETECTION = \"3_Binding Site Detection\"\n",
    "            SIMILARITY_SEARCH = \"4_Ligand Similarity Search\"\n",
    "            DOCKING = \"5_Docking\"\n",
    "            INTERACTION_ANALYSIS = \"6_Interaction Analysis\"\n",
    "            VISUALIZATION = \"7_Visualizations\"\n",
    "            OPTIMIZED_LIGANDS = \"8_Optimized Ligands\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***IO*** Helper-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Now to be able to read and process the input data, we will first define a helper class called <b><i>IO</i></b>, which contains all the necessary functions for handling the input and output data, e.g. for creating a dataframe from the input CSV file, extracting specific information from the dataframe, or creating folders for storing the output data. <br><br><b><i>Note:</i></b> For demonstration of each function, see the <a href=\"#io_demo\">corresponding section</a> in <i>Supplementary Information</i>.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class IO:\n",
    "    \"\"\"\n",
    "    Set of functions for handling the input/output data.\n",
    "    \"\"\"\n",
    "\n",
    "    @staticmethod\n",
    "    def create_dataframe_from_csv_input_file(\n",
    "        input_data_filepath, list_of_index_column_names, list_of_columns_to_keep\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Read a CSV file and create a pandas DataFrame with given specifications.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_data_filepath : str or pathlib.Path object\n",
    "            Path of the CSV data file.\n",
    "        list_of_index_column_names : list of strings\n",
    "            List of column names in the CSV file to be used as indices for the dataframe.\n",
    "        list_of_columns_to_keep : list of strings\n",
    "            List of column names from the CSV file to keep in the dataframe.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            Pandas DataFrame\n",
    "        \"\"\"\n",
    "        input_df = pd.read_csv(input_data_filepath)\n",
    "        input_df.set_index(list_of_index_column_names, inplace=True)\n",
    "        input_df.drop(input_df.columns.difference(list_of_columns_to_keep), 1, inplace=True)\n",
    "        return input_df\n",
    "\n",
    "    @staticmethod\n",
    "    def copy_series_from_dataframe(input_df, index_name, column_name):\n",
    "        \"\"\"\n",
    "        Take a multi-index dataframe, and make a copy of the data\n",
    "        corresponding to a given index and column.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_df : Pandas DataFrame\n",
    "            The dataframe to extract the data from.\n",
    "        index_name : str\n",
    "            The index-value of the needed rows.\n",
    "        column_name : str\n",
    "            The column-name of the needed values.\n",
    "        Returns\n",
    "        -------\n",
    "            Pandas Series\n",
    "            Copy of the data corresponding to given index- and column-name.\n",
    "        \"\"\"\n",
    "        subject_data = input_df.xs(index_name, level=0, axis=0)[column_name].copy()\n",
    "        return subject_data\n",
    "\n",
    "    @staticmethod\n",
    "    def create_folder(folder_name, folder_path=\"\"):\n",
    "        \"\"\"\n",
    "        Create a folder with a given name in a given path.\n",
    "        Also creates all non-existing parent folders.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        folder_name : str\n",
    "            Name of the folder to be created.\n",
    "\n",
    "        folder_path : str (optional; default: current path)\n",
    "            Either relative or absolute path of the folder to be created.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            pathlib.Path object\n",
    "            Full path of the created folder.\n",
    "        \"\"\"\n",
    "        path = Path(folder_path) / folder_name\n",
    "        # Creating the folder and all non-existing parent folders.\n",
    "        path.mkdir(parents=True, exist_ok=True)\n",
    "        return path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Implementing The Pipeline's ***Specs*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "We can now create the <b><i>Specs</i></b> class of the pipeline, using the functionalities in the <b><i>IO</i></b> helper-class. This class is responsible for automatically reading and internalizing all the input data contained in the input file, so that the rest of the program needs only to communicate with this class and not the input dataframe for getting the required data. It also contains some logic, e.g. to check if all the necessary data for a specific project have been inputted by the user, and to fill in some default values if needed. Furthermore, it creates the necessary folders for the output data, and stores their paths.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class Specs:\n",
    "    \"\"\"\n",
    "    Data-class containing all the input data and output paths of the project.\n",
    "    Take the filepath to the CSV input-data as well as output path, and\n",
    "    internalize all the data as instance attributes.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    input_data_filepath : str or pathlib.Path object\n",
    "        Relative or absolute local path of the input CSV-data-file for the project.\n",
    "    output_data_root_folder_path : str or pathlib.Path object\n",
    "        Relative or absolute local path of root folder to store output data in.\n",
    "    \"\"\"\n",
    "\n",
    "    # Defining the default values for all optional entries\n",
    "    # -----------------------------------------------------\n",
    "    class BindingSiteDefaults(Enum):\n",
    "        DEFINITION_METHOD = Consts.BindingSite.DefinitionMethods.DETECTION\n",
    "        DETECTION_METHOD = Consts.BindingSite.DetectionMethods.DOGSITESCORER\n",
    "        SELECTION_METHOD = Consts.BindingSite.SelectionMethods.SORTING\n",
    "        SELECTION_CRITERIA_SORTING = [\n",
    "            Consts.BindingSite.SelectionCriteria.LIGAND_COVERAGE.value,\n",
    "            Consts.BindingSite.SelectionCriteria.POCKET_COVERAGE.value,\n",
    "        ]\n",
    "        SELECTION_CRITERIA_FUNCTION = f\"(df[{Consts.BindingSite.SelectionCriteria.DRUG_SCORE.value}] + df[{Consts.BindingSite.SelectionCriteria.SIMPLE_SCORE.value}]) / df[{Consts.BindingSite.SelectionCriteria.VOLUME}]\"\n",
    "        PROTEIN_CHAIN_ID = \"\"\n",
    "        PROTEIN_LIGAND_ID = \"\"\n",
    "\n",
    "    class LigandSimilaritySearchDefaults(Enum):\n",
    "        SEARCH_ENGINE = Consts.LigandSimilaritySearch.SearchEngines.PUBCHEM\n",
    "        MIN_SIMILARITY_PERCENT = 80\n",
    "        MAX_NUM_RESULTS = 100\n",
    "        MAX_NUM_DRUGLIKE = 30\n",
    "\n",
    "    class DockingDefaults(Enum):\n",
    "        PROGRAM = Consts.Docking.Programs.SMINA\n",
    "        NUM_POSES_PER_LIGAND = 5\n",
    "        EXHAUSTIVENESS = 10\n",
    "        RANDOM_SEED = 1111\n",
    "\n",
    "    class InteractionAnalysisDefaults(Enum):\n",
    "        PROGRAM = Consts.InteractionAnalysis.Programs.PLIP\n",
    "\n",
    "    class OptimizedLigandsDefaults(Enum):\n",
    "        NUM_RESULTS = 1\n",
    "        SELECTION_METHOD = Consts.OptimizedLigands.SelectionMethods.SORTING\n",
    "        SELECTION_CRITERIA_SORTING = [\n",
    "            Consts.OptimizedLigands.SelectionCriteria.BINDING_AFFINITY.value,\n",
    "            Consts.OptimizedLigands.SelectionCriteria.NUM_ALL_INTERACTIONS.value,\n",
    "        ]\n",
    "        SELECTION_CRITERIA_FUNCTION = f\"-2*df[{Consts.OptimizedLigands.SelectionCriteria.BINDING_AFFINITY.value}] + df[{Consts.OptimizedLigands.SelectionCriteria.NUM_ALL_INTERACTIONS}] * df[{Consts.OptimizedLigands.SelectionCriteria.DRUG_SCORE_TOTAL}]\"\n",
    "\n",
    "    # -----------------------------------------------------\n",
    "\n",
    "    def __init__(self, input_data_filepath, output_data_root_folder_path):\n",
    "        self.RawData = self.RawData(input_data_filepath)\n",
    "        self.Protein = self.Protein(self.RawData.protein)\n",
    "        self.Ligand = self.Ligand(self.RawData.ligand)\n",
    "        self.BindingSite = self.BindingSite(self.RawData.binding_site)\n",
    "        self.LigandSimilaritySearch = self.LigandSimilaritySearch(\n",
    "            self.RawData.ligand_similarity_search\n",
    "        )\n",
    "        self.Docking = self.Docking(self.RawData.docking)\n",
    "        self.InteractionAnalysis = self.InteractionAnalysis(self.RawData.interaction_analysis)\n",
    "        self.OptimizedLigands = self.OptimizedLigands(self.RawData.optimized_ligand)\n",
    "        self.OutputPaths = self.OutputPaths(output_data_root_folder_path)\n",
    "\n",
    "    # Defining a sub-class for each part of the pipeline\n",
    "    # -----------------------------------------------------\n",
    "    class RawData:\n",
    "        def __init__(self, input_data_filepath):\n",
    "            self.filepath = input_data_filepath\n",
    "            self.all_data = IO.create_dataframe_from_csv_input_file(\n",
    "                input_data_filepath=input_data_filepath,\n",
    "                list_of_index_column_names=[\n",
    "                    Consts.DataFrame.ColumnNames.SUBJECT.value,\n",
    "                    Consts.DataFrame.ColumnNames.PROPERTY.value,\n",
    "                ],\n",
    "                list_of_columns_to_keep=[Consts.DataFrame.ColumnNames.VALUE.value],\n",
    "            )\n",
    "\n",
    "            for subject_name in Consts.DataFrame.SubjectNames:\n",
    "                subject_data = IO.copy_series_from_dataframe(\n",
    "                    input_df=self.all_data,\n",
    "                    index_name=subject_name.value,\n",
    "                    column_name=Consts.DataFrame.ColumnNames.VALUE.value,\n",
    "                )\n",
    "                setattr(self, subject_name.name.lower(), subject_data)\n",
    "\n",
    "    class Protein:\n",
    "        def __init__(self, input_protein_data):\n",
    "            self.input_type = Consts.Protein.InputTypes(\n",
    "                input_protein_data[Consts.Protein.Properties.INPUT_TYPE.value]\n",
    "            )\n",
    "            self.input_value = input_protein_data[Consts.Protein.Properties.INPUT.value]\n",
    "\n",
    "    class Ligand:\n",
    "        def __init__(self, input_ligand_data):\n",
    "            self.input_type = Consts.Ligand.InputTypes(\n",
    "                input_ligand_data[Consts.Ligand.Properties.INPUT_TYPE.value]\n",
    "            )\n",
    "            self.input_value = input_ligand_data[Consts.Ligand.Properties.INPUT.value]\n",
    "\n",
    "    class BindingSite:\n",
    "        def __init__(self, input_binding_site_data):\n",
    "            definition_method = input_binding_site_data[\n",
    "                Consts.BindingSite.Properties.DEFINITION_METHOD.value\n",
    "            ]\n",
    "            # check if definition method is given, if not set to default (i.e. detection)\n",
    "            self.definition_method = (\n",
    "                Specs.BindingSiteDefaults.DEFINITION_METHOD\n",
    "                if definition_method is np.nan\n",
    "                else Consts.BindingSite.DefinitionMethods(definition_method)\n",
    "            )\n",
    "\n",
    "            if self.definition_method is Consts.BindingSite.DefinitionMethods.COORDINATES:\n",
    "                coordinates_as_string = input_binding_site_data[\n",
    "                    Consts.BindingSite.Properties.COORDINATES.value\n",
    "                ]\n",
    "                coordinates = coordinates_as_string.split(\" \")\n",
    "                self.coordinates = {\"center\": coordinates[:3], \"size\": coordinates[3:]}\n",
    "\n",
    "            elif self.definition_method is Consts.BindingSite.DefinitionMethods.LIGAND:\n",
    "                self.ligand = input_binding_site_data[Consts.BindingSite.Properties.LIGAND.value]\n",
    "\n",
    "            elif self.definition_method is Consts.BindingSite.DefinitionMethods.DETECTION:\n",
    "                detection_method = input_binding_site_data[\n",
    "                    Consts.BindingSite.Properties.DETECTION_METHOD.value\n",
    "                ]\n",
    "                self.detection_method = (\n",
    "                    Specs.BindingSiteDefaults.DETECTION_METHOD\n",
    "                    if detection_method is np.nan\n",
    "                    else Consts.BindingSite.DetectionMethods(detection_method)\n",
    "                )\n",
    "\n",
    "                protein_chain_id = input_binding_site_data[\n",
    "                    Consts.BindingSite.Properties.PROTEIN_CHAIN_ID.value\n",
    "                ]\n",
    "                self.protein_chain_id = (\n",
    "                    Specs.BindingSiteDefaultValues.PROTEIN_CHAIN_ID\n",
    "                    if protein_chain_id is np.nan\n",
    "                    else protein_chain_id\n",
    "                )\n",
    "\n",
    "                protein_ligand_id = input_binding_site_data[\n",
    "                    Consts.BindingSite.Properties.PROTEIN_LIGAND_ID.value\n",
    "                ]\n",
    "                self.protein_ligand_id = (\n",
    "                    Specs.BindingSiteDefaultValues.PROTEIN_LIGAND_ID\n",
    "                    if protein_ligand_id is np.nan\n",
    "                    else protein_ligand_id\n",
    "                )\n",
    "\n",
    "                selection_method = input_binding_site_data[\n",
    "                    Consts.BindingSite.Properties.SELECTION_METHOD.value\n",
    "                ]\n",
    "                self.selection_method = (\n",
    "                    Specs.BindingSiteDefaultValues.SELECTION_METHOD\n",
    "                    if selection_method is np.nan\n",
    "                    else Consts.BindingSite.SelectionMethods(selection_method)\n",
    "                )\n",
    "\n",
    "                if self.selection_method is Consts.BindingSite.SelectionMethods.SORTING:\n",
    "                    selection_criteria = input_binding_site_data[\n",
    "                        Consts.BindingSite.Properties.SELECTION_CRITERIA.value\n",
    "                    ]\n",
    "                    if selection_criteria is np.nan:\n",
    "                        self.selection_criteria = (\n",
    "                            Specs.BindingSiteDefaults.SELECTION_CRITERIA_SORTING.value\n",
    "                        )\n",
    "                    else:\n",
    "                        # pass the column names through the SelectionCriteria enumeration class to make sure they are valid\n",
    "                        self.selection_criteria = [\n",
    "                            Consts.BindingSite.SelectionCriteria(pocket_property.strip()).value\n",
    "                            for pocket_property in selection_criteria.split(\",\")\n",
    "                        ]\n",
    "\n",
    "                elif self.selection_method is Consts.BindingSite.SelectionMethods.FUNCTION:\n",
    "                    selection_criteria = input_binding_site_data[\n",
    "                        Consts.BindingSite.Properties.SELECTION_CRITERIA.value\n",
    "                    ]\n",
    "                    self.selection_criteria = (\n",
    "                        Specs.BindingSiteDefaults.SELECTION_CRITERIA_FUNCTION.value\n",
    "                        if selection_criteria is np.nan\n",
    "                        else selection_criteria\n",
    "                    )\n",
    "\n",
    "    class LigandSimilaritySearch:\n",
    "        def __init__(self, similarity_search_data):\n",
    "\n",
    "            search_engine = similarity_search_data[\n",
    "                Consts.LigandSimilaritySearch.Properties.SEARCH_ENGINE.value\n",
    "            ]\n",
    "            self.search_engine = (\n",
    "                Specs.LigandSimilaritySearchDefaults.SEARCH_ENGINE\n",
    "                if search_engine is np.nan\n",
    "                else Consts.LigandSimilaritySearch.SearchEngines(search_engine)\n",
    "            )\n",
    "\n",
    "            min_similarity_percent = similarity_search_data[\n",
    "                Consts.LigandSimilaritySearch.Properties.MIN_SIMILARITY_PERCENT.value\n",
    "            ]\n",
    "            self.min_similarity_percent = (\n",
    "                Specs.LigandSimilaritySearchDefaults.MIN_SIMILARITY_PERCENT\n",
    "                if min_similarity_percent is np.nan\n",
    "                else min_similarity_percent\n",
    "            )\n",
    "\n",
    "            max_num_results = similarity_search_data[\n",
    "                Consts.LigandSimilaritySearch.Properties.MAX_NUM_RESULTS.value\n",
    "            ]\n",
    "            self.max_num_results = (\n",
    "                Specs.LigandSimilaritySearchDefaults.MAX_NUM_RESULTS\n",
    "                if max_num_results is np.nan\n",
    "                else max_num_results\n",
    "            )\n",
    "\n",
    "            max_num_druglike = int(\n",
    "                similarity_search_data[\n",
    "                    Consts.LigandSimilaritySearch.Properties.MAX_NUM_DRUGLIKE.value\n",
    "                ]\n",
    "            )\n",
    "            self.max_num_druglike = (\n",
    "                Specs.LigandSimilaritySearchDefaults.MAX_NUM_DRUGLIKE\n",
    "                if max_num_druglike is np.nan\n",
    "                else max_num_druglike\n",
    "            )\n",
    "\n",
    "    class Docking:\n",
    "        def __init__(self, docking_data):\n",
    "\n",
    "            program = docking_data[Consts.Docking.Properties.PROGRAM.value]\n",
    "            self.program = (\n",
    "                Specs.DockingDefaults.PROGRAM\n",
    "                if program is np.nan\n",
    "                else Consts.Docking.Programs(program)\n",
    "            )\n",
    "\n",
    "            num_poses_per_ligand = docking_data[\n",
    "                Consts.Docking.Properties.NUM_POSES_PER_LIGAND.value\n",
    "            ]\n",
    "            self.num_poses_per_ligand = (\n",
    "                Specs.DockingDefaults.NUM_POSES_PER_LIGAND.value\n",
    "                if num_poses_per_ligand is np.nan\n",
    "                else num_poses_per_ligand\n",
    "            )\n",
    "\n",
    "            exhaustiveness = docking_data[Consts.Docking.Properties.EXHAUSTIVENESS.value]\n",
    "            self.exhaustiveness = (\n",
    "                Specs.DockingDefaults.EXHAUSTIVENESS.value\n",
    "                if exhaustiveness is np.nan\n",
    "                else exhaustiveness\n",
    "            )\n",
    "\n",
    "            random_seed = docking_data[Consts.Docking.Properties.RANDOM_SEED.value]\n",
    "            self.random_seed = (\n",
    "                Specs.DockingDefaults.RANDOM_SEED.value if random_seed is np.nan else random_seed\n",
    "            )\n",
    "\n",
    "    class InteractionAnalysis:\n",
    "        def __init__(self, interaction_analysis_data):\n",
    "\n",
    "            program = interaction_analysis_data[\n",
    "                Consts.InteractionAnalysis.Properties.PROGRAM.value\n",
    "            ]\n",
    "            self.program = (\n",
    "                Specs.InteractionAnalysisDefaults.PROGRAM\n",
    "                if program is np.nan\n",
    "                else Consts.InteractionAnalysis.Programs(program)\n",
    "            )\n",
    "\n",
    "    class OptimizedLigands:\n",
    "        def __init__(self, optimized_ligand_data):\n",
    "            num_results = optimized_ligand_data[\n",
    "                Consts.OptimizedLigands.Properties.NUM_RESULTS.value\n",
    "            ]\n",
    "            self.num_results = (\n",
    "                Specs.OptimizedLigandsDefaults.NUM_RESULTS.value\n",
    "                if num_results is np.nan\n",
    "                else int(num_results)\n",
    "            )\n",
    "\n",
    "            selection_method = optimized_ligand_data[\n",
    "                Consts.OptimizedLigands.Properties.SELECTION_METHOD.value\n",
    "            ]\n",
    "            self.selection_method = (\n",
    "                Specs.OptimizedLigandsDefaults.SELECTION_METHOD\n",
    "                if selection_method is np.nan\n",
    "                else Consts.OptimizedLigands.SelectionMethods(selection_method)\n",
    "            )\n",
    "\n",
    "            if self.selection_method is Consts.OptimizedLigands.SelectionMethods.SORTING:\n",
    "                selection_criteria = optimized_ligand_data[\n",
    "                    Consts.OptimizedLigands.Properties.SELECTION_CRITERIA.value\n",
    "                ]\n",
    "                if selection_criteria is np.nan:\n",
    "                    self.selection_criteria = (\n",
    "                        Specs.OptimizedLigandsDefaults.SELECTION_CRITERIA_SORTING.value\n",
    "                    )\n",
    "                else:\n",
    "                    # pass the column names through the SelectionCriteria enumeration class to make sure they are valid\n",
    "                    self.selection_criteria = [\n",
    "                        Consts.OptimizedLigands.SelectionCriteria(criterion.strip()).value\n",
    "                        for criterion in selection_criteria.split(\",\")\n",
    "                    ]\n",
    "\n",
    "            elif self.selection_method is Consts.OptimizedLigands.SelectionMethods.FUNCTION:\n",
    "                selection_criteria = optimized_ligand_data[\n",
    "                    Consts.OptimizedLigands.Properties.SELECTION_CRITERIA.value\n",
    "                ]\n",
    "                self.selection_criteria = (\n",
    "                    Specs.OptimizedLigandsDefaults.SELECTION_CRITERIA_FUNCTION.value\n",
    "                    if selection_criteria is np.nan\n",
    "                    else selection_criteria\n",
    "                )\n",
    "\n",
    "    class OutputPaths:\n",
    "        \"\"\"\n",
    "        Data-class containing all the output paths for different parts of the pipeline.\n",
    "        Take a main output path, and create all required parent folders,\n",
    "        as well as sub-folders for each part of the pipeline.\n",
    "        \"\"\"\n",
    "\n",
    "        def __init__(self, output_path):\n",
    "            self.root = Path(output_path)\n",
    "            for folder_name in Consts.Output.FolderNames:\n",
    "                folder_path = IO.create_folder(folder_name.value, output_path)\n",
    "                setattr(self, folder_name.name.lower(), folder_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Processing The Input Data: Instantiating The ***Specs*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "We can now instantiate the <b><i>Specs</i></b> class by specifying the filepath of the input CSV file, and the output path for storing the output data. As discussed, we will assign the created instance to our project, just to have everything organized in one place .</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Specs = Specs(\n",
    "    input_data_filepath=\"data/PipelineInputData_Project1.csv\",\n",
    "    output_data_root_folder_path=\"data/Outputs/\" + project1.name,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\"> \n",
    "All the available data in the input CSV file of our project are now internalized in the project's <b><i>Specs</i></b> instance, and can be accessed using the corresponding instance attributes. Some few examples are shown below.\n",
    "    <br><br>\n",
    "    <b><i>Note:</i></b> Another advantage of storing the data as instance attributes and assigning them to <b><i>project1</i></b> is that everywhere in the code we can directly see all the project's data and know how to access them; Just write <span style=\"background-color: #D3D3D3\">project1.</span> and press the <i>tab</i> button. Code completion will then display a list of all available options to choose from. In contrast, for example storing the data in a dataframe requires you to have the dataframe always in front of you to know the exact adress of each specific data you may need to use in another part of the code.  \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Specs.Protein.input_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Specs.Ligand.input_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project1.Specs.Docking.num_poses_per_ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Specs.OutputPaths.binding_site_detection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is also possible to see the raw input data in its entirety:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Specs.RawData.all_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='3_practical'></a>\n",
    "### Processing the Input Protein Data \n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "As mentioned above, it is a good idea to use the input protein data to make a <b><i>Protein</i></b> object, which would have its own attributes and methods. In order to implement this, we first need to define some functions for processing protein data, such as fetching PDB files from the PDB webserver, parsing PDB files to extract useful data etc. For this purpose, we first define a helper-class called <b><i>PDB</i></b>, which implements these functionalities. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***PDB*** Helper-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "This helper-class contains all the necessary functions for handling protein data, especially via processing PDB files. <br><br><b><i>Note:</i></b> For demonstration of each function, see the <a href=\"#pdb_demo\">corresponding section</a> in <i>Supplementary Information</i>.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PDB:\n",
    "    \"\"\"\n",
    "    Set of functions for working with PDB files.\n",
    "    \"\"\"\n",
    "\n",
    "    @staticmethod\n",
    "    def read_pdb_file_content(input_type, input_value):\n",
    "        \"\"\"\n",
    "        Read the content of a PDB file either from a local path or via fetching the file from PDB webserver.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_type : str\n",
    "            Either 'pdb_code' or 'pdb_filepath'.\n",
    "\n",
    "        input_value : str\n",
    "            Either a valid PDB-code, or a local filepath of a PDB file.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        str\n",
    "            Content of the PDB file as a single string.\n",
    "        \"\"\"\n",
    "        if input_type == \"pdb_code\":\n",
    "            pdb_file_content = pypdb.get_pdb_file(input_value)\n",
    "        elif input_type == \"pdb_filepath\":\n",
    "            with open(input_value) as f:\n",
    "                pdb_file_content = f.read()\n",
    "        return pdb_file_content\n",
    "\n",
    "    @staticmethod\n",
    "    def fetch_and_save_pdb_file(pdb_code, output_filepath):\n",
    "        \"\"\"\n",
    "        Fetch a PDB file from the PDB webserver and save locally.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pdb_code : str\n",
    "            PDB-code of the protein structure.\n",
    "\n",
    "        output_filepath : str or pathlib.Path object\n",
    "            Local file path (including file name, but not the extension) to save the PDB file in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        pathlib.Path object\n",
    "            The full path of the saved PDB file.\n",
    "        \"\"\"\n",
    "        pdb_file_content = pypdb.get_pdb_file(pdb_code)\n",
    "        full_filepath = str(output_filepath) + \".pdb\"\n",
    "        with open(full_filepath, \"w\") as f:\n",
    "            f.write(pdb_file_content)\n",
    "        return Path(full_filepath)\n",
    "\n",
    "    @staticmethod\n",
    "    def extract_molecule_from_pdb_file(molecule_name, input_filepath, output_filepath):\n",
    "        \"\"\"\n",
    "        Extract a specific molecule (i.e. the protein or a ligand)\n",
    "        from a local PDB file and save as a new PDB file in a given path.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        molecule_name : str\n",
    "            Name of the molecule to be extracted.\n",
    "            For the protein, enter 'protein'. For a ligand, enter the ligand-ID.\n",
    "\n",
    "        input_filepath : str or pathlib.Path object\n",
    "            Local path of the original PDB file.\n",
    "\n",
    "        output_filepath : str or pathlib.Path object\n",
    "            Local file path (including file name) to save the PDB file of the extracted molecule in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            <Universe> Structure object\n",
    "            Structure object of the extracted molecule.\n",
    "        \"\"\"\n",
    "\n",
    "        pdb_structure = Structure.from_string(input_filepath)\n",
    "        molecule_name = f\"resname {molecule_name}\" if molecule_name != \"protein\" else molecule_name\n",
    "        extracted_structure = pdb_structure.select_atoms(molecule_name)\n",
    "        extracted_structure.write(output_filepath)\n",
    "        return extracted_structure\n",
    "\n",
    "    @staticmethod\n",
    "    def load_pdb_file_as_dataframe(pdb_file_text_content):\n",
    "        \"\"\"\n",
    "        Transform the textual content of a PDB file into a dictionary of Pandas DataFrames.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pdb_file_text_content : str\n",
    "            Textual content of a PDB file as a single string.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Dictionary of Pandas DataFrames.\n",
    "        The dictionary has four entries with following keys: 'ATOM', 'HETATM', 'ANISOU' and 'OTHERS'.\n",
    "        Each value is a Pandas DataFrame corresponding to the specific information described by the key.\n",
    "        \"\"\"\n",
    "        ppdb = PandasPdb()\n",
    "        pdb_df = ppdb._construct_df(pdb_file_text_content.splitlines(True))\n",
    "        # TODO: Change _construct_df to read_pdb_from_lines once biopandas\n",
    "        # cuts a new release (currently: 0.2.7), see https://github.com/rasbt/biopandas/pull/72\n",
    "        return pdb_df\n",
    "\n",
    "    @staticmethod\n",
    "    def extract_info_from_pdb_file_content(pdb_file_text_content):\n",
    "        \"\"\"\n",
    "        Extract some useful information from the contents of a PDB file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pdb_file_text_content : str\n",
    "            Textual content of a PDB file as a single string.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            dict\n",
    "            Dictionary of the successfully extracted information.\n",
    "            Possible keys are:\n",
    "                'structure_title' : str\n",
    "                    Title of the PDB structure.\n",
    "                'name' : str\n",
    "                    Name of the protein.\n",
    "                'chains' : list of strings\n",
    "                    List of chain-IDs of the available chains in the protein.\n",
    "                'ligands' : list of lists of strings\n",
    "                    List of ligand information: [ligand-ID, chain-ID+residue number, number of heavy atoms]\n",
    "        \"\"\"\n",
    "\n",
    "        pdb_content = pdb_file_text_content.strip().split(\"\\n\")\n",
    "        for index in range(len(pdb_content)):\n",
    "            pdb_content[index] = pdb_content[index].split(\" \", 1)\n",
    "            try:\n",
    "                pdb_content[index][1] = pdb_content[index][1].strip()\n",
    "                if pdb_content[index][1][-1] == \";\":\n",
    "                    pdb_content[index][1] = pdb_content[index][1][:-1]\n",
    "                if (\n",
    "                    pdb_content[index][0] in [\"REMARK\", \"COMPND\"]\n",
    "                    and pdb_content[index][1][0].isdigit()\n",
    "                ):\n",
    "                    try:\n",
    "                        pdb_content[index][1] = pdb_content[index][1].split(\" \", 1)[1]\n",
    "                    except:\n",
    "                        pass\n",
    "            except:\n",
    "                pdb_content[index].append(\" \")\n",
    "\n",
    "        info = {}\n",
    "        ligands = []\n",
    "        for content in pdb_content:\n",
    "            if content[0] == \"TITLE\":\n",
    "                info[\"Structure Title\"] = content[1]\n",
    "            if content[0] == \"COMPND\" and content[1].startswith(\"MOLECULE: \"):\n",
    "                info[\"Name\"] = content[1].split(\"MOLECULE: \")[1]\n",
    "            if content[0] == \"COMPND\" and content[1].startswith(\"CHAIN: \"):\n",
    "                info[\"Chains\"] = content[1].split(\"CHAIN: \")[1].split(\", \")\n",
    "            if content[0] == \"HET\":\n",
    "                lig = list(filter(lambda x: x != \"\", content[1].split(\" \")))\n",
    "                lig[-1] = int(lig[-1])\n",
    "                ligands.append(lig)\n",
    "        info[\"Ligands\"] = ligands\n",
    "        return info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***NGLView*** Helper-Class\n",
    "<br>\n",
    "We also implement another helper-class, containing all the necessary functions for visualizing protein-related data, such as the protein structure, and later the protein binding site, docking poses of ligands and the protein-ligand interactions present in them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NGLView:\n",
    "    @staticmethod\n",
    "    def protein(input_type, input_value, output_image_filename=None):\n",
    "        \"\"\"\n",
    "        Visualize the protein.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_type : str\n",
    "            Either \"pdb_code\" or a file extension e.g. \"pdb\".\n",
    "        input_value: str or pathlib.Path object\n",
    "            Either the PDB-code of the protein, or a local filepath.\n",
    "        output_image_filename : str (optional; default: None)\n",
    "            Filename to save a static image of the protein.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLViewer object\n",
    "            Interactive NGL viewer of the given Protein\n",
    "            and (if available) its co-crystallized ligand.\n",
    "        \"\"\"\n",
    "\n",
    "        if input_type == \"pdb_code\":\n",
    "            viewer = nv.show_pdbid(input_value, height=\"600px\")\n",
    "        else:\n",
    "            with open(input_value) as f:\n",
    "                viewer = nv.show_file(\n",
    "                    f, ext=input_type, height=\"600px\", default_representation=False\n",
    "                )\n",
    "                viewer.add_representation(\"cartoon\", selection=\"protein\")\n",
    "\n",
    "        viewer.add_representation(repr_type=\"ball+stick\", selection=\"hetero and not water\")\n",
    "        viewer.center(\"protein\")\n",
    "\n",
    "        if output_image_filename != None:\n",
    "            viewer.render_image(trim=True, factor=2)\n",
    "            viewer._display_image()\n",
    "            viewer.download_image(output_image_filename)\n",
    "        return viewer\n",
    "\n",
    "    @staticmethod\n",
    "    def binding_site(protein_input_type, protein_input_value, ccp4_filepath):\n",
    "        \"\"\"\n",
    "        3D visualization of a binding pocket using a CCP4 file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        protein_input_type : str\n",
    "            Either \"pdb_code\" or a file extension e.g. \"pdb\".\n",
    "        protein_input_value: str or pathlib.Path object\n",
    "            Either the PDB-code of the protein, or a local filepath.\n",
    "        ccp4_filepath : str\n",
    "            Local file path of the output of the Binding Site Detection.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        NGL viewer that visualizes the selected pocket at its respective position.\n",
    "        \"\"\"\n",
    "        viewer = NGLView.protein(protein_input_type, protein_input_value)\n",
    "        with open(ccp4_filepath, \"rb\") as f:\n",
    "            viewer.add_component(f, ext=\"ccp4\")\n",
    "        viewer.center()\n",
    "\n",
    "        return viewer\n",
    "\n",
    "    @staticmethod\n",
    "    def docking(\n",
    "        protein_filepath,\n",
    "        protein_file_extension,\n",
    "        list_docking_poses_filepaths,\n",
    "        docking_poses_file_extension,\n",
    "        list_docking_poses_labels,\n",
    "        list_docking_poses_affinities,\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Visualize a list of docking poses\n",
    "        in the protein structure, using NGLView.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        protein_filepath : str or pathlib.Path object\n",
    "            Filepath of the extracted protein structure used in docking experiment.\n",
    "        protein_file_extension : str\n",
    "            File extension of the protein file, e.g. \"pdb\", \"pdbqt\" etc.\n",
    "        list_docking_poses_filepaths : list of strings/pathlib.Path objects\n",
    "            List of filepaths for the separated docking poses.\n",
    "        docking_poses_file_extension : str\n",
    "            File extension of the docking-pose files, e.g. \"pdb\", \"pdbqt\" etc.\n",
    "        list_docking_poses_labels : list of strings\n",
    "            List of labels for docking poses to be used for the selection menu.\n",
    "        list_docking_poses_affinities : list of strings/numbers\n",
    "            List of binding affinities in kcal/mol, to be viewed for each docking pose.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLView viewer\n",
    "            Interactive viewer containing the protein structure and all docking poses,\n",
    "            with menu to select between docking poses.\n",
    "        \"\"\"\n",
    "\n",
    "        # JavaScript code needed to update residues around the ligand\n",
    "        # because this part is not exposed in the Python widget\n",
    "        # Based on: http://nglviewer.org/ngl/api/manual/snippets.html\n",
    "        _RESIDUES_AROUND = \"\"\"\n",
    "        var protein = this.stage.compList[0];\n",
    "        var ligand_center = this.stage.compList[{index}].structure.atomCenter();\n",
    "        var around = protein.structure.getAtomSetWithinPoint(ligand_center, {radius});\n",
    "        var around_complete = protein.structure.getAtomSetWithinGroup(around);\n",
    "        var last_repr = protein.reprList[protein.reprList.length-1];\n",
    "        protein.removeRepresentation(last_repr);\n",
    "        protein.addRepresentation(\"licorice\", {{sele: around_complete.toSeleString()}});\n",
    "        \"\"\"\n",
    "        print(\"Docking modes\")\n",
    "        print(\"(CID - mode)\")\n",
    "        # Create viewer widget\n",
    "        viewer = nv.NGLWidget(height=\"860px\")\n",
    "        viewer.add_component(protein_filepath, ext=protein_file_extension)\n",
    "        # viewer.add_representation(\"cartoon\", selection=\"protein\")\n",
    "        # Select first atom in molecule (@0) so it holds the affinity label\n",
    "        label_kwargs = dict(\n",
    "            labelType=\"text\",\n",
    "            sele=\"@0\",\n",
    "            showBackground=True,\n",
    "            backgroundColor=\"black\",\n",
    "        )\n",
    "        list_docking_poses_affinities = list(\n",
    "            map(lambda x: str(x) + \" kcal/mol\", list_docking_poses_affinities)\n",
    "        )\n",
    "        for docking_pose_filepath, ligand_label in zip(\n",
    "            list_docking_poses_filepaths, list_docking_poses_affinities\n",
    "        ):\n",
    "            ngl_ligand = viewer.add_component(\n",
    "                docking_pose_filepath, ext=docking_poses_file_extension\n",
    "            )\n",
    "            ngl_ligand.add_label(labelText=[str(ligand_label)], **label_kwargs)\n",
    "\n",
    "        # Create selection widget\n",
    "        #   Options is a list of (text, value) tuples. When we click on select, the value will be passed\n",
    "        #   to the callable registered in `.observe(...)`\n",
    "        selector = Select(\n",
    "            options=[(label, i) for (i, label) in enumerate(list_docking_poses_labels, 1)],\n",
    "            description=\"\",\n",
    "            rows=len(list_docking_poses_filepaths)\n",
    "            if len(list_docking_poses_filepaths) <= 52\n",
    "            else 52,\n",
    "            layout=Layout(flex=\"flex-grow\", width=\"auto\"),\n",
    "        )\n",
    "\n",
    "        # Arrange GUI elements\n",
    "        # The selection box will be on the left, the viewer will occupy the rest of the window\n",
    "        display(AppLayout(left_sidebar=selector, center=viewer, pane_widths=[1, 6, 1]))\n",
    "\n",
    "        # This is the event handler - action taken when the user clicks on the selection box\n",
    "        # We need to define it here so it can \"see\" the viewer variable\n",
    "        def _on_selection_change(change):\n",
    "            # Update only if the user clicked on a different entry\n",
    "            if change[\"name\"] == \"value\" and (change[\"new\"] != change[\"old\"]):\n",
    "                viewer.hide(\n",
    "                    list(range(1, len(list_docking_poses_filepaths) + 1))\n",
    "                )  # Hide all ligands (components 1-n)\n",
    "                component = getattr(viewer, f\"component_{change['new']}\")\n",
    "                component.show()  # Display the selected one\n",
    "                component.center(500)  # Zoom view\n",
    "                # Call the JS code to show sidechains around ligand\n",
    "                viewer._execute_js_code(_RESIDUES_AROUND.format(index=change[\"new\"], radius=6))\n",
    "\n",
    "        # Register event handler\n",
    "        selector.observe(_on_selection_change)\n",
    "        # Trigger event manually to focus on the first solution\n",
    "        _on_selection_change({\"name\": \"value\", \"new\": 1, \"old\": None})\n",
    "        return viewer\n",
    "\n",
    "    @staticmethod\n",
    "    def interactions(\n",
    "        protein_filepath,\n",
    "        protein_file_extension,\n",
    "        list_docking_poses_filepaths,\n",
    "        docking_poses_file_extension,\n",
    "        list_docking_poses_labels,\n",
    "        list_docking_poses_affinities,\n",
    "        list_docking_poses_plip_dicts,\n",
    "    ):\n",
    "\n",
    "        color_map = {\n",
    "            \"hydrophobic\": [0.90, 0.10, 0.29],\n",
    "            \"hbond\": [0.26, 0.83, 0.96],\n",
    "            \"waterbridge\": [1.00, 0.88, 0.10],\n",
    "            \"saltbridge\": [0.67, 1.00, 0.76],\n",
    "            \"pistacking\": [0.75, 0.94, 0.27],\n",
    "            \"pication\": [0.27, 0.60, 0.56],\n",
    "            \"halogen\": [0.94, 0.20, 0.90],\n",
    "            \"metal\": [0.90, 0.75, 1.00],\n",
    "        }\n",
    "\n",
    "        # Create selection widget\n",
    "        # Options is a list of (text, value) tuples.\n",
    "        # When we click on select, the value will be passed\n",
    "        # to the callable registered in `.observe(...)`\n",
    "        selector = Select(\n",
    "            options=[(label, i) for (i, label) in enumerate(list_docking_poses_labels, 1)],\n",
    "            description=\"\",\n",
    "            rows=len(list_docking_poses_filepaths)\n",
    "            if len(list_docking_poses_filepaths) <= 52\n",
    "            else 52,\n",
    "            layout=Layout(flex=\"flex-grow\", width=\"auto\"),\n",
    "        )\n",
    "\n",
    "        # Arrange GUI elements\n",
    "        # The selection box will be on the left,\n",
    "        # the viewer will occupy the rest of the window (but it will be added later)\n",
    "        app = AppLayout(\n",
    "            left_sidebar=selector,\n",
    "            center=None,\n",
    "            pane_widths=[1, 6, 1],\n",
    "            height=\"860px\",\n",
    "        )\n",
    "\n",
    "        # Show color-map\n",
    "        fig, axs = plt.subplots(nrows=2, ncols=4, figsize=(12, 1))\n",
    "        plt.subplots_adjust(hspace=1)\n",
    "        fig.suptitle(\"Color-map of interactions\", size=10, y=1.3)\n",
    "        for ax, (interaction, color) in zip(fig.axes, color_map.items()):\n",
    "            ax.imshow(np.zeros((1, 5)), cmap=colors.ListedColormap(color_map[interaction]))\n",
    "            ax.set_title(interaction, loc=\"center\", fontsize=10)\n",
    "            ax.set_axis_off()\n",
    "        plt.show()\n",
    "\n",
    "        list_docking_poses_affinities = list(\n",
    "            map(lambda x: str(x) + \" kcal/mol\", list_docking_poses_affinities)\n",
    "        )\n",
    "\n",
    "        # This is the event handler - action taken when the user clicks on the selection box\n",
    "        # We need to define it here so it can \"see\" the viewer variable\n",
    "        def _on_selection_change(change):\n",
    "            # Update only if the user clicked on a different entry\n",
    "            if change[\"name\"] == \"value\" and (change[\"new\"] != change[\"old\"]):\n",
    "                if app.center is not None:\n",
    "                    app.center.close()\n",
    "\n",
    "                # NGL Viewer\n",
    "                app.center = viewer = nv.NGLWidget(height=\"860px\", default=True, gui=True)\n",
    "                prot_component = viewer.add_component(\n",
    "                    protein_filepath, ext=protein_file_extension, default_representation=False\n",
    "                )  # add protein\n",
    "                prot_component.add_representation(\"cartoon\")\n",
    "                time.sleep(1)\n",
    "\n",
    "                label_kwargs = dict(\n",
    "                    labelType=\"text\",\n",
    "                    sele=\"@0\",\n",
    "                    showBackground=True,\n",
    "                    backgroundColor=\"black\",\n",
    "                )\n",
    "                lig_component = viewer.add_component(\n",
    "                    list_docking_poses_filepaths[change[\"new\"]], ext=docking_poses_file_extension\n",
    "                )  # add selected ligand\n",
    "                lig_component.add_label(\n",
    "                    labelText=[str(list_docking_poses_affinities[change[\"new\"]])], **label_kwargs\n",
    "                )\n",
    "                time.sleep(1)\n",
    "                lig_component.center(duration=500)\n",
    "\n",
    "                # Add interactions\n",
    "                interactions = list_docking_poses_plip_dicts[change[\"new\"]]\n",
    "\n",
    "                interacting_residues = []\n",
    "\n",
    "                for interaction_type, interaction_list in interactions.items():\n",
    "                    color = color_map[interaction_type]\n",
    "                    if len(interaction_list) == 1:\n",
    "                        continue\n",
    "                    df_interactions = pd.DataFrame.from_records(\n",
    "                        interaction_list[1:], columns=interaction_list[0]\n",
    "                    )\n",
    "                    for _, interaction in df_interactions.iterrows():\n",
    "                        name = interaction_type\n",
    "                        viewer.shape.add_cylinder(\n",
    "                            interaction[\"LIGCOO\"],\n",
    "                            interaction[\"PROTCOO\"],\n",
    "                            color,\n",
    "                            [0.1],\n",
    "                            name,\n",
    "                        )\n",
    "                        interacting_residues.append(interaction[\"RESNR\"])\n",
    "                # Display interacting residues\n",
    "                res_sele = \" or \".join([f\"({r} and not _H)\" for r in interacting_residues])\n",
    "                res_sele_nc = \" or \".join(\n",
    "                    [f\"({r} and ((_O) or (_N) or (_S)))\" for r in interacting_residues]\n",
    "                )\n",
    "\n",
    "                prot_component.add_ball_and_stick(\n",
    "                    sele=res_sele, colorScheme=\"chainindex\", aspectRatio=1.5\n",
    "                )\n",
    "                prot_component.add_ball_and_stick(\n",
    "                    sele=res_sele_nc, colorScheme=\"element\", aspectRatio=1.5\n",
    "                )\n",
    "\n",
    "        # Register event handler\n",
    "        selector.observe(_on_selection_change)\n",
    "        # Trigger event manually to focus on the first solution\n",
    "        _on_selection_change({\"name\": \"value\", \"new\": 1, \"old\": None})\n",
    "        return app"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Pipeline's ***Protein*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> Now using the functionalities defined in the helper-classes, we can implement the <b><i>Protein</i></b> class of the pipeline, which takes in the protein input data and creates an object with extended attributes and methods.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class Protein:\n",
    "    \"\"\"\n",
    "    Protein object with properties as attributes and methods to visualize and work with the protein.\n",
    "    Take a protein identifier type and corresponding value,\n",
    "    and create a Protein object, while assigning some properties as attributes.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    identifier_type : enum 'InputTypes' from the 'Consts.Protein' class\n",
    "        Type of the protein identifier, e.g. InputTypes.PDB_CODE.\n",
    "    indentifier_value : str\n",
    "        Value of the protein identifier, e.g. its PDB-code.\n",
    "    protein_output_path : str or pathlib.Path object\n",
    "        Output path of the project for protein data.\n",
    "    \"\"\"\n",
    "\n",
    "    class Consts:\n",
    "        # Available properties that are assigned as instance attributes upon instantiation.\n",
    "        class Properties(Enum):\n",
    "            STRUCTURE_TITLE = \"Structure Title\"\n",
    "            NAME = \"Name\"\n",
    "            CHAINS = \"Chains\"\n",
    "            LIGANDS = \"Ligands\"\n",
    "            RESIDUE_NUMBER_FIRST = \"First Residue Number\"\n",
    "            RESIDUE_NUMBER_LAST = \"Last Residue Number\"\n",
    "            RESIDUES_LENGTH = \"Number of Residues\"\n",
    "\n",
    "    def __init__(self, identifier_type, identifier_value, protein_output_path):\n",
    "\n",
    "        setattr(self, identifier_type.name.lower(), identifier_value)\n",
    "\n",
    "        self.file_content = PDB.read_pdb_file_content(identifier_type.value, identifier_value)\n",
    "\n",
    "        dict_of_dataframes = PDB.load_pdb_file_as_dataframe(self.file_content)\n",
    "        for key, value in dict_of_dataframes.items():\n",
    "            setattr(self, f\"dataframe_PDBcontent_{key.lower()}\", value)\n",
    "\n",
    "        self.residue_number_first = self.dataframe_PDBcontent_atom.iloc[0][\"residue_number\"]\n",
    "        self.residue_number_last = self.dataframe_PDBcontent_atom.iloc[-1][\"residue_number\"]\n",
    "        self.residues_length = self.residue_number_last - self.residue_number_first + 1\n",
    "\n",
    "        protein_info = PDB.extract_info_from_pdb_file_content(self.file_content)\n",
    "\n",
    "        for protein_property in self.Consts.Properties:\n",
    "            if protein_property.value in protein_info:\n",
    "                setattr(\n",
    "                    self,\n",
    "                    protein_property.name.lower(),\n",
    "                    protein_info[protein_property.value],\n",
    "                )\n",
    "\n",
    "        if identifier_type is Consts.Protein.InputTypes.PDB_CODE:\n",
    "            self.pdb_filepath = PDB.fetch_and_save_pdb_file(\n",
    "                identifier_value, str(protein_output_path) + \"/\" + identifier_value\n",
    "            )\n",
    "\n",
    "    def __call__(self):\n",
    "        for protein_property in self.Consts.Properties:\n",
    "            if hasattr(self, protein_property.name.lower()):\n",
    "                display(\n",
    "                    Markdown(\n",
    "                        f\"<span style='color:black'>&nbsp;&nbsp;&nbsp;&nbsp;{protein_property.value}: </span><span style='color:black'>**{getattr(self, protein_property.name.lower())}**</span>\"\n",
    "                    )\n",
    "                )\n",
    "        if hasattr(self, \"pdb_code\"):\n",
    "            viewer = NGLView.protein(\"pdb_code\", self.pdb_code)\n",
    "        else:\n",
    "            viewer = NGLView.protein(\"pdb\", self.pdb_filepath)\n",
    "\n",
    "        return viewer\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f\"<Protein: {self.name}>\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Creating a ***Protein*** Object From The Protein Input Data: Instantiating The ***Protein*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "    We now create an instance of the <b><i>Protein</i></b> class, by inputting the protein data of our project, i.e. the protein input type, the corresponding input value, and the output path of the project for storing the protein data. We then assign this instance to our project:</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Protein = Protein(\n",
    "    identifier_type=project1.Specs.Protein.input_type,\n",
    "    identifier_value=project1.Specs.Protein.input_value,\n",
    "    protein_output_path=project1.Specs.OutputPaths.protein,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have implemented a ***\\_\\_call\\_\\_*** method for our ***Protein***, which prints out some useful information and visualizes the protein's structure, simply by calling the object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Protein()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\"> \n",
    "All of these information and some other properties are also stored separately as instance attributes. For example, a list of information on all co-crystallized ligands:\n",
    "<br><br>\n",
    "    <b><i>Note:</i></b> Each sub-list corresponds to one ligand, where the first entry is the ligand-ID, the second entry is the chain-ID (to which the ligand is bound) followed by the ligand residue number, and the third and last entry is the number of heavy atoms in the ligand. For example, here the first ligand has the ID 'W32', is on chain 'A' at residue number '1101', and has 39 heavy atoms.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Protein.ligands"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Moreover, when the protein is inputted by its PDB-code, the PDB file will also be automatically downloaded and stored in the defined output path for the protein output data. The full path is also accessible via the attribute ***pdb_filepath***: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Protein.pdb_filepath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='4_practical'></a>\n",
    "### Processing the Input Ligand Data\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "    Similar to the protein, we can also use the ligand input data to create a <b><i>Ligand</i></b> object with some extended attributes and methods. For this, we first need to define a helper-class to be able to obtain some new information on the input ligand, and another helper-class for calculating ligand's properties, creating 3D conformations, visualizations etc. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***PubChem*** Helper-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Here, we define a class called <b><i>PubChem</i></b>, which contains all the necessary functions to use the <i>PubChem</i> web-service APIs, in order to obtain new information on ligands, such as other identifies (e.g. trivial name, IUPAC name, SMILES, CID, InChI and InChIKey), physiochemical properties, descriptions etc. <i>PubChem</i> has also the ability to perform similarity searches on a given ligand, which we will also implement here, and use later in the <b><i>LigandSimilaritySearch</i></b> class of the pipeline.<br><br><b><i>Note:</i></b> For demonstration of each function, see the <a href=\"#pubchem_demo\">corresponding section</a> in <i>Supplementary Information</i>.  </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PubChem:\n",
    "    \"\"\"\n",
    "    Implementation of the functionalities of PubChem PUG REST API.\n",
    "    \"\"\"\n",
    "\n",
    "    # -----------------------------------------------------------------------------\n",
    "    # Constants for API requests\n",
    "    class APIConsts:\n",
    "        \"\"\"\n",
    "        Constants for API requests.\n",
    "        Request URLs should have the format:\n",
    "            APIConsts.URLs.PROLOG + APIConsts.URLs.Inputs.<type>.value + ...\n",
    "            ... <input_value> + APIConsts.URLs.Operations.GET_<property>.value + ...\n",
    "            ... APIConsts.URLs.Outputs.<type>.value + <?optional parameters>\n",
    "        \"\"\"\n",
    "\n",
    "        class URLs:\n",
    "            PROLOG = \"https://pubchem.ncbi.nlm.nih.gov/rest/pug/\"\n",
    "\n",
    "            class Inputs(Enum):\n",
    "                CID = \"compound/cid/\"\n",
    "                NAME = \"compound/name/\"\n",
    "                SMILES = \"compound/smiles/\"\n",
    "                INCHI = \"compound/inchi/\"\n",
    "                INCHIKEY = \"compound/inchikey/\"\n",
    "                SIMILARITY_FROM_SMILES = \"compound/similarity/smiles/\"\n",
    "                SIMILARITY_RESULTS = \"compound/listkey/\"\n",
    "\n",
    "            class Operations(Enum):\n",
    "                GET_CID = \"/cids/\"\n",
    "                GET_NAME = \"/property/title/\"\n",
    "                GET_SMILES = \"/property/CanonicalSMILES/\"\n",
    "                GET_INCHI = \"/property/InChI/\"\n",
    "                GET_INCHIKEY = \"/property/InChIKey/\"\n",
    "                GET_IUPAC_NAME = \"/property/IUPACName/\"\n",
    "                GET_DESCRIPTION = \"/description/\"\n",
    "                GET_RECORD = \"/record/\"\n",
    "\n",
    "            class Outputs(Enum):\n",
    "                TXT = \"TXT\"\n",
    "                JSON = \"JSON\"\n",
    "                PNG = \"PNG\"\n",
    "                CSV = \"CSV\"\n",
    "                SDF = \"SDF\"\n",
    "                XML = \"XML\"\n",
    "\n",
    "        class ResponseMsgs:\n",
    "            class SimilaritySearch(Enum):\n",
    "                JOBKEY_KEY1 = \"Waiting\"\n",
    "                JOBKEY_KEY2 = \"ListKey\"\n",
    "                RESULT_KEY1 = \"PropertyTable\"\n",
    "                RESULT_KEY2 = \"Properties\"\n",
    "\n",
    "            class GetRecords(Enum):\n",
    "                RESPONSE_KEY = \"PC_Compounds\"\n",
    "\n",
    "            class GetDescription(Enum):\n",
    "                RESPONSE_KEY1 = \"InformationList\"\n",
    "                RESPONSE_KEY2 = \"Information\"\n",
    "\n",
    "    # -----------------------------------------------------------------------------\n",
    "\n",
    "    @staticmethod\n",
    "    def send_request(partial_url, response_type=\"txt\", optional_params=\"\"):\n",
    "        \"\"\"\n",
    "        Send an API request to PubChem and get the response data.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        partial_url : str\n",
    "            The URL part of the request consisting of input-type, input-value and operation.\n",
    "            E.g. 'compound/cid/2244/property/CanonicalSMILES/' requests the SMILES of\n",
    "            the compound with an CID of 2244.\n",
    "        response_type : str (optional; default: txt)\n",
    "            Expected response-type of the API request.\n",
    "            Valid values are 'txt', 'json', 'png', 'csv' and 'sdf'.\n",
    "            Valid values are stored in: PubChem.APIConsts.URLs.Outputs\n",
    "        optional_params : str\n",
    "            The URL part of the request consisting of optional parameters.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            Datatype depends on the value of input parameter 'response_type'.\n",
    "            The response data of the API request.\n",
    "        \"\"\"\n",
    "        full_url = (\n",
    "            PubChem.APIConsts.URLs.PROLOG\n",
    "            + partial_url\n",
    "            + getattr(PubChem.APIConsts.URLs.Outputs, response_type.upper()).value\n",
    "            + f\"?{optional_params}\"\n",
    "        )\n",
    "        response = requests.get(full_url)\n",
    "        response.raise_for_status()\n",
    "        if response_type == \"txt\":\n",
    "            response_data = response.text\n",
    "        elif response_type == \"json\":\n",
    "            response_data = response.json()\n",
    "        else:\n",
    "            response_data = response.content\n",
    "        return response_data\n",
    "\n",
    "    @staticmethod\n",
    "    def convert_compound_identifier(\n",
    "        input_id_type, input_id_value, output_id_type, output_data_type=\"txt\"\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Convert an identifier to another identifier, e.g. CID to SMILES, SMILES to IUPAC-name etc.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_id_type : str\n",
    "            Type of the input identifier.\n",
    "            Valid values are: 'name', 'cid', 'smiles', 'inchi' and 'inchikey'.\n",
    "            Valid values are stored in: PubChem.APIConsts.URLs.Inputs\n",
    "        input_id_value : str, integer, list of strings or list of integers\n",
    "            Value of the input identifier.\n",
    "        output_id_type : str\n",
    "            Type of the ouput identifier.\n",
    "            Valid values are: 'name', 'cid', 'smiles', 'inchi', 'inchikey', 'iupac_name'.\n",
    "            Valid values are stored in: PubChem.APIConsts.URLs.Operations\n",
    "        output_data_type : str (optional; default: 'txt')\n",
    "            Datatype of the output data.\n",
    "            Valid values are 'txt', 'json', 'csv'.\n",
    "            A list of all valid values are stored in: PubChem.APIConsts.URLs.Outputs\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            Datatype depends on the value of input parameter 'response_type'\n",
    "            The response data of the API request.\n",
    "        \"\"\"\n",
    "        if isinstance(input_id_value, list):\n",
    "            input_id_value = \",\".join(map(str, input_id_value))\n",
    "        url = (\n",
    "            getattr(PubChem.APIConsts.URLs.Inputs, input_id_type.upper()).value\n",
    "            + str(input_id_value)\n",
    "            + getattr(PubChem.APIConsts.URLs.Operations, f\"GET_{output_id_type}\".upper()).value\n",
    "        )\n",
    "        response_data = PubChem.send_request(url, output_data_type)\n",
    "        if isinstance(input_id_value, list):\n",
    "            return response_data.strip().split(\"\\n\")\n",
    "        else:\n",
    "            return response_data.strip()\n",
    "\n",
    "    @staticmethod\n",
    "    def get_compound_record(input_id_type, input_id_value, output_data_type=\"json\"):\n",
    "        \"\"\"\n",
    "        Get a full record of all physiochemical properties of the compound.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_id_type : str\n",
    "            Type of the input identifier.\n",
    "            Valid values are: 'name', 'cid', 'smiles', 'inchi' and 'inchikey'.\n",
    "            Valid values are stored in: PubChem.APIConsts.URLs.Inputs\n",
    "        input_id_value : str or integer\n",
    "            Value of the input identifier.\n",
    "        output_data_type : str (optional; default: 'txt')\n",
    "            Datatype of the output data.\n",
    "            Valid values are 'txt', 'json', 'csv'.\n",
    "            A list of all valid values are stored in: PubChem.APIConsts.URLs.Outputs\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            dict\n",
    "            Dictionary keys are: 'id', 'atoms', 'bonds', 'coords', 'charge', 'props', 'count'\n",
    "        \"\"\"\n",
    "        url = (\n",
    "            getattr(PubChem.APIConsts.URLs.Inputs, input_id_type.upper()).value\n",
    "            + str(input_id_value)\n",
    "            + getattr(PubChem.APIConsts.URLs.Operations, \"GET_RECORD\").value\n",
    "        )\n",
    "        response_data = PubChem.send_request(url, output_data_type)[\n",
    "            PubChem.APIConsts.ResponseMsgs.GetRecords.RESPONSE_KEY.value\n",
    "        ][0]\n",
    "        return response_data\n",
    "\n",
    "    @staticmethod\n",
    "    def get_description_from_smiles(smiles, output_data_type=\"json\", printout=False):\n",
    "        \"\"\"\n",
    "        Get a textual description of a molecule, including its usage, properties, source etc.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        smiles : str\n",
    "            SMILES of the molecule.\n",
    "        output_data_type : str (optional; default: 'txt')\n",
    "            Datatype of the output data.\n",
    "            Valid values are 'txt', 'json', 'csv'.\n",
    "            A list of all valid values are stored in: PubChem.APIConsts.URLs.Outputs\n",
    "        printout : bool\n",
    "            Whether to print the descriptions, or return the data.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            list of dicts\n",
    "            When the parameter printout is set to False, the raw data is\n",
    "            returned as a list of dicts, where each element of the list\n",
    "            corresponds to a description from a specific source (e.g. a journal article).\n",
    "        \"\"\"\n",
    "        url = (\n",
    "            PubChem.APIConsts.URLs.Inputs.SMILES.value\n",
    "            + smiles\n",
    "            + PubChem.APIConsts.URLs.Operations.GET_DESCRIPTION.value\n",
    "        )\n",
    "        response_data = PubChem.send_request(url, output_data_type)[\n",
    "            PubChem.APIConsts.ResponseMsgs.GetDescription.RESPONSE_KEY1.value\n",
    "        ][PubChem.APIConsts.ResponseMsgs.GetDescription.RESPONSE_KEY2.value]\n",
    "\n",
    "        if printout:\n",
    "            for entry in response_data:\n",
    "                try:\n",
    "                    print(entry[\"Description\"] + \"\\n\")\n",
    "                except:\n",
    "                    pass\n",
    "        else:\n",
    "            return response_data\n",
    "\n",
    "    @staticmethod\n",
    "    def similarity_search(\n",
    "        smiles,\n",
    "        min_similarity=80,\n",
    "        max_num_results=100,\n",
    "        output_data_type=\"json\",\n",
    "        max_num_attempts=30,\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Run a similarity search on a molecule and get all the similar ligands.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        smiles : str\n",
    "            The canonical SMILES string for the given compound.\n",
    "        min_similarity : int (optional; default: 80)\n",
    "            The threshold of similarity in percent.\n",
    "        max_num_results : int (optional; default: 100)\n",
    "            The maximum number of feedback records.\n",
    "        output_data_type : str (optional; default: 'json')\n",
    "            Datatype of the output data.\n",
    "            Valid values are 'txt', 'json', 'csv'.\n",
    "            A list of all valid values are stored in: PubChem.APIConsts.URLs.Outputs\n",
    "        max_num_attempts : int (optional; default: 30)\n",
    "            Maximum number of attempts to fetch the API response, after the job has been submitted.\n",
    "            Each failed attempt is followed by a 10-second pause.\n",
    "        Returns\n",
    "        -------\n",
    "            Datatype depends on the 'output_data_type' parameter\n",
    "            E.g. when set to \"json\", returns a list of dicts\n",
    "            Each dictionary in the list corresponds to a similar compound,\n",
    "            which has a 'CID' and a 'CanonicalSMILES' key.\n",
    "        \"\"\"\n",
    "        escaped_smiles = quote(smiles).replace(\"/\", \".\")\n",
    "        url = PubChem.APIConsts.URLs.Inputs.SIMILARITY_FROM_SMILES.value + escaped_smiles + \"/\"\n",
    "        response_data = PubChem.send_request(\n",
    "            url,\n",
    "            output_data_type,\n",
    "            f\"Threshold={min_similarity}&MaxRecords={max_num_results}\",\n",
    "        )\n",
    "        job_key = response_data[PubChem.APIConsts.ResponseMsgs.SimilaritySearch.JOBKEY_KEY1.value][\n",
    "            PubChem.APIConsts.ResponseMsgs.SimilaritySearch.JOBKEY_KEY2.value\n",
    "        ]\n",
    "\n",
    "        url = (\n",
    "            PubChem.APIConsts.URLs.Inputs.SIMILARITY_RESULTS.value\n",
    "            + job_key\n",
    "            + PubChem.APIConsts.URLs.Operations.GET_SMILES.value\n",
    "        )\n",
    "\n",
    "        num_attempts = 0\n",
    "        while num_attempts < max_num_attempts:\n",
    "            response_data = PubChem.send_request(url, output_data_type)\n",
    "            if PubChem.APIConsts.ResponseMsgs.SimilaritySearch.RESULT_KEY1.value in response_data:\n",
    "                similar_compounds = response_data[\n",
    "                    PubChem.APIConsts.ResponseMsgs.SimilaritySearch.RESULT_KEY1.value\n",
    "                ][PubChem.APIConsts.ResponseMsgs.SimilaritySearch.RESULT_KEY2.value]\n",
    "                break\n",
    "            time.sleep(10)\n",
    "            num_attempts += 1\n",
    "        else:\n",
    "            raise ValueError(f\"Could not find matches in the response URL: {url}\")\n",
    "        return similar_compounds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***RDKit*** Helper-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "We will also define another helper-class for working with ligands, called <b><i>RDKit</i></b>. This class uses the <i>RDKit</i> python library to implement some useful functionalities for ligands, such as calculating properties and descriptors, e.g. molecular weight, partition coefficients, number of hydrogen-bond acceptors/donors, number of rotatable bonds etc. These properties will then also be used to calculate several drug-likeness scores, for example the Lipinski's rule of 5, and the quantitative estimate of drug-likeness (QED). Other useful functionalities that will be implemented here include visualization of molecular structures, and saving molecules to file, either as an image or an SDF file. We will also define a function to calculate the similarity between two molecules, based on the Dice similarity metric using circular (Morgan) fingerprints. This can be used later to assess the similarity of each of the ligand's analogs found by the similarity search performed using <i>PubChem</i>, since <i>PubChem</i> does not include the exact similarity value of the search results, but only accepts a threshold value. \n",
    "    <br><br>\n",
    "    <b><i>Note:</i></b> For demonstration of each function, see the <a href=\"#rdkit_demo\">corresponding section</a> in <i>Supplementary Information</i>.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class RDKit:\n",
    "    @staticmethod\n",
    "    def create_molecule_object(input_type, input_value):\n",
    "        \"\"\"\n",
    "        Create an RDKit molecule object from various sources.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        input_type : str\n",
    "            Type of the input.\n",
    "            Allowed input-types are: 'smiles', 'inchi', 'smarts', 'pdb_files'\n",
    "        input_value : str\n",
    "            Value of the corresponding input type.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            RDKit molecule object.\n",
    "        \"\"\"\n",
    "\n",
    "        functions = {\n",
    "            \"smiles\": Chem.MolFromSmiles,\n",
    "            \"inchi\": Chem.MolFromInchi,\n",
    "            \"smarts\": Chem.MolFromSmarts,\n",
    "            \"pdb_file\": Chem.MolFromPDBFile,\n",
    "        }\n",
    "        Molobj = functions[input_type](input_value)\n",
    "        return Molobj\n",
    "\n",
    "    @staticmethod\n",
    "    def draw_molecules(\n",
    "        list_mol_objs,\n",
    "        list_legends=None,\n",
    "        mols_per_row=3,\n",
    "        sub_img_size=(550, 550),\n",
    "        filepath=None,\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Take a list of RDKit molecule objects and draws them as a grid image.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        list_mol_objs: list\n",
    "            List of RDKit molecule objects to be drawn.\n",
    "        list_legends: list (optional)\n",
    "            List of legends for the molecules.\n",
    "            If not provided, the list indices (+1) will be used as legend.\n",
    "        mols_per_row : int (optional; default: 3)\n",
    "            Number of structures to show per row.\n",
    "        sub_img_size : tuple (int, int)\n",
    "            Size of each structure.\n",
    "        filepath : str or pathlib.Path object\n",
    "            Full filepath to save the image in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            RDKit MolsToGridImage object.\n",
    "        \"\"\"\n",
    "        if list_legends == None:\n",
    "            list_legends = list(range(1, len(list_mol_objs) + 1))\n",
    "        figure = Draw.MolsToGridImage(\n",
    "            list_mol_objs,\n",
    "            molsPerRow=mols_per_row,\n",
    "            subImgSize=sub_img_size,\n",
    "            legends=list_legends,\n",
    "        )\n",
    "        if filepath != None:\n",
    "            with open(str(filepath) + \".png\", \"wb\") as f:\n",
    "                f.write(figure.data)\n",
    "        return figure\n",
    "\n",
    "    @staticmethod\n",
    "    def save_molecule_image_to_file(mol_obj, filepath):\n",
    "        \"\"\"\n",
    "        Save the image of a single molecule as a PNG file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        mol_obj : RDKit Molecule objects\n",
    "            The molecule to be saved as image.\n",
    "        filepath : str or pathlib.Path object\n",
    "            Full filpath to save the image in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        Draw.MolToFile(mol_obj, str(filepath) + \".png\")\n",
    "\n",
    "    @staticmethod\n",
    "    def save_3D_molecule_to_SDfile(mol_obj, filepath):\n",
    "        \"\"\"\n",
    "        Generate a 3D conformer and save as SDF file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        mol_obj : RDKit Molecule objects\n",
    "            The molecule to be saved as SDF file.\n",
    "        filepath : str or pathlib.Path object\n",
    "            Full filpath to save the image in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        mol = Chem.AddHs(mol_obj)\n",
    "        embedding = AllChem.EmbedMolecule(mol, maxAttempts=1000, clearConfs=True)\n",
    "        uffoptim = AllChem.UFFOptimizeMolecule(mol, maxIters=1000)\n",
    "        # check if calculations converged (both should return 0 when converged)\n",
    "        if embedding + uffoptim != 0:\n",
    "            raise ValueError(\"Embedding/Optimization failed to converge.\")\n",
    "        session = Chem.SDWriter(str(filepath) + \".sdf\")\n",
    "        session.write(mol)\n",
    "        session.close()\n",
    "\n",
    "    @staticmethod\n",
    "    def calculate_similarity_dice(mol_obj1, mol_obj2):\n",
    "        \"\"\"\n",
    "        Calculate the Dice similarity between two molecules,\n",
    "        based on 4096-bit Morgan fingerprints with a radius of 2.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        mol_obj1 : RDKit Molecule objects\n",
    "            The first molecule.\n",
    "        mol_obj2 : RDKit Molecule objects\n",
    "            The second molecule.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            float\n",
    "            Dice similarity between the two molecules\n",
    "        \"\"\"\n",
    "        morgan_fp_mol1 = AllChem.GetMorganFingerprintAsBitVect(mol_obj1, radius=2, nBits=4096)\n",
    "        morgan_fp_mol2 = AllChem.GetMorganFingerprintAsBitVect(mol_obj2, radius=2, nBits=4096)\n",
    "        dice_similarity = round(\n",
    "            AllChem.DataStructs.DiceSimilarity(morgan_fp_mol1, morgan_fp_mol2), 2\n",
    "        )\n",
    "        return dice_similarity\n",
    "\n",
    "    @staticmethod\n",
    "    def calculate_druglikeness(mol_obj):\n",
    "        \"\"\"\n",
    "        Calculate several molecular properties and drug-likeness scores,\n",
    "        from an RDKit molecule object.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        MolObj: RDKit molecule object\n",
    "            Molecule object of interest.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            dict\n",
    "            The calculated values are returned in a dictionary with following keys:\n",
    "            MolWt, NumHAcceptors, NumHDonors, MolLogP, TPSA, NumRotBonds, Saturation,\n",
    "            lipinski_score, custom_drug_score, qed_score, total_drug_score\n",
    "        \"\"\"\n",
    "        properties = {\n",
    "            \"mol_weight\": round(Descriptors.MolWt(mol_obj), 3),\n",
    "            \"num_H_acceptors\": Descriptors.NumHAcceptors(mol_obj),\n",
    "            \"num_H_donors\": Descriptors.NumHDonors(mol_obj),\n",
    "            \"logp\": round(Descriptors.MolLogP(mol_obj), 2),\n",
    "            \"tpsa\": round(Descriptors.TPSA(mol_obj), 2),\n",
    "            \"num_rot_bonds\": Descriptors.NumRotatableBonds(mol_obj),\n",
    "            \"saturation\": round(Descriptors.FractionCSP3(mol_obj), 2),\n",
    "            \"drug_score_qed\": round(Descriptors.qed(mol_obj), 2),\n",
    "        }\n",
    "\n",
    "        # Calculating Lipinski score\n",
    "        l1 = int(properties[\"mol_weight\"] < 500)\n",
    "        l2 = int(properties[\"num_H_acceptors\"] <= 10)\n",
    "        l3 = int(properties[\"num_H_donors\"] <= 5)\n",
    "        l4 = int(properties[\"logp\"] < 5)\n",
    "        properties[\"drug_score_lipinski\"] = round((l1 + l2 + l3 + l4) / 4, 2)\n",
    "\n",
    "        # Calculating druglikeness score with custom scoring functions\n",
    "        # derived from Hopkins paper\n",
    "        def molWt_score(molWt):\n",
    "            if molWt <= 440:\n",
    "                return np.exp(-((molWt - 300) ** 2) / 15000)\n",
    "            else:\n",
    "                return np.exp(-(molWt - 180) / 190) + 0.01\n",
    "\n",
    "        def molLogP_score(molLogP):\n",
    "            return np.exp(-((molLogP - 2.5) ** 2) / 9)\n",
    "\n",
    "        def numHDonors_score(numHDonors):\n",
    "            if numHDonors == 0:\n",
    "                return 0.6\n",
    "            elif numHDonors < 5:\n",
    "                return np.exp(-((numHDonors - 1) ** 2) / 5)\n",
    "            else:\n",
    "                return np.exp(-((numHDonors - 1) ** 2) / 5) + (0.4 / numHDonors)\n",
    "\n",
    "        def numHAcceptors_score(numHAcceptors):\n",
    "            if numHAcceptors < 4:\n",
    "                return np.exp(-((numHAcceptors - 3) ** 2) / 3)\n",
    "            else:\n",
    "                return np.exp(-0.3 * numHAcceptors / 0.8 + 1.4)\n",
    "\n",
    "        def TPSA_score(TPSA):\n",
    "            if TPSA < 50:\n",
    "                return 0.015 * TPSA + 0.25\n",
    "            else:\n",
    "                return np.exp(-((TPSA - 50) ** 2) / 8000)\n",
    "\n",
    "        def numRotBonds_score(numRotBonds):\n",
    "            if numRotBonds < 10:\n",
    "                return np.exp(-((numRotBonds - 4) ** 2) / 19)\n",
    "            else:\n",
    "                return np.exp(-((numRotBonds - 4) ** 2) / 19) + (1.5 / numRotBonds ** 1.5)\n",
    "\n",
    "        def saturation_score(saturation):\n",
    "            return np.exp(-((saturation - 0.625) ** 2) / 0.05)\n",
    "\n",
    "        d1 = molWt_score(properties[\"mol_weight\"])\n",
    "        d2 = numHAcceptors_score(properties[\"num_H_acceptors\"])\n",
    "        d3 = numHDonors_score(properties[\"num_H_donors\"])\n",
    "        d4 = molLogP_score(properties[\"logp\"])\n",
    "        d5 = TPSA_score(properties[\"tpsa\"])\n",
    "        d6 = numRotBonds_score(properties[\"num_rot_bonds\"])\n",
    "        d7 = saturation_score(properties[\"saturation\"])\n",
    "        properties[\"drug_score_custom\"] = round((d1 + d2 + d3 + d4 + d5 + d6 + d7) / 7, 2)\n",
    "\n",
    "        properties[\"drug_score_total\"] = round(\n",
    "            (\n",
    "                3 * properties[\"drug_score_qed\"]\n",
    "                + 2 * properties[\"drug_score_custom\"]\n",
    "                + properties[\"drug_score_lipinski\"]\n",
    "            )\n",
    "            / 6,\n",
    "            2,\n",
    "        )\n",
    "        return properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Pipeline's ***Ligand*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\">Now using the defined functionalities in the <b><i>PubChem</i></b> and <b><i>RDKit</i></b> helper-classes, we can implement the pipeline's <b><i>Ligand</i></b> class. Similar to the <b><i>Protein</i></b> class, this class also takes in the ligand's input data, and creates an object with extended attributes and methods to work with ligands.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Ligand:\n",
    "    \"\"\"\n",
    "    Ligand object with properties as attributes and methods to visualize and work with ligands.\n",
    "    Take a ligand identifier type and corresponding value,\n",
    "    and create a Ligand object, while assigning some properties as attributes.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    identifier_type : enum 'InputTypes' from the 'Consts.Ligand' class\n",
    "        Type of the ligand identifier, e.g. InputTypes.SMILES.\n",
    "\n",
    "    indentifier_value : str\n",
    "        Value of the ligand identifier, e.g. its SMILES.\n",
    "    \"\"\"\n",
    "\n",
    "    class Consts:\n",
    "        # Available properties that are assigned as instance attributes upon instantiation.\n",
    "        class IdentifierTypes(Enum):\n",
    "            NAME = \"name\"\n",
    "            IUPAC_NAME = \"iupac_name\"\n",
    "            SMILES = \"smiles\"\n",
    "            CID = \"cid\"\n",
    "            INCHI = \"inchi\"\n",
    "            INCHIKEY = \"inchikey\"\n",
    "\n",
    "    def __init__(self, identifier_type, identifier_value, ligand_output_path):\n",
    "\n",
    "        self.dataframe = pd.DataFrame(columns=[\"Value\"])\n",
    "        self.dataframe.index.name = \"Property\"\n",
    "\n",
    "        setattr(self, identifier_type.name.lower(), identifier_value)\n",
    "        for identifier in self.Consts.IdentifierTypes:\n",
    "            try:\n",
    "                new_id = PubChem.convert_compound_identifier(\n",
    "                    identifier_type.value, identifier_value, identifier.value\n",
    "                )\n",
    "                setattr(self, identifier.value, new_id)\n",
    "                self.dataframe.loc[identifier.value] = new_id\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "        self.rdkit_obj = RDKit.create_molecule_object(\"smiles\", self.smiles)\n",
    "\n",
    "        dict_of_properties = RDKit.calculate_druglikeness(self.rdkit_obj)\n",
    "        for property_ in dict_of_properties:\n",
    "            setattr(self, property_, dict_of_properties[property_])\n",
    "            self.dataframe.loc[property_] = dict_of_properties[property_]\n",
    "\n",
    "        self.save_as_image(ligand_output_path / (\"CID_\" + self.cid))\n",
    "        self.dataframe.to_csv(ligand_output_path / (\"CID_\" + self.cid + \".csv\"))\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f\"<Ligand CID: {self.cid}>\"\n",
    "\n",
    "    def __call__(self):\n",
    "        df = pd.DataFrame(columns=[\"smiles\"])\n",
    "        df.loc[1] = self.smiles\n",
    "        PandasTools.AddMoleculeColumnToFrame(df, smilesCol=\"smiles\")\n",
    "        romol = df.loc[1, \"ROMol\"]\n",
    "\n",
    "        return pd.concat({romol: self.dataframe}, names=[\"Structure\"])\n",
    "\n",
    "    def remove_counterion(self):\n",
    "        \"\"\"\n",
    "        Remove the counter-ion from the SMILES of a salt.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            str\n",
    "            SMILES of the molecule without its counter-ion.\n",
    "        \"\"\"\n",
    "        if (\n",
    "            \".\" in self.smiles\n",
    "        ):  # SMILES of salts contain a dot, separating the anion and the cation\n",
    "            ions = self.smiles.split(\".\")\n",
    "            length_ions = list(map(len, ions))\n",
    "            molecule_index = length_ions.index(\n",
    "                max(length_ions)\n",
    "            )  # The parent molecule is almost always larger than its corresponding counter-ion\n",
    "            molecule_smiles = ions[molecule_index]\n",
    "        else:\n",
    "            molecule_smiles = self.smiles\n",
    "        return molecule_smiles\n",
    "\n",
    "    def dice_similarity(self, mol_obj):\n",
    "        \"\"\"\n",
    "        Calculate Dice similarity between the ligand and another input ligand,\n",
    "        based on 4096-bit Morgan fingerprints with a radius of 2.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        mol_obj : RDKit molecule object\n",
    "            The molecule to calculate the Dice similarity with.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            float\n",
    "            Dice similarity between the two ligands.\n",
    "        \"\"\"\n",
    "        return RDKit.calculate_similarity_dice(self.rdkit_obj, mol_obj)\n",
    "\n",
    "    def save_as_image(self, filepath):\n",
    "        \"\"\"\n",
    "        Save the ligand as image.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        filepath : str or pathlib.Path object\n",
    "            Full filepath of the image to be saved.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        RDKit.save_molecule_image_to_file(self.rdkit_obj, filepath)\n",
    "\n",
    "    def save_3D_structure_as_SDF_file(self, filepath):\n",
    "        \"\"\"\n",
    "        Generate a 3D conformer and save as SDF file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        filepath : str or pathlib.Path object\n",
    "            Full filpath to save the image in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        RDKit.save_3D_molecule_to_SDfile(self.rdkit_obj, filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Creating a ***Ligand*** Object From The Ligand Input Data: Instantiating The ***Ligand*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "We can now create an instance of the <b><i>Ligand</i></b> class, using the input data of our project's ligand, and assign it to our project. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand = Ligand(\n",
    "    identifier_type=project1.Specs.Ligand.input_type,\n",
    "    identifier_value=project1.Specs.Ligand.input_value,\n",
    "    ligand_output_path=project1.Specs.OutputPaths.ligand,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similar to the ***Protein*** obejct, we have implemented a ***\\_\\_call\\_\\_*** method for our ***Ligand***, which prints out some useful information and visualizes the ligand's structure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All of these information and some other properties are also stored separately as instance attributes. For example, the ligand's identifiers:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.iupac_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.cid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or its physiochemical properties:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.mol_weight"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.logp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\">\n",
    "We have also implemented some useful functions as methods for the <b><i>Ligand</i></b> class. For example, the <b><i>remove_counterion</i></b> method can be used to remove the counter-ion of salt compounds from the main molecule in the SMILES. This is necessary for the docking process, since <i>Smina</i> has some problems processing PDBQT files of salt compounds. Calling this method will simply return the modified SMILES (In the case of our ligand, which is not charged and does not have a counter-ion, the original SMILES is returned): </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.remove_counterion()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='5_practical'></a>\n",
    "### Binding Site Detection\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "Now that the processing of all input data is completed, we can commence with implementing the pipeline's processes. We start with the binding-site detection process, where a suitable binding site should be identified for the target protein. In the input CSV file, the user has the option to select between three definition methods for the binding site:\n",
    "</div>\n",
    "\n",
    "1. ***coordinates***: The user must specify the coordinates of the binding site, in which case there is no need for binding-site detection. \n",
    "2. ***ligand***: The user should specify the ID of a co-crystallized ligand in the protein structure, which will then be used here to define the binding site. \n",
    "3. ***detection***: The user should specify a ***detection method***. \n",
    "\n",
    "<div style=\"text-align: justify\">\n",
    "For this talktorial we are only going to implement one detection method, using the <b><i>DoGSiteScorer</i></b> functionality of the <a href=\"https://proteins.plus\"><b><i>ProteinsPlus</i></b></a> webserver. Therefore, we first need to create a helper-class for communicating with the webserver's API and submitting detection jobs. Afterwards, we will implement the pipeline's <b><i>BindingSiteDetection</i></b> class, which processes the input specification data on the binding site, and acts accordingly. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***DoGSiteScorer*** Helper-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"><i>DoGSiteScorer</i> provides a web-service <a href=\"https://proteins.plus/help/dogsite_rest\">API</a>, which can be used to submit binding-site detection jobs, either by providing the PDB-code of protein structure, or by uploading its PDB file. It will then return a table of all detected pockets and sub-pockets, and their corresponding descriptors. Moreover, for each detected (sub-)pocket, a PDB file and a CCP4 file is generated, which can be downloaded and used to define the coordinates of the (sub-)pocket (needed for the docking experiment), as well as for visualization purposes. We will also define the function <b><i>select_best_pocket</i></b>, which provides several methods for selecting the most suitable binding-site.\n",
    "    <br><br> <i><b>Note:</b></i> For demonstration of each function, see the <a href=\"#dogsitescorer_demo\">corresponding section</a> in <i>Supplementary Information</i>.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DoGSiteScorer:\n",
    "    \"\"\"\n",
    "    Class Containing all the required functions and constants\n",
    "    to communicate with the DoGSiteScorer's Rest-API.\n",
    "    \"\"\"\n",
    "\n",
    "    class APIConsts:\n",
    "        # See https://proteins.plus/help/\n",
    "        # and https://proteins.plus/help/dogsite_rest\n",
    "        # for API specifications.\n",
    "\n",
    "        class FileUpload:\n",
    "            URL = \"https://proteins.plus/api/pdb_files_rest\"\n",
    "            REQUEST_MSG = \"pdb_file[pathvar]\"\n",
    "            RESPONSE_MSG = {\n",
    "                \"status\": \"status_code\",\n",
    "                \"status_codes\": {\"accepted\": \"accepted\", \"denied\": \"bad_request\"},\n",
    "                \"message\": \"message\",\n",
    "                \"url_of_id\": \"location\",\n",
    "            }\n",
    "            RESPONSE_MSG_FETCH_ID = {\"message\": \"message\", \"id\": \"id\"}\n",
    "\n",
    "        class SubmitJob:\n",
    "            URL = \"https://proteins.plus/api/dogsite_rest\"\n",
    "            QUERY_HEADERS = {\n",
    "                \"Content-type\": \"application/json\",\n",
    "                \"Accept\": \"application/json\",\n",
    "            }\n",
    "\n",
    "            RESPONSE_MSG = {\"url_of_job\": \"location\"}\n",
    "\n",
    "            RESPONSE_MSG_FETCH_BINDING_SITES = {\n",
    "                \"result_table\": \"result_table\",\n",
    "                \"pockets_pdb_files\": \"residues\",\n",
    "                \"pockets_ccp4_files\": \"pockets\",\n",
    "            }\n",
    "\n",
    "    @staticmethod\n",
    "    def upload_pdb_file(filepath):\n",
    "        \"\"\"\n",
    "        Upload a PDB file to DoGSiteScorer webserver using their API\n",
    "        and get back a dummy PDB-code, which can be used to submit a detection job.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        filepath : str\n",
    "            Relative or absolute path of the PDB file.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            str\n",
    "            Dummy PDB-code of the uploaded structure,\n",
    "            which can then be used instead of a real PDB-code.\n",
    "        \"\"\"\n",
    "        url = DoGSiteScorer.APIConsts.FileUpload.URL  # Read API URL from Constants\n",
    "        request_msg = (\n",
    "            DoGSiteScorer.APIConsts.FileUpload.REQUEST_MSG\n",
    "        )  # Read API request message from Constants\n",
    "        with open(filepath, \"rb\") as f:  # Open the local PDB file for reading in binary mode\n",
    "            response = requests.post(\n",
    "                url, files={request_msg: f}\n",
    "            )  # Post API query and get the response\n",
    "        response.raise_for_status()  # Raise HTTPError if one occured during query\n",
    "        if response.ok:\n",
    "            response_values = response.json()  # Turn the response values into a dict\n",
    "            # If the request is accepted, the response will contain a URL,\n",
    "            # from which the needed ID of the uploaded protein can be obtained.\n",
    "            # Here, we store this URL from the response values in the url_of_id variable.\n",
    "            url_of_id = response_values[\n",
    "                DoGSiteScorer.APIConsts.FileUpload.RESPONSE_MSG[\"url_of_id\"]\n",
    "            ]\n",
    "        else:\n",
    "            raise ValueError(\n",
    "                \"Uploading PDB file failed.\\n\"\n",
    "                + f\"The response values are as follows: {response_values}\"\n",
    "            )\n",
    "        # After getting the URL, it may take some time for the server to process the uploaded file\n",
    "        # and return an ID. Thus, we try 30 times in intervals of 5 seconds to query the URL,\n",
    "        # until we get the ID\n",
    "        for try_nr in range(30):\n",
    "            id_response = requests.get(url_of_id)  # Query the URL containing the ID\n",
    "            id_response_values = id_response.json()  # Turn the response values into a dict\n",
    "            # The response should contain the ID keyword:\n",
    "            if id_response.ok & (\n",
    "                DoGSiteScorer.APIConsts.FileUpload.RESPONSE_MSG_FETCH_ID[\"id\"]\n",
    "                in id_response_values\n",
    "            ):\n",
    "                id_response_values = id_response.json()\n",
    "                protein_id = id_response_values[\n",
    "                    DoGSiteScorer.APIConsts.FileUpload.RESPONSE_MSG_FETCH_ID[\"id\"]\n",
    "                ]\n",
    "                break\n",
    "            else:\n",
    "                time.sleep(5)\n",
    "        if not (\n",
    "            id_response.ok\n",
    "            & (\n",
    "                DoGSiteScorer.APIConsts.FileUpload.RESPONSE_MSG_FETCH_ID[\"id\"]\n",
    "                in id_response_values\n",
    "            )\n",
    "        ):\n",
    "            raise ValueError(\n",
    "                \"Fetching the ID of uploaded protein failed.\\n\"\n",
    "                + f\"The response values are as follows: {id_response_values}\"\n",
    "            )\n",
    "        return protein_id\n",
    "\n",
    "    @staticmethod\n",
    "    def submit_job(pdb_id, ligand_id=\"\", chain_id=\"\", num_attempts=30):\n",
    "        \"\"\"\n",
    "        Submit a protein structure to DoGSiteScorer webserver using their API\n",
    "        and get back all the information on the detected binding-sites.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pdb_id : str\n",
    "            Either a valid 4-letter PDB-code (e.g. '3w32'),\n",
    "            or a dummy PDB-code of an uploaded PDB file.\n",
    "        ligand_id : str\n",
    "            DogSiteScorer-name of the co-crystallized ligand of interest, e.g. 'W32_A_1101'.\n",
    "            DogSiteScorer naming convention is: <PDB ligand-ID>_<chain-ID>_<PDB residue number of the ligand>\n",
    "        chain_id : str (optional; default: none)\n",
    "            Chain-ID to limit the binding-site detection to.\n",
    "        num_attempts : int (optional; default: 30)\n",
    "            Number of times to attempt to fetch the results after the job has been submitted.\n",
    "            After each failed attempt there is a 10-second pause.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            Pandas DataFrame\n",
    "            Dataframe containing all the information on all detected binding-sites.\n",
    "        \"\"\"\n",
    "        response = requests.post(\n",
    "            DoGSiteScorer.APIConsts.SubmitJob.URL,\n",
    "            json={\n",
    "                \"dogsite\": {\n",
    "                    \"pdbCode\": pdb_id,  # PDB code of protein\n",
    "                    \"analysisDetail\": \"1\",  # 1 = include subpockets in results\n",
    "                    \"bindingSitePredictionGranularity\": \"1\",  # 1 = include drugablity scores\n",
    "                    \"ligand\": ligand_id,  # if name is specified, ligand coverage is calculated\n",
    "                    \"chain\": chain_id,  # if chain is specified, calculation is only performed on this chain\n",
    "                }\n",
    "            },\n",
    "            headers=DoGSiteScorer.APIConsts.SubmitJob.QUERY_HEADERS,\n",
    "        )\n",
    "\n",
    "        response.raise_for_status()\n",
    "        response_values = response.json()\n",
    "        url_of_job = response_values[DoGSiteScorer.APIConsts.SubmitJob.RESPONSE_MSG[\"url_of_job\"]]\n",
    "\n",
    "        attempt_count = 0\n",
    "        while attempt_count <= num_attempts:\n",
    "            job_response = requests.get(url_of_job)\n",
    "            job_response.raise_for_status()\n",
    "            job_response_values = job_response.json()\n",
    "\n",
    "            if (\n",
    "                DoGSiteScorer.APIConsts.SubmitJob.RESPONSE_MSG_FETCH_BINDING_SITES[\"result_table\"]\n",
    "                in job_response_values\n",
    "            ):\n",
    "                binding_site_data_url = job_response_values[\n",
    "                    DoGSiteScorer.APIConsts.SubmitJob.RESPONSE_MSG_FETCH_BINDING_SITES[\n",
    "                        \"result_table\"\n",
    "                    ]\n",
    "                ]\n",
    "                binding_sites_pdb_files_urls = job_response_values[\n",
    "                    DoGSiteScorer.APIConsts.SubmitJob.RESPONSE_MSG_FETCH_BINDING_SITES[\n",
    "                        \"pockets_pdb_files\"\n",
    "                    ]\n",
    "                ]\n",
    "                binding_sites_ccp4_files_urls = job_response_values[\n",
    "                    DoGSiteScorer.APIConsts.SubmitJob.RESPONSE_MSG_FETCH_BINDING_SITES[\n",
    "                        \"pockets_ccp4_files\"\n",
    "                    ]\n",
    "                ]\n",
    "                break\n",
    "            attempt_count += 1\n",
    "            time.sleep(10)\n",
    "        else:\n",
    "            raise ValueError(\n",
    "                \"Fetching the binding-site data failed.\\n\"\n",
    "                + f\"The response values are as follows: {job_response_values}\"\n",
    "            )\n",
    "\n",
    "        binding_site_data_table = requests.get(binding_site_data_url).text\n",
    "        binding_site_data_file = io.StringIO(binding_site_data_table)\n",
    "        binding_site_df = pd.read_csv(binding_site_data_file, sep=\"\\t\").set_index(\"name\")\n",
    "        binding_site_df[\"pdb_file_url\"] = binding_sites_pdb_files_urls\n",
    "        binding_site_df[\"ccp4_file_url\"] = binding_sites_ccp4_files_urls\n",
    "        return binding_site_df\n",
    "\n",
    "    @staticmethod\n",
    "    def save_binding_sites_to_file(binding_site_df, output_path):\n",
    "        \"\"\"\n",
    "        download and save the PDB and CCP4 files corresponding to the calculated binding-sites.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        binding_site_df : Pandas DataFrame\n",
    "            Binding-site data retrieved from the DoGSiteScorer webserver.\n",
    "        output_path : str or pathlib.Path object\n",
    "            Local file path to save the files in.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        for binding_site in binding_site_df.index:\n",
    "            for column in [\"pdb_file_url\", \"ccp4_file_url\"]:\n",
    "                response = requests.get(binding_site_df.loc[binding_site, column])\n",
    "                response.raise_for_status()\n",
    "                if column == \"pdb_file_url\":\n",
    "                    response_file_content = response.content\n",
    "                    file_extension = \".pdb\"\n",
    "                else:\n",
    "                    response_file_content = gzip.decompress(response.content)\n",
    "                    file_extension = \".ccp4\"\n",
    "\n",
    "                file_name = binding_site + file_extension\n",
    "                with open(output_path / file_name, \"wb\") as f:\n",
    "                    f.write(response_file_content)\n",
    "        return\n",
    "\n",
    "    @staticmethod\n",
    "    def select_best_pocket(binding_site_df, selection_method, selection_criteria, ascending=False):\n",
    "        \"\"\"\n",
    "        Select the best binding-site from the table of all detected binding-sites,\n",
    "        either by sorting the binding-sites based on a set of properties in the table,\n",
    "        or by applying a function on the property values.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        binding_site_df : Pandas DataFrame\n",
    "            Binding-site data retrieved from the DoGSiteScorer webserver.\n",
    "        selection_method : str\n",
    "            Selection method for selecting the best binding-site.\n",
    "            Either 'sorting' or 'function'.\n",
    "        selection_criteria : str or list\n",
    "            When 'selection_method' is 'sorting':\n",
    "                List of one or several property names.\n",
    "            When 'selection_method' is 'function':\n",
    "                Any valid python syntax that generates a list-like object\n",
    "                with the same length as the number of detected binding-sites.\n",
    "        ascending : bool (optional; default: False)\n",
    "            If set to True, the binding-site with the lowest value will be selected,\n",
    "            otherwise, the binding-site with the highest value is selected.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            str\n",
    "            Name of the selected binding-site.\n",
    "        \"\"\"\n",
    "        df = binding_site_df\n",
    "        if selection_method == \"sorting\":\n",
    "            sorted_df = df.sort_values(by=selection_criteria, ascending=ascending)\n",
    "        elif selection_method == \"function\":\n",
    "            df[\"function_score\"] = eval(selection_criteria)\n",
    "            sorted_df = df.sort_values(by=\"function_score\", ascending=ascending)\n",
    "\n",
    "        selected_pocket_name = sorted_df.iloc[0].name\n",
    "        return selected_pocket_name\n",
    "\n",
    "    @staticmethod\n",
    "    def calculate_pocket_coordinates_from_pocket_pdb_file(filepath):\n",
    "        \"\"\"\n",
    "        Calculate the coordinates of a binding-site using the binding-site's PDB file\n",
    "        downloaded from DoGSiteScorer.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        filepath : str or pathlib.Path object\n",
    "            Local filepath (including filename, without extension) of the binding-site's PDB file.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            dict of lists of integers\n",
    "            Binding-site coordinates in format:\n",
    "            {'center': [x, y, z], 'size': [x, y, z]}\n",
    "        \"\"\"\n",
    "        with open(str(filepath) + \".pdb\") as f:\n",
    "            pdb_file_text_content = f.read()\n",
    "        pdb_file_df = PDB.load_pdb_file_as_dataframe(pdb_file_text_content)\n",
    "        pocket_coordinates_data = pdb_file_df[\"OTHERS\"].loc[5, \"entry\"]\n",
    "        coordinates_data_as_list = pocket_coordinates_data.split(\" \")\n",
    "        coordinates = []\n",
    "        for elem in coordinates_data_as_list:\n",
    "            try:\n",
    "                coordinates.append(float(elem))\n",
    "            except:\n",
    "                pass\n",
    "        pocket_coordinates = {\n",
    "            \"center\": coordinates[:3],\n",
    "            \"size\": [coordinates[-1] * 2 for dim in range(3)],\n",
    "        }\n",
    "        return pocket_coordinates\n",
    "\n",
    "    @staticmethod\n",
    "    def get_pocket_residues(pocket_residues_url):\n",
    "        \"\"\"\n",
    "        Gets residue IDs and names of a specified pocket (via URL).\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pocket_residues_url : str\n",
    "            URL of selected pocket file on the DoGSiteScorer web server.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            pandas.DataFrame\n",
    "            Table of residues names and IDs for the selected binding site.\n",
    "        \"\"\"\n",
    "        # Retrieve PDB file content from URL\n",
    "        result = requests.get(pocket_residues_url)\n",
    "        # Get content of PDB file\n",
    "        pdb_residues = result.text\n",
    "        # Load PDB format as DataFrame\n",
    "        ppdb = PandasPdb()\n",
    "        # TODO: Change _construct_df to read_pdb_from_lines once biopandas\n",
    "        # cuts a new release (currently: 0.2.7), see https://github.com/rasbt/biopandas/pull/72\n",
    "        pdb_df = ppdb._construct_df(pdb_residues.splitlines(True))[\"ATOM\"]\n",
    "        # Drop duplicates\n",
    "        # PDB file contains per atom entries, we only need per residue info\n",
    "        pdb_df.sort_values(\"residue_number\", inplace=True)\n",
    "        pdb_df.drop_duplicates(subset=\"residue_number\", keep=\"first\", inplace=True)\n",
    "        return pdb_df[[\"residue_number\", \"residue_name\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Pipeline's ***BindingSiteDetection*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "Now we can automate the binding-site detection process by implementing the <b><i>BindingSiteDetection</i></b> class of the pipeline, where depending on the input binding site specification data, the corresponding processes are carried out to output the coordinates of the selected binding site based on the input data.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class BindingSiteDetection:\n",
    "    \"\"\"\n",
    "    Automated binding-site detection process of the pipeline.\n",
    "    Take in the Protein object, Specs.BindingSite object and\n",
    "    the corresponding output path, and automatically run all the necessary calculations\n",
    "    to output the suitable binding-site coordinates based on the input specifications of the project.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Protein : Protein object\n",
    "        The Protein object of the project.\n",
    "    BindingSiteSpecs : Specs.BindingSite object\n",
    "        The binding-site specification data-class of the project.\n",
    "    binding_site_output_path : str or pathlib.Path object\n",
    "        Output path of the project's binding-site information.\n",
    "    \"\"\"\n",
    "\n",
    "    class Consts:\n",
    "        class DefinitionMethods(Enum):\n",
    "            DETECTION = \"detection\"\n",
    "            LIGAND = \"ligand\"\n",
    "            COORDINATES = \"coordinates\"\n",
    "\n",
    "    def __init__(self, Protein, BindingSiteSpecs, binding_site_output_path):\n",
    "\n",
    "        self.output_path = binding_site_output_path\n",
    "        self.Protein = Protein\n",
    "        # derive the relevant function name from definition method\n",
    "        definition_method_name = \"compute_by_\" + BindingSiteSpecs.definition_method.name.lower()\n",
    "        # get the function from its name\n",
    "        definition_method = getattr(self, definition_method_name)\n",
    "        # call the function\n",
    "        definition_method(Protein, BindingSiteSpecs, binding_site_output_path)\n",
    "\n",
    "    def compute_by_coordinates(self, Protein, BindingSiteSpecs, binding_site_output_path):\n",
    "        Protein.binding_site_coordinates = BindingSiteSpecs.coordinates\n",
    "\n",
    "    def compute_by_ligand(self, Protein, BindingSiteSpecs, binding_site_output_path):\n",
    "        ligand_object = PDB.extract_molecule_from_pdb_file(\n",
    "            BindingSiteSpecs.protein_ligand_id,\n",
    "            Protein.pdb_filepath,\n",
    "            binding_site_output_path / BindingSiteSpecs.protein_ligand_id,\n",
    "        )\n",
    "        # calculate the geometric center of the molecule,\n",
    "        # which represents the center of the rectangular box,\n",
    "        # as well as the length of the molecule in each dimension,\n",
    "        # which corresponds to the edge lengths of the rectangular box.\n",
    "        # Also, we will add a 5 Å buffer in each dimension to allow the\n",
    "        # correct placements of ligands that are bigger than the co-crystallized ligands\n",
    "        # or bind in a different fashion.\n",
    "        pocket_center = (\n",
    "            ligand_object.positions.max(axis=0) + ligand_object.positions.min(axis=0)\n",
    "        ) / 2\n",
    "        pocket_size = ligand_object.positions.max(axis=0) - ligand_object.positions.min(axis=0) + 5\n",
    "        pocket_coordinates = {\n",
    "            \"center\": pocket_center.tolist(),\n",
    "            \"size\": [pocket_size.tolist()],\n",
    "        }\n",
    "        Protein.binding_site_coordinates = pocket_coordinates\n",
    "        return\n",
    "\n",
    "    def compute_by_detection(self, Protein, BindingSiteSpecs, binding_site_output_path):\n",
    "        # derive the relevant function name from detection method\n",
    "        detection_method_name = \"detect_by_\" + BindingSiteSpecs.detection_method.name.lower()\n",
    "        # get the function from its name\n",
    "        detection_method = getattr(self, detection_method_name)\n",
    "        # call the function\n",
    "        detection_method(Protein, BindingSiteSpecs, binding_site_output_path)\n",
    "\n",
    "    def detect_by_dogsitescorer(self, Protein, BindingSiteSpecs, binding_site_output_path):\n",
    "\n",
    "        if hasattr(Protein, \"pdb_code\"):\n",
    "            self.dogsitescorer_pdb_id = Protein.pdb_code\n",
    "        elif hasattr(Protein, \"pdb_filepath\"):\n",
    "            self.dogsitescorer_pdb_id = DoGSiteScorer.upload_pdb_file(Protein.pdb_filepath)\n",
    "\n",
    "        # try to get the chain_id for binding-site detection if it's available in input data\n",
    "        if hasattr(BindingSiteSpecs, \"protein_chain_id\"):\n",
    "            self.dogsitescorer_chain_id = BindingSiteSpecs.protein_chain_id\n",
    "        else:\n",
    "            # if chain_id is not in input data, try to set it to the first chain found in pdb file\n",
    "            try:\n",
    "                self.dogsitescorer_chain_id = Protein.chains[0]\n",
    "            # if no chain is found in pdb file either, leave the chain_id empty\n",
    "            except:\n",
    "                self.dogsitescorer_chain_id = \"\"\n",
    "\n",
    "        # try to get the ligand_id for detection calculation if it's available in input data\n",
    "        if hasattr(BindingSiteSpecs, \"protein_ligand_id\"):\n",
    "            ligand_id = BindingSiteSpecs.protein_ligand_id\n",
    "            # check if the given ligand_id is already given in the DoGSiteScorer format\n",
    "            if \"_\" in ligand_id:\n",
    "                self.dogsitescorer_ligand_id = ligand_id\n",
    "                self.dogsitescorer_chain_id = ligand_id.split(\"_\")[1][0]\n",
    "            else:\n",
    "                try:\n",
    "                    for ligand in Protein.ligands:\n",
    "                        if (ligand[0] == ligand_id) and (\n",
    "                            ligand[1][0] == self.dogsitescorer_chain_id\n",
    "                        ):\n",
    "                            self.dogsitescorer_ligand_id = (\n",
    "                                ligand_id + \"_\" + self.dogsitescorer_chain_id + \"_\" + ligand[1][1:]\n",
    "                            )\n",
    "                            break\n",
    "                except:\n",
    "                    self.dogsitescorer_ligand_id = \"\"\n",
    "\n",
    "        self.dogsitescorer_binding_sites_df = DoGSiteScorer.submit_job(\n",
    "            self.dogsitescorer_pdb_id,\n",
    "            self.dogsitescorer_ligand_id,\n",
    "            self.dogsitescorer_chain_id,\n",
    "        )\n",
    "\n",
    "        DoGSiteScorer.save_binding_sites_to_file(\n",
    "            self.dogsitescorer_binding_sites_df, binding_site_output_path\n",
    "        )\n",
    "\n",
    "        self.best_binding_site_name = DoGSiteScorer.select_best_pocket(\n",
    "            self.dogsitescorer_binding_sites_df,\n",
    "            BindingSiteSpecs.selection_method.value,\n",
    "            BindingSiteSpecs.selection_criteria,\n",
    "        )\n",
    "\n",
    "        self.best_binding_site_data = self.dogsitescorer_binding_sites_df.loc[\n",
    "            self.best_binding_site_name\n",
    "        ]\n",
    "\n",
    "        self.best_binding_site_coordinates = (\n",
    "            DoGSiteScorer.calculate_pocket_coordinates_from_pocket_pdb_file(\n",
    "                binding_site_output_path / (self.best_binding_site_name)\n",
    "            )\n",
    "        )\n",
    "\n",
    "        Protein.binding_site_coordinates = self.best_binding_site_coordinates\n",
    "        return\n",
    "\n",
    "    def visualize(self, pocket_name):\n",
    "        \"\"\"\n",
    "        Visualize a detected pocket.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pocket_name : str\n",
    "            Name of the detected pocket, which has a corresponding CCP4 file\n",
    "            with the same name stored in the output folder for binding site data.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLView viewer\n",
    "            Viewer showing the given pocket.\n",
    "        \"\"\"\n",
    "        if hasattr(self.Protein, \"pdb_code\"):\n",
    "            viewer = NGLView.binding_site(\n",
    "                \"pdb_code\",\n",
    "                self.Protein.pdb_code,\n",
    "                str(self.output_path / pocket_name) + \".ccp4\",\n",
    "            )\n",
    "        else:\n",
    "            viewer = NGLView.protein(\n",
    "                \"pdb\",\n",
    "                self.Protein.pdb_filepath,\n",
    "                str(self.output_path / pocket_name) + \".ccp4\",\n",
    "            )\n",
    "        return viewer\n",
    "\n",
    "    def visualize_best(self):\n",
    "        \"\"\"\n",
    "        Visualize the selected binding pocket.\n",
    "        The binding pocket should have a corresponding CCP4 file\n",
    "        with the same name stored in the output folder for binding site data.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLView viewer\n",
    "            Viewer showing the given pocket.\n",
    "        \"\"\"\n",
    "        return self.visualize(self.best_binding_site_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Automatically Calculating the Binding-Site: Instantiating The ***BindingSiteDetection*** Class\n",
    "We can now instantiate the ***BindingSiteDetection*** class, using the ***Protein*** object, the ***Specs.BindingSite*** object, and the binding-site output path of our project. This will automatically run all the processes needed to result in the coordinates of the selected binding-site, based on the specifications in the input data. The results will then be stored as instance attributes in the instantiated ***BindingSiteDetection*** object. We thus assign this object to our project: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.BindingSiteDetection = BindingSiteDetection(\n",
    "    project1.Protein,\n",
    "    project1.Specs.BindingSite,\n",
    "    project1.Specs.OutputPaths.binding_site_detection,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All intermediate information leading to the selected binding-pocket's coordinates are now stored in the ***BindingSiteDetection*** instance of our project. For example, a dataframe containing all retrieved information on all detected binding sites:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.BindingSiteDetection.dogsitescorer_binding_sites_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Name of the selected binding site:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.BindingSiteDetection.best_binding_site_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The coordinates of the selected binding site are also assigned to the ***Protein*** object in the project:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Protein.binding_site_coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also visualize the selected binding pocket:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project1.BindingSiteDetection.visualize_best()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or any other binding pocket, by providing its name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.BindingSiteDetection.visualize(\"P_0\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='6_practical'></a>\n",
    "### Ligand Similarity Search\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "With the coordinates of the protein's binding-site in hand, we now focus on the ligand similarity-search part of the pipeline. For performing the similarity-search, we will use the <i>PubChem</i> webserver. We have already implemented all the required functions to perform the similarity-search in the <b><i>PubChem</i></b> helper-class that we used for developing the <b><i>Ligand</i></b> class. Therefore, we can now directly commence with writing the <b><i>LigandSimilaritySearch</i></b> class of the pipeline. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Pipeline's ***LigandSimilaritySearch*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Here, we will automate the similarity-search process of the pipeline. This class will take in the <b><i>Ligand</i></b> object and the <b><i>Specs.LigandSimilaritySearch</i></b> object of the pipeline, and initialize a similarity-search using <i>PubChem</i> webserver. Subsequently, several drug-likeness scores are calculated for each of the initially retrieved analogs. Using these scores, a given number (specified in the input file) of most drug-like analogs are then selected and used to create <b><i>Ligand</i></b> objects with the help of the <b><i>Ligand</i></b> class that we defined earlier. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class LigandSimilaritySearch:\n",
    "    \"\"\"\n",
    "    Automated ligand similarity-search process of the pipeline.\n",
    "    Take in input Ligand object, Specs.LigandSimilaritySearch object,\n",
    "    and the corresponding output path, and automatically run all the necessary\n",
    "    processes to output a set of analogs with the highest drug-likeness scores.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Ligand_obj : Ligand object\n",
    "        The Protein object of the project.\n",
    "    BindingSiteSpecs : Specs.BindingSite object\n",
    "        The binding-site specification data-class of the project.\n",
    "    binding_site_output_path : str or pathlib.Path object\n",
    "        Output path of the project's binding-site information.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, Ligand_obj, SimilaritySearchSpecs, similarity_search_output_path):\n",
    "\n",
    "        if (\n",
    "            SimilaritySearchSpecs.search_engine\n",
    "            is Consts.LigandSimilaritySearch.SearchEngines.PUBCHEM\n",
    "        ):\n",
    "            analogs_info = PubChem.similarity_search(\n",
    "                Ligand_obj.smiles,\n",
    "                SimilaritySearchSpecs.min_similarity_percent,\n",
    "                SimilaritySearchSpecs.max_num_results,\n",
    "            )\n",
    "\n",
    "        # create dataframe from initial results\n",
    "        all_analog_identifiers_df = pd.DataFrame(analogs_info)\n",
    "        all_analog_identifiers_df[\"Mol\"] = all_analog_identifiers_df[\"CanonicalSMILES\"].apply(\n",
    "            lambda smiles: RDKit.create_molecule_object(\"smiles\", smiles)\n",
    "        )\n",
    "        all_analog_identifiers_df[\"dice_similarity\"] = all_analog_identifiers_df[\"Mol\"].apply(\n",
    "            lambda mol: RDKit.calculate_similarity_dice(Ligand_obj.rdkit_obj, mol)\n",
    "        )\n",
    "        all_analog_properties_df = pd.DataFrame(\n",
    "            (\n",
    "                all_analog_identifiers_df[\"Mol\"].apply(\n",
    "                    lambda mol: RDKit.calculate_druglikeness(mol)\n",
    "                )\n",
    "            ).tolist()\n",
    "        )\n",
    "        all_analogs_df = pd.concat([all_analog_identifiers_df, all_analog_properties_df], axis=1)\n",
    "        all_analogs_df.index = all_analogs_df[\"CID\"]\n",
    "        all_analogs_df.drop(\"CID\", axis=1, inplace=True)\n",
    "        all_analogs_df.sort_values(by=\"dice_similarity\", ascending=False, inplace=True)\n",
    "        self.all_analogs = all_analogs_df\n",
    "        all_analogs_df.drop(\"Mol\", axis=1).to_csv(\n",
    "            similarity_search_output_path / \"analogs_all.csv\"\n",
    "        )\n",
    "\n",
    "        analogs_dict = {}\n",
    "        for analog_cid in (\n",
    "            self.all_analogs.sort_values(by=\"drug_score_total\", ascending=False)\n",
    "            .head(SimilaritySearchSpecs.max_num_druglike)\n",
    "            .index\n",
    "        ):\n",
    "            new_analog_object = Ligand(\n",
    "                Consts.Ligand.InputTypes.CID,\n",
    "                analog_cid,\n",
    "                similarity_search_output_path,\n",
    "            )\n",
    "\n",
    "            new_analog_object.dice_similarity = RDKit.calculate_similarity_dice(\n",
    "                Ligand_obj.rdkit_obj, new_analog_object.rdkit_obj\n",
    "            )\n",
    "\n",
    "            new_analog_object.dataframe.loc[\"similarity\"] = new_analog_object.dice_similarity\n",
    "\n",
    "            analogs_dict[new_analog_object.cid] = new_analog_object\n",
    "\n",
    "        Ligand_obj.analogs = analogs_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Finding Structural Analogs of the Input Ligand: Instantiating The ***LigandSimilaritySearch*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "    By instantiating the <b><i>LigandSimilaritySearch</i></b> class with the <b><i>Ligand</i></b> object and the <b><i>Specs.LigandSimilaritySearch</i></b> data-class of the project, analogs are automatically found and turned to <b><i>Ligand</i></b> objects, which are then assigned as an instance attribute to the <b><i>LigandSimilaritySearch</i></b> class, as well as the input <b><i>Ligand</i></b> object. We instantiate the class and assign it to our project:\n",
    "<br><br>\n",
    "<b><i>Note:</i></b> This process will take about 2 minutes to complete.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.LigandSimilaritySearch = LigandSimilaritySearch(\n",
    "    project1.Ligand,\n",
    "    project1.Specs.LigandSimilaritySearch,\n",
    "    project1.Specs.OutputPaths.similarity_search,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can view the full list of all fetched analogs, and their calculated physiochemical properties and drug-likeness scores (here showing only the first 5 entries):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.LigandSimilaritySearch.all_analogs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From these analogs, a certain number of most drug-like compounds are selected according to the input specifications.  These selected analogs are then turned into <b><i>Ligand</i></b> objects and assigned to the input ***Ligand*** under the attribute name ***analogs***: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen, the list contains ***Ligand*** objects, each corresponding to a found analog. Each analog has thus its own attributes and methods (just like our input ***Ligand***), which can then be accessed separately, via the analog's CID. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"65997\"]()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='7_practical'></a>\n",
    "### Molecular Docking \n",
    "<br>\n",
    "<div style=\"text-align: justify\">\n",
    "We now have successfully defined the binding site of our input protein, and found a list of analogs for our input ligand. The next step in the pipeline is thus to perform docking experiments on the input protein's binding site, using the ligand analogs. But before that, we need to implement some helper functions that are needed to prepare the protein and the analogs for docking.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***OBabel*** Helper-Class\n",
    "Here we will use the <b><i><a href=\"https://github.com/pybel/pybel\">Pybel</a></i></b> package of the <b><i><a href=\"http://openbabel.org/wiki/Main_Page\">OpenBabel</a></i></b> program in order to implement the functions needed to prepare the protein and the ligand analogs for docking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class OBabel:\n",
    "    \"\"\"\n",
    "    A set of functions based on the OpenBabel's pybel package,\n",
    "    for preparing proteins and ligands for the docking experiment.\n",
    "    \"\"\"\n",
    "\n",
    "    @staticmethod\n",
    "    def optimize_structure_for_docking(\n",
    "        pybel_structure_object,\n",
    "        add_hydrogens=True,\n",
    "        protonate_for_pH=7.4,\n",
    "        calculate_partial_charges=True,\n",
    "        generate_3d_structure=False,\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Take a pybel structure object and prepare for docking.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pybel_structure_object : object\n",
    "            The structure to optimize.\n",
    "        add_hydrogens : bool (optional; default: True)\n",
    "            Whether to add hydrogen atoms to the structure.\n",
    "        protonate_for_pH : float (optional; default: 7.4)\n",
    "            pH value to protonate the structure at.\n",
    "            If set to 0 or False, will not protonate.\n",
    "        calculate_partial_charges : bool (optional; default: True)\n",
    "            Whether to calculate partial charges for each atom\n",
    "        generate_3d_structure : bool (optional; default: False)\n",
    "            Whether to generate a 3D conformation.\n",
    "            Must be set to True if the pybel structure is 2D,\n",
    "            for example is structure is made from SMILES string.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "            The structure object is optimized in place.\n",
    "        \"\"\"\n",
    "        if add_hydrogens:\n",
    "            pybel_structure_object.addh()\n",
    "        if protonate_for_pH:\n",
    "            pybel_structure_object.OBMol.CorrectForPH(protonate_for_pH)\n",
    "        if generate_3d_structure:\n",
    "            pybel_structure_object.make3D(forcefield=\"mmff94s\", steps=10000)\n",
    "        if calculate_partial_charges:\n",
    "            for atom in pybel_structure_object.atoms:\n",
    "                atom.OBAtom.GetPartialCharge()\n",
    "        return\n",
    "\n",
    "    @staticmethod\n",
    "    def create_pdbqt_from_pdb_file(pdb_filepath, pdbqt_filepath, pH=7.4):\n",
    "        \"\"\"\n",
    "        Convert a PDB file to a PDBQT file,\n",
    "        while adding hydrogen atoms, correcting the protonation state,\n",
    "        and assigning partial charges.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pdb_filepath: str or pathlib.Path\n",
    "            Path to input PDB file.\n",
    "        pdbqt_filepath: str or pathlib.path\n",
    "            Path to output PDBQT file.\n",
    "        pH: float\n",
    "            pH value for defining the protonation state of the atoms.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            Pybel Molecule object\n",
    "            Molecule object of PDB file optimized for docking.\n",
    "        \"\"\"\n",
    "        # readfile() provides an iterator over the Molecules in a file.\n",
    "        # To access the first (and possibly only) molecule in a file,\n",
    "        # we use .next()\n",
    "        molecule = next(pybel.readfile(\"pdb\", str(pdb_filepath)))\n",
    "        OBabel.optimize_structure_for_docking(molecule, protonate_for_pH=pH)\n",
    "        molecule.write(\"pdbqt\", str(pdbqt_filepath), overwrite=True)\n",
    "        return molecule\n",
    "\n",
    "    @staticmethod\n",
    "    def create_pdbqt_from_smiles(smiles, pdbqt_path, pH=7.4):\n",
    "        \"\"\"\n",
    "        Convert a SMILES string to a PDBQT file,\n",
    "        while adding hydrogen atoms, correcting the protonation state, assigning partial charges,\n",
    "        and generating a 3D conformer.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        smiles: str\n",
    "            SMILES string.\n",
    "        pdbqt_path: str or pathlib.path\n",
    "            Path to output PDBQT file.\n",
    "        pH: float\n",
    "            Protonation at given pH.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        molecule = pybel.readstring(\"smi\", smiles)\n",
    "        OBabel.optimize_structure_for_docking(\n",
    "            molecule, protonate_for_pH=pH, generate_3d_structure=True\n",
    "        )\n",
    "        molecule.write(\"pdbqt\", str(pdbqt_path), overwrite=True)\n",
    "        return\n",
    "\n",
    "    @staticmethod\n",
    "    def split_multistructure_file(filetype, filepath, output_folder_path=None):\n",
    "        \"\"\"\n",
    "        Split a multi-structure file into seperate files (with the same format)\n",
    "        for each structure.Each file is named with consecutive numbers (starting at 1)\n",
    "        at the end of the original filename.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        filetype : str\n",
    "            Type of the multimodel file to be split.\n",
    "            Examples: 'sdf', 'pdb', 'pdbqt' etc.\n",
    "            For a full list of acceptable file-types, call pybel.informats\n",
    "        filepath : str or pathlib.Path\n",
    "            Path of the file to be split.\n",
    "        output_folder_path : str or pathlib.Path\n",
    "            (optional; default: same folder as input filepath)\n",
    "            Path of the output folder to save the split files.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            list of pathlib.Path objects\n",
    "            List of the full-paths for each split file.\n",
    "        \"\"\"\n",
    "        filepath = Path(filepath)\n",
    "        filename = filepath.stem\n",
    "        if output_folder_path == None:\n",
    "            output_folder_path = filepath.parent\n",
    "        else:\n",
    "            output_folder_path = Path(output_folder_path)\n",
    "            output_folder_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "        structures = pybel.readfile(filetype, str(filepath))\n",
    "        output_filepaths = []\n",
    "        for i, structure in enumerate(structures, 1):\n",
    "            output_filepath = output_folder_path / f\"{filename}_{i}.{filetype}\"\n",
    "            output_filepaths.append(output_filepath)\n",
    "            structure.write(filetype, str(output_filepath), overwrite=True)\n",
    "        return output_filepaths\n",
    "\n",
    "    @staticmethod\n",
    "    def merge_molecules_to_single_file(\n",
    "        list_of_pybel_molecule_objects, output_filetype, output_filepath\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Create a single file containing several molecules.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        list_of_pybel_molecule_objects : list of pybel Molecule objects\n",
    "            List of molecule ojects to be merged into a single file.\n",
    "        output_filetype : str\n",
    "            Type of the output file.\n",
    "            Examples: 'sdf', 'pdb', 'pdbqt' etc.\n",
    "            For a full list of acceptable file-types, call pybel.outformats\n",
    "        output_filepath : str or pathlib.Path\n",
    "            Path of the output file including file-name, but excluding extension.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            pathlib.Path object\n",
    "            Full-path (including extension) of the output file.\n",
    "        \"\"\"\n",
    "        fullpath = Path(str(output_filepath) + \".\" + output_filetype)\n",
    "\n",
    "        merged_molecule_file = pybel.Outputfile(output_filetype, str(fullpath))\n",
    "\n",
    "        for pybel_molecule_object in list_of_pybel_molecule_objects:\n",
    "            merged_molecule_file.write(pybel_molecule_object)\n",
    "\n",
    "        merged_molecule_file.close()\n",
    "        return fullpath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***Smina*** Helper-Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\">For the docking, we are going to use the <a href=\"https://sourceforge.net/projects/smina/\"><b><i>Smina</i></b></a> program, which does not have a Python-API. However, we can simply communicate with the program via the <b><i>subprocess</i></b> library, which can execute shell commands. We also need to write a function to read the output log of the program and extract the useful data. These functions are implemented here in the <b><i>Smina</i></b> helper-class.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Smina:\n",
    "    \"\"\"\n",
    "    Set of functions for communicating with the Smina docking program,\n",
    "    and extracting data from its log file.\n",
    "    \"\"\"\n",
    "\n",
    "    @staticmethod\n",
    "    def dock(\n",
    "        ligand_path,\n",
    "        protein_path,\n",
    "        pocket_center,\n",
    "        pocket_size,\n",
    "        output_path,\n",
    "        output_format=\"pdbqt\",\n",
    "        num_poses=10,\n",
    "        exhaustiveness=10,\n",
    "        random_seed=\"\",\n",
    "        log=True,\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Perform docking with Smina.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ligand_path : str or pathlib.Path\n",
    "            Path to ligand PDBQT file that should be docked.\n",
    "        protein_path : str or pathlib.Path\n",
    "            Path to protein PDBQT file that should be docked to.\n",
    "        pocket_center : iterable of float or int\n",
    "            Coordinates defining the center of the binding site.\n",
    "        pocket_size : iterable of float or int\n",
    "            Lengths of edges defining the binding site.\n",
    "        output_path : str or pathlib.Path\n",
    "            Path to which docking poses should be saved, SDF or PDB format.\n",
    "        num_poses : int or str\n",
    "            Maximum number of poses to generate.\n",
    "        exhaustiveness : int or str\n",
    "            Accuracy of docking calculations.\n",
    "        random_seed : int or str\n",
    "            Seed number to make the docking deterministic for reproducibility.\n",
    "        log : bool (optional; default: True)\n",
    "            Whether to also write a log-file in the same output path for each docking.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        output_text: str\n",
    "            The output log of the docking calculation.\n",
    "        \"\"\"\n",
    "        smina_command = (\n",
    "            [\n",
    "                \"smina\",\n",
    "                \"--ligand\",\n",
    "                str(ligand_path),\n",
    "                \"--receptor\",\n",
    "                str(protein_path),\n",
    "                \"--out\",\n",
    "                str(output_path) + \".\" + output_format,\n",
    "                \"--center_x\",\n",
    "                str(pocket_center[0]),\n",
    "                \"--center_y\",\n",
    "                str(pocket_center[1]),\n",
    "                \"--center_z\",\n",
    "                str(pocket_center[2]),\n",
    "                \"--size_x\",\n",
    "                str(pocket_size[0]),\n",
    "                \"--size_y\",\n",
    "                str(pocket_size[1]),\n",
    "                \"--size_z\",\n",
    "                str(pocket_size[2]),\n",
    "                \"--num_modes\",\n",
    "                str(num_poses),\n",
    "                \"--exhaustiveness\",\n",
    "                str(exhaustiveness),\n",
    "            ]\n",
    "            + ([\"--log\", str(output_path) + \"_log.txt\"] if log else [])\n",
    "            + ([\"--seed\", str(random_seed)] if random_seed != \"\" else [])\n",
    "        )\n",
    "\n",
    "        output_text = subprocess.check_output(\n",
    "            smina_command,\n",
    "            universal_newlines=True,\n",
    "        )\n",
    "        return output_text\n",
    "\n",
    "    @staticmethod\n",
    "    def convert_log_to_dataframe(raw_log):\n",
    "        \"\"\"\n",
    "        Convert docking's raw output log into a Pandas DataFrame.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        raw_log : str\n",
    "            Raw output log generated after docking.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            Pandas DataFrame\n",
    "            DataFrame containing columns 'mode', 'affinity[kcal/mol]',\n",
    "            'dist from best mode_rmsd_l.b', and 'dist from best mode_rmsd_u.b'\n",
    "            for each generated docking pose.\n",
    "        \"\"\"\n",
    "\n",
    "        # Remove the unnecessary parts and extract the results table as list of lines\n",
    "        # The table starts after the line containing: -----+------------+----------+----------\n",
    "        # and ends before the word \"Refine\"\n",
    "        log = (\n",
    "            raw_log.split(\"-----+------------+----------+----------\")[1]\n",
    "            .split(\"Refine\")[0]\n",
    "            .strip()\n",
    "            .split(\"\\n\")\n",
    "        )\n",
    "\n",
    "        # parse each line and remove everything except the numbers\n",
    "        for index in range(len(log)):\n",
    "            # turn each line into a list\n",
    "            log[index] = log[index].strip().split(\" \")\n",
    "            # First element is the mode, which is an int.\n",
    "            # The rest of the elements are either empty strings, or floats\n",
    "            # Elements that are not empty strings should be extracted\n",
    "            # (first element as int and the rest as floats)\n",
    "            log[index] = [int(log[index][0])] + [\n",
    "                float(value) for value in log[index][1:] if value != \"\"\n",
    "            ]\n",
    "\n",
    "        df = pd.DataFrame(\n",
    "            log,\n",
    "            columns=[\n",
    "                \"mode\",\n",
    "                \"affinity[kcal/mol]\",\n",
    "                \"dist from best mode_rmsd_l.b\",\n",
    "                \"dist from best mode_rmsd_u.b\",\n",
    "            ],\n",
    "        )\n",
    "        df.index = df[\"mode\"]\n",
    "        df.drop(\"mode\", axis=1, inplace=True)\n",
    "        return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Pipeline's ***Docking*** Class\n",
    "We can now implement the ***Docking*** class, which automatically performs docking experiments on all provided ligands, and processes the results in several useful ways. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Docking:\n",
    "    \"\"\"\n",
    "    Automated docking process of the pipeline.\n",
    "    Take in a Protein and a list of ligands, and\n",
    "    dock each ligand on the protein, using the given specifications.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Protein_object : object; instance of Protein class\n",
    "        The protein to perform docking on.\n",
    "    list_Ligand_objects : list of Ligand objects (instances of Ligand class)\n",
    "        List of ligands to dock on the protein.\n",
    "    DockingSpecs_object : object; instance of Specs.Docking class\n",
    "        Specifications for the docking experiment.\n",
    "    docking_output_path : str or pathlib.Path object\n",
    "        Output folder path to store the docking data in.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        Protein_object,\n",
    "        list_Ligand_objects,\n",
    "        DockingSpecs_object,\n",
    "        docking_output_path,\n",
    "    ):\n",
    "        self.pdb_filepath_extracted_protein = docking_output_path / (\n",
    "            Protein_object.pdb_code + \"_extracted_protein.pdb\"\n",
    "        )\n",
    "        Protein_object.Universe = PDB.extract_molecule_from_pdb_file(\n",
    "            \"protein\", Protein_object.pdb_filepath, self.pdb_filepath_extracted_protein\n",
    "        )\n",
    "\n",
    "        self.pdbqt_filepath_extracted_protein = docking_output_path / (\n",
    "            Protein_object.pdb_code + \"_extracted_protein_ready_for_docking.pdbqt\"\n",
    "        )\n",
    "\n",
    "        OBabel.create_pdbqt_from_pdb_file(\n",
    "            self.pdb_filepath_extracted_protein, self.pdbqt_filepath_extracted_protein\n",
    "        )\n",
    "\n",
    "        temp_list_results_df = []\n",
    "        temp_list_master_df = []\n",
    "\n",
    "        for ligand in list_Ligand_objects:\n",
    "            ligand.pdbqt_filepath = docking_output_path / (\"CID_\" + ligand.cid + \".pdbqt\")\n",
    "            OBabel.create_pdbqt_from_smiles(ligand.remove_counterion(), ligand.pdbqt_filepath)\n",
    "\n",
    "            ligand.docking_poses_filepath = docking_output_path / (\n",
    "                \"CID_\" + ligand.cid + \"_docking_poses.pdbqt\"\n",
    "            )\n",
    "\n",
    "            raw_log = Smina.dock(\n",
    "                ligand.pdbqt_filepath,\n",
    "                self.pdbqt_filepath_extracted_protein,\n",
    "                Protein_object.binding_site_coordinates[\"center\"],\n",
    "                Protein_object.binding_site_coordinates[\"size\"],\n",
    "                str(ligand.docking_poses_filepath).split(\".\")[0],\n",
    "                output_format=\"pdbqt\",\n",
    "                num_poses=DockingSpecs_object.num_poses_per_ligand,\n",
    "                exhaustiveness=DockingSpecs_object.exhaustiveness,\n",
    "                random_seed=DockingSpecs_object.random_seed,\n",
    "                log=True,\n",
    "            )\n",
    "\n",
    "            ligand.docking_poses_split_filepaths = OBabel.split_multistructure_file(\n",
    "                \"pdbqt\", ligand.docking_poses_filepath\n",
    "            )\n",
    "\n",
    "            # Assigning the the dataframe of the Smina output\n",
    "            # to the ligand's attribute 'dataframe_docking'\n",
    "            df = Smina.convert_log_to_dataframe(raw_log)\n",
    "            ligand.dataframe_docking = df.copy()\n",
    "\n",
    "            # Extracting some useful information from the Smina-output dataframe\n",
    "            # and assigning them as separate ligand attributes\n",
    "            # Adding the same summarized information the ligand's general dataframe as well\n",
    "            ligand.dataframe.loc[\"binding_affinity_best\"] = ligand.binding_affinity_best = df[\n",
    "                \"affinity[kcal/mol]\"\n",
    "            ].min()\n",
    "            ligand.dataframe.loc[\"binding_affinity_mean\"] = ligand.binding_affinity_mean = df[\n",
    "                \"affinity[kcal/mol]\"\n",
    "            ].mean()\n",
    "            ligand.dataframe.loc[\"binding_affinity_std\"] = ligand.binding_affinity_std = df[\n",
    "                \"affinity[kcal/mol]\"\n",
    "            ].std()\n",
    "            ligand.dataframe.loc[\n",
    "                \"docking_poses_dist_rmsd_lb_mean\"\n",
    "            ] = ligand.docking_poses_dist_rmsd_lb_mean = df[\"dist from best mode_rmsd_l.b\"].mean()\n",
    "            ligand.dataframe.loc[\n",
    "                \"docking_poses_dist_rmsd_lb_std\"\n",
    "            ] = ligand.docking_poses_dist_rmsd_lb_std = df[\"dist from best mode_rmsd_l.b\"].std()\n",
    "            ligand.dataframe.loc[\n",
    "                \"docking_poses_dist_rmsd_ub_mean\"\n",
    "            ] = ligand.docking_poses_dist_rmsd_ub_mean = df[\"dist from best mode_rmsd_u.b\"].mean()\n",
    "            ligand.dataframe.loc[\n",
    "                \"docking_poses_dist_rmsd_ub_std\"\n",
    "            ] = ligand.docking_poses_dist_rmsd_ub_std = df[\"dist from best mode_rmsd_u.b\"].std()\n",
    "\n",
    "            df[\"CID\"] = ligand.cid\n",
    "            df[\"drug_score_total\"] = ligand.drug_score_total\n",
    "            df.set_index([\"CID\", df.index], inplace=True)\n",
    "\n",
    "            master_df = df.copy()\n",
    "            master_df[\"filepath\"] = ligand.docking_poses_split_filepaths\n",
    "\n",
    "            temp_list_results_df.append(df)\n",
    "            temp_list_master_df.append(master_df)\n",
    "\n",
    "        self.results_dataframe = pd.concat(temp_list_results_df)\n",
    "        self.master_df = pd.concat(temp_list_master_df)\n",
    "        self.results_dataframe.to_csv(docking_output_path / \"Results_Summary.csv\")\n",
    "\n",
    "    def visualize_all_poses(self):\n",
    "        \"\"\"\n",
    "        Visualize docking poses of a all analogs, using NGLView.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLViewer object\n",
    "            Interactive viewer of all analogs' docking poses,\n",
    "            sorted by their binding affinities.\n",
    "        \"\"\"\n",
    "        df = self.master_df.sort_values(by=[\"affinity[kcal/mol]\", \"CID\", \"mode\"])\n",
    "        self.visualize(df)\n",
    "        return\n",
    "\n",
    "    def visualize_analog_poses(self, cid):\n",
    "        \"\"\"\n",
    "        Visualize docking poses of a certain analog, using NGLView.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        cid : str or int\n",
    "            CID of the analog.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLViewer object\n",
    "            Interactive viewer of given analog's docking poses,\n",
    "            sorted by their binding affinities.\n",
    "        \"\"\"\n",
    "        df = self.master_df.xs(str(cid), level=0, axis=0, drop_level=False)\n",
    "        self.visualize(df)\n",
    "        return\n",
    "\n",
    "    def visualize(self, fitted_master_df):\n",
    "        \"\"\"\n",
    "        Visualize any collection of docking poses, using NGLView.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        fitted_master_df : Pandas DataFrame\n",
    "            Any section of the master docking dataframe,\n",
    "            stored under self.master_df.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            NGLViewer object\n",
    "            Interactive viewer of given analog's docking poses,\n",
    "            sorted by their binding affinities.\n",
    "        \"\"\"\n",
    "        list_docking_poses_labels = list(\n",
    "            map(lambda x: x[0] + \" - \" + str(x[1]), fitted_master_df.index.tolist())\n",
    "        )\n",
    "        NGLView.docking(\n",
    "            self.pdb_filepath_extracted_protein,\n",
    "            \"pdb\",\n",
    "            fitted_master_df[\"filepath\"].tolist(),\n",
    "            \"pdbqt\",\n",
    "            list_docking_poses_labels,\n",
    "            fitted_master_df[\"affinity[kcal/mol]\"].tolist(),\n",
    "        )\n",
    "        return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Performing Docking Experiments on The Selected Analogs: Instantiating The ***Docking*** Class\n",
    "<br>\n",
    "<div style=\"text-align: justify\">By instantiating the <b><i>Docking</i></b> class with the <b><i>Protein</i></b> object, the list of analogs, and the <b><i>Specs.Docking</i></b> data-class of the project, analogs are automatically docked onto the protein, and the results are stored separately for each docking pose. Moreover, some meaningful information is extracted from the results of all docking poses for each analog, and stored separately. We instantiate the class and assign it to our project:\n",
    "<br><br>\n",
    "<b><i>Note:</i></b> This part is the most computationally intense process of the pipeline, and will take between 5-10 minutes (for 20 ligands) to complete.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project1.Docking = Docking(\n",
    "    project1.Protein,\n",
    "    list(project1.Ligand.analogs.values()),\n",
    "    project1.Specs.Docking,\n",
    "    project1.Specs.OutputPaths.docking,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now view all the calculated docking parameters for each docking pose of each analog (here showing only the 5 first enteries):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Docking.results_dataframe.sort_values(by=\"affinity[kcal/mol]\").head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, by accessing a specific analog, we can view the full results for that analog using the attribute ***dataframe_docking***:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"11292933\"].dataframe_docking"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A summary of the docking results (e.g. highest/mean binding affinities) are also added to the main dataframe of each analog, and can be viewed, e.g. by calling its object (here showing only the relevant rows, i.e. the last 7 rows):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"11292933\"]().tail(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same summary results are also added as instance attributes for each object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"11292933\"].binding_affinity_best"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now with the help of the <b><i>NGLView</i></b> helper-class we had defined earlier, we have also added methods to the <b><i>Docking</i></b> class for visualization of the docking poses. For example, we can just view all docking poses together in an interactive way, using the <b><i>visualize_all_poses</i></b> method. In this method, the docking poses are sorted by their binding affinities, and labeled by their CID and respective docking pose number (i.e. mode). By selecting an analog from the menu below, the viewer automatically shows the protein residues in close proximity (i.e. 6 Å) of the ligand, as well as its corresponding binding affinity.\n",
    "<br><br>\n",
    "<b><i>Note:</i></b> If we are interested in visualization of a certain analog's docking poses, we can use the <b><i>visualize_analog_poses</i></b> method instead, and provide the analog's CID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Docking.visualize_all_poses()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's also dock the input ligand in order to be able to compare the results later and see whether any of the analogs has a higher binding affinity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.Docking = Docking(\n",
    "    project1.Protein,\n",
    "    [project1.Ligand],\n",
    "    project1.Specs.Docking,\n",
    "    project1.Specs.OutputPaths.ligand,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similar to the analogs, the docking results of the input ligand is also stored in it's object. For example, to see the docking dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.dataframe_docking"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='8_practical'></a>\n",
    "### Analysis of Protein–Ligand Interactions\n",
    "With the calculated docking poses of each analog in hand, we can now focus on analyzing the protein-ligand interactions in each docking pose of each analog. For the analysis we use the ***PLIP*** package, for which we first define a helper-class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Required Functions: The ***PLIP*** Helper-Class\n",
    "Here will use the [***PLIP***](https://github.com/pharmai/plip) package to implement the functions needed to analyze the protein-ligand interactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PLIP:\n",
    "    \"\"\"\n",
    "    Set of functions required to analyze protein-ligand interactions using the PLIP package.\n",
    "    \"\"\"\n",
    "\n",
    "    class Consts:\n",
    "        class InteractionTypes(Enum):\n",
    "            H_BOND = \"hbond\"\n",
    "            HYDROPHOBIC = \"hydrophobic\"\n",
    "            SALT_BRIDGE = \"saltbridge\"\n",
    "            WATER_BRIDGE = \"waterbridge\"\n",
    "            PI_STACKING = \"pistacking\"\n",
    "            PI_CATION = \"pication\"\n",
    "            HALOGEN = \"halogen\"\n",
    "            METAL = \"metal\"\n",
    "\n",
    "    @staticmethod\n",
    "    def calculate_interactions(pdb_filepath):\n",
    "        \"\"\"\n",
    "        Calculate protein-ligand interactions in a PDB file.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        pdb_filepath : str or pathlib.Path object\n",
    "            Filepath of the PDB file containing the protein-ligand complex.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            dict of dicts\n",
    "            Dictionary of all different interaction data for all detected ligands.\n",
    "            The keys of first dictionary correspond to the ligand-IDs of detected ligands in the PDB file.\n",
    "            The keys of each sub-dictionary correspond to interaction types, as defined in PLIP.Consts.InteractionTypes.\n",
    "        \"\"\"\n",
    "        protein_ligand_complex = PDBComplex()\n",
    "        protein_ligand_complex.load_pdb(str(pdb_filepath))\n",
    "\n",
    "        for ligand in protein_ligand_complex.ligands:\n",
    "            protein_ligand_complex.characterize_complex(ligand)\n",
    "\n",
    "        all_ligands_interactions = {}\n",
    "\n",
    "        for ligand, ligand_binding_site in sorted(protein_ligand_complex.interaction_sets.items()):\n",
    "\n",
    "            interaction_object = BindingSiteReport(\n",
    "                ligand_binding_site\n",
    "            )  # collect data about interactions\n",
    "\n",
    "            interaction_data = {\n",
    "                interaction_type.value: (\n",
    "                    [getattr(interaction_object, interaction_type.value + \"_features\")]\n",
    "                    + getattr(interaction_object, interaction_type.value + \"_info\")\n",
    "                )\n",
    "                for interaction_type in PLIP.Consts.InteractionTypes\n",
    "            }\n",
    "\n",
    "            all_ligands_interactions[ligand] = interaction_data\n",
    "\n",
    "        return all_ligands_interactions\n",
    "\n",
    "    @staticmethod\n",
    "    def create_dataframe_of_ligand_interactions(ligand_interaction_data, interaction_type):\n",
    "        \"\"\"\n",
    "        Create a Pandas DataFrame from interaction data of a specific ligand,\n",
    "        for a specific interaction type.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ligand_interaction_data : dict\n",
    "            Interaction data calculated by the 'calculate_interactions' function,\n",
    "            where only one ligand's interaction data is chosen from the output of\n",
    "            that function.\n",
    "        interaction_type : Enum\n",
    "            One of the enumerations from PLIP.Consts.InteractionTypes,\n",
    "            specifying the type of interaction, for which the DataFrame should be created.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            pandas DataFrame\n",
    "            DataFrame containing all the information for the specified interactions\n",
    "            in the input ligand interaction data.\n",
    "        \"\"\"\n",
    "\n",
    "        interaction_df = pd.DataFrame.from_records(\n",
    "            # interaction data are stored after the first element\n",
    "            ligand_interaction_data[interaction_type.value][1:],\n",
    "            # the first element corresponds to interaction parameters,\n",
    "            # which are used here as column names.\n",
    "            columns=ligand_interaction_data[interaction_type.value][0],\n",
    "        )\n",
    "\n",
    "        return interaction_df\n",
    "\n",
    "    @staticmethod\n",
    "    def create_protein_ligand_complex(\n",
    "        protein_pdbqt_filepath, docking_pose_pdbqt_filepath, ligand_id, output_filepath\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Create a protein-ligand-complex PDB file out of separate protein and ligand files.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        protein_pdbqt_filepath : str or pathlib.Path object\n",
    "            Filepath of the PDB file containing the protein.\n",
    "        docking_pose_pdbqt_filepath : str or pathlib.Path object\n",
    "            Filepath of the PDB file containing the ligand.\n",
    "        ligand_id : str\n",
    "            An identifier for the ligand to write into the PDB file.\n",
    "        output_filepath : str or pathlib.Path object\n",
    "            Output filepath of the PDB file containing the protein-ligand complex.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            pathlib.Path object\n",
    "            Complete filepath of the created protein-ligand-complex PDB file.\n",
    "        \"\"\"\n",
    "\n",
    "        def pdbqt_to_pdbblock(pdbqt_filepath):\n",
    "            lines = []\n",
    "            with open(pdbqt_filepath) as file:\n",
    "                for line in file:\n",
    "                    if line[:4] == \"ATOM\":\n",
    "                        lines.append(line[:67].strip())\n",
    "            return \"\\n\".join(lines)\n",
    "\n",
    "        protein_pdbblock = pdbqt_to_pdbblock(protein_pdbqt_filepath)\n",
    "\n",
    "        ligand_pdbblock = pdbqt_to_pdbblock(docking_pose_pdbqt_filepath)\n",
    "        # ligand_pdbblock = ligand_pdbblock.replace('ATOM', 'HETATM')\n",
    "        # ligand_pdbblock = ligand_pdbblock.replace('UNL     ', (ligand_id[:8]+\n",
    "        #                                    \" \"*(8-len(ligand_id))))\n",
    "\n",
    "        full_output_filepath = str(output_filepath) + \".pdb\"\n",
    "        with open(full_output_filepath, \"w\") as file:\n",
    "            file.write(protein_pdbblock)\n",
    "            file.write(f\"\\nCOMPND    {ligand_id}\\n\")\n",
    "            file.write(ligand_pdbblock)\n",
    "\n",
    "        return Path(full_output_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implementing The Pipeline's ***InteractionAnalysis*** Class\n",
    "We can now implement the ***InteractionAnalysis*** class, which automatically calculates all interaction information for each docking pose of each ligand. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class InteractionAnalysis:\n",
    "    \"\"\"\n",
    "    Automated protein-ligand interaction analysis process of the pipeline.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    separated_protein_pdbqt_filepath : str or pathlib.Path object\n",
    "        Filepath of the separated protein PDBQT file used in docking.\n",
    "    list_Ligand_objects : list of Ligand objects (instances of Ligand class)\n",
    "        List of ligands to analyze their interactions with the protein.\n",
    "    docking_results_df : pandas DataFrame\n",
    "        Summary dataframe created by the docking class.\n",
    "    InteractionAnalysisSpecs_object : object; instance of Specs.InteractionAnalysis class\n",
    "        Specifications for the interaction analysis processes.\n",
    "    interaction_analysis_output_path : str or pathlib.Path object\n",
    "        Output folder path to store the analyzed data in.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        separated_protein_pdbqt_filepath,\n",
    "        separated_protein_pdb_filepath,\n",
    "        protein_first_residue_number,\n",
    "        list_Ligand_objects,\n",
    "        docking_master_df,\n",
    "        InteractionAnalysisSpecs_object,\n",
    "        interaction_analysis_output_path,\n",
    "    ):\n",
    "\n",
    "        self._analogs = list_Ligand_objects\n",
    "        self._pdb_filepath_extracted_protein = separated_protein_pdb_filepath\n",
    "\n",
    "        if InteractionAnalysisSpecs_object.program is Consts.InteractionAnalysis.Programs.PLIP:\n",
    "            results_df = docking_master_df.copy()\n",
    "            results_df.drop(\"filepath\", axis=1, inplace=True)\n",
    "            results_df[\"total_num_interactions\"] = 0\n",
    "\n",
    "            interaction_master_df = docking_master_df.copy()\n",
    "\n",
    "            for analog in list_Ligand_objects:\n",
    "                analog.dataframe.loc[\"average_num_total_interactions\", \"Value\"] = 0\n",
    "                for interaction_type in PLIP.Consts.InteractionTypes:\n",
    "                    analog.dataframe.loc[\n",
    "                        f\"average_num_{interaction_type.name.lower()}\", \"Value\"\n",
    "                    ] = 0\n",
    "\n",
    "                for index, docking_pose_filepath in zip(\n",
    "                    range(len(analog.docking_poses_split_filepaths)),\n",
    "                    analog.docking_poses_split_filepaths,\n",
    "                ):\n",
    "                    analog.protein_complex_filepath = PLIP.create_protein_ligand_complex(\n",
    "                        separated_protein_pdbqt_filepath,\n",
    "                        docking_pose_filepath,\n",
    "                        analog.cid,\n",
    "                        interaction_analysis_output_path / (f\"CID_{analog.cid}_{index+1}\"),\n",
    "                    )\n",
    "\n",
    "                    # interaction_data will be dict of dicts, where each of the\n",
    "                    # outer dict's items correspond to a ligand found in the pdb file\n",
    "                    interaction_data = PLIP.calculate_interactions(analog.protein_complex_filepath)\n",
    "\n",
    "                    # Since we are only passing PDB files with a single ligand, the outer dict\n",
    "                    # will only have one item, which we extract:\n",
    "                    ligand_interaction_data = interaction_data[list(interaction_data.keys())[0]]\n",
    "                    # This extracted item is again a dict, where items correspond to\n",
    "                    # different interaction-types.\n",
    "\n",
    "                    # NOTICE: when using pybel to create the protein PDBQT file\n",
    "                    # for the docking experiment, the residue numbers are reset (i.e. start at 1).\n",
    "                    # Since this PDBQT file is also used to create a protein-ligand complext file\n",
    "                    # for interaction-analysis with PLIP, the residue numbers in PLIP are also affected.\n",
    "                    # Now we have to fix this here, before further processing the PLIP data.\n",
    "                    # To do so, we simply loop through all the interaction data to find all the\n",
    "                    # residue numbers, add the protein's first residue number to them, and subtract 1,\n",
    "                    # so that we get back the original residue numbers. Also, since the data are stored\n",
    "                    # in tuples, we have to convert them to a list first, in order to be able to change them.\n",
    "                    for certain_interactions_data in ligand_interaction_data.values():\n",
    "                        for single_interaction_data in range(1, len(certain_interactions_data)):\n",
    "                            list_from_tuple = list(\n",
    "                                certain_interactions_data[single_interaction_data]\n",
    "                            )\n",
    "                            list_from_tuple[0] += protein_first_residue_number - 1\n",
    "                            certain_interactions_data[single_interaction_data] = tuple(\n",
    "                                list_from_tuple\n",
    "                            )\n",
    "\n",
    "                    interaction_master_df.loc[(analog.cid, index + 1), \"plip_dict\"] = [\n",
    "                        interaction_data[list(interaction_data.keys())[0]]\n",
    "                    ]\n",
    "\n",
    "                    for interaction_type in PLIP.Consts.InteractionTypes:\n",
    "                        df = PLIP.create_dataframe_of_ligand_interactions(\n",
    "                            interaction_data[list(interaction_data.keys())[0]],\n",
    "                            interaction_type,\n",
    "                        )\n",
    "                        setattr(\n",
    "                            analog,\n",
    "                            f\"docking_pose_{index+1}_interactions_{interaction_type.name.lower()}\",\n",
    "                            df,\n",
    "                        )\n",
    "\n",
    "                        results_df.loc[\n",
    "                            (analog.cid, index + 1), interaction_type.name.lower()\n",
    "                        ] = len(df)\n",
    "                        analog.dataframe.loc[\n",
    "                            f\"average_num_{interaction_type.name.lower()}\", \"Value\"\n",
    "                        ] += len(df)\n",
    "                        analog.dataframe.loc[\"average_num_total_interactions\", \"Value\"] += len(df)\n",
    "\n",
    "                        results_df[interaction_type.name.lower()] = pd.to_numeric(\n",
    "                            results_df[interaction_type.name.lower()],\n",
    "                            downcast=\"integer\",\n",
    "                        )\n",
    "                        results_df.loc[(analog.cid, index + 1), \"total_num_interactions\"] += len(\n",
    "                            df\n",
    "                        )\n",
    "\n",
    "                analog.num_total_interactions_highest = results_df.loc[\n",
    "                    analog.cid, \"total_num_interactions\"\n",
    "                ].max()\n",
    "                analog.dataframe.loc[\"average_num_total_interactions\", \"Value\"] /= len(\n",
    "                    analog.docking_poses_split_filepaths\n",
    "                )\n",
    "                analog.num_total_interactions_mean = analog.dataframe.loc[\n",
    "                    \"average_num_total_interactions\", \"Value\"\n",
    "                ]\n",
    "                for interaction_type in PLIP.Consts.InteractionTypes:\n",
    "                    analog.dataframe.loc[\n",
    "                        f\"average_num_{interaction_type.name.lower()}\", \"Value\"\n",
    "                    ] /= len(analog.docking_poses_split_filepaths)\n",
    "\n",
    "            results_df[\"total_num_interactions\"] = pd.to_numeric(\n",
    "                results_df[\"total_num_interactions\"], downcast=\"integer\"\n",
    "            )\n",
    "            self.results = results_df\n",
    "            self.master_df = interaction_master_df\n",
    "\n",
    "    def find_poses_with_specific_interactions(self, list_interaction_data, all_or_any):\n",
    "        \"\"\"\n",
    "        Find docking poses containing a specific set of interactions.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        list_interaction_data : list of lists\n",
    "            List of desired interactions. Each sub-list should have the following format:\n",
    "            [interaction_type, residue_number]\n",
    "            interaction_type : str\n",
    "                Type of desired interaction. Allowed values are:\n",
    "                'h_bond', 'hydrophobic', 'salt_bridge', 'water_bridge',\n",
    "                'pi_stacking', 'pi_cation', 'halogen', 'metal'\n",
    "            residue_nr : int\n",
    "                Residue number involved in the given interaction_type\n",
    "            Example: [[\"h_bond\", 793], [\"hydrophobic\", 860]]\n",
    "        all_or_any : str\n",
    "            Allowed values: \"all\", \"any\"\n",
    "            \"all\": all given interactions should be present in a docking pose.\n",
    "            \"any\": it is enough when one of the given interactions is present in a docking pose.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            list of tuples\n",
    "            List of identifiers for the docking poses in the format:\n",
    "            (CID, pose_nr)\n",
    "            CID : str\n",
    "                CID of the analog with the eligible docking pose\n",
    "            pose_nr : int\n",
    "                Docking pose number of the eligible docking pose\n",
    "        \"\"\"\n",
    "        list_eligible_analog_cid_docking_pose_nr_tuple = []\n",
    "        for analog in self._analogs:\n",
    "            for analog_docking_pose_nr in range(1, len(analog.dataframe_docking) + 1):\n",
    "                num_hits_in_analog_docking_pose = 0\n",
    "                for interaction_type, residue_nr in list_interaction_data:\n",
    "                    interaction_df = getattr(\n",
    "                        analog,\n",
    "                        f\"docking_pose_{analog_docking_pose_nr}_interactions_{interaction_type}\",\n",
    "                    )\n",
    "                    if residue_nr in interaction_df[\"RESNR\"].values:\n",
    "                        num_hits_in_analog_docking_pose += 1\n",
    "                if (\n",
    "                    (all_or_any == \"all\")\n",
    "                    and (num_hits_in_analog_docking_pose == len(list_interaction_data))\n",
    "                ) or ((all_or_any == \"any\") and (num_hits_in_analog_docking_pose != 0)):\n",
    "                    list_eligible_analog_cid_docking_pose_nr_tuple.append(\n",
    "                        (analog.cid, analog_docking_pose_nr)\n",
    "                    )\n",
    "        return list_eligible_analog_cid_docking_pose_nr_tuple\n",
    "\n",
    "    def visualize_all_interactions(self):\n",
    "        \"\"\"\n",
    "        Visualize docking poses of a all analogs, using NGLView.\n",
    "        The docking poses are sorted by their binding affinities in a menu.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        df = self.master_df.sort_values(by=[\"affinity[kcal/mol]\", \"CID\", \"mode\"])\n",
    "        view = self.visualize(df)\n",
    "        return view\n",
    "\n",
    "    def visualize_analog_interactions(self, cid):\n",
    "        \"\"\"\n",
    "        Visualize interactions in the docking poses of a certain analog, using NGLView.\n",
    "        The docking poses are sorted by their binding affinities in a menu.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        cid : str or int\n",
    "            CID of the analog.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        df = self.master_df.xs(str(cid), level=0, axis=0, drop_level=False)\n",
    "        view = self.visualize(df)\n",
    "        return view\n",
    "\n",
    "    def visualize_docking_poses_interactions(self, list_of_cid_pose_nr_tuples):\n",
    "        \"\"\"\n",
    "        Visualize interactions in a given set of docking poses, using NGLView.\n",
    "        The docking poses are sorted in a menu, in the same order as the given list.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        list_of_cid_pose_nr_tuples : list of tuples\n",
    "            List of identifiers for the docking poses in the format:\n",
    "            (CID, pose_nr)\n",
    "            CID : str\n",
    "                CID of the analog\n",
    "            pose_nr : int\n",
    "                Docking pose number\n",
    "            Example: [(\"184614\", 1), (\"3628\", 4)]\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        df = self.master_df.loc[list_of_cid_pose_nr_tuples]\n",
    "        view = self.visualize(df)\n",
    "        return view\n",
    "\n",
    "    def visualize(self, fitted_master_df):\n",
    "        \"\"\"\n",
    "        Visualize interactions in any collection of docking poses, using NGLView.\n",
    "        The docking poses are sorted by their given order in a menu.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        fitted_master_df : Pandas DataFrame\n",
    "            Any section of the master InteractionAnalysis dataframe,\n",
    "            stored under self.master_df.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        list_docking_poses_labels = list(\n",
    "            map(lambda x: x[0] + \" - \" + str(x[1]), fitted_master_df.index.tolist())\n",
    "        )\n",
    "        view = NGLView.interactions(\n",
    "            self._pdb_filepath_extracted_protein,\n",
    "            \"pdb\",\n",
    "            fitted_master_df[\"filepath\"].tolist(),\n",
    "            \"pdbqt\",\n",
    "            list_docking_poses_labels,\n",
    "            fitted_master_df[\"affinity[kcal/mol]\"].tolist(),\n",
    "            fitted_master_df[\"plip_dict\"].tolist(),\n",
    "        )\n",
    "        return view\n",
    "\n",
    "    def plot_interaction_affinity_correlation(self):\n",
    "        \"\"\"\n",
    "        View a correlation plot between binding affinity and number of interactions\n",
    "        in each docking pose.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            None\n",
    "        \"\"\"\n",
    "        df = self.results.sort_values(by=\"affinity[kcal/mol]\", ascending=True)\n",
    "\n",
    "        fig, ax1 = plt.subplots()\n",
    "        color = \"tab:red\"\n",
    "        ax1.set_ylabel(\"Binding Affinity [kcal/mol]\", color=color)\n",
    "        ax1.tick_params(axis=\"y\", labelcolor=color)\n",
    "        ax1.plot(\n",
    "            list(map(abs, df[\"affinity[kcal/mol]\"].tolist())),\n",
    "            linewidth=0.5,\n",
    "            linestyle=\"--\",\n",
    "            color=\"r\",\n",
    "            marker=\".\",\n",
    "            markersize=2,\n",
    "            markerfacecolor=\"blue\",\n",
    "            markeredgecolor=\"red\",\n",
    "        )\n",
    "\n",
    "        ax2 = ax1.twinx()\n",
    "        color = \"tab:blue\"\n",
    "        ax2.set_ylabel(\"Number of Interactions\", color=color)\n",
    "        ax2.tick_params(axis=\"y\", labelcolor=color)\n",
    "\n",
    "        ax2.plot(\n",
    "            df[\"total_num_interactions\"].tolist(),\n",
    "            linewidth=0.5,\n",
    "            label=\"Total Interactions\",\n",
    "            color=\"black\",\n",
    "        )\n",
    "        ax2.plot(\n",
    "            df[\"h_bond\"].tolist(),\n",
    "            linewidth=0.5,\n",
    "            label=\"H-bond Interactions\",\n",
    "            color=\"g\",\n",
    "            linestyle=\"--\",\n",
    "        )\n",
    "        ax2.plot(\n",
    "            df[\"hydrophobic\"].tolist(),\n",
    "            linewidth=0.5,\n",
    "            label=\"Hydrophobic Interactions\",\n",
    "            linestyle=\"dashed\",\n",
    "        )\n",
    "\n",
    "        plt.legend(fontsize=7)\n",
    "        plt.show()\n",
    "        return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Analyzing the Protein-Ligand Interactions in Calculated Docking Poses: Instantiating The ***InteractionAnalysis*** Class\n",
    "The ***InteractionAnalysis*** class can be instantiated by providing the filepaths of the separated protein structure, a list of all analogs (as ***Ligand*** objects), the results dataframe of the docking process, and the output path for storing the interaction analysis data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project1.InteractionAnalysis = InteractionAnalysis(\n",
    "    project1.Docking.pdbqt_filepath_extracted_protein,\n",
    "    project1.Docking.pdb_filepath_extracted_protein,\n",
    "    project1.Protein.residue_number_first,\n",
    "    list(project1.Ligand.analogs.values()),\n",
    "    project1.Docking.master_df,\n",
    "    project1.Specs.InteractionAnalysis,\n",
    "    project1.Specs.OutputPaths.interaction_analysis,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The interactions can now be inspected collectively for all docking poses of all analogs. Here, only the number of interactions are recorded for each interaction type (showing only the 5 first entries):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.InteractionAnalysis.results.sort_values(\n",
    "    by=\"total_num_interactions\", ascending=False\n",
    ").head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we are interested in the details of a specific interaction type for a specific docking pose, we can access these information from the corresponding ***Ligand*** objects of the analogs. For example, accessing the data for hydrophobic interactions of the docking pose 1 of analog 65997 (showing only the 5 first entries):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"65997\"].docking_pose_1_interactions_hydrophobic.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or hydrogen-bond interactions of docking pose 2 of analog 11292933:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"11292933\"].docking_pose_2_interactions_h_bond.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, a summary of interaction analysis data are also added to each analog's main dataframe, and can be viewed, for example by calling the object (here showing only the relevant data, i.e. the last 9 rows):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.analogs[\"11292933\"]().tail(9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also, let's use ***plot_interaction_affinity_correlation*** function see if there is any visible correlation between the calculated binding affinities and the number of interactions for each docking pose:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.InteractionAnalysis.plot_interaction_affinity_correlation()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen, no obvious correlation is visible between the two sets of data; The number of total interactions seems to be very weakly correlated with the binding affinity, however, with several outliers."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now with the help of the <b><i>NGLView</i></b> helper-class we had defined earlier, we have also added methods to the <b><i>InteractionAnalysis</i></b> class for visualization of the protein-ligand interactions. For example, we can just view the interactions for all docking poses in an interactive way. Here, by selecting each docking pose (labeled by their CID and mode-number, and sorted by their binding affinities), all the interacting residues are shown, and the interactions are visualized with colored lines, for which a color-map is also provided.\n",
    "<br><br>\n",
    "<b><i>Note:</i></b> If we are interested in visualization of a certain analog's interactions, we can use the <b><i>visualize_analog_interactions</i></b> method instead, and provide the analog's CID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.InteractionAnalysis.visualize_all_interactions()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also search for docking poses with a specific set of interactions with specific residues of the protein, using the ***find_poses_with_specific_interactions*** method. For example, many EGFR inhibitors are involved in hydrogen-bonding with the M793 residue of the protein. To find the docking poses that exhibit this interaction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "desired_interactions = project1.InteractionAnalysis.find_poses_with_specific_interactions(\n",
    "    [[\"h_bond\", 793]], \"any\"\n",
    ")\n",
    "desired_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we can also visualize the results separately, instead of looking for them in the menu of the viewer above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.InteractionAnalysis.visualize_docking_poses_interactions(desired_interactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's also analyze the interactions in the docking poses of the input ligand:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.InteractionAnalysis = InteractionAnalysis(\n",
    "    project1.Docking.pdbqt_filepath_extracted_protein,\n",
    "    project1.Docking.pdb_filepath_extracted_protein,\n",
    "    project1.Protein.residue_number_first,\n",
    "    [project1.Ligand],\n",
    "    project1.Ligand.Docking.master_df,\n",
    "    project1.Specs.InteractionAnalysis,\n",
    "    project1.Specs.OutputPaths.ligand,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results can be viewed similar to the analogs. For example, for viewing the summary data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.Ligand.InteractionAnalysis.results.sort_values(\n",
    "    by=\"total_num_interactions\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='9_practical'></a>\n",
    "### Selection of the Best Optimized Ligand\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "At this point, we have carried out all the processes in the pipeline, and have gathered all the information in our <b><i>LeadOptimizationPipeline</i></b> instance, i.e. <b><i>project1</i></b>. The project now contains all the possible information required for selecting the most suited analog. The choice of selecting the best analog is highly dependent on the specific project at hand, and can be approached from different angles. To recap, we have gathered a number of analogs to the input ligand, and filtered them to choose those with the highest drug-likeness scores. We then calculated several docking poses and corresponding binding affinities for each analog, and analyzed their interactions with the protein. Now we define a class <b><i>OptimizedLigands</i></b>, which takes in the whole project, and based on the specifications in the input file, selects the best analog. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class OptimizedLigands:\n",
    "    \"\"\"\n",
    "    The automated selection process of optimized analog(s) at the end of the pipeline.\n",
    "    Take in the whole project, create a short summary of results, and select the best\n",
    "    optimized analogs based on user's specifications defined in the input data.\n",
    "    \"\"\"\n",
    "    def __init__(self, project):\n",
    "\n",
    "        self._project = project\n",
    "\n",
    "        df = project.InteractionAnalysis.results.rename(columns={\"affinity[kcal/mol]\": \"affinity\"})\n",
    "\n",
    "        self.higher_affinity_poses = (\n",
    "            df[df[\"affinity\"] < project.Ligand.binding_affinity_best]\n",
    "            .copy()\n",
    "            .sort_values(by=\"affinity\")\n",
    "        )\n",
    "        self.higher_affinity_analogs = [\n",
    "            project.Ligand.analogs[cid]\n",
    "            for cid in self.higher_affinity_poses.index.get_level_values(0).unique()\n",
    "        ]\n",
    "        self.higher_interacting_poses = (\n",
    "            df[\n",
    "                df[\"total_num_interactions\"]\n",
    "                > project.Ligand.InteractionAnalysis.results[\"total_num_interactions\"].max()\n",
    "            ]\n",
    "            .copy()\n",
    "            .sort_values(by=\"total_num_interactions\", ascending=False)\n",
    "        )\n",
    "        self.higher_interacting_analogs = [\n",
    "            project.Ligand.analogs[cid]\n",
    "            for cid in self.higher_interacting_poses.index.get_level_values(0).unique()\n",
    "        ]\n",
    "        self.higher_affinity_and_interacting_poses = df.loc[\n",
    "            self.higher_affinity_poses.index.intersection(self.higher_interacting_poses.index)\n",
    "        ]\n",
    "        self.higher_affinity_and_interacting_analogs = [\n",
    "            project.Ligand.analogs[cid]\n",
    "            for cid in self.higher_affinity_and_interacting_poses.index.get_level_values(\n",
    "                0\n",
    "            ).unique()\n",
    "        ]\n",
    "        self.higher_affinity_and_interacting_and_druglike_analogs = [\n",
    "            analog\n",
    "            for analog in self.higher_affinity_and_interacting_analogs\n",
    "            if analog.drug_score_total > project.Ligand.drug_score_total\n",
    "        ]\n",
    "\n",
    "        if (\n",
    "            project.Specs.OptimizedLigands.selection_method\n",
    "            is Consts.OptimizedLigands.SelectionMethods.SORTING\n",
    "        ):\n",
    "            df[\"affinity\"] = df[\"affinity\"].apply(abs)\n",
    "            final_results_cids = (\n",
    "                df.sort_values(\n",
    "                    by=project.Specs.OptimizedLigands.selection_criteria,\n",
    "                    ascending=False,\n",
    "                )\n",
    "                .index.get_level_values(0)\n",
    "                .unique()\n",
    "            )\n",
    "\n",
    "        elif (\n",
    "            project.Specs.OptimizedLigands.selection_method\n",
    "            is Consts.OptimizedLigands.SelectionMethods.FUNCTION\n",
    "        ):\n",
    "            df[\"function_score\"] = eval(selection_criteria)\n",
    "            final_results_cids = (\n",
    "                df.sort_values(by=\"function_score\", ascending=False)\n",
    "                .index.get_level_values(0)\n",
    "                .unique()\n",
    "            )\n",
    "\n",
    "        self.output = [project.Ligand.analogs[cid] for cid in final_results_cids][\n",
    "            : int(project.Specs.OptimizedLigands.num_results)\n",
    "        ]\n",
    "\n",
    "    def show_higher_affinity_analogs(self):\n",
    "        for analog in self.higher_affinity_analogs:\n",
    "            display(analog())\n",
    "\n",
    "    def show_higher_interacting_analogs(self):\n",
    "        for analog in self.higher_interacting_analogs:\n",
    "            display(analog())\n",
    "\n",
    "    def show_higher_affinity_and_interacting_analogs(self):\n",
    "        for analog in self.higheraffinity_and_interacting_analogs:\n",
    "            display(analog())\n",
    "\n",
    "    def show_higher_affinity_and_interacting_and_druglike_analogs(self):\n",
    "        for analog in self.higher_affinity_and_interacting_and_druglike_analogs:\n",
    "            display(analog)\n",
    "\n",
    "    def show_final_output(self):\n",
    "        for analog in self.output:\n",
    "            display(analog())\n",
    "\n",
    "    def __call__(self):\n",
    "        def pprint(text1, text2):\n",
    "            display(\n",
    "                Markdown(\n",
    "                    f\"<span style='color:blue'>{text1}</span><span style='color:black'>{text2}</span>\"\n",
    "                )\n",
    "            )\n",
    "\n",
    "        pprint(\n",
    "            \"Number of docking poses with higher binding affinity than highest binding affinity of ligand: \",\n",
    "            len(self.higher_affinity_poses),\n",
    "        )\n",
    "        pprint(\n",
    "            \"&nbsp;&nbsp;&nbsp;&nbsp;CIDs of analogs corresponding to these docking poses: \",\n",
    "            [analog.cid for analog in self.higher_affinity_analogs],\n",
    "        )\n",
    "        pprint(\n",
    "            \"Number of docking poses with higher number of total interactions than highest interacting pose of ligand: \",\n",
    "            len(self.higher_interacting_poses),\n",
    "        )\n",
    "        pprint(\n",
    "            \"&nbsp;&nbsp;&nbsp;&nbsp;CIDs of analogs corresponding to these docking poses: \",\n",
    "            [analog.cid for analog in self.higher_interacting_analogs],\n",
    "        )\n",
    "        pprint(\n",
    "            \"Number of docking poses with higher affinity and number of total interactions than best corresponding poses of ligand: \",\n",
    "            len(self.higher_affinity_and_interacting_poses),\n",
    "        )\n",
    "        pprint(\n",
    "            \"&nbsp;&nbsp;&nbsp;&nbsp;CIDs of analogs corresponding to these docking poses: \",\n",
    "            [analog.cid for analog in self.higher_affinity_and_interacting_analogs],\n",
    "        )\n",
    "        pprint(\n",
    "            \"CIDs of analogs with higher binding affinity, number of total interactions and drug-likeness score than ligand: \",\n",
    "            [analog.cid for analog in self.higher_affinity_and_interacting_and_druglike_analogs],\n",
    "        )\n",
    "        pprint(\n",
    "            \"**CIDs of selected analogs as final output:** \",\n",
    "            [analog.cid for analog in self.output],\n",
    "        )\n",
    "\n",
    "        pprint(\"Comparison between the input ligand and optimized analog: \", \"\")\n",
    "\n",
    "        self.comparison_dataframe = df = pd.DataFrame(columns=[\"Input Ligand\", \"Optimized Analog\"])\n",
    "        df.loc[\"Drug-Score\"] = [\n",
    "            self._project.Ligand.drug_score_total,\n",
    "            self._project.OptimizedLigands.output[0].drug_score_total,\n",
    "        ]\n",
    "        df.loc[\"Highest Binding Affinity\"] = [\n",
    "            self._project.Ligand.binding_affinity_best,\n",
    "            self._project.OptimizedLigands.output[0].binding_affinity_best,\n",
    "        ]\n",
    "        df.loc[\"Highest Number of Total Interactions\"] = [\n",
    "            self._project.Ligand.num_total_interactions_highest,\n",
    "            self._project.OptimizedLigands.output[0].num_total_interactions_highest,\n",
    "        ]\n",
    "        display(self.comparison_dataframe)\n",
    "        return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now instantiate the class, and assign it to our project:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.OptimizedLigands = OptimizedLigands(project1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The class has a ***\\_\\_call\\_\\_*** method, which prints out a summary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.OptimizedLigands()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, ***the pipeline successfully found an analog, which is better than the input ligand in all of the three metrics, i.e. drug-likeness, binding affinity, and total number of protein-ligand interactions.***\n",
    "We can also simply visualize each of the above groups of analogs. For example for visualizing the final output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "project1.OptimizedLigands.show_final_output()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='10_practical'></a>\n",
    "### Putting the Pieces Together: A Fully Automated Pipeline\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Now that we have implemented all the necessary parts of our pipeline, we can write a function to automatically run the whole pipeline and display the results. The function thus takes in the project name, the filepath of the input data, and the output path, and performs all the necessary processes to generate the final output. It will also print the intermediate results of each part of the pipeline, so that the process can be followed.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_pipeline(project_name, input_data_filepath, output_data_root_folder_path):\n",
    "    \"\"\"\n",
    "    Automatically run the whole lead optimization pipeline to completion,\n",
    "    and print out a summary in real-time.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    project_name : str\n",
    "        Name of the lead optimization project.\n",
    "    input_data_filepath : str or pathlib.Path object\n",
    "        Filepath of the input CSV file containing all the specifications of the project.\n",
    "    output_data_root_folder_path : str or pathlib.Path object\n",
    "        Root folder path to save the pipeline's data in.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        LeadOptimizationPipeline object\n",
    "        Object containing all the information about the pipeline.\n",
    "    \"\"\"\n",
    "\n",
    "    def pprint(markdown_list):\n",
    "        markdown_command = \"\"\n",
    "        for command in markdown_list:\n",
    "            text, color = command\n",
    "            markdown_command += f\"<span style='color:{color}'>{text}</span>\"\n",
    "        display(Markdown(markdown_command))\n",
    "\n",
    "    def pprint_header(header):\n",
    "        display(\n",
    "            Markdown(\n",
    "                f\"<span style='color:blue'>**{header}:** </span><span style='color:green'>Successful</span>\"\n",
    "            )\n",
    "        )\n",
    "\n",
    "    project = LeadOptimizationPipeline(project_name=project_name)\n",
    "    pprint_header(\"1. Initializing Project\")\n",
    "    pprint([(f\"&nbsp;&nbsp;&nbsp;&nbsp;Project name: **{project.name}**\", \"black\")])\n",
    "\n",
    "    project.Specs = Specs(\n",
    "        input_data_filepath=input_data_filepath,\n",
    "        output_data_root_folder_path=f\"{output_data_root_folder_path}/{project.name}\",\n",
    "    )\n",
    "    pprint_header(\"2. Initializing Input/Output\")\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Input data read from: **{input_data_filepath}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Output folders created at: **{output_data_root_folder_path}/{project.name}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    project.Protein = Protein(\n",
    "        identifier_type=project.Specs.Protein.input_type,\n",
    "        identifier_value=project.Specs.Protein.input_value,\n",
    "        protein_output_path=project.Specs.OutputPaths.protein,\n",
    "    )\n",
    "    pprint_header(\"3. Processing Protein Data\")\n",
    "    display(project.Protein())\n",
    "\n",
    "    project.Ligand = Ligand(\n",
    "        identifier_type=project.Specs.Ligand.input_type,\n",
    "        identifier_value=project.Specs.Ligand.input_value,\n",
    "        ligand_output_path=project.Specs.OutputPaths.ligand,\n",
    "    )\n",
    "    pprint_header(\"4. Processing Ligand Data\")\n",
    "    display(project.Ligand())\n",
    "\n",
    "    project.BindingSiteDetection = BindingSiteDetection(\n",
    "        project.Protein,\n",
    "        project.Specs.BindingSite,\n",
    "        project.Specs.OutputPaths.binding_site_detection,\n",
    "    )\n",
    "    pprint_header(\"5. Binding Site Detection\")\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Binding site definition method: **{project.Specs.BindingSite.definition_method.name}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    if (\n",
    "        project.Specs.BindingSite.definition_method\n",
    "        is Consts.BindingSite.DefinitionMethods.DETECTION\n",
    "    ):\n",
    "        pprint(\n",
    "            [\n",
    "                (\n",
    "                    f\"&nbsp;&nbsp;&nbsp;&nbsp;Binding site detection method: **{project.Specs.BindingSite.detection_method.name}**\",\n",
    "                    \"black\",\n",
    "                )\n",
    "            ]\n",
    "        )\n",
    "        pprint(\n",
    "            [\n",
    "                (\n",
    "                    f\"&nbsp;&nbsp;&nbsp;&nbsp;Selection method for best binding site: **{project.Specs.BindingSite.selection_method.name}**\",\n",
    "                    \"black\",\n",
    "                )\n",
    "            ]\n",
    "        )\n",
    "        pprint(\n",
    "            [\n",
    "                (\n",
    "                    f\"&nbsp;&nbsp;&nbsp;&nbsp;Selection criteria for best binding site: **{project.Specs.BindingSite.selection_criteria}**\",\n",
    "                    \"black\",\n",
    "                )\n",
    "            ]\n",
    "        )\n",
    "        pprint(\n",
    "            [\n",
    "                (\n",
    "                    f\"&nbsp;&nbsp;&nbsp;&nbsp;Name of selected binding site: **{project.BindingSiteDetection.best_binding_site_name}**\",\n",
    "                    \"black\",\n",
    "                )\n",
    "            ]\n",
    "        )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Binding site coordinates - center: **{project.Protein.binding_site_coordinates['center']}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Binding site coordinates - size: **{project.Protein.binding_site_coordinates['size']}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    display(project.BindingSiteDetection.visualize_best())\n",
    "\n",
    "    project.LigandSimilaritySearch = LigandSimilaritySearch(\n",
    "        project.Ligand,\n",
    "        project.Specs.LigandSimilaritySearch,\n",
    "        project.Specs.OutputPaths.similarity_search,\n",
    "    )\n",
    "    pprint_header(\"6. Ligand Similarity Search\")\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Search engine: **{project.Specs.LigandSimilaritySearch.search_engine.name}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Number of fetched analogs with a similarity higher than **{project.Specs.LigandSimilaritySearch.min_similarity_percent}%**: **{len(project.LigandSimilaritySearch.all_analogs)}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Dice-similarity range of fetched analogs: **{project.LigandSimilaritySearch.all_analogs['dice_similarity'].min()} - {project.LigandSimilaritySearch.all_analogs['dice_similarity'].max()}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;CID of analog with the highest Dice-similarity: **{project.LigandSimilaritySearch.all_analogs.sort_values(by='dice_similarity', ascending=False).head(1).index.values[0]}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Number of selected drug-like analogs: **{len(project.Ligand.analogs)}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    sorted_analogs_df = project.LigandSimilaritySearch.all_analogs.sort_values(\n",
    "        by=\"drug_score_total\", ascending=False\n",
    "    ).head(len(project.Ligand.analogs))\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Range of drug-likeness score in selected analogs: **{sorted_analogs_df['drug_score_total'].min()} - {sorted_analogs_df['drug_score_total'].max()}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;CID of analog with the highest drug-likeness score: **{sorted_analogs_df.head(1).index.values[0]}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    project.Docking = Docking(\n",
    "        project.Protein,\n",
    "        list(project.Ligand.analogs.values()),\n",
    "        project.Specs.Docking,\n",
    "        project.Specs.OutputPaths.docking,\n",
    "    )\n",
    "    project.Ligand.Docking = Docking(\n",
    "        project.Protein,\n",
    "        [project.Ligand],\n",
    "        project.Specs.Docking,\n",
    "        project.Specs.OutputPaths.ligand,\n",
    "    )\n",
    "    pprint_header(\"7. Docking Experiment\")\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Highest binding affinity of input ligand: **{project.Ligand.binding_affinity_best}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Binding affinity range of analogs: **{project.Docking.results_dataframe['affinity[kcal/mol]'].max()} - {project.Docking.results_dataframe['affinity[kcal/mol]'].min()}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Number of analog docking poses with higher affinity than ligand: **{len(project.Docking.results_dataframe[project.Docking.results_dataframe['affinity[kcal/mol]']<project.Ligand.binding_affinity_best].sort_values(by=['affinity[kcal/mol]']))}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Number of analogs with higher affinity than ligand: **{len(set(project.Docking.results_dataframe[project.Docking.results_dataframe['affinity[kcal/mol]']<project.Ligand.binding_affinity_best].sort_values(by=['affinity[kcal/mol]']).index.get_level_values(0)))}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    highest_aff_cids = set(\n",
    "        project.Docking.results_dataframe[\n",
    "            project.Docking.results_dataframe[\"affinity[kcal/mol]\"]\n",
    "            < project.Ligand.binding_affinity_best\n",
    "        ]\n",
    "        .sort_values(by=[\"affinity[kcal/mol]\"])\n",
    "        .index.get_level_values(0)\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;CIDs of analogs with higher affinity than ligand: **{highest_aff_cids}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    project.Docking.visualize_all_poses()\n",
    "\n",
    "    project.InteractionAnalysis = InteractionAnalysis(\n",
    "        project.Docking.pdbqt_filepath_extracted_protein,\n",
    "        project.Docking.pdb_filepath_extracted_protein,\n",
    "        project.Protein.residue_number_first,\n",
    "        list(project.Ligand.analogs.values()),\n",
    "        project.Docking.master_df,\n",
    "        project.Specs.InteractionAnalysis,\n",
    "        project.Specs.OutputPaths.interaction_analysis,\n",
    "    )\n",
    "\n",
    "    project.Ligand.InteractionAnalysis = InteractionAnalysis(\n",
    "        project.Docking.pdbqt_filepath_extracted_protein,\n",
    "        project.Docking.pdb_filepath_extracted_protein,\n",
    "        project.Protein.residue_number_first,\n",
    "        [project.Ligand],\n",
    "        project.Ligand.Docking.master_df,\n",
    "        project.Specs.InteractionAnalysis,\n",
    "        project.Specs.OutputPaths.ligand,\n",
    "    )\n",
    "    pprint_header(\"8. Protein-Ligand Interaction Analysis\")\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Highest number of total interactions in a docking pose of input ligand: **{project.Ligand.InteractionAnalysis.results.sort_values(by='total_num_interactions',ascending=False).head(1)['total_num_interactions'].values[0]}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Range of total number of interactions in docking poses of analogs: **{project.InteractionAnalysis.results['total_num_interactions'].min()} - {project.InteractionAnalysis.results['total_num_interactions'].max()}**\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    pprint(\n",
    "        [\n",
    "            (\n",
    "                f\"&nbsp;&nbsp;&nbsp;&nbsp;Correlation plot between binding affinity and number of interactions in docking poses of analogs:\",\n",
    "                \"black\",\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    project.InteractionAnalysis.plot_interaction_affinity_correlation()\n",
    "\n",
    "    display(project.InteractionAnalysis.visualize_all_interactions())\n",
    "\n",
    "    pprint_header(\"9. Selecting The Optimized Analog\")\n",
    "    project.OptimizedLigands = OptimizedLigands(project)\n",
    "    project.OptimizedLigands()\n",
    "    print(\"\\n\")\n",
    "    pprint([(\"**Selected analogs as final output:**\", \"blue\")])\n",
    "    project.OptimizedLigands.show_final_output()\n",
    "\n",
    "    cid_pose_nr_list = []\n",
    "    for final_analog in project.OptimizedLigands.output:\n",
    "        for docking_pose_nr in range(1, len(final_analog.dataframe_docking) + 1):\n",
    "            cid_pose_nr_list.append((final_analog.cid, docking_pose_nr))\n",
    "    display(project.InteractionAnalysis.visualize_docking_poses_interactions(cid_pose_nr_list))\n",
    "\n",
    "    pprint_header(\"10. Pipeline Completed\")\n",
    "\n",
    "    return project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By calling the above function with the necessary parameters, the whole pipeline will automatically run, and we will get a full summary of the pipeline processes and obtain the final results.\n",
    "\n",
    "For this demonstration, as the input ligand let's use the optimized ligand we just found, to see whether we can one up our previous optimization results and obtain a yet better analog. We first need to create the corresponding input file. To do so, we can simply open the input file of our project, modify the ligand input value, and save the file as a new input file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_input_df = pd.read_csv(project1.Specs.RawData.filepath)\n",
    "new_input_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_input_df.loc[3, \"Value\"] = project1.OptimizedLigands.output[0].smiles\n",
    "new_input_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_input_df.to_csv(\"data/PipelineInputData_Project2.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"text-align: justify\"> \n",
    "Now let's run the pipeline again, this time fully automated and with the already optimized ligand. Other than printing out a short summary of results in real-time, the function also returns the whole project at the end. Thus, by assigning the return value to a variable (here <b><i>project2</i></b>), we can later further investigate the the information generated by the pipeline in more detail.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "project2 = run_pipeline(\n",
    "    \"Project2_EGFR_CID11292933\",\n",
    "    input_data_filepath=\"data/PipelineInputData_Project2.csv\",\n",
    "    output_data_root_folder_path=\"data/Outputs\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id='discussion'></a>\n",
    "## Discussion\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "In this talktorial, we have successfully implemented a <b>fully-automated virtual screening pipeline, particularly targeted at hit-expansion and lead-optimization phases of a drug discovery project</b>. As the input, the pipeline accepts a single CSV data-file, specifying a protein structure (either as PDB-code or local PDB-file), a ligand (as one of the many possible identifiers, e.g. SMILES, CID, InChI etc.), and several other settings regarding the different processes of the pipeline. It then <b>performs all the necessary processes in order to derive at an optimized analog of the provided ligand, in terms of binding affinity, number of protein-ligand interactions and drug-likeness</b>. The process start with <b>binding-site definition</b>, which can be carried out via several provided methods, including binding-site detection by the <i>DoGSiteScorer</i> functionality of the <i>ProteinsPlus</i> webserver. The next step is <b>ligand similarity-search</b>, which is performed using the <i>PubChem</i> web-services via the provided API backend. Subsequently, the structural analogs detected via the similarity search are filtered based on their physiochemical properties, in order to <b>select the most drug-like analogs</b>. These are then assessed in a virtual screening process using the <i>Smina</i> program, in order to <b>calculate their binding affinities and corresponding binding-poses</b>. Using the <i>PLIP</i> package, these generated docking poses can be <b>analyzed in terms of their specific non-covalent interactions with the target protein</b>, and analogs displaying a specific set of desired interactions can be selected, e.g. in order to achieve <b>higher selectivities for the target protein</b>. After all required processes are carried out, the pipeline can <b>output one or several optimized analogs</b>, based on the criteria defined by the user in the input data, such as binding affinity, physiochemical and pharmacokinetic properties, and specific interactions with the protein structure. \n",
    "<br><br><div style=\"text-align: justify\"> \n",
    "The code is specifically structured in such a way that it will be easy to digest, maintain, and expand. For each functionality of the pipeline, one or several helper classes are defined, containing static methods that are required to develop a certain part of the pipeline. While this may have added some extra workload in the development of the pipeline, the advantage is that, first, long classes that make the demonstration impractical are avoided, making the code more readable and communicative. Moreover, these <b>static methods can be easily adopted for other projects as well</b>. On the other hand, the pipeline is also contained in a single class, with sub-classes for every process, which makes the access to all parts of the data much more easier. Furthermore, after each part of the pipeline is separately developed at demonstrated, at the end of the pipeline a function is provided, which puts the whole program together, and <b>automatically runs a given project to completion, merely by providing an input specification file</b>. Since the pipeline provides much more functionalities and options than what the scope of this talktorial allows to demonstrate, we have also provided a <i>Supplementary Information</i> section, which showcases the possibilities of the pipeline in more detail. Furthermore, we have structured the code to be easily expandable, so it can be freely adopted and built upon for various situations and needs.<br><br>\n",
    "During the talktorial we demonstrated the pipeline with a sample project, using EGFR (PDB-code: 3W32) as the target protein, and a promising inhibitor with the ChEMBL-ID CHEMBL328216. The abilities of the pipeline was successfully displayed with this example, where at the end of the pipeline <b>we were able to identify an analog with much higher binding affinity and drug-likeness score</b> (Figure 9). Moreover, to showcase the fully automated version of the pipeline at the end, we re-ran the pipeline using this optimized ligand as the input. We were then delighted to see that <b>a second run could still improve the ligand and result in yet another analog with higher binding affinity</b>. The automated function at the end of the pipeline also prints out a full report and interactive visualization of the processes while running, allowing the user to have a better understanding of the inner workings of the program in real-time. \n",
    "</div><br>\n",
    "<p style=\"text-align:center;\">\n",
    "<img src=\"images/fig9.png\" width=\"950\" height=auto class=\"center\"/>\n",
    "</p>\n",
    "<b>Figure 9.</b> Results of the pipeline's performance in the demonstrated examples; Starting with a ligand with a binding affinity of -8.7 kcal/mol, the pipeline was able to find an optimized ligand with an improved affinity of -10.3 kcal/mol. Subjecting this optimized ligand to the same process further improved the binding affinity, and a yet more suitable analog was found with a binding affinity of -10.9 kcal/mol. <i><b>Note:</b></i> The results of the pipeline are not fully deterministic, due to several randomized events taking place during the processes. Therefore, re-evaluating the pipeline under the same conditions may sometimes result in different outputs. </div>\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='quiz'></a>\n",
    "## Quiz\n",
    "**Conceptual Questions:**\n",
    "1. Describe the processes involved in the hit-to-lead and lead optimization phases of a drug design pipeline. How are these processes implemented in the virtual pipeline described in this talktorial?\n",
    "2. Describe the importance of binding-site definition in a virtual screening pipeline. What options are available for binding-site definition and -selection in this pipeline? \n",
    "3. What are the possibilities for a virtual ligand-derivatization process? Which one is used in this pipeline? \n",
    "4. Why do some of the obtained analogs have a lower calculated similarity measure than the threshold specified in the input data?\n",
    "5. Which criteria are used in this pipeline to filter the initially obtained analogs for the docking experiment?\n",
    "6. Why don't we carry out the docking experiment for all initially obtained analogs, so that we can choose the best candidate between all available options at the end?\n",
    "7. What are the general preparative steps for performing a docking experiment?\n",
    "8. Why do we calculate protein-ligand interactions for each docking pose, although these interactions have been implicitly taken into account by the scoring function of the docking algorithm?\n",
    "9. How can the interaction-analysis data be used to select an analog with higher selectivity for the target protein?\n",
    "10. Why do we have to visualize the docking poses and their corresponding interactions and manually inspect them? What criteria should we be looking for in these inspections?\n",
    "11. What criteria are generally used for selecting the best optimized ligand at the end of the pipeline? What options do our pipeline offer?\n",
    "**Exercises:**\n",
    "1. In order to prevent the talktorial from being too lengthy, we only demonstrated a sub-set of the options and information that the pipeline offers. Execute the pipeline and explore its other options by looking up the attributes and methods of different classes comprising the pipeline. Since we have organized the whole project in a ***LeadOptimizationPipeline*** instance, this should be a fairly easy task. \n",
    "2. Try to implement a loop, where the final output of a pipeline is re-entered as the input for a new pipeline (i.e. try to optimize your initial input ligand through several runs, same as we did for the demonstration of the ***run_pipeline*** function). On average, after how many cycles does the pipeline reach a plateau where no better analog can be found anymore?\n",
    "3. Generate your own input CSV-files, each time choosing a different set of specifications, and compare the results. Can you find a set of specifications that perform considerably better than the others?\n",
    "4. Since this was a relatively large project and we have also offered many different options for each process of the pipeline, it may be the case that some options are not fully thought-out and developed. Try to run the pipeline with all possible combinations of specifications, and see if you can find a problem in any of the branches. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='supp'></a>\n",
    "## Supplementary Information\n",
    "<br>\n",
    "<div style=\"text-align: justify\"> \n",
    "Here we will also demonstrate the functions of each helper-class used in the pipeline, to provide a better understanding of the inner workings of the code.<br><br>\n",
    "    <b><i>Note:</i></b> Some cells use variables defined in an early cell. Therefore, in order for all cells to work you have to run them sequentially.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='io_demo'></a>\n",
    "### Demonstration of The ***IO*** Helper-Class\n",
    "Using this class, the input CSV file can be easily imported via the ***create_dataframe_from_csv_input_file*** function, by specifying the corresponding filepath, the name of the columns to be used as indices for the dataframe, and the name of the other columns we want to keep:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_input_df = IO.create_dataframe_from_csv_input_file(\n",
    "    input_data_filepath=\"data/PipelineInputData_Project1.csv\",\n",
    "    list_of_index_column_names=[\n",
    "        Consts.DataFrame.ColumnNames.SUBJECT.value,\n",
    "        Consts.DataFrame.ColumnNames.PROPERTY.value,\n",
    "    ],\n",
    "    list_of_columns_to_keep=[Consts.DataFrame.ColumnNames.VALUE.value],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that for inputting the ***list_of_index_column_names*** and ***list_of_column_names_to_keep*** we didn't write the name of the columns (i.e. '***Subject***', '***Property***' and '***Value***') explicitly (although we could), but used the respective constants stored in the ***Consts*** class. In this way, if the column names in the input file are changed in the future, they only need to be corrected in the ***Consts*** class, and nowhere else in the code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now view the imported data in its entirety:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_input_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Moreover, we can also extract all the data corresponding to a specific ***Subject***, using the ***copy_series_from_dataframe*** function and specifying the main dataframe and the name of the index-level and column that we want to extract. For example, to extract the binding-site specification data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "example_input_protein_data = IO.copy_series_from_dataframe(\n",
    "    input_df=example_input_df,\n",
    "    index_name=Consts.DataFrame.SubjectNames.BINDING_SITE.value,\n",
    "    column_name=Consts.DataFrame.ColumnNames.VALUE.value,\n",
    ")\n",
    "\n",
    "# Displaying the extracted data:\n",
    "example_input_protein_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "Later, for storing the output data we will also need the ***create_folder*** function, which can create a folder (and all its necessary parent folders) with a given name at a given path. The function then returns the full path of the created folder. We now create a folder named *Examples*, which we will use to store the files used in the demonstration sections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "example_output_path = IO.create_folder(folder_name=\"Examples\", folder_path=\"data/Outputs\")\n",
    "\n",
    "# Displaying the full path of the generated folder:\n",
    "example_output_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='pdb_demo'></a>\n",
    "### Demonstration of The ***PDB*** Helper-Class\n",
    "We can use the ***fetch_and_save_pdb_file*** function to download a protein's PDB file using its PDB-code. The function takes in a full filepath, including the name of the file (but excluding the .PDB extension), and returns the full filepath (including the .PDB extension)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_downloaded_protein_filepath = PDB.fetch_and_save_pdb_file(\n",
    "    pdb_code=\"3w32\", output_filepath=example_output_path / \"3W32\"\n",
    ")\n",
    "\n",
    "example_downloaded_protein_filepath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The PDB file is now downloaded on the disk. Now, the file content of the PDB file can be obtained as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "example_pdb_file_content = PDB.read_pdb_file_content(\n",
    "    input_type=\"pdb_filepath\", input_value=example_downloaded_protein_filepath\n",
    ")\n",
    "\n",
    "# Here we are only showing the first 800 characters, since the full content is too long:\n",
    "example_pdb_file_content[:800]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that alternatively we also could have set the ***input_type*** to 'pdb_code' and entered a valid PDB-code as the ***input_value*** to directly fetch and read the file contents of a protein PDB file from the PDB webserver. However, in that case, the PDB file itself would not have been saved on the disk."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This file content can the be used to extract some useful information from the PDB file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PDB.extract_info_from_pdb_file_content(example_pdb_file_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or directly without storing the PDB file content in a variable (here using another PDB-code as example):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PDB.extract_info_from_pdb_file_content(\n",
    "    pdb_file_text_content=PDB.read_pdb_file_content(input_type=\"pdb_code\", input_value=\"5gty\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ***load_pdb_file_as_dataframe*** function can also be used, but it does not contain all the information that the ***extract_info_from_pdb_file_content*** can extract. The function return a dictionary, where each value is a *Pandas DataFrame*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_pdb_dataframe = PDB.load_pdb_file_as_dataframe(example_pdb_file_content)\n",
    "example_pdb_dataframe.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example, all the information regarding the atoms in the protein can be accessed via the ***ATOM*** key. This data was used in the pipeline, for example to extract the range of residue numbers of the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_pdb_dataframe[\"ATOM\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='pubchem_demo'></a>\n",
    "### Demonstration of The ***PubChem*** Helper-Class\n",
    "The ***convert_compound_identifier*** function can be used to get a specific compound's identifier, by providing another identifier. For example, getting the SMILES of Aspirin, providing its name: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PubChem.convert_compound_identifier(\n",
    "    input_id_type=\"name\", input_id_value=\"aspirin\", output_id_type=\"smiles\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or getting it's IUPAC name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PubChem.convert_compound_identifier(\n",
    "    input_id_type=\"name\", input_id_value=\"aspirin\", output_id_type=\"iupac_name\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or getting its name from its SMILES:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PubChem.convert_compound_identifier(\n",
    "    input_id_type=\"smiles\",\n",
    "    input_id_value=\"CC(=O)OC1=CC=CC=C1C(=O)O\",\n",
    "    output_id_type=\"name\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or getting its CID from its name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PubChem.convert_compound_identifier(\n",
    "    input_id_type=\"name\", input_id_value=\"aspirin\", output_id_type=\"cid\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the ***get_compound_record*** function, all available records on the molecule can be obtained in a dictionary with following keys: '***id***', '***atoms***', '***bonds***', '***coords***', '***charge***', '***props***', '***count***'.\n",
    "As an example, we access the '***props***' key, which contains the physiochemical properties of the compound: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Only showing the first 2 entries, since the full list is too long:\n",
    "PubChem.get_compound_record(input_id_type=\"name\", input_id_value=\"aspirin\")[\"props\"][:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ***get_description_from_smiles*** function provides a textual description of the compound. However, it is not available for every compound. If the parameter ***printout*** is set to False, the function will return a dictionary containing the descriptions, sources, and other information. By setting it to True, only the description part will be printed out. For example, for aspirin: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PubChem.get_description_from_smiles(smiles=\"CC(=O)OC1=CC=CC=C1C(=O)O\", printout=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='rdkit_demo'></a>\n",
    "### Demonstration of The ***RDKit*** Helper-Class\n",
    "To use any other function in the ***RDKit*** class, we first have to generate an RDKit molecule object, using the ***create_molecule_object*** function, which can accept several different identifiers as input, for example the SMILES (here again using Aspirin as an example);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_mol_obj = RDKit.create_molecule_object(\n",
    "    input_type=\"smiles\", input_value=\"CC(=O)OC1=CC=CC=C1C(=O)O\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The RDKit molecule objects can be called directly, which simply visualizes the molecule:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_mol_obj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ***calculate_druglikeness*** function can then be used to calculate several molecular properties and drug-likeness scores, which will be returned as a dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RDKit.calculate_druglikeness(mol_obj=example_mol_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ***calculate_similarity_dice*** function is used to calculate the Dice similarity metric between two ligands, using 4096-bit Morgan fingerprints with a raidus of 2. Here, as an example we will input the same molecule twice, which results in a similarity score of 1 (i.e. 100% similarity): "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RDKit.calculate_similarity_dice(mol_obj1=example_mol_obj, mol_obj2=example_mol_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the ***save_molecule_image_to_file*** function, an image file containing the ligand's structure can be saved at a given path as well: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RDKit.save_molecule_image_to_file(\n",
    "    mol_obj=example_mol_obj, filepath=example_output_path / \"aspirin\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition, we can use the ***save_3D_molecule_to_SDfile*** function, which first prepares the molecule for docking experiments (i.e. adds hydrogen atoms, creates 3D conformation and runs energy minimizations etc.) and then saves it to an SDF file: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RDKit.save_3D_molecule_to_SDfile(mol_obj=example_mol_obj, filepath=example_output_path / \"aspirin\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id='dogsitescorer_demo'></a>\n",
    "### Demonstration of The ***DoGSiteScorer*** Class\n",
    "If we want to detect the binding sites of a protein from a local PDB file, we should first upload the file to the *DogSiteScorer* webserver, using the ***upload_pdb_file*** function. For this, we will use the PDB file we downloaded earlier: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_dummy_pdb_id = DoGSiteScorer.upload_pdb_file(filepath=example_downloaded_protein_filepath)\n",
    "\n",
    "example_dummy_pdb_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ***upload_pdb_file*** returns a dummy PDB-code for the uploaded structure, which can be used in place of a valid PDB-code to submit a binding-site detection job. Notice that the ***submit_job*** function can also take in a chain-ID to limit the detection on that specific chain, and a ligand-ID to also calculate the coverage of each detected binding-site. However, the ligand-ID that *DoGSiteScorer* accepts has its own format, and is not the same as the ligand-ID in the protein PDB file. Nevertheless, it follows the following format: (ligand-ID)_(chain-ID)_(ligand residue-ID). When implementing the ***BindingSiteDetection*** class, we will circumvent this by automatically generating the *DogSiteScorer* ligand-ID from the normal ligand-ID, so that the user does not have to manually look up and enter the ligand-ID in this specific format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_binding_site_data = DoGSiteScorer.submit_job(\n",
    "    pdb_id=example_dummy_pdb_id, ligand_id=\"W32_A_1101\", chain_id=\"A\"\n",
    ")\n",
    "\n",
    "example_binding_site_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, we can also use a valid PDB-code to directly submit a binding-site detection job:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DoGSiteScorer.submit_job(\n",
    "    pdb_id=project1.Protein.pdb_code,\n",
    "    ligand_id=\"W32_A_1101\",\n",
    "    chain_id=project1.Protein.chains[0],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen, the ***submit_job*** function returns a DataFrame containing all the detected binding-sites and their respective properties. It also contains the URLs of the PDB and CCP4 files for each of the detected binding-sites, which will be used to download the files later. Among the calculated properties, the most important are ***drugScore***, ***simpleScore***, ***volume***, as well as ***lig_cov*** and ***poc_cov*** for when a ligand-ID is also inputted. The list of all properties can be accessed by retrieving the column names of the DataFrame: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_binding_site_data.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ***simpleScore*** is a simple druggability score, based on a linear combination of the three descriptors describing volume, hydrophobicity and enclosure. In addition, a subset of meaningful descriptors is incorporated in a support vector machine (libsvm) to predict a druggability score called ***drugScore***, which has a value between 0 and 1 (the higher the score the more druggable the binding-site is estimated to be). ***lig_cov*** gives the percentage of the ligand volume that is covered by the binding-site, and ***poc_cov*** gives the percentage of the binding-site volume that is covered by the ligand. <br>\n",
    "\n",
    "Depending on the specifics of the project, these values can be used to select the most-suitable detected binding-site. Here, we have implemented two possibilities for this selection process in the ***select_best_pocket*** function. The first option is to provide a list of properties in order of importance, based on which the binding-site with the highest or lowest value is to be chosen. The function then returns the name of the selected binding-site (The list of properties should be inputted as a list):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_best_pocket_name = DoGSiteScorer.select_best_pocket(\n",
    "    binding_site_df=example_binding_site_data,\n",
    "    selection_method=\"sorting\",\n",
    "    selection_criteria=[\"lig_cov\", \"poc_cov\"],\n",
    "    ascending=False,\n",
    ")\n",
    "\n",
    "example_best_pocket_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example, the above ***selection_criteria*** sorts the binding-sites by their ***lig_cov*** values, and if there are two or more binding-sites with the same ***lig_cov*** value, it then sorts them by their ***poc_cov*** values.<br> \n",
    "\n",
    "Another possibility for selecting a binding site is to provide any valid python expression, which generates a list-like object with the same length as the number of detected binding-site. This python expression is inputted as string, so that the user can directly input a python expression in the input CSV file, and it will be evaluated during the runtime. For example, using this method we can perform any calculation on the properties in the binding-site DataFrame (note that for referring to the binding-site dataframe, **'df'** should be used):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DoGSiteScorer.select_best_pocket(\n",
    "    binding_site_df=example_binding_site_data,\n",
    "    selection_method=\"function\",\n",
    "    selection_criteria=\"((df['lig_cov'] + df['poc_cov']) / 2) * ((df['drugScore'] + df['simpleScore']) / 2) / df['volume']\",\n",
    "    ascending=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For example, the above ***selection_criteria*** is a function that calculates the average of ***lig_cov*** and ***poc_cov***, as well as the average of ***drugScore*** and ***simpleScore***, multiplies the two together and then divides the result by the ***volume***. The binding-site that has the highest value is then chosen (If we want to have the lowest value, we can set the ***ascending*** parameter to False). Note that the dataframe should always be referred to as **'df'**. Another example is: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DoGSiteScorer.select_best_pocket(\n",
    "    binding_site_df=example_binding_site_data,\n",
    "    selection_method=\"function\",\n",
    "    selection_criteria=\"df[['drugScore', 'simpleScore']].min(axis=1) * df[['poc_cov', 'lig_cov']].max(axis=1)\",\n",
    "    ascending=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above example calculates the minimum value between ***drugScore*** and ***simpleScore***, and multiplies the results with the maximum value between the ***poc_cov*** and ***lig_cov***. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now want to calculate the coordinates of the selected binding-site, so that we can use it in the docking process. For this, we first need to download the PDB files of the detected binding-sites, using the ***save_binding_sites_to_file*** function: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DoGSiteScorer.save_binding_sites_to_file(\n",
    "    binding_site_df=example_binding_site_data, output_path=example_output_path\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having downloaded the necessary binding-site files, we can now calculate the coordinates of the selected binding-site: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DoGSiteScorer.calculate_pocket_coordinates_from_pocket_pdb_file(\n",
    "    filepath=example_output_path / example_best_pocket_name\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Copy of T04-Short presentation.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "toc-autonumbering": true,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
