{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# T026 · Kinase similarity: Interaction fingerprints\n",
    "\n",
    "Authors:\n",
    "\n",
    "- Talia B. Kimber, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)\n",
    "- Dominique Sydow, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)\n",
    "- Andrea Volkamer, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aim of this talktorial\n",
    "\n",
    "We will assess the similarity between a set of kinases based on detected protein-ligand interactions in available complex structures. The [KLIFS](https://klifs.net/) interaction fingerprint (IFP), which describes the interactions seen in a structurally resolved kinase-ligand complex, will be used for this exercise.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contents in *Theory*\n",
    "\n",
    "* Kinase dataset\n",
    "* Kinase similarity descriptor: KLIFS interaction fingerprint\n",
    "* Fetching KLIFS data with `opencadd.databases.klifs`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contents in *Practical*\n",
    "\n",
    "* Define the kinases of interest\n",
    "* Retrieve and preprocess data\n",
    "* Show kinase coverage\n",
    "* Compare structures\n",
    "* Map structure to kinase distance matrix\n",
    "* Save kinase distance matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "\n",
    "* Kinase dataset: [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629) \n",
    "* Protein Data Bank\n",
    "  * PDB URL: http://www.rcsb.org/\n",
    "  * PDB database: [<i>Acta Cryst.</i> (2002), <b>D58</b>, 899-907](https://doi.org/10.1107/S0907444902003451) and [<i>Structure</i> (2012), <b>20(3)</b>, 391-396](https://doi.org/10.1016/j.str.2012.01.010)\n",
    "* KLIFS\n",
    "  * KLIFS URL: https://klifs.net/\n",
    "  * KLIFS database: [<i>Nucleic Acid Res.</i> (2020), <b>49(D1)</b>, D562-D569](https://doi.org/10.1093/nar/gkaa895)\n",
    "  * KLIFS binding site definition and interaction fingerprint calculation: [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w)\n",
    "* Interaction fingerprint (IFP): [<i>J. Chem. Inf. Model.</i> (2007), <b>71(1)</b>, 195-207](https://doi.org/10.1021/ci600342e)\n",
    "* `opencadd`, a Python library for structural cheminformatics\n",
    "  * GitHub repository: https://github.com/volkamerlab/opencadd\n",
    "  * Documentation: https://opencadd.readthedocs.io"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Theory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use nine kinases as investigated in [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629). In the study diverse kinase similarity measures were analyzed for different combinations of kinase on- and off-targets to explore the limits of multi-kinase screenings.\n",
    "\n",
    "> We aggregated the investigated kinases in “profiles”. Profile 1 combined **EGFR** and **ErbB2** as targets and **BRAF** as a (general) anti-target. Out of similar considerations, Profile 2 consisted of EGFR and **PI3K** as targets and BRAF as anti-target. This profile is expected to be more challenging as PI3K is an atypical kinase and thus less similar to EGFR than for example ErbB2 used in Profile 1. Profile 3, comprised of EGFR and **VEGFR2** as targets and BRAF as anti-target, was contrasted with the hit rate that we found with a standard docking against the single target VEGFR2 (Profile 4).\n",
    "> To broaden the comparison and obtain an estimate for the promiscuity of each compound, the kinases **CDK2**, **LCK**, **MET** and **p38α** were included in the experimental assay panel and the structure-based bioinformatics comparison as commonly used anti-targets.\n",
    "\n",
    "*Table 1:* \n",
    "Kinases used in this notebook, adapted from [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629), with their synonyms, UniProt IDs, and kinase groups.\n",
    "\n",
    "| Kinase                     | Synonyms               | UniProt ID | Group    | Full kinase name                                 |\n",
    "|----------------------------|------------------------|------------|----------|--------------------------------------------------|\n",
    "| EGFR                       | ErbB1                  | P00533     | TK       | Epidermal growth factor receptor                 |\n",
    "| ErbB2                      | Her2                   | P04626     | TK       | Erythroblastic leukemia viral oncogene homolog 2 |\n",
    "| PI3K                       | PI3KCA, p110a          | P42336     | Atypical | Phosphatidylinositol-3-kinase                    |\n",
    "| VEGFR2                     | KDR                    | P35968     | TK       | Vascular endothelial growth factor receptor 2    |\n",
    "| BRAF                       | -                      | P15056     | TKL      | Rapidly accelerated fibrosarcoma isoform B       |\n",
    "| CDK2                       | -                      | P24941     | CMGC     | Cyclic-dependent kinase 2                        |\n",
    "| LCK                        | -                      | P06239     | TK       | Lymphocyte-specific protein tyrosine kinase      |\n",
    "| MET                        | -                      | P08581     | TK       | Mesenchymal-epithelial transition factor         |\n",
    "| p38a                       | MAPK14                 | Q16539     | CMGC     | p38 mitogen activated protein kinase α           |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase similarity descriptor: KLIFS interaction fingerprints\n",
    "\n",
    "Interaction fingerprints encode the binding mode of a ligand in a binding site, i.e., the protein-ligand interactions that are present in a structurally resolved complex. If a ligand can form similar interaction patterns in proteins other (off-targets) than its designated protein (on-target), it is possible that this ligand will cause unintended side effects. Knowing about similarities between proteins can therefore help to avoid such off-target effects; or to exploit the found similarities for polypharmacology effects where one ligand could intentionally target multiple proteins at once.\n",
    "\n",
    "In this talktorial, we are interested in assessing the similarity between a set of kinases based on the similarity between their co-crystallized ligands' binding modes. We will make use of pre-calculated interaction fingerprints provided by the kinase-focused resource KLIFS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The KLIFS database ([<i>Nucleic Acid Res.</i> (2020), <b>49(D1)</b>, D562-D569](https://doi.org/10.1093/nar/gkaa895), [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w)) fetches all kinase structures deposited in the structural database PDB ([<i>Acta Cryst.</i> (2002), <b>D58</b>, 899-907](https://doi.org/10.1107/S0907444902003451), [<i>Structure</i> (2012), <b>20(3)</b>, 391-396](https://doi.org/10.1016/j.str.2012.01.010)) and processes them as follows: All multi-chain structures in the PDB are split into monomers and aligned to each other with a special focus on a pre-defined binding site of 85 residues (Figure 1). For example, this means that the conserved gatekeeper (GK) residue at KLIFS position 45 can be easily and quickly looked up in any of the over 10,000 monomeric kinase structures in KLIFS. \n",
    "\n",
    "![KLIFS binding site](https://klifs.net/images/faq/xcolors.png.pagespeed.ic.dprMuoZGzn.webp)\n",
    "*Figure 1:* \n",
    "Kinase binding site residues as defined by KLIFS.\n",
    "Figure and description taken from: [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The KLIFS authors calculate for each of these structures &mdash; that are co-crystallized with a ligand &mdash; an interaction fingerprint (IFP) as proposed by Marcou and Rognan ([<i>JCIM</i> (2007), <b>71(1)</b>, 195-207](https://pubs.acs.org/doi/10.1021/ci600342e)). This is how the IFP is described in the respective KLIFS pulication [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w):\n",
    "\n",
    "> For each amino acid in the catalytic cleft, seven types of protein−ligand interactions are determined. The presence of a certain type of interaction results in the type-setting ofa “1” in the bit-string; otherwise a “0” is used to indicate the absence of the interaction. The following seven types of interactions are summarized in one bit-string; bit 1 = hydrophobic contact (HYD); bit 2 = face to face aromatic interactions (F−F); bit 3 = face to edge aromatic interactions (F−E); bit 4 = protein H-bond donor (DON); bit 5 = protein H- bond acceptor (ACC); bit 6 = protein cationic interactions (ION+), and bit 7 = protein anionic interactions (ION−).\n",
    "\n",
    "This results in a $85 \\times 7 = 595$ bit vector, which describes the interactions seen in a ligand-kinase complex. Since the binding site is aligned across all kinases, each bit position in this IFP can be directly compared across all IFPs in KLIFS. This is what we will do now in the practical part of this tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetching KLIFS data with `opencadd.databases.klifs`\n",
    "\n",
    "`opencadd` is a Python library for structural cheminformatics developed by the Volkamer lab at the Charité in Berlin. \n",
    "This library is a growing collection of modules that help facilitate and standardize common tasks in structural bioinformatics and cheminformatics. Today, we will use the module `opencadd.databases.klifs`, which allows us to fetch the KLIFS IFP as `pandas` DataFrame.\n",
    "\n",
    "- GitHub repository: https://github.com/volkamerlab/opencadd\n",
    "- Documentation: https://opencadd.readthedocs.io\n",
    "\n",
    "For more information about this library and the KLIFS OpenAPI, please refer to __Talktorial T012__."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Practical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:opencadd.databases.klifs.api:If you want to see an non-truncated version of the DataFrames in this module, use `pd.set_option('display.max_columns', 50)` in your notebook.\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.metrics import pairwise_distances\n",
    "from scipy.spatial import distance\n",
    "from opencadd.databases.klifs import setup_remote"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "HERE = Path(_dh[-1])\n",
    "DATA = HERE / \"data\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define the kinases of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's define our kinases of interest and all their synonyms as described in Table 1 in the global variable `query_kinases`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_kinases = [\n",
    "    \"EGFR\",\n",
    "    \"ErbB2\",\n",
    "    \"Her2\",\n",
    "    \"PI3K\",\n",
    "    \"PI3KCA\",\n",
    "    \"p110a\",\n",
    "    \"BRAF\",\n",
    "    \"CDK2\",\n",
    "    \"LCK\",\n",
    "    \"MET\",\n",
    "    \"p38a\",\n",
    "    \"MAPK14\",\n",
    "    \"VEGFR2\",\n",
    "    \"KDR\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve and preprocess data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we query the KLIFS database using the `opencadd.databases.klifs` module to generate our IFP dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Set up a remote KLIFS session"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from opencadd.databases.klifs import setup_remote"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:opencadd.databases.klifs.api:Set up remote session...\n",
      "INFO:opencadd.databases.klifs.api:Remote session is ready!\n"
     ]
    }
   ],
   "source": [
    "klifs_session = setup_remote()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fetch all structures describing these kinases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 2484\n",
      "Kinases: CDK2 p38a EGFR ErbB2 MET LCK KDR BRAF p110a\n"
     ]
    }
   ],
   "source": [
    "structures = klifs_session.structures.by_kinase_name(kinase_names=query_kinases)\n",
    "print(f\"Number of structures: {len(structures)}\")\n",
    "print(\"Kinases:\", *structures[\"kinase.klifs_name\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter structures\n",
    "\n",
    "We filter the structures by different criteria:\n",
    "\n",
    "- Species: human\n",
    "- Conformation: DFG-in (the active kinase conformation)\n",
    "- Resolution: $\\le 3$ Angström\n",
    "- Quality score*: $\\ge 6$\n",
    "\n",
    "\\* The KLIFS quality score takes into account the quality of the alignment and the number of missing residues and atoms (the higher the better the structure quality)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 1632\n",
      "Kinases: CDK2 p38a EGFR ErbB2 MET LCK KDR BRAF p110a\n"
     ]
    }
   ],
   "source": [
    "structures = structures[\n",
    "    (structures[\"species.klifs\"] == \"Human\")\n",
    "    & (structures[\"structure.dfg\"] == \"in\")\n",
    "    & (structures[\"structure.resolution\"] <= 3)\n",
    "    & (structures[\"structure.qualityscore\"] >= 6)\n",
    "]\n",
    "print(f\"Number of structures: {len(structures)}\")\n",
    "print(\"Kinases:\", *structures[\"kinase.klifs_name\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the structure KLIFS IDs for the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 1632\n"
     ]
    }
   ],
   "source": [
    "structure_klifs_ids = structures[\"structure.klifs_id\"].to_list()\n",
    "print(f\"Number of structures: {len(structure_klifs_ids)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To make it easier for us to maintain the talktorials, we will load now a set of frozen structure KLIFS IDs (2021-08-23) and continue to work with those. If you would like to work with the latest KLIFS data, please uncomment the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 1620\n"
     ]
    }
   ],
   "source": [
    "# Load frozen dataset\n",
    "structure_klifs_ids = pd.read_csv(DATA / \"frozen_structure_klifs_ids.csv\")[\"structure.klifs_id\"].to_list()\n",
    "print(f\"Number of structures: {len(structure_klifs_ids)}\")\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### Fetch the structures' IFPs (if available)\n",
    "\n",
    "We fetch the IFPs for the set of structures. Not all structures will have an IFP because not all structures have a co-crystallized ligand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures with IFPs: 1466\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th>interaction.fingerprint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>777</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>778</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>779</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>782</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   structure.klifs_id                            interaction.fingerprint\n",
       "0                 775  0000000000000010000000000000000000000000000000...\n",
       "1                 777  0000000000000010000001000000000000000000000000...\n",
       "2                 778  0000000000000010000000000000000000000000000000...\n",
       "3                 779  0000000000000010000001000000000000000000000000...\n",
       "4                 782  0000000000000010000001000000000000000000000000..."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifps = klifs_session.interactions.by_structure_klifs_id(structure_klifs_ids)\n",
    "print(f\"Number of structures with IFPs: {len(ifps)}\")\n",
    "ifps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge structural and IFP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures with IFPs: 1466\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th>interaction.fingerprint</th>\n",
       "      <th>structure.pdb_id</th>\n",
       "      <th>structure.alternate_model</th>\n",
       "      <th>structure.chain</th>\n",
       "      <th>species.klifs</th>\n",
       "      <th>kinase.klifs_id</th>\n",
       "      <th>kinase.klifs_name</th>\n",
       "      <th>kinase.family</th>\n",
       "      <th>kinase.group</th>\n",
       "      <th>...</th>\n",
       "      <th>structure.bp_ii_b_in</th>\n",
       "      <th>structure.bp_ii_out</th>\n",
       "      <th>structure.bp_ii_b</th>\n",
       "      <th>structure.bp_iii</th>\n",
       "      <th>structure.bp_iv</th>\n",
       "      <th>structure.bp_v</th>\n",
       "      <th>structure.grich_distance</th>\n",
       "      <th>structure.grich_angle</th>\n",
       "      <th>structure.grich_rotation</th>\n",
       "      <th>structure.filepath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "      <td>4r5s</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>Human</td>\n",
       "      <td>406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>21.684799</td>\n",
       "      <td>71.330002</td>\n",
       "      <td>42.087002</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>777</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "      <td>3vjo</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>Human</td>\n",
       "      <td>406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>778</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "      <td>3w2s</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>Human</td>\n",
       "      <td>406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>16.591600</td>\n",
       "      <td>53.096100</td>\n",
       "      <td>37.676601</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>779</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "      <td>2itu</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>Human</td>\n",
       "      <td>406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>20.174601</td>\n",
       "      <td>65.518700</td>\n",
       "      <td>21.431101</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>782</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "      <td>3w33</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>Human</td>\n",
       "      <td>406</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>14.820200</td>\n",
       "      <td>48.159698</td>\n",
       "      <td>43.046001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   structure.klifs_id                            interaction.fingerprint  \\\n",
       "0                 775  0000000000000010000000000000000000000000000000...   \n",
       "1                 777  0000000000000010000001000000000000000000000000...   \n",
       "2                 778  0000000000000010000000000000000000000000000000...   \n",
       "3                 779  0000000000000010000001000000000000000000000000...   \n",
       "4                 782  0000000000000010000001000000000000000000000000...   \n",
       "\n",
       "  structure.pdb_id structure.alternate_model structure.chain species.klifs  \\\n",
       "0             4r5s                         -               A         Human   \n",
       "1             3vjo                         -               A         Human   \n",
       "2             3w2s                         A               A         Human   \n",
       "3             2itu                         A               A         Human   \n",
       "4             3w33                         A               A         Human   \n",
       "\n",
       "   kinase.klifs_id kinase.klifs_name kinase.family kinase.group  ...  \\\n",
       "0              406              EGFR          <NA>         <NA>  ...   \n",
       "1              406              EGFR          <NA>         <NA>  ...   \n",
       "2              406              EGFR          <NA>         <NA>  ...   \n",
       "3              406              EGFR          <NA>         <NA>  ...   \n",
       "4              406              EGFR          <NA>         <NA>  ...   \n",
       "\n",
       "  structure.bp_ii_b_in structure.bp_ii_out structure.bp_ii_b structure.bp_iii  \\\n",
       "0                False               False             False            False   \n",
       "1                False               False             False            False   \n",
       "2                False               False             False            False   \n",
       "3                False               False             False            False   \n",
       "4                False               False             False            False   \n",
       "\n",
       "  structure.bp_iv structure.bp_v structure.grich_distance  \\\n",
       "0           False          False                21.684799   \n",
       "1           False          False                 0.000000   \n",
       "2           False          False                16.591600   \n",
       "3           False          False                20.174601   \n",
       "4           False          False                14.820200   \n",
       "\n",
       "   structure.grich_angle  structure.grich_rotation  structure.filepath  \n",
       "0              71.330002                 42.087002                <NA>  \n",
       "1               0.000000                  0.000000                <NA>  \n",
       "2              53.096100                 37.676601                <NA>  \n",
       "3              65.518700                 21.431101                <NA>  \n",
       "4              48.159698                 43.046001                <NA>  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structures_with_ifps = ifps.merge(structures, on=\"structure.klifs_id\", how=\"inner\")\n",
    "print(f\"Number of structures with IFPs: {len(structures_with_ifps)}\")\n",
    "structures_with_ifps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Show kinase coverage\n",
    "\n",
    "Let's get the number of structures that describe our kinases (kinase coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "kinase.klifs_name\n",
       "ErbB2      4\n",
       "KDR        6\n",
       "LCK       30\n",
       "p110a     45\n",
       "BRAF      57\n",
       "MET       95\n",
       "p38a     127\n",
       "EGFR     339\n",
       "CDK2     763\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use pandas' groupby method to count the number of structures (rows) per kinase\n",
    "n_structures_per_kinase = structures_with_ifps.groupby(\"kinase.klifs_name\").size().sort_values()\n",
    "n_structures_per_kinase\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's plot our results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZoAAAEGCAYAAABcolNbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAoq0lEQVR4nO3de5xVdb3/8debGUERhZIgBHXCSJDbqARqHiUJQ+NUpImIHhTUsuMpTS2sc47Wr46kXTTpVJgXjnLELG9HEFPBMFNxlM3FlEhFQVG8onERGD6/P9Ya2ozDsIFZe++ZeT8fj/2Yvb7ru9Z670H8sL5r7e9SRGBmZpaVNqUOYGZmLZsLjZmZZcqFxszMMuVCY2ZmmXKhMTOzTFWWOkCxde7cOaqqqkodw8ys2ejcuTP33XfffRExYme2b3WFpqqqipqamlLHMDNrViR13tltPXRmZmaZcqExM7NMudCYmVmmXGjMzCxTre5mgEUvr6Zq4oxSxzAzK6plkz5XsmP7jMbMzDLlQmNm1gosWbKE6urqLa+9996bq666CoBrrrmGgw46iL59+/Ktb30LgHnz5m3pO3DgQIBOO3vsog6dSfoocBXwSeB9YBlwPrAAeBbYHXgP+EVETE23OQMYFBHnSWoD3ADUAv8G/BY4MF3+v4iYWLxPY2bWfBx00EHkcjkAamtr6d69O6NGjWLOnDncddddLFy4kHbt2rFq1SoA+vXrR01NDZWVlaxcuZJ99933AEmVEbFpR49dtEIjScAdwNSIOCVtqwa6As9FxCFpW0/gdkltIuKGetv/CtgNOJOkKP04IuZIags8KOn4iLi3WJ/JzKw5evDBBznwwAM54IADuPjii5k4cSLt2rUDoEuXLgC0b99+S//169fv0vGKOXT2aWBjRPyqriEicsDy/E4R8TzwTeDr9ba/GtgH+JeI2BwRayNiTrrNBuApoEd28c3MWobp06czZswYAP7617/y8MMPM2TIEI455hieeOKJLf0ef/xx+vbtS//+/QFe3JmzGShuoekHPFlg36eA3nnLpwKHAac09EEldQL+GXiwoZ1JOkdSjaSa2rWrdyi0mVlLsmHDBu6++26+/OUvA7Bp0ybefvttHnvsMa688kpOPvlk6p68PGTIEJ5++um64tNN0u47c8xyvRlA9ZafAg4ABn+go1QJ3AL8PD0b+oCImBIRgyJiUEX7jk0e1sysubj33ns59NBD6dq1KwA9evTgS1/6EpIYPHgwbdq04Y033thqmz59+kByLbzfzhyzmIXmaZKzkkIcAjyTt/wscDJwq6S+9fpOAZZGxFW7nNDMrIW75ZZbtgybAXzxi19k9uzZQDKMtmHDBjp37swLL7zApk3JANKLL74IyXXxZTtzzGIWmtlAO0ln1zVI+iTJmQp5bVXAj4Fr8tsj4s/AV4EZkvZP+/4A6Ehy55qZmTVi7dq13H///XzpS1/a0jZ+/Hief/55+vXrxymnnMLUqVORxJ/+9CcGDhxIdXU1o0aNAngpIt7Y5s4bobqxuGKQtC/J7c2HAev5x+3NC9n69uZf1t1xln97c7p8JnAxcHy6/bMkt0oDTI6I3zSWoV23XtFt3FVN9pnMzJqDXZ0ZQNKTETFoZ7Yt6vdoIuIVkiGw+vZoZJsbgRvzlm8g+S4NfPBajpmZlZlWN9dZ/+4dqSnhnD9mZq1Nud51ZmZmLYQLjZmZZcqFxszMMuVCY2ZmmXKhMTOzTLnQmJlZplxozMwsUy40ZmaWKRcaMzPLlAuNmZllyoXGzMwy5UJjZmaZanWTai56eTVVE2eUOoY1I7s6vbpZa+czGrMCrF+/nsGDBzNw4ED69u3LpZdeCsB//Md/MGDAAKqrqznuuON45ZVXgOS57GeeeSb9+/dn4MCBPPTQQyVMb1ZaRS00kmol5fJeE9P2Skn/JWlp3rrvNrJdlaShklZLmi/pWUk/LuZnsdalXbt2zJ49mwULFpDL5Zg1axaPPfYYF198MQsXLiSXyzFy5Ei+//3vA3DttdcCsGjRIu6//34uvPBCNm/eXMqPYFYyxR46WxcR1Q20/wD4KNA/ItZL2gu4sLHt0kc+PxwRIyXtAcyXdEdEPJJNdGvNJNGhQwcANm7cyMaNG5HE3nvvvaXPmjVrkJJn8f3lL39h2LBhAHTp0oVOnTpRU1PD4MGDix/erMRKPnQmqT1wNvBvEbEeICLei4jLCt1HRKwDckD3LDKaAdTW1lJdXU2XLl0YPnw4Q4YMAeC73/0u++23H9OmTdtyRjNw4EDuuusuNm3axAsvvMCTTz7J8uXLSxnfrGSKXWj2qDcENhr4OPBSRLxX4HZ31F8p6UNAL2BuRrnNqKioIJfLsWLFCubNm8fixYsB+OEPf8jy5csZO3YskydPBmD8+PH06NGDQYMGcf7553PkkUdSWdnq7r0xA4pfaNZFRHXe69b6HSSdmRaU5ZL2a2C7UXnd/0nSQuBV4J6IeLWhg0o6R1KNpJrataub/lNZq9KpUyeGDh3KrFmztmo/9dRT+f3vfw9AZWUlP/vZz8jlctx1112888479OrVqxRxzUqu5ENnwN+A/dPrMkTEDen1mNVAxXa2fTgiBgD9gXMlVTfUKSKmRMSgiBhU0b5j0yW3VuP111/nnXfeAWDdunU88MAD9O7dm6VLl27pc/fdd9O7d28A1q5dy5o1awC4//77qays5OCDDy56brNyUPJz+YhYK+k6YLKkr6Q3A1QAbXdgH3+VdDnwbWBMVlmt9Vq5ciXjxo2jtraWzZs3c/LJJzNy5EhOPPFElixZQps2bTjggAP41a9+BcCqVav47Gc/S5s2bejevTs33XRTiT+BWekUu9DsISmXtzwrIiYC3wX+H7BY0nvAOmAq8MoO7PtXwEWSPhYRLzRVYDOAAQMGMH/+/A+01w2V1VdVVcWSJUuyjmXWLBS10EREg0NhEbERmJi+GlrfoYG2h4CH8pbX4bvOzMzKTjlcozEzsxas5Ndoiq1/947UeO4qM7Oi8RmNmZllyoXGzMwy5UJjZmaZcqExM7NMudCYmVmmXGjMzCxTLjRmZpYpFxozM8uUC42ZmWXKhcbMzDLlQmNmZplqdXOdLXp5NVUTZ5Q6RpNa5rnbzKyM+YzGzMwy5ULTQowfP54uXbrQr1+/LW0XX3wxvXv3ZsCAAYwaNWrLo4inTZtGdXX1llebNm3I5XKlCW5mLV7ZFBpJ10laIGmhpN9J6pC2d5T0f+m6pyWdWeqs5eiMM85g1qxZW7UNHz6cxYsXs3DhQj7xiU9w+eWXAzB27FhyuRy5XI6bbrqJqqoqqqurS5DazFqDsik0wAURMTAiBgAvAeel7f8K/CUiBgJDgZ9IaluijGXr6KOP5sMf/vBWbccddxyVlclluMMPP5wVK1Z8YLtbbrmFMWPGFCWjmbVORS80kqokPStpat7ZS/uIeDddL2APINJNAtgrbe8AvAVsSvveKenJ9EznnGJ/lubk+uuv5/jjj/9A+6233upCY2aZKtUZzUHAlPTs5V3gawCSbgBeBXoD16R9JwN9gFeARcA3ImJzum58RBwGDAK+Lmmfhg4m6RxJNZJqateuzuozla0f/vCHVFZWMnbs2K3aH3/8cdq3b7/VdR0zs6ZWqkKzPCIeSd/fDBwFEBFnAvsCzwCj0/WfBXJpezUwWdLe6bqvS1oAPAbsB/Rq6GARMSUiBkXEoIr2HZv+05SxqVOncs899zBt2jSSk8J/mD59us9mzCxzpfoeTWxrOSJqJd0KXAzcAJwJTIqIAP4m6QWgt6T2wGeAIyJiraSHgN2Lkr6ZmDVrFj/60Y/44x//SPv27bdat3nzZm677Tbmzp1bonRm1lqU6oxmf0lHpO/HAH+S9HHYco3mn4Fn0/UvAcPSdV1Jht2eBzoCb6dFpjdweBHzl50xY8ZwxBFHsGTJEnr06MF1113Heeedx3vvvcfw4cOprq7mq1/96pb+c+fOpUePHvTs2bOEqc2sNVByolDEA0pVwExgLnAksBQ4Hbgf2BsQsAA4NyLelbQvcCPQLV03KSJultQOuBPoDiwBPgJcFhEPNXb8dt16RbdxVzX1xyopzwxgZlmT9GREDNqZbUs1dLY5Ir5ar+1TDXWMiFeA4xpofx/44G1UZmZWVlrdXGf9u3ekxmcAZmZFU/RCExHLAN9Pa2bWSpTTzABmZtYCudCYmVmmXGjMzCxTLjRmZpYpFxozM8uUC42ZmWXKhcbMzDLlQmNmZplyoTEzs0wVXGgk7ZllEDMza5m2OwWNpCOB35A8Rnl/SQOBr0TE17IOl4VFL6+mauKMUsdokGdhNrOWqJAzmp+RPOXyTYCIWAAcnWUoMzNrOQoaOouI5fWaajPIYqmrr76afv360bdvX6666ioALrvsMrp37051dTXV1dXMnDmztCHNzApUyOzNy9Phs5DUFvg68ExThpAUwM0RcXq6XAmsBB6PiJGSzgCuBF7O22wcMDV9vz+wOn29ERGfacp8xbR48WKuvfZa5s2bR9u2bRkxYgSf+1wypHbBBRdw0UUXlTihmdmOKaTQfBW4muRJliuAPwD/2sQ51gD9JO0REeuA4WxdVABujYjz6rVVA0i6EbgnIn7XxLmK7plnnuHwww+nffv2ABxzzDHccccdJU5lZrbztjt0FhFvRMTYiOgaEV0i4rSIeDODLPcCdVfDxwC3ZHCMstevXz/mzp3Lm2++ydq1a5k5cybLlycjl5MnT2bAgAGMHz+et99+u8RJzcwKs91CI+ljkn4q6XZJd9e9MsgyHThF0u7AAODxeutHS8rlvfYodMeSzpFUI6mmdu3qpszc5Pr06cO3v/1thg8fzogRIxg4cCCVlZWce+65PPfcc+RyObp168aFF15Y6qhmZgUpZOjsTuA64P+AzVkFiYiFkqpIzmYautLd0NBZofueAkwBaNetV+x0yCKZMGECEyZMAOA73/kOPXr0oGvXrlvWn3322YwcObJU8czMdkghhWZ9RPw88ySJu4EfA0OBfYp0zLKzatUqunTpwksvvcTtt9/Oo48+ysqVK+nWrRsAd9xxB/36+WnYZtY8FFJorpZ0KclNAO/XNUbEUxnkuR5YHRGLJA3NYP/Nwoknnsibb77Jbrvtxi9+8Qs+9KEPcfrpp5PL5ZBEVVUVv/71r0sd08ysIIUUmv7A6cCx/GPoLNLlJhURK0jucGvIaElH5S1/LSL+3NQZysHDDz/8gbabbrqpBEnMzHZdIYVmFNAzIjZkFSIiOjTQ9hDwUPr+RuDGRrY/I5NgZma2ywopNAuATsCqbKMUR//uHanxnGJmZkVTSKHpCjwr6Qm2vkbz+cxSmZlZi1FIobk08xRmZtZibbfQRMQfixHEzMxapkJmBjhc0hOS/i5pg6RaSe8WI5yZmTV/hTwmYDLJt/WXAnsAZ6VtZmZm21XINRoi4m+SKiKiFrhBUov8/oqZmTW9QgrN2vQ5NDlJV5A8J2bPbGOZmVlLUcjQ2elpv/NInhuzH3BilqHMzKzlKOSusxfTt+uB72Ubx8zMWprtFhpJnwIuAw7I7x8RPbOLZWZmLUUh12iuAy4AngRqs41jZmYtTSGFZnVE3Jt5kiJZ9PJqqibOyGz/yzyPmpnZVgopNHMkXQncTvbPozEzsxamkEIzJP05KK8tk+fRtBRVVVXstddeVFRUUFlZSU1NDaNHj2bJkiUAvPPOO3Tq1IlcLlfaoGZmRVDIXWefbsoDSqoFFgEiueZzXkT8WVIV8AywBGgL1AATImJjul0l8CpwbURckre/h4BuwLq06QcR8bumzLwz5syZQ+fOnbcs33rrrVveX3jhhXTs2LEUsczMiq6gmQGa2LqIqAaQ9FngcuCYdN1zEVEtqQK4HzgZmJauO46kCJ0s6TsREXn7HBsRNUVJv4sigt/+9rfMnj271FHMzIqikC9sZmlv4O36jelUN/OA7nnNY0ge8/wScHhR0u0kSRx33HEcdthhTJkyZat1Dz/8MF27dqVXr14lSmdmVlylOKPZQ1IO2J1kyOsD13ok7U5ybegb6fIewDDgKyRP+xwDPJq3yTRJdUNnwyLizXr7Owc4B6Bi74804Udp2COPPMK+++7LqlWrGD58OL179+boo48G4JZbbmHMmDGZZzAzKxeFPCagvaT/kHRtutxL0shdOOa6iKiOiN7ACOB/JCldd2BahN4EXoqIhWn7SGBORKwFfg+MSofX6oxN91ldv8gARMSUiBgUEYMq2md/bWTfffcFoEuXLowaNYp58+YBsGnTJm6//XZGjx6deQYzs3JRyNDZDSS3NR+RLq8AftAUB4+IR4HOQN1pxnPp9ZuPA4dLqntc9BjgM5KWkXxxdB+gSW9SaCpr1qzhvffe2/L+D3/4A/369QPggQceoHfv3vTo0aOUEc3MiqqQobMDI2K0pDEAEbEu7wxkl0jqDVSQnMG0r2uPiJWSJgKXpHeVHQXsFxHvp9udSVJ8HmiKHE3ptddeY9SoUUByBnPqqacyYsQIAKZPn+5hMzNrdQopNBvSayQBIOlA8r64uRPqrtFAcovzuIiobaB23Ukyx9o3gNl1RSZ1F3CFpHa7kCMTPXv2ZMGCBQ2uu/HGG4sbxsysDBRSaC4FZgH7SZoGfAo4Y2cPGBEV22hfBvTLWw5g4Db6vsU/htuG7mwWMzPLnrb+Oso2Okn7kNxSLOCxiHgj62BZGTRoUNTUNIuv3JiZlQ1JT0bEoO33/KBC7jr7FLA+ImaQ3Fr8HUkH7MzBzMys9SnkrrNfkjzOeSBwMfAi8D+ZpjIzsxajkEKzKb1e8gXg5xFxNbBXtrHMzKylKORmgPckXQKcBhydflFyt2xjmZlZS1HIGc1oktuZJ0TEqyTzj12ZaSozM2sxCnlMwKvAT/OWX8LXaMzMrECF3HV2uKQnJP1d0gZJtZJWFyOcmZk1f4UMnU0mme5lKbAHcBbwiyxDmZlZy1HQYwIi4m+SKtLnxNwg6c8Z5zIzsxaikEKzVlJbICfpCmAlsGe2sczMrKUoZOjs9LTfecAaYD/gxCxDmZlZy1HQXGctSbtuvaLbuKuaZF/LJn2uSfZjZlbudmWus+0OnaVznV0GHJDfPyJ67swBzcysdSlk6Ow6ku/RHAV8Mu9lqdraWg455BBGjkyecH3ZZZfRvXt3qqurqa6uZubMmSVOaGZWOoUUmtURcW9ErIqIN+teO3tASedJ+pukkNQ5r723pEclvS/ponrbjJC0JN1u4s4eOytXX301ffr02artggsuIJfLkcvlOOGEE0qUzMys9AopNHMkXSnpCEmH1r124ZiPAJ8hmQU631vA14Ef5zemc6v9AjgeOBgYI+ngXTh+k1qxYgUzZszgrLPOKnUUM7OyVEihGQIMAv4L+En6+nGjWwCSqiQ9K2mqpIWSfiepfUTMT5+muZX0jOkJYGO9VYOBv0XE8xGxAZhOMpM0ks5OZy1YIOn3ktoX8Hma1Pnnn88VV1xBmzZb/yonT57MgAEDGD9+PG+//XaxY5mZlY3tFpqI+HQDr2ML3P9BwJSIGAC8C3xtJzJ2B5bnLa9I2wBuj4hPRsRA4BlgQkM7kHSOpBpJNbVrm272nHvuuYcuXbpw2GGHbdV+7rnn8txzz5HL5ejWrRsXXnhhkx3TzKy52eZdZ5JOi4ibJX2zofUR8dOG2utZHhGPpO9vpoGhsQKoocOnP/tJ+gHJkz87APc1tIOImAJMgeT25h08/jY98sgj3H333cycOZP169fz7rvvctppp3HzzTdv6XP22WdvuUnAzKw1auyMpu7b/3s18OpQ4P7r/099Z/4nv4LkS6J1egCvpO9vBM6LiP7A94Ddd2L/O+3yyy9nxYoVLFu2jOnTp3Psscdy8803s3Llyi197rjjDvr161fMWGZmZaWx79HMAIiI79VfIemfC9z//pKOiIhHSSbm/NOOR+QJoJekjwEvA6cAp6br9gJWStoNGJuuL7lvfetb5HI5JFFVVcWvf/3rUkcyMyuZbc4MIGkJ8Nn6F+4lnQn8e0Qc2OiOpSpgJjAXOJJk9ufTSWZ//hbwUWAVMDMizpL0UaAG2BvYDPwdODgi3pV0AnAVUAFcHxE/TI9xbrqvF4FFwF4RcUZjuTwzgJnZjstqZoALgPslnRARS9MDXUJyNnFMgfvfHBFfrdf28/S1lfQBaz0a2klEzCQpWvXbfwn8ssAsZmZWAtssNBExU9L7wL2SvkhyJvJJ4OiIaLb36/bv3pEan4mYmRVNo7c3R8SDwBnAQ0BPYFihRSYilkWEr4KbmbVyjd3e/B7JXWIC2gHDgFWSBERE7F2ciGZm1pw1NnS2VzGDmJlZy1TIFDRmZmY7zYXGzMwy5UJjZmaZcqExM7NMudCYmVmmXGjMzCxTLjRmZpYpFxozM8tUY5NqtkiLXl5N1cQZBff3DM1mZrvGZzRmZpYpF5oCrF+/nsGDBzNw4ED69u3LpZdeCsBbb73F8OHD6dWrF8OHD+ftt5vtpNZmZpkpq0Ij6e/baP8XSYslPS3pL5IuSttvlHRS+v7DkuanD2ZrUu3atWP27NksWLCAXC7HrFmzeOyxx5g0aRLDhg1j6dKlDBs2jEmTJjX1oc3Mmr2yKjQNkXQ8cD5wXET0BQ4FVtfr0xG4D5gSETdkkIEOHToAsHHjRjZu3Igk7rrrLsaNGwfAuHHjuPPOO5v60GZmzV7ZFxrgEuCiiHgFICLWR8S1ees7APcC/5s+cTMTtbW1VFdX06VLF4YPH86QIUN47bXX6NatGwDdunVj1apVWR3ezKzZag6Fph/wZCPrfwr8KSJ+tq0Oks6RVCOppnbt6m11a1RFRQW5XI4VK1Ywb948Fi9evFP7MTNrbZpDodme2cAXJHXZVoeImBIRgyJiUEX7jrt0sE6dOjF06FBmzZpF165dWblyJQArV66kS5dtRjAza7WaQ6F5GjiskfXTgV8CMyVl8rC2119/nXfeeQeAdevW8cADD9C7d28+//nPM3XqVACmTp3KF77whSwOb2bWrDWHL2xeDlwhaWREvCqpHfCViPh5XYeIuEpSN+AOSSdExIamDLBy5UrGjRtHbW0tmzdv5uSTT2bkyJEcccQRnHzyyVx33XXsv//+3HbbbU15WDOzFqHcCk17SSvyln8aET+V1BV4QJKAAK6vv2FEfFvSDcBNksZExOamCjVgwADmz5//gfZ99tmHBx98sKkOY2bWIpVVoYmIBofy0luWP3DbckScUW+5yb9DY2Zmu6asCk0x9O/ekRrPX2ZmVjTN4WYAMzNrxlxozMwsUy40ZmaWKRcaMzPLlAuNmZllyoXGzMwy5UJjZmaZcqExM7NMudCYmVmmXGjMzCxTLjRmZpYpFxozM8tUqys0i15eTdXEGaWOYWbWarS6QlPfO++8w0knnUTv3r3p06cPjz76aKkjmZm1KGVRaCT9Pe/9CZKWStpf0mWSXpaUS9tul3RwXt+HJC2RtEDSE5Kqd/TY3/jGNxgxYgTPPvssCxYsoE+fPk30qczMDMqk0NSRNAy4BhgRES+lzT+LiOqI6AXcCsyW9JG8zcZGxEDgv4Erd+R47777LnPnzmXChAkAtG3blk6dOu3qxzAzszxlU2gk/RNwLfC5iHiuoT4RcSvwB+DUBlY/CnTfkWM+//zzfOQjH+HMM8/kkEMO4ayzzmLNmjU7Gt3MzBpRLoWmHXAX8MWIeHY7fZ8CejfQPgK4c0cOumnTJp566inOPfdc5s+fz5577smkSZN2ZBdmZrYd5VJoNgJ/BiYU0Ff1lqdJWgF8m2TY7YMbSOdIqpFUU7t29Zb2Hj160KNHD4YMGQLASSedxFNPPbUz+c3MbBvKpdBsBk4GPinpO9vpewjwTN7yWOBjwP8Cv2hog4iYEhGDImJQRfuOW9o/+tGPst9++7FkyRIAHnzwQQ4++OCGdmFmZjupstQB6kTEWkkjgYclvRYR19XvI+lE4DjgwnrbbpT078BzkvpExDP1t92Wa665hrFjx7JhwwZ69uzJDTfcsKsfxczM8pRNoQGIiLckjQDmSnojbb5A0mnAnsBi4NiIeL2BbddJ+glwEYUNwQFQXV1NTU1NE6Q3M7OGlEWhiYgOee+XkwyFQXKDwGWNbDe03vJPMohnZma7oFyu0ZiZWQvV6gpN/+4dWTbpc6WOYWbWarS6QmNmZsXlQmNmZplyoTEzs0y50JiZWaZcaMzMLFMuNGZmlikXGjMzy5QLjZmZZcqFxszMMuVCY2ZmmXKhMTOzTLW6QrPo5dXb72RmZk2m1RUaMzMrrlZfaGpraznkkEMYOXJkqaOYmbVImRcaSbWScnmvidvo95CkQQ20nyHp9XTbpyX9TlL7dN03Jf1F0kJJD0o6YEfzXX311fTp02fHP5iZmRWkGGc06yKiOu81qX4HSRXb2cet6bZ9gQ3A6LR9PjAoIgYAvwOu2JFgK1asYMaMGZx11lk7spmZme2Akg2dSVom6T8l/Qn4ctp8mqQ/S1osaXAD21QCewJvA0TEnIhYm65+DOixIxnOP/98rrjiCtq0afUjiGZmmSnG/2H3qDd0Njpv3fqIOCoipqfLe0bEkcDXgOvz+o2WlANeBj4M/F8Dx5kA3NtQAEnnSKqRVFO7Nrnr7J577qFLly4cdthhu/bpzMysUaUYOrs1b92t9freAhARc4G9JXWq6xcR1cBHgUXAxfkbSToNGARc2VCAiJgSEYMiYlBF+44APPLII9x9991UVVVxyimnMHv2bE477bRd+6RmZvYBpR4zWlNvORpbjoggOZs5uq5N0meA7wKfj4j3Cz3w5ZdfzooVK1i2bBnTp0/n2GOP5eabb96x9GZmtl2lLjT1jQaQdBSwOiIa+nblUcBzab9DgF+TFJlVRUtpZmYFqyzCMfZIr6/UmRURDd7iDLwt6c/A3sD4vPbRafFpA6wAzkjbrwQ6ALdJAngpIj6/owGHDh3K0KFDd3QzMzMrgJLRqNajXbde8f7KpaWOYWbWrEh6MiI+8F3HQpTb0Fnm+nfvWOoIZmatSqsrNGZmVlwuNGZmlikXGjMzy5QLjZmZZcqFxszMMtXqbm+W9B6wpNQ5tqMz8EapQxTAOZuWczad5pARmk/O3YEVETFiZzZujYWmZmfvBS+W5pARnLOpOWfTaQ4ZofXk9NCZmZllyoXGzMwy1RoLzZRSByhAc8gIztnUnLPpNIeM0EpytrprNGZmVlyt8YzGzMyKyIXGzMwy1WoKjaQRkpZI+pukbT0Pp1hZrpe0StLivLYPS7pf0tL054fy1l2S5l4i6bNFyrifpDmSnpH0tKRvlGnO3SXNk7Qgzfm9csyZd+wKSfMl3VOuOSUtk7RIUk5STRnn7CTpd5KeTf87PaKccko6KP0d1r3elXR+OWXMO+4F6d+fxZJuSf9eNV3OiGjxL6CC5KmcPYG2wALg4BLmORo4FFic13YFMDF9PxH4Ufr+4DRvO+Bj6eeoKELGbsCh6fu9gL+mWcotp4AO6fvdgMeBw8stZ17ebwL/C9xTjn/u6bGXAZ3rtZVjzqnAWen7tkCncsyZHr8CeBU4oNwyAt2BF4A90uXfkjxcsslyFuWXXOoXcARwX97yJcAlJc5UxdaFZgnQLX3fDVjSUFbgPuCIEuS9CxhezjmB9sBTwJByzAn0AB4EjuUfhaYccy7jg4WmrHKSPIX3BdIbmso1Z97xjgMeKceMJIVmOfBhkqcu35PmbbKcrWXorO4XWWdF2lZOukbESoD0Z5e0veTZJVUBh5CcLZRdznQ4KgesAu6PiLLMCVwFfAvYnNdWjjkD+IOkJyWdU6Y5ewKvAzekQ5G/kbRnGeascwpwS/q+rDJGxMvAj4GXgJXA6oj4Q1PmbC2FRg20NZf7ukuaXVIH4PfA+RHxbmNdG2grSs6IqI2IapIzhsGS+jXSvSQ5JY0EVkXEk4Vu0kBbsf7cPxURhwLHA/8q6ehG+pYqZyXJ8PMvI+IQYA3J8M62lOz3Kakt8Hngtu11baCtGP9tfgj4Askw2L7AnpJOa2yTBtoazdlaCs0KYL+85R7AKyXKsi2vSeoGkP5clbaXLLuk3UiKzLSIuL1cc9aJiHeAh4ARlF/OTwGfl7QMmA4cK+nmMsxJRLyS/lwF3AEMLsOcK0gmeXw8Xf4dSeEpt5yQFOynIuK1dLncMn4GeCEiXo+IjcDtwJFNmbO1FJongF6SPpb+6+IU4O4SZ6rvbmBc+n4cyTWRuvZTJLWT9DGgFzAv6zCSBFwHPBMRPy3jnB+R1Cl9vwfJX5pnyy1nRFwSET0ioorkv7/ZEXFaueWUtKekverek4zVLy63nBHxKrBc0kFp0zDgL+WWMzWGfwyb1WUpp4wvAYdLap/+vR8GPNOkOYt1MazUL+AEkjunngO+W+Ist5CMhW4k+dfBBGAfkgvFS9OfH87r/9009xLg+CJlPIrkdHghkEtfJ5RhzgHA/DTnYuA/0/ayylkv81D+cTNAWeUkufaxIH09Xfd3pdxypsetBmrSP/s7gQ+VW06SG1TeBDrmtZVVxvS43yP5B9pi4CaSO8qaLKenoDEzs0y1lqEzMzMrERcaMzPLlAuNmZllyoXGzMwy5UJjZmaZcqGxZktSSPpJ3vJFki5ron3fKOmkptjXdo7z5XTm4TkF9v9OEx67k6SvNdX+zLbFhcaas/eBL0nqXOog+SRV7ED3CcDXIuLTBfZvsNAosaN/nzsBO1xodvDzmbnQWLO2ieRZ5hfUX1H/jETS39OfQyX9UdJvJf1V0iRJY5U802aRpAPzdvMZSQ+n/Uam21dIulLSE5IWSvpK3n7nSPpfYFEDecak+18s6Udp23+SfDH2V5KurNe/m6S5Sp5jsljSP0maBOyRtk2TVJWeDf03yazV+9V9znQfJ0m6MX3fVdIdSp7bs0DSkcAk4MB0f1emn+GevO0nSzojfb9M0n9K+hPwZUnHSXpU0lOSblMyJx7p7/Mv6e/mx4X9MVpLV1nqAGa76BfAQklX7MA2A4E+wFvA88BvImKwkoe7/RtwftqvCjgGOBCYI+njwL+QzG77SUntgEck/SHtPxjoFxEv5B9M0r7Aj4DDgLdJZkb+YkR8X9KxwEURUVMv46kkj7b4YXoG0T4iHpZ0XiQTiNbNqn0QcGZEfC1t29Zn/jnwx4gYle6vA8kklP3y9jd0O7+39RFxVHoGeTvwmYhYI+nbwDclTQZGAb0jIuqmBjJzobFmLSLelfQ/wNeBdQVu9kSk059Leg6oKxSLgPwhrN9GxGZgqaTngd4kc38NyDtb6kgy19MGYF79IpP6JPBQRLyeHnMaycPv7mwsI3C9kolN74yI3Db6vRgRjzX2YVPHkhRJIqIWWK28JyYW6Nb05+EkD796JC1sbYFHgXeB9cBvJM0gea6JmYfOrEW4iuRax555bZtI//tOJwpsm7fu/bz3m/OWN7P1P77qz88UJFOk/1tEVKevj0Xy7A5IpqpvyDZPM7YlIuaSFKOXgZsk/cs2utY/Zn7m3XfwsFt+Z9vYvu5YInnuT93v4OCImBARm0jO6n4PfBGYtYPHtxbKhcaavYh4i+TxsxPympeRDFVB8qyN3XZi11+W1Ca9btOTZALB+4Bz0zMNJH1CySzHjXkcOEZS53TYagzwx8Y2kHQAyfNrriWZRfvQdNXGumNvw2uS+qQ3BozKa38QODfdd4WkvYH3SB7TXedF4GAls/J2JJnFtyGPAZ9KhxJRMuvvJ9LrNB0jYibJ8GN1Y5/RWg8XGmspfgLk3312Lcn/3OeRPNp5W2cbjVlCUhDuBb4aEeuB35BMR/+UpMXAr9nOEHQ6THcJMIdkVuSnIuKuxrYhmeE5J2k+cCJwddo+heSa1LRtbDeRZMhqNskM4XW+AXxa0iLgSaBvRLxJMvy1WNKVEbGcpGAvBKaRzIrd0Od5neSZ8rdIWkhSeHqTFK170rY/0sBNGtY6efZmMzPLlM9ozMwsUy40ZmaWKRcaMzPLlAuNmZllyoXGzMwy5UJjZmaZcqExM7NM/X9NBGuncZxNrQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "n_structures_per_kinase.plot(kind=\"barh\", ax=ax)\n",
    "ax.set_xlabel(\"Number of structures\")\n",
    "ax.set_ylabel(\"Kinase name\")\n",
    "for i, value in enumerate(n_structures_per_kinase):\n",
    "    ax.text(value, i, str(value), va=\"center\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that our dataset is highly imbalanced. While some kinases are structurally resolved very often, other kinases are not: See e.g. CDK2 and EGFR vs. KDR and ErbB2. We will have to keep this in mind when interpreting our results later."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare structures\n",
    "\n",
    "We will compare the structures' IFP pairwise using the Tanimoto/Jaccard distance as implemented in `sklearn.metrics.pairwise_distances`, which uses under the hood the method `scipy.spatial.distance`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare IFPs as `numpy` array\n",
    "\n",
    "KLIFS deposits the IFP as a string of 0 and 1. We have to prepare our IFPs as an array of boolean vectors (required by `scipy.spatial.distance` to be able to use the Jaccard distance). Each row in this array refers to one IFP, each columns to one of the IFP's features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "structure.klifs_id  kinase.klifs_name\n",
       "775                 EGFR                 0000000000000010000000000000000000000000000000...\n",
       "777                 EGFR                 0000000000000010000001000000000000000000000000...\n",
       "778                 EGFR                 0000000000000010000000000000000000000000000000...\n",
       "779                 EGFR                 0000000000000010000001000000000000000000000000...\n",
       "782                 EGFR                 0000000000000010000001000000000000000000000000...\n",
       "Name: interaction.fingerprint, dtype: string"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is the KLIFS format of the IFP (structure KLIFS ID and kinase name set as index)\n",
    "ifp_series = structures_with_ifps.set_index([\"structure.klifs_id\", \"kinase.klifs_name\"])[\"interaction.fingerprint\"]\n",
    "ifp_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cast string to list of \"0\" and \"1\"\n",
    "# ifps_series = ifp_series.apply(list)\n",
    "# ifp_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "structure.klifs_id  kinase.klifs_name\n",
       "775                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "777                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "778                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "779                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "782                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "Name: interaction.fingerprint, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cast \"0\" and \"1\" to boolean False and True\n",
    "ifp_series = ifp_series.apply(lambda x: [True if i == \"1\" else False for i in x])\n",
    "ifp_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       ...,\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert to numpy array\n",
    "ifps_array = np.array(ifp_series.to_list())\n",
    "ifps_array"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate pairwise Jaccard distances\n",
    "\n",
    "The Jaccard distance is a common distance measure in case of binary fingerprints. \n",
    "\n",
    "$d_J(A,B) = 1 - J(A,B) = \\frac{\\mid A \\mid \\cup \\mid B \\mid - \\mid A \\mid \\cap \\mid B \\mid}{\\mid A \\mid + \\mid B \\mid}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "structure_distance_matrix = pairwise_distances(ifps_array, metric=\"jaccard\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure distance matrix size: (1466, 1466)\n",
      "Show matrix subset:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th>775</th>\n",
       "      <th>777</th>\n",
       "      <th>778</th>\n",
       "      <th>779</th>\n",
       "      <th>782</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.483871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.703704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.411765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.655172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>0.483871</td>\n",
       "      <td>0.703704</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "structure.klifs_id       775       777       778       779       782\n",
       "structure.klifs_id                                                  \n",
       "775                 0.000000  0.720000  0.542857  0.500000  0.483871\n",
       "777                 0.720000  0.000000  0.741935  0.500000  0.703704\n",
       "778                 0.542857  0.741935  0.000000  0.696970  0.411765\n",
       "779                 0.500000  0.500000  0.696970  0.000000  0.655172\n",
       "782                 0.483871  0.703704  0.411765  0.655172  0.000000"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create DataFrame with structure KLIFS IDs as index/columns\n",
    "structure_klifs_ids = ifp_series.index.get_level_values(0)\n",
    "structure_distance_matrix = pd.DataFrame(\n",
    "    structure_distance_matrix, index=structure_klifs_ids, columns=structure_klifs_ids\n",
    ")\n",
    "print(f\"Structure distance matrix size: {structure_distance_matrix.shape}\")\n",
    "print(\"Show matrix subset:\")\n",
    "structure_distance_matrix.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Map structure to kinase distance matrix\n",
    "\n",
    "First &mdash; as an intermediate step &mdash; we will create a structure distance matrix but instead of labeling the data with structure KLIFS IDs we add the corresponding kinase name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Show matrix subset:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>kinase.klifs_name</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kinase.klifs_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.483871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.703704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.411765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.655172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.483871</td>\n",
       "      <td>0.703704</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "kinase.klifs_name      EGFR      EGFR      EGFR      EGFR      EGFR\n",
       "kinase.klifs_name                                                  \n",
       "EGFR               0.000000  0.720000  0.542857  0.500000  0.483871\n",
       "EGFR               0.720000  0.000000  0.741935  0.500000  0.703704\n",
       "EGFR               0.542857  0.741935  0.000000  0.696970  0.411765\n",
       "EGFR               0.500000  0.500000  0.696970  0.000000  0.655172\n",
       "EGFR               0.483871  0.703704  0.411765  0.655172  0.000000"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Copy distance matrix to kinase matrix\n",
    "kinase_distance_matrix = structure_distance_matrix\n",
    "# Replace structure KLIFS IDs with the structures' kinase names\n",
    "kinase_names = ifp_series.index.get_level_values(1)\n",
    "kinase_distance_matrix.index = kinase_names\n",
    "kinase_distance_matrix.columns = kinase_names\n",
    "print(\"Show matrix subset:\")\n",
    "kinase_distance_matrix.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A kinase pair can be represented by many different structure pairs which are associated with different distance values. For example, if we compare all EGFR structure to each other, the range of distance values is already quite high because we can observe quite different binding modes of the co-crystallized ligands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAALOElEQVR4nO3dQYxd51mH8edvW1kgSOuQAYEdYi/cgi01CIbABlGEyNhlYSGxSIqoiCpZlhrEMt5Ai7pigYRQUyyrsiI2eEMEBhk8K+iiVPJYKmndytXIlePBSJlgC6SyiNy+LHxT3dxezz3j3vH1vHl+0khzzvnm3teLPPl8PGcmVYUkaffbs+gBJEnzYdAlqQmDLklNGHRJasKgS1IT+xb1xk8//XQdOnRoUW8vSbvS1atX36mqpWnXFhb0Q4cOsba2tqi3l6RdKcnNB13zloskNWHQJakJgy5JTRh0SWrCoEtSEwZdGrOyssKePXtIwp49e1hZWVn0SNJgBl0aWVlZYXV1lfd+AmlVsbq6atS1axh0aWR1dXVb56XHjUGXpCYMuiQ1YdClCceOHePmzZscO3Zs0aNI27Kwn+UiPa6uXbvGs88+u+gxpG1zhy5JTRh0SWrCoEsjL7zwwrbOS48b76GrvSQ/0tevrq4Oeo33HkiSFsWgq72HCW0SA61dx1suktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1MSjoSY4nuZ5kPcmZKdc/lOQfk/xHkmtJXp7/qJKkrcwMepK9wGvACeAo8FKSoxPLPgN8s6qeAz4O/EWSJ+Y8qyRpC0N26M8D61V1o6reBS4AJyfWFPATSQL8OHAHuDfXSSVJWxoS9APArbHjjdG5cV8AfgG4DXwd+OOq+v7kCyU5lWQtydrm5uZDjixJmmZI0DPlXE0crwBfA34W+EXgC0me/KEvqjpXVctVtby0tLTNUSVJWxkS9A3gmbHjg9zfiY97GXij7lsHvgP8/HxGlCQNMSToV4AjSQ6P/qHzReDixJq3gN8CSPLTwEeBG/McVJK0tX2zFlTVvSSvAJeBvcD5qrqW5PTo+lng88DrSb7O/Vs0r1bVOzs4tyRpwsygA1TVJeDSxLmzY5/fBl6Y72iSpO3wSVFJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJvYtegBpO5566inu3r37SN4ryY6+/v79+7lz586Ovoc+WAy6dpW7d+9SVYseYy52+n8Y+uDxloskNWHQJamJQUFPcjzJ9STrSc48YM3Hk3wtybUk/zbfMSVJs8y8h55kL/Aa8NvABnAlycWq+ubYmg8DXwSOV9VbSX5qh+aVJD3AkB3688B6Vd2oqneBC8DJiTWfBN6oqrcAqurt+Y4pSZplSNAPALfGjjdG58Z9BNif5F+TXE3yqWkvlORUkrUka5ubmw83sSRpqiFBn/a9VZPfN7YP+GXgd4AV4E+SfOSHvqjqXFUtV9Xy0tLStoeVJD3YkO9D3wCeGTs+CNyesuadqvou8N0kXwaeA749lyklSTMN2aFfAY4kOZzkCeBF4OLEmn8Afj3JviQ/Bvwq8K35jipJ2srMHXpV3UvyCnAZ2Aucr6prSU6Prp+tqm8l+RfgTeD7wJeq6hs7Obgk6f2yqMeol5eXa21tbSHvrd0rSatH/7v8WfToJLlaVcvTrvmkqCQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEv4JOu0p99kn43IcWPcZc1GefXPQIasaga1fJn/1vm4dxklCfW/QU6sRbLpLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IS/sUi7TpJFjzAX+/fvX/QIasaga1d5VL9+LkmbX3WnDw5vuUhSEwZdkpow6JLUhEGXpCYGBT3J8STXk6wnObPFul9J8r0kvze/ESVJQ8wMepK9wGvACeAo8FKSow9Y9+fA5XkPKUmabcgO/XlgvapuVNW7wAXg5JR1fwT8HfD2HOeTJA00JOgHgFtjxxujcz+Q5ADwu8DZrV4oyakka0nWNjc3tzurJGkLQ4I+7bG8yScu/hJ4taq+t9ULVdW5qlququWlpaWBI0qShhjypOgG8MzY8UHg9sSaZeDC6JHsp4FPJLlXVX8/jyElSbMNCfoV4EiSw8B/Ai8CnxxfUFWH3/s8yevAPxlzSXq0Zga9qu4leYX7372yFzhfVdeSnB5d3/K+uSTp0Rj0w7mq6hJwaeLc1JBX1R/+6GNJkrbLJ0UlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpiUFBT3I8yfUk60nOTLn++0neHH18Jclz8x9VkrSVmUFPshd4DTgBHAVeSnJ0Ytl3gN+oqo8BnwfOzXtQSdLWhuzQnwfWq+pGVb0LXABOji+oqq9U1d3R4VeBg/MdU5I0y5CgHwBujR1vjM49yKeBf552IcmpJGtJ1jY3N4dPKUmaaUjQM+VcTV2Y/Cb3g/7qtOtVda6qlqtqeWlpafiUkqSZ9g1YswE8M3Z8ELg9uSjJx4AvASeq6r/nM54kaaghO/QrwJEkh5M8AbwIXBxfkOTngDeAP6iqb89/TEnSLDN36FV1L8krwGVgL3C+qq4lOT26fhb4U+AngS8mAbhXVcs7N7YkaVKqpt4O33HLy8u1tra2kPeWZknCov7bkLaS5OqDNsxD7qFLHxijv2G+73PDrt3CR/+lkfGYDzkvPW7coau9eQR5yGu4k9eiGXS1NzS0W0XbWGs38JaLJDVh0CWpCYMuSU0YdElqwqBLUhMGXZKaMOiS1IRBl6QmDLokNWHQJakJgy5JTRh0SWrCoEtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklNGHRJasKgS1ITBl2SmjDoktSEQZekJgy6JDVh0CWpCYMuSU0YdElqwqBLUhODgp7keJLrSdaTnJlyPUn+anT9zSS/NP9RpZ1VVds6Lz1uZgY9yV7gNeAEcBR4KcnRiWUngCOjj1PAX895TmnHJdnWeelxM2SH/jywXlU3qupd4AJwcmLNSeBv6r6vAh9O8jNznlV6JKrqBx/SbjIk6AeAW2PHG6Nz211DklNJ1pKsbW5ubndWSdIWhgR92t83J7cuQ9ZQVeeqarmqlpeWlobMJ0kaaN+ANRvAM2PHB4HbD7FG2hW8Z67dasgO/QpwJMnhJE8ALwIXJ9ZcBD41+m6XXwP+p6r+a86zSjvK73LRbjdzh15V95K8AlwG9gLnq+paktOj62eBS8AngHXg/4CXd25kaecYb+1mQ265UFWXuB/t8XNnxz4v4DPzHU2StB0+KSpJTRh0SWrCoEtSEwZdkprIov5VP8kmcHMhby7N9jTwzqKHkKZ4tqqmPpm5sKBLj7Mka1W1vOg5pO3wloskNWHQJakJgy5Nd27RA0jb5T10SWrCHbokNWHQJakJgy6NSXI+ydtJvrHoWaTtMujS+70OHF/0ENLDMOjSmKr6MnBn0XNID8OgS1ITBl2SmjDoktSEQZekJgy6NCbJ3wL/Dnw0yUaSTy96JmkoH/2XpCbcoUtSEwZdkpow6JLUhEGXpCYMuiQ1YdAlqQmDLklN/D94L39H1DcCMwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Select EGFR-EGFR structure pairs only\n",
    "D = kinase_distance_matrix.loc[\"EGFR\", \"EGFR\"]\n",
    "# Extract all pairwise distances without identical structure pairs\n",
    "# = lower triangular matrix without the diagonal\n",
    "D_condensed = distance.squareform(D)\n",
    "# Plot pairwise distances\n",
    "plt.boxplot(D_condensed);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this talktorial, we will consider per kinase pair the two structures that show the most similar binding mode for their co-crystallized ligands. Hence, we select the structure pair with the minimum IFP distance as representative for a kinase pair."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We unstack the matrix (each pairwise comparison in a single row)\n",
    "# We group by kinase names (level=[0, 1] ensures that the order of the kinases is ignored\n",
    "# We take the minimum value in each kinase pair group\n",
    "# We unstack the remaining data points\n",
    "kinase_distance_matrix = (\n",
    "    kinase_distance_matrix.unstack().groupby(level=[0, 1]).min().unstack(level=1)\n",
    ")\n",
    "kinase_distance_matrix.index.name = None\n",
    "kinase_distance_matrix.columns.name = None\n",
    "\n",
    "# Show matrix with background gradient\n",
    "cm = sns.light_palette(\"green\", as_cmap=True)\n",
    "kinase_distance_matrix.style.background_gradient(cmap=cm).format(\"{:.3f}\") \n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Save kinase distance matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kinase_distance_matrix.to_csv(DATA / \"kinase_distance_matrix_ifp.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Discussion\n",
    "\n",
    "In this talktorial, we have used the KLIFS interaction fingerprints, which describe binding modes of co-crystallized ligands in kinase structures, to assess kinase similarity.\n",
    "\n",
    "The approach shown here is a rather simple one. We only compare the two closest binding modes per kinase pair, although we have - at least for kinases such as EGFR and CDK2 - much more data available on binding modes. As an alternative, we could instead aggregate the information from multiple binding modes per kinase in a fingerprint that reports all interactions seen in any structure. However, this approach has its drawbacks itself: Some kinases have much higher coverage than others, leading to an imbalance in information content.\n",
    "\n",
    "The kinase distance matrix above will be reloaded in **Talktorial T028**, where we compare kinase similarities from different perspectives, including the pocket structure perspective we have talked about in this talktorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quiz\n",
    "\n",
    "1. Why is the structural coverage between kinases so different?\n",
    "2. At one point, we are mapping structure pairs to kinase pairs. We use here the minimum structure distance. What other mapping strategies can you think of? What advantages and disadvantages could they have?\n",
    "3. What is the difference between the distances that we show in the heatmap and the dendrogram? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc-autonumbering": true,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
