{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Talktorial 11\n",
    "\n",
    "# CADD web services that can be used via a Python API\n",
    "\n",
    "__Developed at AG Volkamer, Charité__\n",
    "\n",
    "Dr. Jaime Rodríguez-Guerra\n",
    "\n",
    "## Aim of this talktorial\n",
    "\n",
    "Web services are a convenient way of using software because it frees the user from any installation hassles. A web UI is usually available for easy usage, at the cost of losing the possibility to automate a workflow. Fortunately, the number of web services that provide an API for automated access has been increasing. Some examples in the field of Computer Aided Drug Design include:\n",
    "\n",
    "- ChEMBL\n",
    "- RCSB PDB\n",
    "- KLIFS\n",
    "- Proteins.plus\n",
    "- SwissDock\n",
    "\n",
    "In this notebook, you will learn how to programmatically use online web-services from Python, always in the context of drug design. The end goal will be to build a full pipeline that exclusively relies on web-services, without any local execution!\n",
    "\n",
    "## Learning goals\n",
    "\n",
    "### Theory\n",
    "\n",
    "- Types of programmatic access\n",
    "- Description of the pipeline and the involved webservices\n",
    "\n",
    "### Practical\n",
    "\n",
    "- Building the pipeline\n",
    "- Case study: XXXXXXXXXXX\n",
    "\n",
    "### Discussion\n",
    "\n",
    "Pending.\n",
    "\n",
    "### Quiz\n",
    "\n",
    "Pending.\n",
    "\n",
    "***\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Theory\n",
    "\n",
    "## Types of programmatic access\n",
    "\n",
    "### Programmatic APIs\n",
    "\n",
    "Modern webservices are able to provide standardized ways to access their data, especially when it comes to databases. This usually means that you can access a specific URL in their site to request results that are __machine readable__.\n",
    "\n",
    "For example, [UniProt](https://www.uniprot.org/) is a database for all types of information concerning proteins. If you look for a specific protein, like `Src` (involved in cancer), you will arrive to [this beautiful webpage](https://www.uniprot.org/uniprot/P12931) with well organized content sections. If you add `.fasta` to the URL, however, you will obtain the protein sequence in `FASTA` format.\n",
    "\n",
    "```\n",
    "https://www.uniprot.org/uniprot/P12931 -> https://www.uniprot.org/uniprot/P12931.fasta\n",
    "---\n",
    "\n",
    ">sp|P12931|SRC_HUMAN Proto-oncogene tyrosine-protein kinase Src OS=Homo sapiens OX=9606 GN=SRC PE=1 SV=3\n",
    "MGSNKSKPKDASQRRRSLEPAENVHGAGGGAFPASQTPSKPASADGHRGPSAAFAPAAAE\n",
    "PKLFGGFNSSDTVTSPQRAGPLAGGVTTFVALYDYESRTETDLSFKKGERLQIVNNTEGD\n",
    "WWLAHSLSTGQTGYIPSNYVAPSDSIQAEEWYFGKITRRESERLLLNAENPRGTFLVRES\n",
    "ETTKGAYCLSVSDFDNAKGLNVKHYKIRKLDSGGFYITSRTQFNSLQQLVAYYSKHADGL\n",
    "CHRLTTVCPTSKPQTQGLAKDAWEIPRESLRLEVKLGQGCFGEVWMGTWNGTTRVAIKTL\n",
    "KPGTMSPEAFLQEAQVMKKLRHEKLVQLYAVVSEEPIYIVTEYMSKGSLLDFLKGETGKY\n",
    "LRLPQLVDMAAQIASGMAYVERMNYVHRDLRAANILVGENLVCKVADFGLARLIEDNEYT\n",
    "ARQGAKFPIKWTAPEAALYGRFTIKSDVWSFGILLTELTTKGRVPYPGMVNREVLDQVER\n",
    "GYRMPCPPECPESLHDLMCQCWRKEPEERPTFEYLQAFLEDYFTSTEPQYQPGENL\n",
    "\n",
    "```\n",
    "\n",
    "This is a way to provide programmatic access to a webservice: through specific URL schemes. However, each webservice would have to come up with their own scheme, which will force the developers to implement their scripts in case-by-case basis.\n",
    "\n",
    "Fortunately, there are some standardized ways to provide this kind of programmatic access:\n",
    "\n",
    "- HTTP-based RESTful APIs ([wiki](https://en.wikipedia.org/wiki/Representational_state_transfer#Applied_to_Web_services)).\n",
    "- GraphQL\n",
    "- SOAP\n",
    "- gRPC\n",
    "\n",
    "In this notebook, we will only discuss REST APIs\n",
    "\n",
    "#### HTTP-based RESTful APIs\n",
    "\n",
    "This type of programmatic access/provision defines a specific entry point for clients (scripts, libraries, programs) that require programmatic access, something like `api.webservice.com`. They can be versioned, so the provider can update the scheme without disrupting existing implementations (`api.webservice.com/v1` will still work even when `api.webservice.com/v2` has been deployed).\n",
    "\n",
    "This kind of APIs are usually accompanyed by well-written documentation explaining all the available actions in the platform. For example, look at the [GitHub REST API for listing repositories](https://developer.github.com/v3/repos/#list-organization-repositories). You can see how every argument and option is documented, along with usage examples. \n",
    "\n",
    "The only difficulty is to build the needed URLs in the correct way:\n",
    "\n",
    "```\n",
    "https://api.github.com/users/volkamerlab/repos\n",
    "```\n",
    "\n",
    "<details>\n",
    "    <summary>\n",
    "        Returns\n",
    "    </summary>\n",
    "\n",
    "```\n",
    "[\n",
    "  {\n",
    "    \"id\": 156864288,\n",
    "    \"node_id\": \"MDEwOlJlcG9zaXRvcnkxNTY4NjQyODg=\",\n",
    "    \"name\": \"TeachOpenCADD\",\n",
    "    \"full_name\": \"volkamerlab/TeachOpenCADD\",\n",
    "    \"private\": false,\n",
    "    \"owner\": {\n",
    "      \"login\": \"volkamerlab\",\n",
    "      \"id\": 44878588,\n",
    "      \"node_id\": \"MDEyOk9yZ2FuaXphdGlvbjQ0ODc4NTg4\",\n",
    "      \"avatar_url\": \"https://avatars2.githubusercontent.com/u/44878588?v=4\",\n",
    "      \"gravatar_id\": \"\",\n",
    "      \"url\": \"https://api.github.com/users/volkamerlab\",\n",
    "      \"html_url\": \"https://github.com/volkamerlab\",\n",
    "      \"followers_url\": \"https://api.github.com/users/volkamerlab/followers\",\n",
    "      \"following_url\": \"https://api.github.com/users/volkamerlab/following{/other_user}\",\n",
    "      \"gists_url\": \"https://api.github.com/users/volkamerlab/gists{/gist_id}\",\n",
    "      \"starred_url\": \"https://api.github.com/users/volkamerlab/starred{/owner}{/repo}\",\n",
    "      \"subscriptions_url\": \"https://api.github.com/users/volkamerlab/subscriptions\",\n",
    "      \"organizations_url\": \"https://api.github.com/users/volkamerlab/orgs\",\n",
    "      \"repos_url\": \"https://api.github.com/users/volkamerlab/repos\",\n",
    "      \"events_url\": \"https://api.github.com/users/volkamerlab/events{/privacy}\",\n",
    "      \"received_events_url\": \"https://api.github.com/users/volkamerlab/received_events\",\n",
    "      \"type\": \"Organization\",\n",
    "      \"site_admin\": false\n",
    "    },\n",
    "    \"html_url\": \"https://github.com/volkamerlab/TeachOpenCADD\",\n",
    "    \"description\": \"TeachOpenCADD: a teaching platform for computer-aided drug design (CADD) using open source packages and data\",\n",
    "    \"fork\": false,\n",
    "    \"url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD\",\n",
    "    \"forks_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/forks\",\n",
    "    \"keys_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/keys{/key_id}\",\n",
    "    \"collaborators_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/collaborators{/collaborator}\",\n",
    "    \"teams_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/teams\",\n",
    "    \"hooks_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/hooks\",\n",
    "    \"issue_events_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues/events{/number}\",\n",
    "    \"events_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/events\",\n",
    "    \"assignees_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/assignees{/user}\",\n",
    "    \"branches_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/branches{/branch}\",\n",
    "    \"tags_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/tags\",\n",
    "    \"blobs_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/blobs{/sha}\",\n",
    "    \"git_tags_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/tags{/sha}\",\n",
    "    \"git_refs_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/refs{/sha}\",\n",
    "    \"trees_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/trees{/sha}\",\n",
    "    \"statuses_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/statuses/{sha}\",\n",
    "    \"languages_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/languages\",\n",
    "    \"stargazers_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/stargazers\",\n",
    "    \"contributors_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/contributors\",\n",
    "    \"subscribers_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/subscribers\",\n",
    "    \"subscription_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/subscription\",\n",
    "    \"commits_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/commits{/sha}\",\n",
    "    \"git_commits_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/commits{/sha}\",\n",
    "    \"comments_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/comments{/number}\",\n",
    "    \"issue_comment_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues/comments{/number}\",\n",
    "    \"contents_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/contents/{+path}\",\n",
    "    \"compare_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/compare/{base}...{head}\",\n",
    "    \"merges_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/merges\",\n",
    "    \"archive_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/{archive_format}{/ref}\",\n",
    "    \"downloads_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/downloads\",\n",
    "    \"issues_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues{/number}\",\n",
    "    \"pulls_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/pulls{/number}\",\n",
    "    \"milestones_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/milestones{/number}\",\n",
    "    \"notifications_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/notifications{?since,all,participating}\",\n",
    "    \"labels_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/labels{/name}\",\n",
    "    \"releases_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/releases{/id}\",\n",
    "    \"deployments_url\": \"https://api.github.com/repos/volkamerlab/TeachOpenCADD/deployments\",\n",
    "    \"created_at\": \"2018-11-09T13:15:15Z\",\n",
    "    \"updated_at\": \"2019-07-18T02:36:48Z\",\n",
    "    \"pushed_at\": \"2019-05-03T15:02:03Z\",\n",
    "    \"git_url\": \"git://github.com/volkamerlab/TeachOpenCADD.git\",\n",
    "    \"ssh_url\": \"git@github.com:volkamerlab/TeachOpenCADD.git\",\n",
    "    \"clone_url\": \"https://github.com/volkamerlab/TeachOpenCADD.git\",\n",
    "    \"svn_url\": \"https://github.com/volkamerlab/TeachOpenCADD\",\n",
    "    \"homepage\": null,\n",
    "    \"size\": 28121,\n",
    "    \"stargazers_count\": 39,\n",
    "    \"watchers_count\": 39,\n",
    "    \"language\": \"Jupyter Notebook\",\n",
    "    \"has_issues\": true,\n",
    "    \"has_projects\": true,\n",
    "    \"has_downloads\": true,\n",
    "    \"has_wiki\": true,\n",
    "    \"has_pages\": false,\n",
    "    \"forks_count\": 13,\n",
    "    \"mirror_url\": null,\n",
    "    \"archived\": false,\n",
    "    \"disabled\": false,\n",
    "    \"open_issues_count\": 0,\n",
    "    \"license\": {\n",
    "      \"key\": \"cc-by-4.0\",\n",
    "      \"name\": \"Creative Commons Attribution 4.0 International\",\n",
    "      \"spdx_id\": \"CC-BY-4.0\",\n",
    "      \"url\": \"https://api.github.com/licenses/cc-by-4.0\",\n",
    "      \"node_id\": \"MDc6TGljZW5zZTI1\"\n",
    "    },\n",
    "    \"forks\": 13,\n",
    "    \"open_issues\": 0,\n",
    "    \"watchers\": 39,\n",
    "    \"default_branch\": \"master\"\n",
    "  }\n",
    "]\n",
    "```\n",
    "</details>\n",
    "\n",
    ", which happens to be a [JSON](https://en.wikipedia.org/wiki/JSON)-formatted dictionary! This is easily parsed into a Python dictionary using the `json` library. The best news is that you don't even need that. Using `requests` the following operation can be done in three lines:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 156864288,\n",
       "  'node_id': 'MDEwOlJlcG9zaXRvcnkxNTY4NjQyODg=',\n",
       "  'name': 'TeachOpenCADD',\n",
       "  'full_name': 'volkamerlab/TeachOpenCADD',\n",
       "  'private': False,\n",
       "  'owner': {'login': 'volkamerlab',\n",
       "   'id': 44878588,\n",
       "   'node_id': 'MDEyOk9yZ2FuaXphdGlvbjQ0ODc4NTg4',\n",
       "   'avatar_url': 'https://avatars2.githubusercontent.com/u/44878588?v=4',\n",
       "   'gravatar_id': '',\n",
       "   'url': 'https://api.github.com/users/volkamerlab',\n",
       "   'html_url': 'https://github.com/volkamerlab',\n",
       "   'followers_url': 'https://api.github.com/users/volkamerlab/followers',\n",
       "   'following_url': 'https://api.github.com/users/volkamerlab/following{/other_user}',\n",
       "   'gists_url': 'https://api.github.com/users/volkamerlab/gists{/gist_id}',\n",
       "   'starred_url': 'https://api.github.com/users/volkamerlab/starred{/owner}{/repo}',\n",
       "   'subscriptions_url': 'https://api.github.com/users/volkamerlab/subscriptions',\n",
       "   'organizations_url': 'https://api.github.com/users/volkamerlab/orgs',\n",
       "   'repos_url': 'https://api.github.com/users/volkamerlab/repos',\n",
       "   'events_url': 'https://api.github.com/users/volkamerlab/events{/privacy}',\n",
       "   'received_events_url': 'https://api.github.com/users/volkamerlab/received_events',\n",
       "   'type': 'Organization',\n",
       "   'site_admin': False},\n",
       "  'html_url': 'https://github.com/volkamerlab/TeachOpenCADD',\n",
       "  'description': 'TeachOpenCADD: a teaching platform for computer-aided drug design (CADD) using open source packages and data',\n",
       "  'fork': False,\n",
       "  'url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD',\n",
       "  'forks_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/forks',\n",
       "  'keys_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/keys{/key_id}',\n",
       "  'collaborators_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/collaborators{/collaborator}',\n",
       "  'teams_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/teams',\n",
       "  'hooks_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/hooks',\n",
       "  'issue_events_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues/events{/number}',\n",
       "  'events_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/events',\n",
       "  'assignees_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/assignees{/user}',\n",
       "  'branches_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/branches{/branch}',\n",
       "  'tags_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/tags',\n",
       "  'blobs_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/blobs{/sha}',\n",
       "  'git_tags_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/tags{/sha}',\n",
       "  'git_refs_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/refs{/sha}',\n",
       "  'trees_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/trees{/sha}',\n",
       "  'statuses_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/statuses/{sha}',\n",
       "  'languages_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/languages',\n",
       "  'stargazers_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/stargazers',\n",
       "  'contributors_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/contributors',\n",
       "  'subscribers_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/subscribers',\n",
       "  'subscription_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/subscription',\n",
       "  'commits_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/commits{/sha}',\n",
       "  'git_commits_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/commits{/sha}',\n",
       "  'comments_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/comments{/number}',\n",
       "  'issue_comment_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues/comments{/number}',\n",
       "  'contents_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/contents/{+path}',\n",
       "  'compare_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/compare/{base}...{head}',\n",
       "  'merges_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/merges',\n",
       "  'archive_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/{archive_format}{/ref}',\n",
       "  'downloads_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/downloads',\n",
       "  'issues_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues{/number}',\n",
       "  'pulls_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/pulls{/number}',\n",
       "  'milestones_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/milestones{/number}',\n",
       "  'notifications_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/notifications{?since,all,participating}',\n",
       "  'labels_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/labels{/name}',\n",
       "  'releases_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/releases{/id}',\n",
       "  'deployments_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/deployments',\n",
       "  'created_at': '2018-11-09T13:15:15Z',\n",
       "  'updated_at': '2019-07-18T02:36:48Z',\n",
       "  'pushed_at': '2019-05-03T15:02:03Z',\n",
       "  'git_url': 'git://github.com/volkamerlab/TeachOpenCADD.git',\n",
       "  'ssh_url': 'git@github.com:volkamerlab/TeachOpenCADD.git',\n",
       "  'clone_url': 'https://github.com/volkamerlab/TeachOpenCADD.git',\n",
       "  'svn_url': 'https://github.com/volkamerlab/TeachOpenCADD',\n",
       "  'homepage': None,\n",
       "  'size': 28121,\n",
       "  'stargazers_count': 39,\n",
       "  'watchers_count': 39,\n",
       "  'language': 'Jupyter Notebook',\n",
       "  'has_issues': True,\n",
       "  'has_projects': True,\n",
       "  'has_downloads': True,\n",
       "  'has_wiki': True,\n",
       "  'has_pages': False,\n",
       "  'forks_count': 13,\n",
       "  'mirror_url': None,\n",
       "  'archived': False,\n",
       "  'disabled': False,\n",
       "  'open_issues_count': 0,\n",
       "  'license': {'key': 'cc-by-4.0',\n",
       "   'name': 'Creative Commons Attribution 4.0 International',\n",
       "   'spdx_id': 'CC-BY-4.0',\n",
       "   'url': 'https://api.github.com/licenses/cc-by-4.0',\n",
       "   'node_id': 'MDc6TGljZW5zZTI1'},\n",
       "  'forks': 13,\n",
       "  'open_issues': 0,\n",
       "  'watchers': 39,\n",
       "  'default_branch': 'master'}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "response = requests.get(\"https://api.github.com/users/volkamerlab/repos\")\n",
    "response.raise_for_status()\n",
    "result = response.json()\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you parameterize the URL with a `f-string`, then you can design a function that will list the repositories of any user:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 156864288,\n",
       "  'node_id': 'MDEwOlJlcG9zaXRvcnkxNTY4NjQyODg=',\n",
       "  'name': 'TeachOpenCADD',\n",
       "  'full_name': 'volkamerlab/TeachOpenCADD',\n",
       "  'private': False,\n",
       "  'owner': {'login': 'volkamerlab',\n",
       "   'id': 44878588,\n",
       "   'node_id': 'MDEyOk9yZ2FuaXphdGlvbjQ0ODc4NTg4',\n",
       "   'avatar_url': 'https://avatars2.githubusercontent.com/u/44878588?v=4',\n",
       "   'gravatar_id': '',\n",
       "   'url': 'https://api.github.com/users/volkamerlab',\n",
       "   'html_url': 'https://github.com/volkamerlab',\n",
       "   'followers_url': 'https://api.github.com/users/volkamerlab/followers',\n",
       "   'following_url': 'https://api.github.com/users/volkamerlab/following{/other_user}',\n",
       "   'gists_url': 'https://api.github.com/users/volkamerlab/gists{/gist_id}',\n",
       "   'starred_url': 'https://api.github.com/users/volkamerlab/starred{/owner}{/repo}',\n",
       "   'subscriptions_url': 'https://api.github.com/users/volkamerlab/subscriptions',\n",
       "   'organizations_url': 'https://api.github.com/users/volkamerlab/orgs',\n",
       "   'repos_url': 'https://api.github.com/users/volkamerlab/repos',\n",
       "   'events_url': 'https://api.github.com/users/volkamerlab/events{/privacy}',\n",
       "   'received_events_url': 'https://api.github.com/users/volkamerlab/received_events',\n",
       "   'type': 'Organization',\n",
       "   'site_admin': False},\n",
       "  'html_url': 'https://github.com/volkamerlab/TeachOpenCADD',\n",
       "  'description': 'TeachOpenCADD: a teaching platform for computer-aided drug design (CADD) using open source packages and data',\n",
       "  'fork': False,\n",
       "  'url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD',\n",
       "  'forks_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/forks',\n",
       "  'keys_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/keys{/key_id}',\n",
       "  'collaborators_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/collaborators{/collaborator}',\n",
       "  'teams_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/teams',\n",
       "  'hooks_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/hooks',\n",
       "  'issue_events_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues/events{/number}',\n",
       "  'events_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/events',\n",
       "  'assignees_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/assignees{/user}',\n",
       "  'branches_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/branches{/branch}',\n",
       "  'tags_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/tags',\n",
       "  'blobs_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/blobs{/sha}',\n",
       "  'git_tags_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/tags{/sha}',\n",
       "  'git_refs_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/refs{/sha}',\n",
       "  'trees_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/trees{/sha}',\n",
       "  'statuses_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/statuses/{sha}',\n",
       "  'languages_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/languages',\n",
       "  'stargazers_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/stargazers',\n",
       "  'contributors_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/contributors',\n",
       "  'subscribers_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/subscribers',\n",
       "  'subscription_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/subscription',\n",
       "  'commits_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/commits{/sha}',\n",
       "  'git_commits_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/git/commits{/sha}',\n",
       "  'comments_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/comments{/number}',\n",
       "  'issue_comment_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues/comments{/number}',\n",
       "  'contents_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/contents/{+path}',\n",
       "  'compare_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/compare/{base}...{head}',\n",
       "  'merges_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/merges',\n",
       "  'archive_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/{archive_format}{/ref}',\n",
       "  'downloads_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/downloads',\n",
       "  'issues_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/issues{/number}',\n",
       "  'pulls_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/pulls{/number}',\n",
       "  'milestones_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/milestones{/number}',\n",
       "  'notifications_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/notifications{?since,all,participating}',\n",
       "  'labels_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/labels{/name}',\n",
       "  'releases_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/releases{/id}',\n",
       "  'deployments_url': 'https://api.github.com/repos/volkamerlab/TeachOpenCADD/deployments',\n",
       "  'created_at': '2018-11-09T13:15:15Z',\n",
       "  'updated_at': '2019-07-18T02:36:48Z',\n",
       "  'pushed_at': '2019-05-03T15:02:03Z',\n",
       "  'git_url': 'git://github.com/volkamerlab/TeachOpenCADD.git',\n",
       "  'ssh_url': 'git@github.com:volkamerlab/TeachOpenCADD.git',\n",
       "  'clone_url': 'https://github.com/volkamerlab/TeachOpenCADD.git',\n",
       "  'svn_url': 'https://github.com/volkamerlab/TeachOpenCADD',\n",
       "  'homepage': None,\n",
       "  'size': 28121,\n",
       "  'stargazers_count': 39,\n",
       "  'watchers_count': 39,\n",
       "  'language': 'Jupyter Notebook',\n",
       "  'has_issues': True,\n",
       "  'has_projects': True,\n",
       "  'has_downloads': True,\n",
       "  'has_wiki': True,\n",
       "  'has_pages': False,\n",
       "  'forks_count': 13,\n",
       "  'mirror_url': None,\n",
       "  'archived': False,\n",
       "  'disabled': False,\n",
       "  'open_issues_count': 0,\n",
       "  'license': {'key': 'cc-by-4.0',\n",
       "   'name': 'Creative Commons Attribution 4.0 International',\n",
       "   'spdx_id': 'CC-BY-4.0',\n",
       "   'url': 'https://api.github.com/licenses/cc-by-4.0',\n",
       "   'node_id': 'MDc6TGljZW5zZTI1'},\n",
       "  'forks': 13,\n",
       "  'open_issues': 0,\n",
       "  'watchers': 39,\n",
       "  'default_branch': 'master'}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def repos(username):\n",
    "    response = requests.get(f\"https://api.github.com/users/{username}/repos\")\n",
    "    response.raise_for_status()\n",
    "    return response.json()\n",
    "\n",
    "repos(\"volkamerlab\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generating a client for any API\n",
    "\n",
    "Did you find that convenient? Well, we are not done yet!\n",
    "\n",
    "The REST API schema can be expressed programmatically in a document called \"Swagger/OpenAPI definitions\", which allows to dynamically generate a Python client for any REST API that implements the Swagger/OpenAPI schema. [This the one for GitHub](https://api.apis.guru/v2/specs/github.com/v3/swagger.json).\n",
    "\n",
    "Of course, there are libraries for doing that in Python:\n",
    "\n",
    "- `bravado`\n",
    "- `agithub`\n",
    "\n",
    "Using Bravado:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this if you get an ImportError below\n",
    "!pip install bravado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SwaggerClient(https://api.github.com/)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bravado.client import SwaggerClient\n",
    "GITHUB_SWAGGER = \"https://api.apis.guru/v2/specs/github.com/v3/swagger.json\"\n",
    "client = SwaggerClient.from_url(GITHUB_SWAGGER)\n",
    "client"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, you can have fun inspecting the `client` object for all the API actions as methods. We will see an actual example with [KLIFS](http://klifs.vu-compmedchem.nl) in the case study.\n",
    "\n",
    "Other standards like SOAP operate on a very similar principle: all you need to provide is a definition file to be processed by a client-generating library. For SOAP, the definition files are formatted as `.wsdl`. One of the most popular libraries is `zeep`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!pip install suds-py3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Suds ( https://github.com/cackharot/suds-py3 )  version: 1.3.3.0 IN  build: 20170311\n",
      "\n",
      "Service ( SwissdockInterfaceWrapper ) tns=\"http://swissdock.vital-it.ch/soap/\"\n",
      "   Prefixes (1)\n",
      "      ns0 = \"http://swissdock.vital-it.ch/soap/\"\n",
      "   Ports (1):\n",
      "      (SwissdockInterfaceWrapper)\n",
      "         Methods (21):\n",
      "            forget(xs:string jobID, )\n",
      "            getNewDockingPool()\n",
      "            getPredictedDocking(xs:string jobID, )\n",
      "            getPredictedDockingAllFiles(xs:string jobID, )\n",
      "            getPredictedDockingFile(xs:string jobID, xs:string filename, )\n",
      "            getPredictedDockingPool(xs:string dockingPoolID, xs:boolean mergedock4, )\n",
      "            getPredictedDockingPoolFile(xs:string dockingPoolID, xs:string filename, )\n",
      "            getPreparedLigand(xs:string jobID, )\n",
      "            getPreparedTarget(xs:string jobID, )\n",
      "            isDockingKnown(xs:string jobID, )\n",
      "            isDockingPoolKnown(xs:string jobID, )\n",
      "            isDockingPoolRunning(xs:string currentPoolID, )\n",
      "            isDockingPoolTerminated(xs:string currentPoolID, )\n",
      "            isDockingRunning(xs:string jobID, )\n",
      "            isDockingTerminated(xs:string jobID, )\n",
      "            isLigandPrepared(xs:string jobID, )\n",
      "            isTargetPrepared(xs:string jobID, )\n",
      "            prepareLigand(xs:string ligand, )\n",
      "            prepareTarget(xs:string target, )\n",
      "            startDocking(xs:string targetPsf, xs:string targetCrd, xs:string ligandPdb, stringArray ligandRtf, stringArray ligandPar, xs:string jobName, xs:string emailAddress, xs:string password, xs:float xmin, xs:float xmax, xs:float ymin, xs:float ymax, xs:float zmin, xs:float zmax, xs:int wantedConfs, xs:int nbFactsEval, xs:int nbSeeds, xs:int sdSteps, xs:int abnrSteps, xs:float clusteringRadius, xs:int maxClusterSize, xs:float passiveFlexibilityDistance, xs:boolean ignorePocketBias, xs:boolean keepLigandInPlace, xs:string dockingPoolID, )\n",
      "            startDockingPool(xs:string dockingPoolID, )\n",
      "         Types (43):\n",
      "            forget\n",
      "            forgetResponse\n",
      "            getNewDockingPool\n",
      "            getNewDockingPoolResponse\n",
      "            getPredictedDocking\n",
      "            getPredictedDockingAllFiles\n",
      "            getPredictedDockingAllFilesResponse\n",
      "            getPredictedDockingFile\n",
      "            getPredictedDockingFileResponse\n",
      "            getPredictedDockingPool\n",
      "            getPredictedDockingPoolFile\n",
      "            getPredictedDockingPoolFileResponse\n",
      "            getPredictedDockingPoolResponse\n",
      "            getPredictedDockingResponse\n",
      "            getPreparedLigand\n",
      "            getPreparedLigandResponse\n",
      "            getPreparedTarget\n",
      "            getPreparedTargetResponse\n",
      "            isDockingKnown\n",
      "            isDockingKnownResponse\n",
      "            isDockingPoolKnown\n",
      "            isDockingPoolKnownResponse\n",
      "            isDockingPoolRunning\n",
      "            isDockingPoolRunningResponse\n",
      "            isDockingPoolTerminated\n",
      "            isDockingPoolTerminatedResponse\n",
      "            isDockingRunning\n",
      "            isDockingRunningResponse\n",
      "            isDockingTerminated\n",
      "            isDockingTerminatedResponse\n",
      "            isLigandPrepared\n",
      "            isLigandPreparedResponse\n",
      "            isTargetPrepared\n",
      "            isTargetPreparedResponse\n",
      "            prepareLigand\n",
      "            prepareLigandResponse\n",
      "            prepareTarget\n",
      "            prepareTargetResponse\n",
      "            startDocking\n",
      "            startDockingPool\n",
      "            startDockingPoolResponse\n",
      "            startDockingResponse\n",
      "            stringArray\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "import suds\n",
    "wsdl = 'http://www.swissdock.ch/soap/wsdl'\n",
    "client = suds.Client(wsdl)\n",
    "print(client)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Document parsing\n",
    "\n",
    "Sometimes you won't be as lucky and the webservice won't provide a standardized API that produces machine-readable documents. Instead, you will have to use the regular webpage and parse through the HTML code to obtain the information you need. This is called (web) __scraping__, which usually involves finding the right HTML tags and IDs that contain the valuable data (ignoring things such as the sidebars, top menus, footers, ads, etc).\n",
    "\n",
    "In scraping, you basically do two things:\n",
    "\n",
    "1. Access the webpage with `requests` and obtain the HTML contents.\n",
    "2. Parse the HTML string with `BeautifulSoup` or `requests-html`.\n",
    "\n",
    "\n",
    "Let's parse the proteinogenic amino acids table in this [Wikipedia article](https://en.wikipedia.org/wiki/Proteinogenic_amino_acid):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['A', 'Ala', 89.09404, 6.01, 2.35, 9.87],\n",
       " ['C', 'Cys', 121.15404, 5.05, 1.92, 10.7],\n",
       " ['D', 'Asp', 133.10384, 2.85, 1.99, 9.9],\n",
       " ['E', 'Glu', 147.13074, 3.15, 2.1, 9.47],\n",
       " ['F', 'Phe', 165.19184, 5.49, 2.2, 9.31],\n",
       " ['G', 'Gly', 75.06714, 6.06, 2.35, 9.78],\n",
       " ['H', 'His', 155.15634, 7.6, 1.8, 9.33],\n",
       " ['I', 'Ile', 131.17464, 6.05, 2.32, 9.76],\n",
       " ['K', 'Lys', 146.18934, 9.6, 2.16, 9.06],\n",
       " ['L', 'Leu', 131.17464, 6.01, 2.33, 9.74],\n",
       " ['M', 'Met', 149.20784, 5.74, 2.13, 9.28],\n",
       " ['N', 'Asn', 132.11904, 5.41, 2.14, 8.72],\n",
       " ['O', 'Pyl', 255.31, '?', '?', '?'],\n",
       " ['P', 'Pro', 115.13194, 6.3, 1.95, 10.64],\n",
       " ['Q', 'Gln', 146.14594, 5.65, 2.17, 9.13],\n",
       " ['R', 'Arg', 174.20274, 10.76, 1.82, 8.99],\n",
       " ['S', 'Ser', 105.09344, 5.68, 2.19, 9.21],\n",
       " ['T', 'Thr', 119.12034, 5.6, 2.09, 9.1],\n",
       " ['U', 'Sec', 168.053, 5.47, 1.91, 10.0],\n",
       " ['V', 'Val', 117.14784, 6.0, 2.39, 9.74],\n",
       " ['W', 'Trp', 204.22844, 5.89, 2.46, 9.41],\n",
       " ['Y', 'Tyr', 181.19124, 5.64, 2.2, 9.21]]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "r = requests.get(\"https://en.wikipedia.org/wiki/Proteinogenic_amino_acid\")\n",
    "r.raise_for_status()\n",
    "\n",
    "# To guess the correct steps here, you will have to inspect the HTML code by hand\n",
    "# Tip: use right-click + inspect content in any webpage to land in the HTML definition ;)\n",
    "html = BeautifulSoup(r.text)\n",
    "header = html.find('span', id=\"General_chemical_properties\")\n",
    "table = header.find_all_next()[4]\n",
    "table_body = table.find('tbody')\n",
    "\n",
    "data = []\n",
    "for row in table_body.find_all('tr'):\n",
    "    cells = row.find_all('td')\n",
    "    if cells:\n",
    "        data.append([])\n",
    "    for cell in cells:\n",
    "        cell_content = cell.text.strip()\n",
    "        try:  # convert to float if possible\n",
    "            cell_content = float(cell_content)\n",
    "        except ValueError:\n",
    "            pass\n",
    "        data[-1].append(cell_content)\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Browser remote control\n",
    "\n",
    "The trend some years ago was to build servers that dynamically-generated HTML document with some JavaScript here and there (such as Wikipedia). In other words, the HTML is built in the server and sent to the client.\n",
    "\n",
    "However, latest trends are pointing towards full applications built entirely with JavaScript frameworks. This means that the HTML content is dynamically generated in the client, so traditional parsing won't work: you will only download placeholder HTML to host the JavaScript framework. To work around this, the HTML must be rendered with a client-side JavaScript engine.\n",
    "\n",
    "We won't cover this in this notebook, but you can check these projects if you are interested:\n",
    "\n",
    "- [puppeteer](https://github.com/GoogleChrome/puppeteer)\n",
    "- [selenium](https://www.seleniumhq.org/)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description of the pipeline and the involved webservices\n",
    "\n",
    "In this talktorial, we will obtain input structures from from _KLIFS_, _PubChem_ and _ChEMBL_, and then perform several quality checks with the _proteins.plus_ web services. Finally, if everything is working correctly, we will perform docking using _SwissDock_.\n",
    "\n",
    "### KLIFS\n",
    "\n",
    "* Role: Kinase-ligand interaction profiles database\n",
    "* Website: http://klifs.vu-compmedchem.nl/\n",
    "* API: Yes, REST-based, Swagger-enabled. No official client. Use `bravado`.\n",
    "* Documentation: http://klifs.vu-compmedchem.nl/swagger/\n",
    "* Literature:\n",
    "    * Nucleic Acids Research, 2016, 44, D1, 6,  D365–D371, https://doi.org/10.1093/nar/gkv1082\n",
    "    * J. Med. Chem. 2014, 57, 2, 249-277. https://pubs.acs.org/doi/abs/10.1021/jm400378w\n",
    "    \n",
    "\n",
    "> Kinase-Ligand Interaction Fingerprints and Structures database (KLIFS), developed at the Division of Medicinal Chemistry - VU University Amsterdam, is a database that revolves around the protein structure of catalytic kinase domains and the way kinase inhibitors can interact with them. Based on the underlying systematic and consistent protocol all (currently human and mouse) kinase structures and the binding mode of kinase ligands can be directly compared to each other. Moreover, because of the classification of an all-encompassing binding site of 85 residues it is possible to compare the interaction patterns of kinase-inhibitors to each other to, for example, identify crucial interactions determining kinase-inhibitor selectivity.\n",
    "\n",
    "### PubChem\n",
    "\n",
    "* Role: Small compounds database\n",
    "* Website: https://pubchem.ncbi.nlm.nih.gov/\n",
    "* API: Yes, REST-based. No official client. Use `bravado`.\n",
    "* Documentation: https://pubchemdocs.ncbi.nlm.nih.gov/pug-rest\n",
    "* Literature:\n",
    "    * Nucleic Acids Res. 2019 Jan 8; 47(D1):D1102-1109. https://academic.oup.com/nar/article/47/D1/D1102/5146201\n",
    "\n",
    "> PubChem is an open chemistry database at the National Institutes of Health (NIH). “Open” means that you can put your scientific data in PubChem and that others may use it. Since the launch in 2004, PubChem has become a key chemical information resource for scientists, students, and the general public. Each month our website and programmatic services provide data to several million users worldwide.\n",
    "\n",
    "### ChEMBL\n",
    "\n",
    "* Role: Small compounds database\n",
    "* Website: https://www.ebi.ac.uk/chembl/\n",
    "* API: Yes, REST-based. [Official client available](https://github.com/chembl/chembl_webresource_client).\n",
    "* Documentation: https://www.ebi.ac.uk/chembl/api/data/docs\n",
    "* Literature:\n",
    "    * Nucleic Acids Res. 2012; 40: D1100–D1107. https://academic.oup.com/nar/article/40/D1/D1100/2903401\n",
    "\n",
    "> ChEMBL is a manually curated database of bioactive molecules with drug-like properties. It brings together chemical, bioactivity and genomic data to aid the translation of genomic information into effective new drugs.\n",
    "\n",
    "### SwissDock\n",
    "\n",
    "* Role: Perform docking calculations\n",
    "* Website: http://www.swissdock.ch\n",
    "* API: Yes, SOAP-based. No official client, use `suds`.\n",
    "* Documentation: http://www.swissdock.ch/pages/soap_access\n",
    "* Literature:\n",
    "    * Nucleic Acids Res. 2011 Jul;39(Web Server issue):W270-7. doi: 10.1093/nar/gkr366. https://academic.oup.com/nar/article/39/suppl_2/W270/2506492\n",
    "    * J Comput Chem. 2011 Jul 30;32(10):2149-59. doi: 10.1002/jcc.21797. https://onlinelibrary.wiley.com/doi/abs/10.1002/jcc.21797\n",
    "\n",
    "> SwissDock, a web service to predict the molecular interactions that may occur between a target protein and a small molecule.\n",
    "> SwissDock is based on the docking software EADock DSS, whose algorithm consists of the following steps:\n",
    "> 1. Many binding modes are generated either in a box (local docking) or in the vicinity of all target cavities (blind docking).\n",
    "> 2. Simultaneously, their CHARMM energies are estimated on a grid.\n",
    "> 3. The binding modes with the most favorable energies are evaluated with FACTS, and clustered.\n",
    "> 4. The most favorable clusters can be visualized online and downloaded on your computer.\n",
    "\n",
    "## Proposed pipeline\n",
    "\n",
    "1. KLIFS will provide protein-ligand pairs that are known to interact with each other.  We can obtain the curated protein structure from this database and use the ligand information to retrieve similar ligands from other databases, like PubChem or ChEMBL.\n",
    "2. Using the ligand information provided by KLIFS, we can query PubChem and ChEMBL for similar compounds.\n",
    "3. After having obtained the protein structure(s) and several candidate ligands, we can dock them online with SwissDock.\n",
    "4. The results can be validated and/or visualized using _proteins.plus_ webservices."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "# Practical\n",
    "\n",
    "## Building the pipeline\n",
    "\n",
    "### Obtaining information from KLIFS\n",
    "\n",
    "Choose a random kinase (in `mol2`) and ligand (`SMILES`) from the KLIFS database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bravado.client import SwaggerClient\n",
    "KLIFS_API_DEFINITIONS = \"http://klifs.vu-compmedchem.nl/swagger/swagger.json\"\n",
    "KLIFS_CLIENT = SwaggerClient.from_url(KLIFS_API_DEFINITIONS, config={'validate_responses': False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _all_kinase_families():\n",
    "    return KLIFS_CLIENT.Information.get_kinase_families().response().result\n",
    "\n",
    "def _kinases_from_family(family, species=\"HUMAN\"):\n",
    "    return KLIFS_CLIENT.Information.get_kinase_names(kinase_family=family, species=species).response().result\n",
    "\n",
    "def _protein_and_ligand_structure(*kinase_ids):\n",
    "    structures = KLIFS_CLIENT.Structures.get_structures_list(kinase_ID=kinase_ids).response().result\n",
    "    protein = KLIFS_CLIENT.Structures.get_structure_get_protein(structure_ID=structures[0].structure_ID).response().result\n",
    "    ligands = KLIFS_CLIENT.Ligands.get_ligands_list(kinase_ID=kinase_ids).response().result\n",
    "    return protein, [ligand.SMILES for ligand in ligands]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Use nglview here\n",
    "import io\n",
    "def show_protein(mol2_text):\n",
    "    f = io.StringIO()\n",
    "    f.write(mol2_text)\n",
    "    f.seek(0)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen MAPK12 kinase with ID 252 from family MAPK\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import time\n",
    "def step_01():\n",
    "    # Get a random kinase from a random family\n",
    "    attempts = 20\n",
    "    families = _all_kinase_families()\n",
    "    while attempts:  # some kinase IDs cannot be found in the structure list...\n",
    "        family = random.choice(families)\n",
    "        kinase = random.choice(_kinases_from_family(family))\n",
    "        try:\n",
    "            protein, ligands = _protein_and_ligand_structure(kinase.kinase_ID)\n",
    "        except:\n",
    "            attempts -= 1\n",
    "            time.sleep(1)\n",
    "        else:                   \n",
    "            print(\"Chosen\", kinase.name, \"kinase with ID\", kinase.kinase_ID, \"from family\", family)\n",
    "            return protein, ligands\n",
    "    print(\"Could not find a valid kinase. Try again!\")\n",
    "    return None, None\n",
    "protein, ligands = step_01()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['P(=O)(OP(=O)(O)NP(=O)(O)O)(OC[C@H]1O[C@@H](N2c3ncnc(N)c3N=C2)[C@H](O)[C@@H]1O)O']"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ligands"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query ChEMBL and PubChem\n",
    "\n",
    "Use the previously obtained `SMILES` string to query similar compounds in PubChem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "def similar_compounds_pubchem(smiles, threshold=75, n_records=10):\n",
    "    # Query PubChem for similar compounds\n",
    "    url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/similarity/smiles/{smiles}/JSON?Threshold={threshold}&MaxRecords={n_records}\"\n",
    "    r = requests.get(url)\n",
    "    r.raise_for_status()\n",
    "    key = r.json()['Waiting']['ListKey']\n",
    "    # Response is async, poll every second for 25 seconds to see if it's done\n",
    "    attempts = 25\n",
    "    while attempts:\n",
    "        r = requests.get(f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/listkey/{key}/cids/JSON\")\n",
    "        r.raise_for_status()\n",
    "        response = r.json()\n",
    "        if 'IdentifierList' in response:\n",
    "            cids = response['IdentifierList']['CID']\n",
    "            break\n",
    "        attempts -= 1\n",
    "        time.sleep(1)\n",
    "    else:\n",
    "        raise IOError(\"Could not find matches for \" + smiles)\n",
    "    # We get the Compound IDs back, but we need the SMILES\n",
    "    url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/{','.join(map(str, cids))}/property/CanonicalSMILES/JSON\"\n",
    "    r = requests.get(url)\n",
    "    r.raise_for_status()\n",
    "    return [item['CanonicalSMILES'] for item in r.json()['PropertyTable']['Properties']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit import Chem\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit.Chem.Draw import IPythonConsole #Needed to show molecules\n",
    "\n",
    "def step_02(ligands):\n",
    "    # Get a random ligands for the current kinase\n",
    "    smiles = random.choice(ligands)\n",
    "    # Look for 10 most similar compounds on pubchem\n",
    "    similar_smiles = similar_compounds_pubchem(smiles, n_records=10)\n",
    "    molecules = []\n",
    "    # Generate 3D conformers with RDKit\n",
    "    for compound in similar_smiles:\n",
    "        m = Chem.AddHs(Chem.MolFromSmiles(compound))\n",
    "        AllChem.EmbedMolecule(m)\n",
    "        molecules.append(m)\n",
    "    return similar_smiles, molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "similar_smiles, molecules = step_02(ligands)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C1=NC(=C2C(=N1)N(C=N2)C3C(C(C(O3)CO)O)O)N\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAbTUlEQVR4nO3deVhU9f4H8PdswMAAgwPKZgYOKIsLIplLSWaldctd6rpDUW6B/kwINEhzwa5d9QEVS8s1l6xumlrZTa8phTsOKDEKKC7JPgzCMNvvjyMICDLIMGcGP6/Hx2dmPJzvB6y333O+y+Ho9XoQQgh5XFy2CyCEEMtGMUoIIW1CMWpepFLpkSNH6t5GR0fPmTOHxXoIIS2iGCWEkDahGCWEkDahGCWEkDbhs10AaSwsLEwgEDCvKysrIyIi2K2HEPJoFKNmJyUlJTQ0lHmdmJjIZimEEANQjJodZ2dnT09P5rVIJNJoNOzWQwh5NLo3SgghbUIxSgghbcKhNfVmpaioqF+/fqGhodu2bWO7FkKIQag3al7S09Nv3LiRn5/PdiGEEENRjJqX9PR0ACEhIWwXQggxFMXofWaymP306dOgGCXEolCMmpczZ84AeOaZZ9guhBBiKIpRM5Kbm3v37l0XFxcvLy+2ayGEGIpi1IzQjVFCLBGtYnqA9cXsdGOUEEtEMfoA64vZmRilG6OEWBaK0QfYXcyu1WrPnTsHoH///qZslxDSRnRv1FxkZWUplUovL6/OnTuzXQshpBUoRptVUlJSWVlpsuaY8SW6oifE4lCM3ieXy0eMGFH31sXFZc+ePevWrTNZATS+RIiFohht2rPPPqvT6VatWlVaWmqaFqk3SoiFohht2osvvjhs2LCysrI1a9aYoLnq6mqZTMbj8YKCgkzQHCHEiChGm/XJJ5+I7OyEf/6JoqL2buvcuXNqtTogIEAkErV3W4QQ46IYbdbAgQMLXn019qefkJTU3m3RFT0hloti9FEcFy0Cl4vkZNy82X6trF+//uDBg6DxJUIsE8XoI/XujTFjUF3dfh3Sffv2zZ49+9ixY6AYJcQy0SqmlixZgqoqTJtmxFPW1NTk5ORkZWVdu3ZNJpM5ODgoFAoul+vo6GjEVgghpkHPYmp3SqXy8uXLWVlZzO9ZWVm5ubk6na7+MRwOR6/Xu7i47N27NzQ0NDo6WqPRJCcns1UzIcRw1Bs1zPr1WLsWeXlwc0N4OOLjweM1fWRZGS5fRmYmrlyBTKbPy3PKydE0DE2BQNCjRw9/f38/P7+AgICePXuOGTPGzc0tLS1t+PDhy5YtM8V3RAgxEopRAyQlYc0abNmCIUNw6RJmzMDt29iwAQBKS3HtGjIzkZV1//fcXNTr4HOA5729i0QiJjH9/Pz8/f19fHzqduRj8Hi8xYsXZ2RkxMXFxcbG9ujRo26vKUKImaOL+pYolXB1xY4dGD36/ifnz6N/f8jleOst/Pln4+NtbdGzJ/z94e8PPz8EBMDbu9muay2pVJqcnDxixIgDBw5MmTKlvLzc2dk5PT2dtsEnxPxRb7Qlp09DrcYbbzz4JCgIXl44fhydO8PaGt27IyAA/v73f+/Zs8XQbFLdptEcDofD4RQVFYWEhHz99dcvvfSSsb4VQkh7oBhtSXExXFzAbTgzzNUVxcXYuRP29sZqp/6m0fHx8b///vu1a9dGjhy5bNmyhQsXcjgcYzVECDEumjfaEokEhYVoOEaEO3cgkRgxQ1G7aTTDyclp5MiRK1eu1Ov1sbGxb731VnNb9pnJc6EJeZJRjLYkJAQCAX744cEn588jNxfPP9/eLcfExPzwww9isXjPnj2DBg26evVqe7dICHkMFKMtEYmweDFmzsThw6iowKlTePNNREbC29sEjb/22mvp6ekBAQEZGRn9+vX7z3/+Y4JGCSGtQvdGDRATA5EIUVH3541GRCA+3rgtyOXy+m/r787n4+OTlpY2Y8aM/fv3jxkzZuHChcuXL+dy6d8/QswFTXiyDHq9ftWqVXFxcTqd7vXXX9++fTuzclQqlRYWFjZ6LrQxlz8Zvu6AkCcVdWosA4fDiYmJOXjwoJOT04EDBwYMGHD58mXmj1JSUi7UmjRpkjFbTUrC0qVYswZFRdi1Czt3gsavCHkIxaglGTlyZHp6emBgYHZ29oABA77//ns0HOI35q7PSiWWLsWGDRg5Evb2GDQIu3dj0ybk5hqtCUI6BIpRCyOVStPS0saPH19RUTF27Nji4mKj3Zb5+28cPYo1a/D223jmGezc2ey6A0JIPTTEZHlEItHevXtXrVoVHx9fVlaWmJg4cOBAsVis1+s1Go2hZ6mpQU4Ozp69vxvA2bO4fbvBAcHBza47IITUQ0NMFuynn3765z//WVJS4uPjEx4enpSUFB4evnr16oePVKvV2dnZmZmZGRkZzykUIw4darSFCgCIxQgMRGAgevdGYCAUCowdi6qqBkkqlWLRIkyf3r7fGCEWhWLUsl2/fn3s2LFnz54VCoVVVVVDhw5lNtK/detWVlZWZmbm2bNnmU1Oq6qqmC+Z3bt3ckYGBAL4+NzfByA4GAEB8PJC/SWnze3JkpNjmjmzhFgKilGLx0xy2rNnDwAej/fMM89kZmYqFIr6x3C5XG9v7169egUGBj7bp8+r/v7w8QG/pVs6D+8QOGzY/R0CCSG1KEY7iE2bNs2aNcvKyorpdTo5Ofn7+wcEBPj7+wcHB/ft2/cxB/FTUh7MG42IwMsv49AhLFli5OoJsWQUox1EYWFh586dhULhwYMH+/TpI5FIjN+GUglPT5SX4/hxE2wpQIiloAlPHcSFCxcA9OvXb9iwYe2SoQBEIsyfDwAffNB4eIqQJxjFaAdx8eJFAH379m3fZubPR5cuSE/HwYPt2xAhloNitINgYrRPnz7t24xIhLg4AIiLa7wHKyFPKorRDoK5qG/33iiA996DtzdkMuzc2e5tEWIJKEY7ApVKlZ2dzePxAgIC2r0xKyskJADARx9BpWr35kgr0QMRTI9itCOQyWRqtdrX19fW1tYU7U2ejD59kJeHzz83RXOEmDeK0Y7AdFf0DC4XH38MAEuWoKLCRI0SYq4oRjsCE40v1TdqFAYNQmEh1q41XaOEmCWK0Y7ARLOdGlm5EgD+9S/a88nchIWFOddKTU1lu5yOj2LU4un1+oyMDJi4NwrguecwYsQ9d/dvN240abukJe34QATSFIpRi5eXl1dWVta5c2dXV1cTN52/YoXjX39NXrasoKDAxE2TR2ivByKQZtC2zRaPGV8KCgoyfdPd+vYdP2HC7t27lyxZsmnTJtMXYIaqq6u3bdvGvGC2iamsrKypqQGgUCi0Wi2A0tJSAFqtltmIq6amprKyEsC9e/dUKtWnn346duxYFr8F0loUoxaPnRujtZYvX/7tt99u2bJl3rx5fn5+rNRQ5/bt20xyqVSqe/fuMR+q1WqlUikQ+KnVHlot6u8gWFoKAF5eeOklo9VQXV397rvvtuUMJSUlxiqGmAbFqMVjYZi+Hi8vr/Dw8I0bNyYkJOzdu5eVGupMnz79559/bvKPQkNPHDvm0eQfjR5tzBgVCoWRkZEArK2tmWm8tra21tbWAOzt7fl8PgCxWMzhcHg8noODAwArKys7Ozvma21sbJydndtSgFwur3ut1erj4j7h8+mZ2O2LNsqzeF5eXnl5eZmZmf7+/qwUcPv2balUWlVVlZaWNmDAAFZqYERERDCb/9dFGAA+n29vb+/tnXjt2mAuF46OD453cgKAXr1YeG60VCpNTk4eMWIE8zY6Olqj0SQnJ5u6DmIM1Bu1bGVlZfn5+UKh0NfXl60a3Nzc5s6dm5SUtHjx4uY6g6axefPmtp6iuBh5eQ/eurnB3b2t5zQeqRQqFeRyWFsDwJw5EInuTzwzyvHk8dBIvWW7ePGiXq8PDAzkt/hEkPYUGxvbqVOnX3755ddff2WxjLa6ezff7Vn071/3S9m9z2fvXGa7rAYqK9Gqfyxaezx5DNQbtWzsji/VEYvFCxYsiIuLi42NTU9P59R/NJ4xmKhXlZXVTS1Xcuyv8ny5XATa54uKi7oVnJwzx08kQkwMANQNUqnVUCoBQKUCM5pVXf1rVVUJAKVSqVarAZSXl+t0OtQOzet0usLCwosXL4aEhBQWFq5bt+67775jWj516pRer4+Pv1lU5AGgpgaVlfeLUqvLqqs5traOe/ZodDrue++pli2zHjeu3MoKKpWIzxcwA2UPc3AAgHnzsGIF3n4bVlbG/nGRWhSjlo3d8aX6oqOj169ff+bMmW+//XbcuHGt+lpDbhQyvapZsx6zPGbsXq/Xl5WVoXYSUmlp6Z9//lld3a2wcIC3d8WLvLPBQEEX/+DCNKGV9mfXtwYWf1NcXHy84IZMpk1Kktc7HxcoBSprf3UC7J2dPy4qynx0GQ4ODgqF4vz58wAOHz7c6E/l8sEqlR5gtimoBpiHufIALQCBAIB8xYq3gcWurt8AG1xdL9y506e55bgXLgDAc8/h2DFs3oyZM1v7MyOGohi1bMykUXOIUaFQGBcXN2vWrJiYGH9//7r5RnVKG/aa+PynSkp8tVoAUCpx4gQqKlBRAY0GMhm0WjDzUN3d8Y9/AE31qv73P/3KlQVcrl6t1lZW6gEoFNBqK6qrNSoVdDp07nzz6tVJ9+7d0+l0QHegX23j1oAtAGdn56KicuBpwAc4FYoLvwF37mRrUa1U/pqZKRgInD59WgYRoAQOAV8CToA9wAc8gZvM6WxsulVX5xcVtfxTavTE1kZUqvyHPxQI+Gq1FgCfz9fpuHZ29sBnlZWpjo7f8vlcGxsIhU2fjVc7Pp+QgMmTERHRcnnk8dBIvQXTaDT29vYqlaqsrIyZOsMutVrt4eFRU1NTXl7e4sFDh847fvyz2ndSIBkYUfs2GtAAycB8D4+CgIDyEye+7Nr141u3ImxsDvL5n5eWLq2pKdHrYwAAXkBui81xubN0upRHHxOKY7/hhSKB3X904/R67lDOcR9tbqwkcpftazxelbPzH2fO/Lv2bGo7u4EaTRafz+fz+f369evUqRMAkUgkEAgAODo6crlcDocjFotRO1sAgI2NjVAoBBAVFTV37tyQkBAHBwcej7d69Wo+n5+YmFg3C6r+ZIM6Uim++AKhoRg2DBMnQiZreYipyeO3b4dEgldfbfHHRgxCvVELdvny5erq6u7du5tDhgI4efJkcXGxQCDw9vZ2YiYT1dPoE2/v7p06gRkYO3wYQUFwdYVIBIEAp05Br8fgwfjmm6M3b166eRNA1V9//QV8oFTuAJIAFaAHwOfburvbVVc/zeNxAXC5sLOzFwj4HA44HIhEQisrHp/P53A4Pj7D6nqLVlaws7v/WiBA3WpJF0VX5UZ7Z3VFBLYxn2jA7/bSpDckz4tESEgIu30bjo6wt4eVlQA405af1YIFC+7cuXPt2jWdTldeXn79+nWdTnf4sI1S2XhiwP/9Hx4ePkxIwNSpeOUVQ5urf7xSifnzUVSEPn0QH4/x42HsW9lPHIpRC2Ym40uMsrKy6dOn63S6uLi4jz76qFVfK5Xi4sWwrCwB87aysjIiIiI5GWPHfqpQKBwcHMLD3RMSUp99tvrdd51Hj74rl9s4OQlWrYppe9kVFRV5eTfu3vUPDESXLt1fP/C/1WHp27bhhRdw9ChyuD0qSp73kQCAUAhv77Y3eF9VVdW6desafajX448/Gh/5/vtNxOjQofD2xv79eOcdg5qrfzyPh7g4rFqFixcxcSKCg/HRR3j9dQrTx0cxasHMZ3wJwKxZs/Lz8319fT/88EPmk1ZNKU9JSQkNDWVeJyYmMi9eqe1uCYXw9fXt3RsrV2LqVOErr4BrpKl6P/7445QpM7Taqn37MG4cLlv3vfWPvru/x/CZ+OMuzp0D/kC3bsZpq77vvvuOWXrP5XIda5cEiMXaoUMbH8lrZglSQgJeeKEVLdYdLxRi3jy89x42bUJSEs6exahRGDs2bcqUv0eNGmX0WRZPAopRC2Y+40s7duz4+uuvuVzuwoULmZuDrcVsSsS8FolEGo2mycNa2wtrkaenp5+fj6enxsmp8f8LTO489ZQxm6szePDgwYMHt/ar6q3zRGgoWhzXeMTxQiGiojBzJr76CkuXQi6PGTPmRGBg4AcffDB58mSusf6ZejLQD8uCMduMsn5RX1BQ8P777wNwdnb28Gh63boRJSTAiHt3DBkyJCMj49Ah/rBhACCXo7ZPfD93tm41ZnPmxsoKkZHIydG9885ET09PmUw2bdq0oKCg/fv30+Cz4ShGLdXNmzfv3r0rFou7du3KYhk6nW7KlCmlpaVjxoxhBqPbw8Pp1q4rGk3cHOtsbLhz5syRy+UpKSldu3bNyMgYP348c0eFnjNqCLqot1TMjdGgoCB2b2YlJSUdO3bM3d39888/HzBgQFhYWN1FPTNSZMhJ6m9KBGDNmjXGL5S0xNraetasWW+//TazgWzdagjSIopRS2Xqp4E25dy5c4mJiRwO54svvpBIJGhmpIhYECsrq6lTp7755pt0UW84ilFLxfow/b179yZNmlRTUzNv3ryRI0cyHxo4UkTMnBWtwG8NilFLxfow/YIFC65cueLv779s2TK2aiAm8Hg3ap4oFKMWKSMjQy6X83g8FxcXVgo4cuTIxo0bra2td+3aJWxuUTfpEOhGTYsoRi3P6tWrY2JiAGi1Wm9v71GjRs2bN2/gwIEmK6CwsHDGjBl6vX7FihX1u8M0UtQh0Y2aFtGEJ0uiUChCQkIWLFig1WptbW1DQ0O1Wu2+ffsGDRo0ZMiQvXv3MttctreIiIg7d+4MHz48KirKBM0RYuYoRi3Gjz/+6OrqeubMGQBDhw4tLi7+7bffrl+/npCQIJFITp48GRYW9tRTTyUmJhYWFrZfGRs2bDhw4IBYLN68eTOtdSEEAPTE7Om02uTJk5n5oQKBYNOmTY0OqKqq2rp1a69evZi/U2tr6wkTJqSlpRm9kpycHJFIBGD37t1GPzkhFopi1OzdvTu9dp2Sr6/v33///YhjT5w4MWHChLrnMgUHB2/durWmpsYohajVaubBn9OnTzfKCQnpGChGzUv37t0PHz5c9zZqzJjZdnY/A04cztdz5xp4kqtXr8bExDAbCQNwc3NLSEgoLCxsY21xcXEAvLy8ysvL23gqQjoSilHz8iBGq6r0MTFRHM5sQP/CCxVXrrT2VBUVFampqYGBgXVX+lOmTLlw4cLjFfb777/zeDwul3vs2LHHOwMhHRUNEZirzz5DUhK4XAwahKNHRT16tPYEIpEoMjLy0qVLzJW+RqPZvn173759+/fvv23btlZNWykvL588ebJWq120aNHQh3fEJOTJRjFqHtavR48esLZGfj527YJWi/nzMW4cwsIQFNTGPYqZuVDZ2dkxMTFOTk5nz56dNm1at27dEhMTi4uLDTnDnDlz8vLygoODFy1a1JZKCOmY2O4OE71+5Uq9q6v+0CG9QtHdw8OBy5XY2EgkEolEYmNjM3v2bCM2pVAoUlNTAwICmL995kr/4sWLj/iSb775BoCdnV12drYRKyGkw6AYZVtFhd7OTv/dd8y77t27b//kkxtc7o1Tp27cuBEREWHcGGXodLpffvllwoQJvNonVAwePJiZvd/oyIKCAmao6uFZVoQQBl3Us+30aajVeOONug+cg4M9vbw8s7M9PT1FdU+tNCoOhzN8+PC9e/deuXIlJiZGLBafPHly4sSJvr6+SUlJJbW7vet0uqlTp5aUlIwaNeodoz9Gg5COgmKUbcXFcHFpfPfT1RWG3bVsI6lUunLlyuvXr6empvr5+eXm5sbGxnp4eEydOlUmk61evfq///1v586dU1NTTVAMIRaKYpRtEgkKC6HTNfjwzh1IJCYrwd7ePjIyUiaTHThw4OWXX1apVNu3b+/du/eHH37I4XC++uqrLl26mKwYQiwOR097XLNLqYSrK3bswOjR9z85fx79+yMnx5iPRW+NnJyc5OTkTZs28fl8pVI5fPjw5cuXh4SEsFIMIeaPeqNsE4mweDFmzsThw6iowKlTePNNREaylaEAfHx81q5dO3v2bKVSaWNjc/To0WHDhpWXl9PTzQhpEu03agZsbTFlCqKikJcHNzdERCA+nu2awIxuRUVFcblcOzs7R0dHtisixExRjLKtoADz5oHPx+3bcHJiu5oHmP1NBALB0qVL2a6FELNGF/Vs++oraLUYPdqsMhQA8/gd0+wDTYhFo94oq/R6bN0KAOHhbJfSGNMbbbT0np5uRsjDKEZZdewY5HJ4euLFF9kupTEmRhv1RunpZoQ8jGKUTb8fPtzP1tY2PBy1izLNB9PrbNQbpaebEfIwujfKmvLy8ldSUjwEgrIZM9iupQlNXtQTQh5GMcqaXbt23bt3r39IiPjpp9mupQlNXtQTQh5GF/Ws2bJlC4Bw8xtcYjx8UU+PoSekSRSj7Lh06dKZM2fEYvHoujWgZiZUJMp67jmOiwvbhRBi7ihG2bF//34AkyZNEgqFbNfStK5aLU6cAG1KQkhLKEbZkZCQEBoaWjfqbY6YpzTTEBMhLaEYZQeHw6mbgGmmmBilISZCWkIj9aZjYTskMauVqDdKSEsoRkkz6KKeEMNQjJJmML1RuqgnpCUUo6QZ1BslxDA0xGRSlrRDEvVGCTEMxahJWdIOSdQbJcQwFKMmZUk7JFFvlBDDUIySZri54d//hrs723UQYu4oRkkzJBJER7NdBCEWgJ5TT5qyfj3Wrr3/pNLwcMTHm+HG0oSYCeqNkockJWHNGmzZgiFDcOkSZszA7dvYsIHtsggxU9QbJQ0plXB1xY4dqNvB7/x59O8PuRxeXqxWRoiZoun3pKHTp6FW4403HnwSFAQvLxw/zl5NhJg1ilHSUHExXFzAbfgfhqsriotZKogQc0cxShqSSFBYCJ2uwYd37kAiYakgQswdxShpKCQEAgF++OHBJ+fPIzcXzz/PXk2EmDUaqScNiURYvBgzZ8La+sFIfWQkvL3ZrowQM0Uj9aQpKSkP5o1GRCA+HgoFnJzYLosQc0QxSlqi1eK997BnDy5fhocH29UQYnbo3ihpCY8HhQIVFViyhO1SCDFH1BslBsjJQUAAdDrIZOjZk+1qCDEvFKPEIJVz566VyfKffjr1yy/ZroUQ80IxSgxy69YtHx+fqqqq06dPBwcHs10OIWaE7o0Sg7i7u8+ePVuv18fHx7NdCyHmhWKUGCo2NtbJyUksFqtUKrZrIcSM0EU9aYXi4mIJrQolpCGKUUIIaRO6qCetI5VKjxw5Uvc2Ojp6zpw5LNZDCOsoRgkhpE0oRgkhpE0oRgkhpE1oozzSamFhYQKBgHldWVkZERHBbj2EsItilLRaSkpKaGgo8zoxMZHNUggxAxSjpNWcnZ09PT2Z1yKRSKPRsFsPIeyie6OEENImFKOEENImtIqJEELa5P8BBSg/mL55fRAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x7fc2287ec0d0>"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(similar_smiles[0])\n",
    "molecules[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform docking in SwissDock"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def job_setup():\n",
    "    pass\n",
    "\n",
    "def prepare_protein(protein):\n",
    "    pass\n",
    "\n",
    "def prepare_ligand(ligand):\n",
    "    pass\n",
    "\n",
    "def dock(protein, ligand):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assess results with proteins.plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_interations():\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Case study: XXXXXXXXXXX\n",
    "Pending."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "# Discussion\n",
    "Pending."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "# Quiz\n",
    "Pending."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
